% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/methods.R
\docType{methods}
\name{ttBulk_SAM_BAM}
\alias{ttBulk_SAM_BAM}
\title{Creates a `tt` object from a list of file names of BAM/SAM}
\usage{
ttBulk_SAM_BAM(file_names, genome = "hg38", ...)
}
\arguments{
\item{file_names}{A character vector}

\item{genome}{A character string}

\item{...}{Further parameters passed to the function Rsubread::featureCounts}
}
\value{
A `ttBulk` object
}
\description{
ttBulk_SAM_BAM() creates a `tt` object from a `tbl` formatted as | <SAMPLE> | <TRANSCRIPT> | <COUNT> | <...> |
}
\details{
This function is based on FeatureCounts package. This function created a ttBulk object and is useful if you want
to avoid to specify .sample, .transcript and .abundance arguments all the times.
The ttBulk object have an attribute called parameters where these three
arguments are stored as metadata. They can be extracted as attr(<object>, "parameters").
}
