<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings transcriptomics to the tidyverse • tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings transcriptomics to the tidyverse">
<meta property="og:description" content="This is a collection of utility functions that allow to perform exploration of and calculations to RNA sequencing data, in a modular, pipe-friendly and tidy fashion.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/comparison_coding.html">Side-by-side comparison with standard interfaces</a>
    </li>
    <li>
      <a href="articles/introduction.html">tidybulk: An R tidy framework for modular transcriptomic data analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="tidybulk-an-r-tidy-framework-for-modular-transcriptomic-data-analysis">tidybulk: An R tidy framework for modular transcriptomic data analysis<a class="anchor" aria-label="anchor" href="#tidybulk-an-r-tidy-framework-for-modular-transcriptomic-data-analysis"></a>
</h1></div>
<!-- badges: start -->

<p><strong>tidybulk</strong> is a powerful R package designed for modular transcriptomic data analysis that brings transcriptomics to the tidyverse.</p>
<div class="section level2">
<h2 id="why-tidybulk">Why tidybulk?<a class="anchor" aria-label="anchor" href="#why-tidybulk"></a>
</h2>
<p>Tidybulk provides a unified interface for comprehensive transcriptomic data analysis with seamless integration of SummarizedExperiment objects and tidyverse principles. It streamlines the entire workflow from raw data to biological insights.</p>
</div>
<div class="section level2">
<h2 id="functionsutilities-available">Functions/utilities available<a class="anchor" aria-label="anchor" href="#functionsutilities-available"></a>
</h2>
<div class="section level3">
<h3 id="abundance-normalization-functions">Abundance Normalization Functions<a class="anchor" aria-label="anchor" href="#abundance-normalization-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="45%">
<col width="54%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/scale_abundance-methods.html">scale_abundance()</a></code></td>
<td>Scale abundance data</td>
</tr>
<tr>
<td><code><a href="reference/quantile_normalise_abundance-methods.html">quantile_normalise_abundance()</a></code></td>
<td>Quantile normalization</td>
</tr>
<tr>
<td><code><a href="reference/adjust_abundance-methods.html">adjust_abundance()</a></code></td>
<td>Adjust abundance for unwanted variation</td>
</tr>
<tr>
<td><code><a href="reference/fill_missing_abundance-methods.html">fill_missing_abundance()</a></code></td>
<td>Fill missing abundance values</td>
</tr>
<tr>
<td><code><a href="reference/impute_missing_abundance-methods.html">impute_missing_abundance()</a></code></td>
<td>Impute missing abundance values</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="filtering-and-selection-functions">Filtering and Selection Functions<a class="anchor" aria-label="anchor" href="#filtering-and-selection-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="30%">
<col width="69%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/identify_abundant-methods.html">identify_abundant()</a></code></td>
<td>Identify abundant transcripts without removing them</td>
</tr>
<tr>
<td><code><a href="reference/keep_abundant.html">keep_abundant()</a></code></td>
<td>Keep abundant transcripts</td>
</tr>
<tr>
<td><code><a href="reference/keep_variable-methods.html">keep_variable()</a></code></td>
<td>Keep variable transcripts</td>
</tr>
<tr>
<td><code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr()</a></code></td>
<td>Filter by expression</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="dimensionality-reduction-functions">Dimensionality Reduction Functions<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/reduce_dimensions-methods.html">reduce_dimensions()</a></code></td>
<td>Reduce dimensions with PCA/MDS/tSNE/UMAP</td>
</tr>
<tr>
<td><code><a href="reference/rotate_dimensions-methods.html">rotate_dimensions()</a></code></td>
<td>Rotate dimensions</td>
</tr>
<tr>
<td><code><a href="reference/remove_redundancy-methods.html">remove_redundancy()</a></code></td>
<td>Remove redundant features</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="clustering-functions">Clustering Functions<a class="anchor" aria-label="anchor" href="#clustering-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/cluster_elements-methods.html">cluster_elements()</a></code></td>
<td>Cluster elements with various methods</td>
</tr>
<tr>
<td><code>kmeans clustering</code></td>
<td>K-means clustering</td>
</tr>
<tr>
<td><code>SNN clustering</code></td>
<td>Shared nearest neighbor clustering</td>
</tr>
<tr>
<td><code>hierarchical clustering</code></td>
<td>Hierarchical clustering</td>
</tr>
<tr>
<td><code>DBSCAN clustering</code></td>
<td>Density-based clustering</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="differential-analysis-functions">Differential Analysis Functions<a class="anchor" aria-label="anchor" href="#differential-analysis-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody><tr>
<td><code><a href="reference/test_differential_expression-methods.html">test_differential_expression()</a></code></td>
<td>Test differential expression with various methods</td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="cellularity-analysis-functions">Cellularity Analysis Functions<a class="anchor" aria-label="anchor" href="#cellularity-analysis-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="38%">
<col width="61%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity()</a></code></td>
<td>Deconvolve cellularity with various methods</td>
</tr>
<tr>
<td><code>cibersort()</code></td>
<td>CIBERSORT analysis</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="gene-enrichment-functions">Gene Enrichment Functions<a class="anchor" aria-label="anchor" href="#gene-enrichment-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/test_gene_enrichment-methods.html">test_gene_enrichment()</a></code></td>
<td>Test gene enrichment</td>
</tr>
<tr>
<td><code><a href="reference/test_gene_overrepresentation-methods.html">test_gene_overrepresentation()</a></code></td>
<td>Test gene overrepresentation</td>
</tr>
<tr>
<td><code><a href="reference/test_gene_rank-methods.html">test_gene_rank()</a></code></td>
<td>Test gene rank</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="utility-functions">Utility Functions<a class="anchor" aria-label="anchor" href="#utility-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/describe_transcript-methods.html">describe_transcript()</a></code></td>
<td>Describe transcript characteristics</td>
</tr>
<tr>
<td><code><a href="reference/get_bibliography-methods.html">get_bibliography()</a></code></td>
<td>Get bibliography</td>
</tr>
<tr>
<td><code><a href="reference/resolve_complete_confounders_of_non_interest.html">resolve_complete_confounders_of_non_interest()</a></code></td>
<td>Resolve confounders</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="validation-and-utility-functions">Validation and Utility Functions<a class="anchor" aria-label="anchor" href="#validation-and-utility-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/check_if_counts_is_na.html">check_if_counts_is_na()</a></code></td>
<td>Check if counts contain NA values</td>
</tr>
<tr>
<td><code><a href="reference/check_if_duplicated_genes.html">check_if_duplicated_genes()</a></code></td>
<td>Check for duplicated genes</td>
</tr>
<tr>
<td><code>check_if_wrong_input()</code></td>
<td>Validate input data</td>
</tr>
<tr>
<td><code><a href="reference/log10_reverse_trans.html">log10_reverse_trans()</a></code></td>
<td>Log10 reverse transformation</td>
</tr>
<tr>
<td><code><a href="reference/logit_trans.html">logit_trans()</a></code></td>
<td>Logit transformation</td>
</tr>
</tbody>
</table>
<p>All functions are directly compatible with <code>SummarizedExperiment</code> objects and follow tidyverse principles for seamless integration with the tidyverse ecosystem.</p>
</div>
<div class="section level3">
<h3 id="scientific-citation">Scientific Citation<a class="anchor" aria-label="anchor" href="#scientific-citation"></a>
</h3>
<p>Mangiola, Stefano, Ramyar Molania, Ruining Dong, Maria A. Doyle, and Anthony T. Papenfuss. 2021. “Tidybulk: An R tidy framework for modular transcriptomic data analysis.” Genome Biology 22 (42). <a href="https://doi.org/10.1186/s13059-020-02233-7" class="external-link uri">https://doi.org/10.1186/s13059-020-02233-7</a></p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7" class="external-link">Genome Biology - tidybulk: an R tidy framework for modular transcriptomic data analysis</a></p>
<p>In this example we will use the <code>airway</code> dataset, a <code>SummarizedExperiment</code> object containing RNA-seq data from an experiment studying the effect of dexamethasone treatment on airway smooth muscle cells. This dataset is available in the <a href="https://bioconductor.org/packages/airway/" class="external-link">airway</a> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparison-of-differential-expression-methods">Comparison of Differential Expression methods<a class="anchor" aria-label="anchor" href="#comparison-of-differential-expression-methods"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Standard differential expression analysis</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span><span class="co"># Use QL method</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>, prefix <span class="op">=</span> <span class="st">"ql__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Use edger_robust_likelihood_ratio</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"edger_robust_likelihood_ratio"</span>, prefix <span class="op">=</span> <span class="st">"lr_robust__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span><span class="co"># Use DESeq2 method</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"DESeq2"</span>, prefix <span class="op">=</span> <span class="st">"deseq2__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Use limma_voom</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"limma_voom"</span>, prefix <span class="op">=</span> <span class="st">"voom__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span><span class="co"># Use limma_voom_sample_weights</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"limma_voom_sample_weights"</span>, prefix <span class="op">=</span> <span class="st">"voom_weights__"</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="histograms-of-p-values-across-methods">Histograms of p-values across methods<a class="anchor" aria-label="anchor" href="#histograms-of-p-values-across-methods"></a>
</h3>
<p>Inspection of the raw p-value histogram provides a rapid check of differential-expression results. When no gene is truly differentially expressed, the p-values follow a uniform U(0,1) distribution across the interval 0–1, so the histogram appears flat <a href="https://bioconductor.org/help/course-materials/2014/useR2014/Workflows.html" class="external-link">Source</a>. In a more realistic scenario where only a subset of genes changes, this uniform background is still present but an obvious spike emerges close to zero, created by the genuine signals.</p>
<p>Thanks to the modularity of the <code>tidybulk</code> workflow, that can multiplex different methods, we can easily compare the p-values across methods.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span></span>
<span>    <span class="va">ql__PValue</span>, </span>
<span>    <span class="va">lr_robust__PValue</span>, </span>
<span>    <span class="va">voom__P.Value</span>, </span>
<span>    <span class="va">voom_weights__P.Value</span>, </span>
<span>    <span class="va">deseq2__pvalue</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu">everything</span><span class="op">(</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"method"</span>, values_to <span class="op">=</span> <span class="st">"pvalue"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pvalue</span>, fill <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Histogram of p-values across methods"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing non-finite outside the scale range</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="do">## (`stat_bin()`).</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-pvalue-histograms-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="compare-results-across-methods">Compare Results Across Methods<a class="anchor" aria-label="anchor" href="#compare-results-across-methods"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summary statistics</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">select</span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"ql|lr_robust|voom|voom_weights|deseq2"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">select</span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"logFC"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span><span class="fu">summarise</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">everything</span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min <span class="op">=</span> <span class="va">min</span>, median <span class="op">=</span> <span class="va">median</span>, max <span class="op">=</span> <span class="va">max</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">## Warning: There was 1 warning in `summarise()`.</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="do">## ℹ In argument: `across(...)`.</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="do">## Caused by warning:</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="do">## ! The `...` argument of `across()` is deprecated as of dplyr 1.1.0.</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="do">## Supply arguments directly to `.fns` through an anonymous function instead.</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="do">##   # Previously</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="do">##   across(a:b, mean, na.rm = TRUE)</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="do">##   # Now</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="do">##   across(a:b, \(x) mean(x, na.rm = TRUE))</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="do">## # A tibble: 1 × 0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pairplot-of-pvalues-across-methods-ggpairs">Pairplot of pvalues across methods (GGpairs)<a class="anchor" aria-label="anchor" href="#pairplot-of-pvalues-across-methods-ggpairs"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/" class="external-link">GGally</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## Warning: package 'GGally' was built under R version 4.5.1</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">ql__PValue</span>, <span class="va">lr_robust__PValue</span>, <span class="va">voom__P.Value</span>, <span class="va">voom_weights__P.Value</span>, <span class="va">deseq2__pvalue</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="reference/log10_reverse_trans.html">log10_reverse_trans</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="reference/log10_reverse_trans.html">log10_reverse_trans</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Pairplot of p-values across methods"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-pvalue-pairplot-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="pairplot-of-effect-sizes-across-methods-ggpairs">Pairplot of effect sizes across methods (GGpairs)<a class="anchor" aria-label="anchor" href="#pairplot-of-effect-sizes-across-methods-ggpairs"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/" class="external-link">GGally</a></span><span class="op">)</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">ql__logFC</span>, <span class="va">lr_robust__logFC</span>, <span class="va">voom__logFC</span>, <span class="va">voom_weights__logFC</span>, <span class="va">deseq2__log2FoldChange</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Pairplot of effect sizes across methods"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-effectsize-pairplot-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>It is important to check the quality of the fit. All methods produce a fit object that can be used for quality control. The fit object produced by each underlying method is stored as an attribute of the <code>airway_mini</code> object. We can use them, for example, to perform quality control of the fit.</p>
<div class="section level4">
<h4 id="for-edger">For edgeR<a class="anchor" aria-label="anchor" href="#for-edger"></a>
</h4>
<p>Plot the biological coefficient of variation (BCV) trend. This plot is helpful in understanding the dispersion of the data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">edgeR_quasi_likelihood_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-edgeR-object-1.png"><!-- --></p>
<p>Plot the log-fold change vs mean plot.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">edgeR_quasi_likelihood_fit</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-edgeR-fit-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="for-deseq2">For DESeq2<a class="anchor" aria-label="anchor" href="#for-deseq2"></a>
</h4>
<p>Plot the mean-variance trend.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">DESeq2_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">plotDispEsts</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-DESeq2-object-1.png"><!-- --></p>
<p>Plot the log-fold change vs the mean plot.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">DESeq2_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotMA.html" class="external-link">plotMA</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-DESeq2-fit-1.png"><!-- --></p>
</div>
</div>
<div class="section level3">
<h3 id="volcano-plots-for-each-method">Volcano Plots for Each Method<a class="anchor" aria-label="anchor" href="#volcano-plots-for-each-method"></a>
</h3>
<p>Visualising the significance and effect size of the differential expression results as a volcano plot, we appreciate that some methods have much lower p-value distributions than other methods, for the same model and data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create volcano plots</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Select the columns we want to plot</span></span>
<span>    <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">select</span><span class="op">(</span></span>
<span>            <span class="va">.feature</span>,</span>
<span>      <span class="va">ql__logFC</span>, <span class="va">ql__PValue</span>,</span>
<span>      <span class="va">lr_robust__logFC</span>, <span class="va">lr_robust__PValue</span>,</span>
<span>      <span class="va">voom__logFC</span>, <span class="va">voom__P.Value</span>,</span>
<span>      <span class="va">voom_weights__logFC</span>, <span class="va">voom_weights__P.Value</span>,</span>
<span>      <span class="va">deseq2__log2FoldChange</span>, <span class="va">deseq2__pvalue</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Pivot longer to get a tidy data frame</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>      <span class="op">-</span> <span class="va">.feature</span>,</span>
<span>      names_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"method"</span>, <span class="st">"stat"</span><span class="op">)</span>,</span>
<span>      values_to <span class="op">=</span> <span class="st">"value"</span>, names_sep <span class="op">=</span> <span class="st">"__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Harmonize column names</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>stat  <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"log2FoldChange"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"logFC"</span>,</span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PValue"</span>, <span class="st">"pvalue"</span>, <span class="st">"P.Value"</span>, <span class="st">"p.value"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"PValue"</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"stat"</span>, values_from <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, size <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="reference/log10_reverse_trans.html">log10_reverse_trans</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="fl">0.5</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Volcano Plots by Method"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="do">## (`geom_point()`).</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-volcano-plots-1-1.png"><!-- --></p>
<p>Plotting independent y-axis scales for the p-values and effect sizes allows us to compare the top genes across methods.</p>
<p><strong>Methods:</strong></p>
<ul>
<li><p><strong>edgeR quasi-likelihood:</strong> Quasi-likelihood F-tests for differential expression</p></li>
<li><p><strong>edgeR robust likelihood ratio:</strong> Robust likelihood ratio tests</p></li>
<li><p><strong>DESeq2:</strong> Negative binomial distribution with dispersion estimation</p></li>
<li><p><strong>limma-voom:</strong> Linear modeling with empirical Bayes moderation</p></li>
<li><p><strong>limma-voom with sample weights:</strong> Enhanced voom with quality weights <strong>References:</strong></p></li>
<li><p>Robinson et al. (2010) edgeR: a Bioconductor package for differential expression analysis</p></li>
<li><p>Chen et al. (2016) From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline</p></li>
<li><p>Love et al. (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2</p></li>
<li><p>Law et al. (2014) voom: precision weights unlock linear model analysis tools for RNA-seq read counts</p></li>
<li><p>Liu et al. (2015) Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="automatic-bibliography">Automatic bibliography<a class="anchor" aria-label="anchor" href="#automatic-bibliography"></a>
</h2>
<p><code>tidybulk</code> allows you to get the bibliography of all methods used in our workflow.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get bibliography of all methods used in our workflow</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/get_bibliography-methods.html">get_bibliography</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="do">##  @Article{tidybulk,</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="do">##   title = {tidybulk: an R tidy framework for modular transcriptomic data analysis},</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="do">##   author = {Stefano Mangiola and Ramyar Molania and Ruining Dong and Maria A. Doyle &amp; Anthony T. Papenfuss},</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="do">##   journal = {Genome Biology},</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="do">##   year = {2021},</span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="do">##   volume = {22},</span></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="do">##   number = {42},</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="do">##   url = {https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7},</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="do">##   }</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a><span class="do">## @article{wickham2019welcome,</span></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a><span class="do">##   title={Welcome to the Tidyverse},</span></span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a><span class="do">##   author={Wickham, Hadley and Averick, Mara and Bryan, Jennifer and Chang, Winston and McGowan, Lucy D'Agostino and Francois, Romain and Grolemund, Garrett and Hayes, Alex and Henry, Lionel and Hester, Jim and others},</span></span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a><span class="do">##   journal={Journal of Open Source Software},</span></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a><span class="do">##   volume={4},</span></span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a><span class="do">##   number={43},</span></span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a><span class="do">##   pages={1686},</span></span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a><span class="do">##   year={2019}</span></span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb19-19"><a href="#cb19-19" tabindex="-1"></a><span class="do">## @article{robinson2010edger,</span></span>
<span id="cb19-20"><a href="#cb19-20" tabindex="-1"></a><span class="do">##   title={edgeR: a Bioconductor package for differential expression analysis of digital gene expression data},</span></span>
<span id="cb19-21"><a href="#cb19-21" tabindex="-1"></a><span class="do">##   author={Robinson, Mark D and McCarthy, Davis J and Smyth, Gordon K},</span></span>
<span id="cb19-22"><a href="#cb19-22" tabindex="-1"></a><span class="do">##   journal={Bioinformatics},</span></span>
<span id="cb19-23"><a href="#cb19-23" tabindex="-1"></a><span class="do">##   volume={26},</span></span>
<span id="cb19-24"><a href="#cb19-24" tabindex="-1"></a><span class="do">##   number={1},</span></span>
<span id="cb19-25"><a href="#cb19-25" tabindex="-1"></a><span class="do">##   pages={139--140},</span></span>
<span id="cb19-26"><a href="#cb19-26" tabindex="-1"></a><span class="do">##   year={2010},</span></span>
<span id="cb19-27"><a href="#cb19-27" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb19-28"><a href="#cb19-28" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb19-29"><a href="#cb19-29" tabindex="-1"></a><span class="do">## @article{robinson2010scaling,</span></span>
<span id="cb19-30"><a href="#cb19-30" tabindex="-1"></a><span class="do">##   title={A scaling normalization method for differential expression analysis of RNA-seq data},</span></span>
<span id="cb19-31"><a href="#cb19-31" tabindex="-1"></a><span class="do">##   author={Robinson, Mark D and Oshlack, Alicia},</span></span>
<span id="cb19-32"><a href="#cb19-32" tabindex="-1"></a><span class="do">##   journal={Genome biology},</span></span>
<span id="cb19-33"><a href="#cb19-33" tabindex="-1"></a><span class="do">##   volume={11},</span></span>
<span id="cb19-34"><a href="#cb19-34" tabindex="-1"></a><span class="do">##   number={3},</span></span>
<span id="cb19-35"><a href="#cb19-35" tabindex="-1"></a><span class="do">##   pages={1--9},</span></span>
<span id="cb19-36"><a href="#cb19-36" tabindex="-1"></a><span class="do">##   year={2010},</span></span>
<span id="cb19-37"><a href="#cb19-37" tabindex="-1"></a><span class="do">##   publisher={BioMed Central}</span></span>
<span id="cb19-38"><a href="#cb19-38" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb19-39"><a href="#cb19-39" tabindex="-1"></a><span class="do">## @incollection{smyth2005limma,</span></span>
<span id="cb19-40"><a href="#cb19-40" tabindex="-1"></a><span class="do">##   title={Limma: linear models for microarray data},</span></span>
<span id="cb19-41"><a href="#cb19-41" tabindex="-1"></a><span class="do">##   author={Smyth, Gordon K},</span></span>
<span id="cb19-42"><a href="#cb19-42" tabindex="-1"></a><span class="do">##   booktitle={Bioinformatics and computational biology solutions using R and Bioconductor},</span></span>
<span id="cb19-43"><a href="#cb19-43" tabindex="-1"></a><span class="do">##   pages={397--420},</span></span>
<span id="cb19-44"><a href="#cb19-44" tabindex="-1"></a><span class="do">##   year={2005},</span></span>
<span id="cb19-45"><a href="#cb19-45" tabindex="-1"></a><span class="do">##   publisher={Springer}</span></span>
<span id="cb19-46"><a href="#cb19-46" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb19-47"><a href="#cb19-47" tabindex="-1"></a><span class="do">## @Manual{,</span></span>
<span id="cb19-48"><a href="#cb19-48" tabindex="-1"></a><span class="do">##     title = {R: A Language and Environment for Statistical Computing},</span></span>
<span id="cb19-49"><a href="#cb19-49" tabindex="-1"></a><span class="do">##     author = {{R Core Team}},</span></span>
<span id="cb19-50"><a href="#cb19-50" tabindex="-1"></a><span class="do">##     organization = {R Foundation for Statistical Computing},</span></span>
<span id="cb19-51"><a href="#cb19-51" tabindex="-1"></a><span class="do">##     address = {Vienna, Austria},</span></span>
<span id="cb19-52"><a href="#cb19-52" tabindex="-1"></a><span class="do">##     year = {2020},</span></span>
<span id="cb19-53"><a href="#cb19-53" tabindex="-1"></a><span class="do">##     url = {https://www.R-project.org/},</span></span>
<span id="cb19-54"><a href="#cb19-54" tabindex="-1"></a><span class="do">##   }</span></span>
<span id="cb19-55"><a href="#cb19-55" tabindex="-1"></a><span class="do">## @article{lund2012detecting,</span></span>
<span id="cb19-56"><a href="#cb19-56" tabindex="-1"></a><span class="do">##   title={Detecting differential expression in RNA-sequence data using quasi-likelihood with shrunken dispersion estimates},</span></span>
<span id="cb19-57"><a href="#cb19-57" tabindex="-1"></a><span class="do">##   author={Lund, Steven P and Nettleton, Dan and McCarthy, Davis J and Smyth, Gordon K},</span></span>
<span id="cb19-58"><a href="#cb19-58" tabindex="-1"></a><span class="do">##   journal={Statistical applications in genetics and molecular biology},</span></span>
<span id="cb19-59"><a href="#cb19-59" tabindex="-1"></a><span class="do">##   volume={11},</span></span>
<span id="cb19-60"><a href="#cb19-60" tabindex="-1"></a><span class="do">##   number={5},</span></span>
<span id="cb19-61"><a href="#cb19-61" tabindex="-1"></a><span class="do">##   year={2012},</span></span>
<span id="cb19-62"><a href="#cb19-62" tabindex="-1"></a><span class="do">##   publisher={De Gruyter}</span></span>
<span id="cb19-63"><a href="#cb19-63" tabindex="-1"></a><span class="do">##     }</span></span>
<span id="cb19-64"><a href="#cb19-64" tabindex="-1"></a><span class="do">## @article{zhou2014robustly,</span></span>
<span id="cb19-65"><a href="#cb19-65" tabindex="-1"></a><span class="do">##   title={Robustly detecting differential expression in RNA sequencing data using observation weights},</span></span>
<span id="cb19-66"><a href="#cb19-66" tabindex="-1"></a><span class="do">##   author={Zhou, Xiaobei and Lindsay, Helen and Robinson, Mark D},</span></span>
<span id="cb19-67"><a href="#cb19-67" tabindex="-1"></a><span class="do">##   journal={Nucleic acids research},</span></span>
<span id="cb19-68"><a href="#cb19-68" tabindex="-1"></a><span class="do">##   volume={42},</span></span>
<span id="cb19-69"><a href="#cb19-69" tabindex="-1"></a><span class="do">##   number={11},</span></span>
<span id="cb19-70"><a href="#cb19-70" tabindex="-1"></a><span class="do">##   pages={e91--e91},</span></span>
<span id="cb19-71"><a href="#cb19-71" tabindex="-1"></a><span class="do">##   year={2014},</span></span>
<span id="cb19-72"><a href="#cb19-72" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb19-73"><a href="#cb19-73" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb19-74"><a href="#cb19-74" tabindex="-1"></a><span class="do">## @article{love2014moderated,</span></span>
<span id="cb19-75"><a href="#cb19-75" tabindex="-1"></a><span class="do">##   title={Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2},</span></span>
<span id="cb19-76"><a href="#cb19-76" tabindex="-1"></a><span class="do">##   author={Love, Michael I and Huber, Wolfgang and Anders, Simon},</span></span>
<span id="cb19-77"><a href="#cb19-77" tabindex="-1"></a><span class="do">##   journal={Genome biology},</span></span>
<span id="cb19-78"><a href="#cb19-78" tabindex="-1"></a><span class="do">##   volume={15},</span></span>
<span id="cb19-79"><a href="#cb19-79" tabindex="-1"></a><span class="do">##   number={12},</span></span>
<span id="cb19-80"><a href="#cb19-80" tabindex="-1"></a><span class="do">##   pages={550},</span></span>
<span id="cb19-81"><a href="#cb19-81" tabindex="-1"></a><span class="do">##   year={2014},</span></span>
<span id="cb19-82"><a href="#cb19-82" tabindex="-1"></a><span class="do">##   publisher={Springer}</span></span>
<span id="cb19-83"><a href="#cb19-83" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb19-84"><a href="#cb19-84" tabindex="-1"></a><span class="do">## @article{law2014voom,</span></span>
<span id="cb19-85"><a href="#cb19-85" tabindex="-1"></a><span class="do">##   title={voom: Precision weights unlock linear model analysis tools for RNA-seq read counts},</span></span>
<span id="cb19-86"><a href="#cb19-86" tabindex="-1"></a><span class="do">##   author={Law, Charity W and Chen, Yunshun and Shi, Wei and Smyth, Gordon K},</span></span>
<span id="cb19-87"><a href="#cb19-87" tabindex="-1"></a><span class="do">##   journal={Genome biology},</span></span>
<span id="cb19-88"><a href="#cb19-88" tabindex="-1"></a><span class="do">##   volume={15},</span></span>
<span id="cb19-89"><a href="#cb19-89" tabindex="-1"></a><span class="do">##   number={2},</span></span>
<span id="cb19-90"><a href="#cb19-90" tabindex="-1"></a><span class="do">##   pages={R29},</span></span>
<span id="cb19-91"><a href="#cb19-91" tabindex="-1"></a><span class="do">##   year={2014},</span></span>
<span id="cb19-92"><a href="#cb19-92" tabindex="-1"></a><span class="do">##   publisher={Springer}</span></span>
<span id="cb19-93"><a href="#cb19-93" tabindex="-1"></a><span class="do">##     }</span></span>
<span id="cb19-94"><a href="#cb19-94" tabindex="-1"></a><span class="do">## @article{liu2015weight,</span></span>
<span id="cb19-95"><a href="#cb19-95" tabindex="-1"></a><span class="do">##   title={Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses},</span></span>
<span id="cb19-96"><a href="#cb19-96" tabindex="-1"></a><span class="do">##   author={Liu, Ruijie and Holik, Aliaksei Z and Su, Shian and Jansz, Natasha and Chen, Kelan and Leong, Huei San and Blewitt, Marnie E and Asselin-Labat, Marie-Liesse and Smyth, Gordon K and Ritchie, Matthew E},</span></span>
<span id="cb19-97"><a href="#cb19-97" tabindex="-1"></a><span class="do">##   journal={Nucleic acids research},</span></span>
<span id="cb19-98"><a href="#cb19-98" tabindex="-1"></a><span class="do">##   volume={43},</span></span>
<span id="cb19-99"><a href="#cb19-99" tabindex="-1"></a><span class="do">##   number={15},</span></span>
<span id="cb19-100"><a href="#cb19-100" tabindex="-1"></a><span class="do">##   pages={e97--e97},</span></span>
<span id="cb19-101"><a href="#cb19-101" tabindex="-1"></a><span class="do">##   year={2015},</span></span>
<span id="cb19-102"><a href="#cb19-102" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb19-103"><a href="#cb19-103" tabindex="-1"></a><span class="do">##     }</span></span>
<span id="cb19-104"><a href="#cb19-104" tabindex="-1"></a><span class="do">## @article{leek2012sva,</span></span>
<span id="cb19-105"><a href="#cb19-105" tabindex="-1"></a><span class="do">##   title={The sva package for removing batch effects and other unwanted variation in high-throughput experiments},</span></span>
<span id="cb19-106"><a href="#cb19-106" tabindex="-1"></a><span class="do">##   author={Leek, Jeffrey T and Johnson, W Evan and Parker, Hilary S and Jaffe, Andrew E and Storey, John D},</span></span>
<span id="cb19-107"><a href="#cb19-107" tabindex="-1"></a><span class="do">##   journal={Bioinformatics},</span></span>
<span id="cb19-108"><a href="#cb19-108" tabindex="-1"></a><span class="do">##   volume={28},</span></span>
<span id="cb19-109"><a href="#cb19-109" tabindex="-1"></a><span class="do">##   number={6},</span></span>
<span id="cb19-110"><a href="#cb19-110" tabindex="-1"></a><span class="do">##   pages={882--883},</span></span>
<span id="cb19-111"><a href="#cb19-111" tabindex="-1"></a><span class="do">##   year={2012},</span></span>
<span id="cb19-112"><a href="#cb19-112" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb19-113"><a href="#cb19-113" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb19-114"><a href="#cb19-114" tabindex="-1"></a><span class="do">## @article{newman2015robust,</span></span>
<span id="cb19-115"><a href="#cb19-115" tabindex="-1"></a><span class="do">##   title={Robust enumeration of cell subsets from tissue expression profiles},</span></span>
<span id="cb19-116"><a href="#cb19-116" tabindex="-1"></a><span class="do">##   author={Newman, Aaron M and Liu, Chih Long and Green, Michael R and Gentles, Andrew J and Feng, Weiguo and Xu, Yue and Hoang, Chuong D and Diehn, Maximilian and Alizadeh, Ash A},</span></span>
<span id="cb19-117"><a href="#cb19-117" tabindex="-1"></a><span class="do">##   journal={Nature methods},</span></span>
<span id="cb19-118"><a href="#cb19-118" tabindex="-1"></a><span class="do">##   volume={12},</span></span>
<span id="cb19-119"><a href="#cb19-119" tabindex="-1"></a><span class="do">##   number={5},</span></span>
<span id="cb19-120"><a href="#cb19-120" tabindex="-1"></a><span class="do">##   pages={453--457},</span></span>
<span id="cb19-121"><a href="#cb19-121" tabindex="-1"></a><span class="do">##   year={2015},</span></span>
<span id="cb19-122"><a href="#cb19-122" tabindex="-1"></a><span class="do">##   publisher={Nature Publishing Group}</span></span>
<span id="cb19-123"><a href="#cb19-123" tabindex="-1"></a><span class="do">##  }</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/tidybulk" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/stemangiola/tidybulk/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/stemangiola/tidybulk/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidybulk</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>   </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidybulk/actions/" class="external-link"><img src="https://github.com/stemangiola/tidybulk/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/tidybulk/" class="external-link"><img src="https://bioconductor.org/shields/build/release/bioc/tidybulk.svg" alt="Bioconductor status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
