<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings transcriptomics to the tidyverse  • tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings transcriptomics to the tidyverse ">
<meta property="og:description" content="This is a collection of utility functions that allow
    to perform exploration of and calculations to RNA sequencing data, in
    a modular, pipe-friendly and tidy fashion.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Overview of the tidybulk package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="tidybulk---part-of-tidytranscriptomics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#tidybulk---part-of-tidytranscriptomics" class="anchor"></a>tidybulk - part of tidyTranscriptomics</h1></div>
<!-- badges: start -->

<p><strong>Brings transcriptomics to the tidyverse!</strong> (SEO: tidy differential expression)</p>
</div>
<div id="section" class="section level1">
<h1 class="hasAnchor">
<a href="#section" class="anchor"></a><img src="inst/logo.png" height="139px" width="120px">
</h1>
<p>website: <a href="http://stemangiola.github.io/tidybulk/">stemangiola.github.io/tidybulk/</a></p>
<p>Please have a look also to</p>
<ul>
<li>
<a href="https://github.com/stemangiola/nanny">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://github.com/stemangiola/tidyHeatmap">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
<!---

[![Build Status](https://travis-ci.org/stemangiola/tidybulk.svg?branch=master)](https://travis-ci.org/stemangiola/tidybulk) [![Coverage Status](https://coveralls.io/repos/github/stemangiola/tidybulk/badge.svg?branch=master)](https://coveralls.io/github/stemangiola/tidybulk?branch=master)

--><div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<colgroup>
<col width="28%">
<col width="71%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>aggregate_duplicates</code></td>
<td>Aggregate abundance and annotation of duplicated transcripts in a robust way</td>
</tr>
<tr class="even">
<td><code>scale_abundance</code></td>
<td>Scale (normalise) abundance for RNA sequencing depth</td>
</tr>
<tr class="odd">
<td><code>reduce_dimensions</code></td>
<td>Perform dimensionality reduction (PCA, MDS, tSNE)</td>
</tr>
<tr class="even">
<td><code>cluster_elements</code></td>
<td>Labels elements with cluster identity (kmeans, SNN)</td>
</tr>
<tr class="odd">
<td><code>remove_redundancy</code></td>
<td>Filter out elements with highly correlated features</td>
</tr>
<tr class="even">
<td><code>adjust_abundance</code></td>
<td>Remove known unwanted variation (Combat)</td>
</tr>
<tr class="odd">
<td><code>test_differential_abundance</code></td>
<td>Differential transcript abundance testing (DE)</td>
</tr>
<tr class="even">
<td><code>deconvolve_cellularity</code></td>
<td>Estimated tissue composition (Cibersort or llsr)</td>
</tr>
<tr class="odd">
<td><code>test_differential_cellularity</code></td>
<td>Differential cell-type abundance testing</td>
</tr>
<tr class="even">
<td><code>keep_variable</code></td>
<td>Filter for top variable features</td>
</tr>
<tr class="odd">
<td><code>keep_abundant</code></td>
<td>Filter out lowly abundant transcripts</td>
</tr>
<tr class="even">
<td><code>test_gene_enrichment</code></td>
<td>Gene enrichment analyses (EGSEA)</td>
</tr>
<tr class="odd">
<td><code>test_gene_overrepresentation</code></td>
<td>Gene enrichment on list of transcript names (no rank)</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="29%">
<col width="70%">
</colgroup>
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>get_bibliography</code></td>
<td>Get the bibliography of your workflow</td>
</tr>
<tr class="even">
<td><code>tidybulk</code></td>
<td>add tidybulk attributes to a tibble object</td>
</tr>
<tr class="odd">
<td><code>tidybulk_SAM_BAM</code></td>
<td>Convert SAM BAM files into tidybulk tibble</td>
</tr>
<tr class="even">
<td><code>pivot_sample</code></td>
<td>Select sample-wise columns/information</td>
</tr>
<tr class="odd">
<td><code>pivot_transcript</code></td>
<td>Select transcript-wise columns/information</td>
</tr>
<tr class="even">
<td><code>rotate_dimensions</code></td>
<td>Rotate two dimensions of a degree</td>
</tr>
<tr class="odd">
<td><code>ensembl_to_symbol</code></td>
<td>Add gene symbol from ensembl IDs</td>
</tr>
<tr class="even">
<td><code>symbol_to_entrez</code></td>
<td>Add entrez ID from gene symbol</td>
</tr>
<tr class="odd">
<td><code>describe_transcript</code></td>
<td>Add gene description from gene symbol</td>
</tr>
<tr class="even">
<td><code>impute_missing_abundance</code></td>
<td>Impute abundance for missing data points using sample groupings</td>
</tr>
<tr class="odd">
<td><code>fill_missing_abundance</code></td>
<td>Fill abundance for missing data points using an arbitrary value</td>
</tr>
</tbody>
</table>
</div>
<div id="minimal-input-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#minimal-input-data-frame" class="anchor"></a>Minimal input data frame</h2>
<table class="table">
<thead><tr class="header">
<th>sample</th>
<th>transcript</th>
<th>abundance</th>
<th>annotation</th>
</tr></thead>
<tbody><tr class="odd">
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td><code>integer</code></td>
<td>…</td>
</tr></tbody>
</table>
</div>
<div id="output-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#output-data-frame" class="anchor"></a>Output data frame</h2>
<table class="table">
<thead><tr class="header">
<th>sample</th>
<th>transcript</th>
<th>abundance</th>
<th>annotation</th>
<th>new information</th>
</tr></thead>
<tbody><tr class="odd">
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td><code>integer</code></td>
<td>…</td>
<td>…</td>
</tr></tbody>
</table>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code> object.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>From Bioconductor</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"tidybulk"</span>)
</pre></div>
<p>From Github</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">devtools</span>::<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"stemangiola/tidybulk"</span>)
</pre></div>
</div>
<div id="create-tidybulk-tibble" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidybulk-tibble" class="anchor"></a>Create <code>tidybulk</code> tibble.</h2>
<p>It memorises key column names</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">tt</span> <span class="op">=</span> <span class="kw">counts</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/tidybulk.html">tidybulk</a></span>(<span class="kw">sample</span>, <span class="kw">transcript</span>, <span class="kw">count</span>)
</pre></div>
</div>
<div id="get-the-bibliography-of-your-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#get-the-bibliography-of-your-workflow" class="anchor"></a>Get the bibliography of your workflow</h2>
<p>First of all, you can cite all articles utilised within your workflow automatically from any tidybulk tibble</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="co"># call analysis functions</span>
    <span class="fu"><a href="reference/get_bibliography.html">get_bibliography</a></span>()
</pre></div>
</div>
<div id="aggregate-duplicated-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#aggregate-duplicated-transcripts" class="anchor"></a>Aggregate duplicated <code>transcripts</code>
</h2>
<p>tidybulk provide the <code>aggregate_duplicates</code> function to aggregate duplicated transcripts (e.g., isoforms, ensembl). For example, we often have to convert ensembl symbols to gene/transcript symbol, but in doing so we have to deal with duplicates. <code>aggregate_duplicates</code> takes a tibble and column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) as arguments and returns a tibble with transcripts with the same name aggregated. All the rest of the columns are appended, and factors and boolean are appended as characters.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<p><code>r yellow tt.aggr = tt %&gt;% aggregate_duplicates()</code></p>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw">temp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(
    symbol = <span class="kw">dge_list</span><span class="op">$</span><span class="kw">genes</span><span class="op">$</span><span class="kw">symbol</span>,
    <span class="kw">dge_list</span><span class="op">$</span><span class="kw">counts</span>
)
<span class="kw">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/by.html">by</a></span>(<span class="kw">temp</span>, <span class="kw">temp</span><span class="op">$</span><span class="kw">symbol</span>,
    <span class="fu">function</span>(<span class="kw">df</span>)
        <span class="co">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw">df</span>[<span class="fl">1</span>,<span class="fl">1</span>])<span class="op">&gt;</span><span class="fl">0</span>)
            <span class="kw">matrixStats</span>:::<span class="fu">colSums</span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw">df</span>[,<span class="op">-</span><span class="fl">1</span>]))
)
<span class="kw">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">"rbind"</span>, <span class="kw">dge_list.nr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">dge_list.nr</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">dge_list</span>)
</pre></div>
</div>
<div style="clear:both;">
</div>
</div>
<div id="scale-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#scale-counts" class="anchor"></a>Scale <code>counts</code>
</h2>
<p>We may want to compensate for sequencing depth, scaling the transcript abundance (e.g., with TMM algorithm, Robinson and Oshlack doi.org/10.1186/gb-2010-11-3-r25). <code>scale_abundance</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a method as arguments and returns a tibble with additional columns with scaled data as <code>&lt;NAME OF COUNT COLUMN&gt;_scaled</code>.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">tt.norm</span> <span class="op">=</span> <span class="kw">tt.aggr</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/scale_abundance.html">scale_abundance</a></span>()
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">library</span>(edgeR)</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a>dgList &lt;-<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">count_m=</span>x,<span class="dt">group=</span>group)</span>
<span id="cb7-4"><a href="#cb7-4"></a>keep &lt;-<span class="st"> </span><span class="kw">filterByExpr</span>(dgList)</span>
<span id="cb7-5"><a href="#cb7-5"></a>dgList &lt;-<span class="st"> </span>dgList[keep,,keep.lib.sizes=<span class="ot">FALSE</span>]</span>
<span id="cb7-6"><a href="#cb7-6"></a>[...]</span>
<span id="cb7-7"><a href="#cb7-7"></a>dgList &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(dgList, <span class="dt">method=</span><span class="st">"TMM"</span>)</span>
<span id="cb7-8"><a href="#cb7-8"></a>norm_counts.table &lt;-<span class="st"> </span><span class="kw">cpm</span>(dgList)</span></code></pre></div>
</div>
<div style="clear:both;">
</div>
<p>We can easily plot the scaled density to check the scaling outcome. On the x axis we have the log scaled counts, on the y axes we have the density, data is grouped by sample and coloured by cell type.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="kw">count_scaled</span> <span class="op">+</span> <span class="fl">1</span>, group=<span class="kw">sample</span>, color=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
    <span class="fu">geom_density</span>() <span class="op">+</span>
    <span class="fu">scale_x_log10</span>() <span class="op">+</span>
    <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot_normalise-1.png"><!-- --></p>
</div>
<div id="filter-variable-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-variable-transcripts" class="anchor"></a>Filter <code>variable transcripts</code>
</h2>
<p>We may want to identify and filter variable transcripts.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">tt.norm.variable</span> <span class="op">=</span> <span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/keep_variable.html">keep_variable</a></span>()
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span>)

<span class="kw">x</span> <span class="op">=</span> <span class="kw">norm_counts.table</span>

<span class="kw">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>((<span class="kw">x</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="kw">x</span>))<span class="op">^</span><span class="fl">2</span>)
<span class="kw">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="kw">s</span>,decreasing=<span class="fl">TRUE</span>)
<span class="kw">x</span> <span class="op">&lt;-</span> <span class="kw">x</span>[<span class="kw">o</span>[<span class="fl">1L</span><span class="op">:</span><span class="kw">top</span>],,drop=<span class="fl">FALSE</span>]

<span class="kw">norm_counts.table</span> <span class="op">=</span> <span class="kw">norm_counts.table</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">x</span>)]

<span class="kw">norm_counts.table</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(
        <span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>,
        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">norm_counts.table</span>)
    ),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">
</div>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce <code>dimensions</code>
</h2>
<p>We may want to reduce the dimensions of our data, for example using PCA or MDS algorithms. <code>reduce_dimensions</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a method (e.g., MDS or PCA) as arguments and returns a tibble with additional columns for the reduced dimensions.</p>
<p><strong>MDS</strong> (Robinson et al., 10.1093/bioinformatics/btp616)</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">tt.norm.MDS</span> <span class="op">=</span>
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/reduce_dimensions.html">reduce_dimensions</a></span>(method=<span class="st">"MDS"</span>, .dims = <span class="fl">6</span>)
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioinf.wehi.edu.au/limma">limma</a></span>)

<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)
<span class="kw">cmds</span> <span class="op">=</span> <span class="kw">limma</span>::<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html">plotMDS</a></span>(ndim = <span class="kw">.dims</span>, plot = <span class="fl">FALSE</span>)

<span class="kw">cmds</span> <span class="op">=</span> <span class="kw">cmds</span> <span class="op">%$%</span> 
    <span class="kw">cmdscale.out</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">"Dim%s"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>))

<span class="kw">cmds</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">cmds</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">
</div>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data is coloured by cell type.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>()  <span class="op">%&gt;%</span> <span class="fu">select</span>(<span class="fu">contains</span>(<span class="st">"Dim"</span>), <span class="fu">everything</span>())
</pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co">## # A tibble: 48 x 15</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="co">##       Dim1   Dim2    Dim3     Dim4    Dim5   Dim6 sample `Cell type` time </span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="co">##      &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt;</span></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="co">##  1 -1.52    0.559 -2.06    0.111    0.125  -0.239 SRR17… b_cell      0 d  </span></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">##  2 -1.51    0.548 -2.08    0.0950   0.0942 -0.230 SRR17… b_cell      1 d  </span></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="co">##  3 -1.49    0.519 -2.00    0.136    0.102  -0.229 SRR17… b_cell      3 d  </span></span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="co">##  4 -1.50    0.527 -2.04    0.157    0.0953 -0.196 SRR17… b_cell      7 d  </span></span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co">##  5  0.150  -1.84   0.0708 -0.0425  -0.781  -0.207 SRR17… dendritic_… 0 d  </span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co">##  6  0.0907 -1.82   0.0922 -0.00628 -0.704  -0.205 SRR17… dendritic_… 1 d  </span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="co">##  7  0.161  -1.82   0.0301 -0.0586  -0.940  -0.179 SRR17… dendritic_… 3 d  </span></span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="co">##  8  0.0450 -1.84   0.0818 -0.00920 -0.714  -0.184 SRR17… dendritic_… 7 d  </span></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="co">##  9  0.407  -1.95   0.200  -0.315    0.742  -0.209 SRR17… monocyte    0 d  </span></span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="co">## 10  0.412  -1.90   0.198  -0.300    0.728  -0.186 SRR17… monocyte    1 d  </span></span>
<span id="cb14-14"><a href="#cb14-14"></a><span class="co">## # … with 38 more rows, and 6 more variables: condition &lt;chr&gt;, batch &lt;dbl&gt;,</span></span>
<span id="cb14-15"><a href="#cb14-15"></a><span class="co">## #   factor_of_interest &lt;chr&gt;, `merged transcripts` &lt;dbl&gt;, TMM &lt;dbl&gt;,</span></span>
<span id="cb14-16"><a href="#cb14-16"></a><span class="co">## #   multiplier &lt;dbl&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
  <span class="kw">GGally</span>::<span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(columns = <span class="fl">10</span><span class="op">:</span><span class="fl">15</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour=<span class="kw">`Cell type`</span>))
</pre></div>
<p><img src="reference/figures/plot_mds-1.png"><!-- --></p>
<p><strong>PCA</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">tt.norm.PCA</span> <span class="op">=</span>
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/reduce_dimensions.html">reduce_dimensions</a></span>(method=<span class="st">"PCA"</span>, .dims = <span class="fl">6</span>)
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)
<span class="kw">pc</span> <span class="op">=</span> <span class="kw">count_m_log</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(scale = <span class="fl">TRUE</span>)
<span class="kw">variance</span> <span class="op">=</span> <span class="kw">pc</span><span class="op">$</span><span class="kw">sdev</span><span class="op">^</span><span class="fl">2</span>
<span class="kw">variance</span> <span class="op">=</span> (<span class="kw">variance</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw">variance</span>))[<span class="fl">1</span><span class="op">:</span><span class="fl">6</span>]
<span class="kw">pc</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">counts</span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">counts</span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">pc</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">
</div>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data is coloured by cell type.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">tt.norm.PCA</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span> <span class="fu">select</span>(<span class="fu">contains</span>(<span class="st">"PC"</span>), <span class="fu">everything</span>())
</pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="co">## # A tibble: 48 x 15</span></span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="co">##        PC1   PC2    PC3     PC4    PC5   PC6 sample `Cell type` time  condition</span></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="co">##      &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;    </span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="co">##  1 -12.8   -1.94 -15.6  -0.578  -0.600 -1.32 SRR17… b_cell      0 d   TRUE     </span></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="co">##  2 -12.8   -1.94 -15.8  -0.279  -0.396 -1.36 SRR17… b_cell      1 d   TRUE     </span></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co">##  3 -12.8   -1.86 -15.2  -0.803  -0.388 -1.20 SRR17… b_cell      3 d   TRUE     </span></span>
<span id="cb19-7"><a href="#cb19-7"></a><span class="co">##  4 -12.7   -1.73 -15.5  -0.850  -0.489 -1.05 SRR17… b_cell      7 d   TRUE     </span></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co">##  5   1.29  13.4    1.79  0.0409  4.82  -1.46 SRR17… dendritic_… 0 d   FALSE    </span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co">##  6   0.794 13.3    1.90 -0.0114  4.32  -1.39 SRR17… dendritic_… 1 d   FALSE    </span></span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="co">##  7   1.43  13.4    1.55 -0.0722  5.59  -1.19 SRR17… dendritic_… 3 d   FALSE    </span></span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="co">##  8   0.412 13.5    1.90 -0.188   4.28  -1.19 SRR17… dendritic_… 7 d   FALSE    </span></span>
<span id="cb19-12"><a href="#cb19-12"></a><span class="co">##  9   3.84  13.7    2.64  2.16   -4.48  -1.36 SRR17… monocyte    0 d   FALSE    </span></span>
<span id="cb19-13"><a href="#cb19-13"></a><span class="co">## 10   3.86  13.3    2.51  2.00   -4.48  -1.49 SRR17… monocyte    1 d   FALSE    </span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="co">## # … with 38 more rows, and 5 more variables: batch &lt;dbl&gt;,</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="co">## #   factor_of_interest &lt;chr&gt;, `merged transcripts` &lt;dbl&gt;, TMM &lt;dbl&gt;,</span></span>
<span id="cb19-16"><a href="#cb19-16"></a><span class="co">## #   multiplier &lt;dbl&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">tt.norm.PCA</span> <span class="op">%&gt;%</span>
     <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
  <span class="kw">GGally</span>::<span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(columns = <span class="fl">10</span><span class="op">:</span><span class="fl">12</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour=<span class="kw">`Cell type`</span>))
</pre></div>
<p><img src="reference/figures/plot_pca-1.png"><!-- --></p>
<p><strong>tSNE</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="kw">tt.norm.tSNE</span> <span class="op">=</span>
    <span class="kw">breast_tcga_mini</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/tidybulk.html">tidybulk</a></span>(       <span class="kw">sample</span>, <span class="kw">ens</span>, <span class="kw">count_scaled</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        method = <span class="st">"tSNE"</span>,
        perplexity=<span class="fl">10</span>,
        pca_scale =<span class="fl">TRUE</span>
    )
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)

<span class="kw">tsne</span> <span class="op">=</span> <span class="kw">Rtsne</span>::<span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">count_m_log</span>),
    perplexity=<span class="fl">10</span>,
        pca_scale =<span class="fl">TRUE</span>
)<span class="op">$</span><span class="kw">Y</span>
<span class="kw">tsne</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">tsne</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">
</div>
<p>Plot</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="kw">tt.norm.tSNE</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">select</span>(<span class="fu">contains</span>(<span class="st">"tSNE"</span>), <span class="fu">everything</span>()) 
</pre></div>
<div class="sourceCode" id="cb24"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co">## # A tibble: 251 x 4</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="co">##     tSNE1  tSNE2 sample                       Call </span></span>
<span id="cb24-3"><a href="#cb24-3"></a><span class="co">##     &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;                        &lt;fct&gt;</span></span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="co">##  1  -4.95   5.57 TCGA-A1-A0SD-01A-11R-A115-07 LumA </span></span>
<span id="cb24-5"><a href="#cb24-5"></a><span class="co">##  2   6.24 -11.3  TCGA-A1-A0SF-01A-11R-A144-07 LumA </span></span>
<span id="cb24-6"><a href="#cb24-6"></a><span class="co">##  3 -11.4    2.97 TCGA-A1-A0SG-01A-11R-A144-07 LumA </span></span>
<span id="cb24-7"><a href="#cb24-7"></a><span class="co">##  4  -6.53  -8.35 TCGA-A1-A0SH-01A-11R-A084-07 LumA </span></span>
<span id="cb24-8"><a href="#cb24-8"></a><span class="co">##  5  -7.29 -12.6  TCGA-A1-A0SI-01A-11R-A144-07 LumB </span></span>
<span id="cb24-9"><a href="#cb24-9"></a><span class="co">##  6   1.78   5.05 TCGA-A1-A0SJ-01A-11R-A084-07 LumA </span></span>
<span id="cb24-10"><a href="#cb24-10"></a><span class="co">##  7  28.4  -13.7  TCGA-A1-A0SK-01A-12R-A084-07 Basal</span></span>
<span id="cb24-11"><a href="#cb24-11"></a><span class="co">##  8  10.1    2.78 TCGA-A1-A0SM-01A-11R-A084-07 LumA </span></span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="co">##  9   9.19   3.94 TCGA-A1-A0SN-01A-11R-A144-07 LumB </span></span>
<span id="cb24-13"><a href="#cb24-13"></a><span class="co">## 10 -17.8    9.12 TCGA-A1-A0SQ-01A-21R-A144-07 LumA </span></span>
<span id="cb24-14"><a href="#cb24-14"></a><span class="co">## # … with 241 more rows</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="kw">tt.norm.tSNE</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(x = <span class="kw">`tSNE1`</span>, y = <span class="kw">`tSNE2`</span>, color=<span class="kw">Call</span>)) <span class="op">+</span> <span class="fu">geom_point</span>() <span class="op">+</span> <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/unnamed-chunk-11-1.png"><!-- --></p>
</div>
<div id="rotate-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#rotate-dimensions" class="anchor"></a>Rotate <code>dimensions</code>
</h2>
<p>We may want to rotate the reduced dimensions (or any two numeric columns really) of our data, of a set angle. <code>rotate_dimensions</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and an angle as arguments and returns a tibble with additional columns for the rotated dimensions. The rotated dimensions will be added to the original data set as <code>&lt;NAME OF DIMENSION&gt; rotated &lt;ANGLE&gt;</code> by default, or as specified in the input arguments.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="kw">tt.norm.MDS.rotated</span> <span class="op">=</span>
  <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/rotate_dimensions.html">rotate_dimensions</a></span>(<span class="kw">`Dim1`</span>, <span class="kw">`Dim2`</span>, rotation_degrees = <span class="fl">45</span>, action=<span class="st">"get"</span>)
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="kw">rotation</span> <span class="op">=</span> <span class="fu">function</span>(<span class="kw">m</span>, <span class="kw">d</span>) {
    <span class="kw">r</span> <span class="op">=</span> <span class="kw">d</span> <span class="op">*</span> <span class="kw">pi</span> <span class="op">/</span> <span class="fl">180</span>
    ((<span class="fu"><a href="reference/arrange.html">bind_rows</a></span>(
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(`1` = <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">cos</a></span>(<span class="kw">r</span>), `2` = <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html">sin</a></span>(<span class="kw">r</span>)),
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(`1` = <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">sin</a></span>(<span class="kw">r</span>), `2` = <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">cos</a></span>(<span class="kw">r</span>))
    ) <span class="op">%&gt;%</span> <span class="kw">as_matrix</span>) <span class="op">%*%</span> <span class="kw">m</span>)
}
<span class="kw">mds_r</span> <span class="op">=</span> <span class="kw">pca</span> <span class="op">%&gt;%</span> <span class="fu">rotation</span>(<span class="kw">rotation_degrees</span>)
<span class="kw">mds_r</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">counts</span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">counts</span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">mds_r</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">
</div>
<p><strong>Original</strong> On the x and y axes axis we have the first two reduced dimensions, data is coloured by cell type.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="kw">tt.norm.MDS.rotated</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`Cell type`</span> )) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot_rotate_1-1.png"><!-- --></p>
<p><strong>Rotated</strong> On the x and y axes axis we have the first two reduced dimensions rotated of 45 degrees, data is coloured by cell type.</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="kw">tt.norm.MDS.rotated</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(x=<span class="kw">`Dim1 rotated 45`</span>, y=<span class="kw">`Dim2 rotated 45`</span>, color=<span class="kw">`Cell type`</span> )) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot_rotate_2-1.png"><!-- --></p>
</div>
<div id="test-differential-abundance" class="section level2">
<h2 class="hasAnchor">
<a href="#test-differential-abundance" class="anchor"></a>Test <code>differential abundance</code>
</h2>
<p>We may want to test for differential transcription between sample-wise factors of interest (e.g., with edgeR). <code>test_differential_abundance</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a formula representing the desired linear model as arguments and returns a tibble with additional columns for the statistics from the hypothesis test (e.g., log fold change, p-value and false discovery rate).</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="kw">tt.de</span> <span class="op">=</span>
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/test_differential_abundance.html">test_differential_abundance</a></span>( <span class="op">~</span> <span class="kw">condition</span>, action=<span class="st">"get"</span>)
<span class="kw">tt.de</span>
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span>)

<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span>(counts=<span class="kw">counts_m</span>,group=<span class="kw">group</span>)
<span class="kw">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span>(<span class="kw">dgList</span>)
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="kw">dgList</span>[<span class="kw">keep</span>,,keep.lib.sizes=<span class="fl">FALSE</span>]
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span>(<span class="kw">dgList</span>)
<span class="kw">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(<span class="op">~</span><span class="kw">group</span>)
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span>(<span class="kw">dgList</span>,<span class="kw">design</span>)
<span class="kw">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span>(<span class="kw">dgList</span>,<span class="kw">design</span>)
<span class="kw">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFTest</a></span>(<span class="kw">fit</span>,coef=<span class="fl">2</span>)
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span>(<span class="kw">qlf</span>, n=<span class="fl">Inf</span>)
</pre></div>
</div>
<div style="clear:both;">
</div>
<p>The functon <code>test_differential_abundance</code> operated with contrasts too. The constrasts hve the name of the design matrix (generally <name_column_covariate><values_of_covariate>)</values_of_covariate></name_column_covariate></p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="kw">tt.de</span> <span class="op">=</span>
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/test_differential_abundance.html">test_differential_abundance</a></span>(
        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="kw">condition</span>,                  
        .contrasts = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>( <span class="st">"conditionTRUE - conditionFALSE"</span>),
        action=<span class="st">"get"</span>
    )
</pre></div>
</div>
<div id="adjust-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#adjust-counts" class="anchor"></a>Adjust <code>counts</code>
</h2>
<p>We may want to adjust <code>counts</code> for (known) unwanted variation. <code>adjust_abundance</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a formula representing the desired linear model where the first covariate is the factor of interest and the second covariate is the unwanted variation, and returns a tibble with additional columns for the adjusted counts as <code>&lt;COUNT COLUMN&gt;_adjusted</code>. At the moment just an unwanted covariated is allowed at a time.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="kw">tt.norm.adj</span> <span class="op">=</span>
    <span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/adjust_abundance.html">adjust_abundance</a></span>(   <span class="op">~</span> <span class="kw">factor_of_interest</span> <span class="op">+</span> <span class="kw">batch</span>)
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">sva</span>)

<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)

<span class="kw">design</span> <span class="op">=</span>
        <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(
            object = <span class="op">~</span> <span class="kw">factor_of_interest</span> <span class="op">+</span> <span class="kw">batch</span>,
            data = <span class="kw">annotation</span>
        )

<span class="kw">count_m_log.sva</span> <span class="op">=</span>
    <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat.html">ComBat</a></span>(
            batch = <span class="kw">design</span>[,<span class="fl">2</span>],
            mod = <span class="kw">design</span>,
            <span class="kw">...</span>
        )

<span class="kw">count_m_log.sva</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="kw">count_m_log.sva</span>) <span class="op">-</span><span class="fl">1</span>)
<span class="kw">count_m_log.sva</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">counts</span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">counts</span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">count_m_log.sva</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">
</div>
</div>
<div id="deconvolve-cell-type-composition" class="section level2">
<h2 class="hasAnchor">
<a href="#deconvolve-cell-type-composition" class="anchor"></a>Deconvolve <code>Cell type composition</code>
</h2>
<p>We may want to infer the cell type composition of our samples (with the algorithm Cibersort; Newman et al., 10.1038/nmeth.3337). <code>deconvolve_cellularity</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and returns a tibble with additional columns for the adjusted cell type proportions.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="kw">tt.cibersort</span> <span class="op">=</span>
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/deconvolve_cellularity.html">deconvolve_cellularity</a></span>(action=<span class="st">"get"</span>, cores=<span class="fl">1</span>)
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a><span class="kw">source</span>(‘CIBERSORT.R’)</span>
<span id="cb36-2"><a href="#cb36-2"></a>count_m <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">write.table</span>(<span class="st">"mixture_file.txt"</span>)</span>
<span id="cb36-3"><a href="#cb36-3"></a>results &lt;-<span class="st"> </span><span class="kw">CIBERSORT</span>(</span>
<span id="cb36-4"><a href="#cb36-4"></a>    <span class="st">"sig_matrix_file.txt"</span>,</span>
<span id="cb36-5"><a href="#cb36-5"></a>    <span class="st">"mixture_file.txt"</span>,</span>
<span id="cb36-6"><a href="#cb36-6"></a>    <span class="dt">perm=</span><span class="dv">100</span>, <span class="dt">QN=</span><span class="ot">TRUE</span></span>
<span id="cb36-7"><a href="#cb36-7"></a>)</span>
<span id="cb36-8"><a href="#cb36-8"></a>results<span class="op">$</span>cell_type =<span class="st"> </span>tidybulk<span class="op">::</span>counts[</span>
<span id="cb36-9"><a href="#cb36-9"></a>    <span class="kw">match</span>(tidybulk<span class="op">::</span>counts<span class="op">$</span>sample, <span class="kw">rownames</span>(results)),</span>
<span id="cb36-10"><a href="#cb36-10"></a>    <span class="st">"Cell type"</span></span>
<span id="cb36-11"><a href="#cb36-11"></a>]</span></code></pre></div>
</div>
<div style="clear:both;">
</div>
<p>With the new annotated data frame, we can plot the distributions of cell types across samples, and compare them with the nominal cell type labels to check for the purity of isolation. On the x axis we have the cell types inferred by Cibersort, on the y axis we have the inferred proportions. The data is facetted and coloured by nominal cell types (annotation given by the researcher after FACS sorting).</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="kw">tt.cibersort</span> <span class="op">%&gt;%</span>
    <span class="fu">select</span>(<span class="fu">contains</span>(<span class="st">"cibersort:"</span>), <span class="fu">everything</span>()) <span class="op">%&gt;%</span>
    <span class="fu">gather</span>(<span class="kw">`Cell type inferred`</span>, <span class="kw">`proportion`</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>) <span class="op">%&gt;%</span>
  <span class="fu">ggplot</span>(<span class="fu">aes</span>(x=<span class="kw">`Cell type inferred`</span>, y=<span class="kw">proportion</span>, fill=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
  <span class="fu">geom_boxplot</span>() <span class="op">+</span>
  <span class="fu">facet_wrap</span>(<span class="op">~</span><span class="kw">`Cell type`</span>) <span class="op">+</span>
  <span class="kw">my_theme</span> <span class="op">+</span>
  <span class="fu">theme</span>(axis.text.x = <span class="fu">element_text</span>(angle = <span class="fl">90</span>, hjust = <span class="fl">1</span>, vjust = <span class="fl">0.5</span>), aspect.ratio=<span class="fl">1</span><span class="op">/</span><span class="fl">5</span>)
</pre></div>
<p><img src="reference/figures/plot_cibersort-1.png"><!-- --></p>
</div>
<div id="test-differential-cell-type-abundance" class="section level2">
<h2 class="hasAnchor">
<a href="#test-differential-cell-type-abundance" class="anchor"></a>Test differential cell-type abundance</h2>
<p>We can also perform a statistical test on the differential cell-type abundance across conditions</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/test_differential_cellularity.html">test_differential_cellularity</a></span>( <span class="op">~</span> <span class="kw">condition</span> )
</pre></div>
<div class="sourceCode" id="cb39"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="co">## # A tibble: 22 x 7</span></span>
<span id="cb39-2"><a href="#cb39-2"></a><span class="co">##    .cell_type cell_type_propo… `estimate_(Inte… estimate_condit…</span></span>
<span id="cb39-3"><a href="#cb39-3"></a><span class="co">##    &lt;chr&gt;      &lt;list&gt;                      &lt;dbl&gt;            &lt;dbl&gt;</span></span>
<span id="cb39-4"><a href="#cb39-4"></a><span class="co">##  1 B cells n… &lt;tibble [48 × 9…            -3.38           3.17  </span></span>
<span id="cb39-5"><a href="#cb39-5"></a><span class="co">##  2 B cells m… &lt;tibble [48 × 9…            -2.98           2.83  </span></span>
<span id="cb39-6"><a href="#cb39-6"></a><span class="co">##  3 Plasma ce… &lt;tibble [48 × 9…            -7.09          -0.508 </span></span>
<span id="cb39-7"><a href="#cb39-7"></a><span class="co">##  4 T cells C… &lt;tibble [48 × 9…            -4.05          -0.445 </span></span>
<span id="cb39-8"><a href="#cb39-8"></a><span class="co">##  5 T cells C… &lt;tibble [48 × 9…            -1.93           0.0167</span></span>
<span id="cb39-9"><a href="#cb39-9"></a><span class="co">##  6 T cells C… &lt;tibble [48 × 9…            -3.07          -0.533 </span></span>
<span id="cb39-10"><a href="#cb39-10"></a><span class="co">##  7 T cells C… &lt;tibble [48 × 9…            -5.97           1.46  </span></span>
<span id="cb39-11"><a href="#cb39-11"></a><span class="co">##  8 T cells f… &lt;tibble [48 × 9…            -5.17          -0.469 </span></span>
<span id="cb39-12"><a href="#cb39-12"></a><span class="co">##  9 T cells r… &lt;tibble [48 × 9…            -5.62          -0.602 </span></span>
<span id="cb39-13"><a href="#cb39-13"></a><span class="co">## 10 T cells g… &lt;tibble [48 × 9…            -5.53           1.02  </span></span>
<span id="cb39-14"><a href="#cb39-14"></a><span class="co">## # … with 12 more rows, and 3 more variables: std.error_conditionTRUE &lt;dbl&gt;,</span></span>
<span id="cb39-15"><a href="#cb39-15"></a><span class="co">## #   statistic_conditionTRUE &lt;dbl&gt;, p.value_conditionTRUE &lt;dbl&gt;</span></span></code></pre></div>
<p>We can also perform regression analysis with censored data (coxph).</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/test_differential_cellularity.html">test_differential_cellularity</a></span>(<span class="kw">survival</span>::<span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span>(<span class="kw">time</span>, <span class="kw">dead</span>) <span class="op">~</span> <span class="kw">.</span>)
</pre></div>
</div>
<div id="cluster-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#cluster-samples" class="anchor"></a>Cluster <code>samples</code>
</h2>
<p>We may want to cluster our data (e.g., using k-means sample-wise). <code>cluster_elements</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and returns a tibble with additional columns for the cluster annotation. At the moment only k-means clustering is supported, the plan is to introduce more clustering methods.</p>
<p><strong>k-means</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb41"><pre class="downlit">
<span class="kw">tt.norm.cluster</span> <span class="op">=</span> <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/cluster_elements.html">cluster_elements</a></span>(method=<span class="st">"kmeans"</span>, centers = <span class="fl">2</span>, action=<span class="st">"get"</span> )
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)

<span class="kw">k</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(<span class="kw">count_m_log</span>, iter.max = <span class="fl">1000</span>, <span class="kw">...</span>)
<span class="kw">cluster</span> <span class="op">=</span> <span class="kw">k</span><span class="op">$</span><span class="kw">cluster</span>

<span class="kw">cluster</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">cluster</span>)),
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Cell type"</span>, <span class="st">"Dim1"</span>, <span class="st">"Dim2"</span>)
]
</pre></div>
</div>
<div style="clear:both;">
</div>
<p>We can add cluster annotation to the MDS dimesion reduced data set and plot.</p>
<div class="sourceCode" id="cb43"><pre class="downlit">
 <span class="kw">tt.norm.cluster</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`cluster kmeans`</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot_cluster-1.png"><!-- --></p>
<p><strong>SNN</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="kw">tt.norm.SNN</span> <span class="op">=</span>
    <span class="kw">tt.norm.tSNE</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/cluster_elements.html">cluster_elements</a></span>(method = <span class="st">"SNN"</span>)
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb45"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://www.satijalab.org/seurat">Seurat</a></span>)

<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="kw">count_m</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span>(
    <span class="kw">snn</span>, display.progress = <span class="fl">TRUE</span>,
    num.cores=<span class="fl">4</span>, do.par = <span class="fl">TRUE</span>
)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="kw">snn</span>, selection.method = <span class="st">"vst"</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="kw">snn</span>, selection.method = <span class="st">"vst"</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="kw">snn</span>, npcs = <span class="fl">30</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="kw">snn</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="kw">snn</span>, method = <span class="st">"igraph"</span>, <span class="kw">...</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="kw">snn</span>[[<span class="st">"seurat_clusters"</span>]]

<span class="kw">snn</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">snn</span>)),
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Cell type"</span>, <span class="st">"Dim1"</span>, <span class="st">"Dim2"</span>)
]
</pre></div>
</div>
<div style="clear:both;">
</div>
<div class="sourceCode" id="cb46"><pre class="downlit">
<span class="kw">tt.norm.SNN</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">select</span>(<span class="fu">contains</span>(<span class="st">"tSNE"</span>), <span class="fu">everything</span>()) 
</pre></div>
<div class="sourceCode" id="cb47"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="co">## # A tibble: 251 x 5</span></span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="co">##     tSNE1  tSNE2 sample                       Call  `cluster SNN`</span></span>
<span id="cb47-3"><a href="#cb47-3"></a><span class="co">##     &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;                        &lt;fct&gt; &lt;fct&gt;        </span></span>
<span id="cb47-4"><a href="#cb47-4"></a><span class="co">##  1  -4.95   5.57 TCGA-A1-A0SD-01A-11R-A115-07 LumA  0            </span></span>
<span id="cb47-5"><a href="#cb47-5"></a><span class="co">##  2   6.24 -11.3  TCGA-A1-A0SF-01A-11R-A144-07 LumA  2            </span></span>
<span id="cb47-6"><a href="#cb47-6"></a><span class="co">##  3 -11.4    2.97 TCGA-A1-A0SG-01A-11R-A144-07 LumA  1            </span></span>
<span id="cb47-7"><a href="#cb47-7"></a><span class="co">##  4  -6.53  -8.35 TCGA-A1-A0SH-01A-11R-A084-07 LumA  0            </span></span>
<span id="cb47-8"><a href="#cb47-8"></a><span class="co">##  5  -7.29 -12.6  TCGA-A1-A0SI-01A-11R-A144-07 LumB  0            </span></span>
<span id="cb47-9"><a href="#cb47-9"></a><span class="co">##  6   1.78   5.05 TCGA-A1-A0SJ-01A-11R-A084-07 LumA  1            </span></span>
<span id="cb47-10"><a href="#cb47-10"></a><span class="co">##  7  28.4  -13.7  TCGA-A1-A0SK-01A-12R-A084-07 Basal 3            </span></span>
<span id="cb47-11"><a href="#cb47-11"></a><span class="co">##  8  10.1    2.78 TCGA-A1-A0SM-01A-11R-A084-07 LumA  2            </span></span>
<span id="cb47-12"><a href="#cb47-12"></a><span class="co">##  9   9.19   3.94 TCGA-A1-A0SN-01A-11R-A144-07 LumB  2            </span></span>
<span id="cb47-13"><a href="#cb47-13"></a><span class="co">## 10 -17.8    9.12 TCGA-A1-A0SQ-01A-21R-A144-07 LumA  1            </span></span>
<span id="cb47-14"><a href="#cb47-14"></a><span class="co">## # … with 241 more rows</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit">
<span class="kw">tt.norm.SNN</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">gather</span>(<span class="kw">source</span>, <span class="kw">Call</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cluster SNN"</span>, <span class="st">"Call"</span>)) <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/distinct.html">distinct</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(x = <span class="kw">`tSNE1`</span>, y = <span class="kw">`tSNE2`</span>, color=<span class="kw">Call</span>)) <span class="op">+</span> <span class="fu">geom_point</span>() <span class="op">+</span> <span class="fu">facet_grid</span>(<span class="op">~</span><span class="kw">source</span>) <span class="op">+</span> <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/SNN_plot-1.png"><!-- --></p>
<div class="sourceCode" id="cb49"><pre class="downlit">
<span class="co"># Do differential transcription between clusters</span>
<span class="kw">tt.norm.SNN</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/mutate.html">mutate</a></span>(factor_of_interest = <span class="kw">`cluster SNN`</span> <span class="op">==</span> <span class="fl">3</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/test_differential_abundance.html">test_differential_abundance</a></span>(
    <span class="op">~</span> <span class="kw">factor_of_interest</span>,
    action=<span class="st">"get"</span>
   )
</pre></div>
<div class="sourceCode" id="cb50"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a><span class="co">## # A tibble: 500 x 8</span></span>
<span id="cb50-2"><a href="#cb50-2"></a><span class="co">##    ens          logFC logCPM      F   PValue      FDR significant lowly_abundant</span></span>
<span id="cb50-3"><a href="#cb50-3"></a><span class="co">##    &lt;chr&gt;        &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;lgl&gt;       &lt;lgl&gt;         </span></span>
<span id="cb50-4"><a href="#cb50-4"></a><span class="co">##  1 ENSG00000… -0.981   10.7  42.9   3.13e-10 8.86e-10 TRUE        FALSE         </span></span>
<span id="cb50-5"><a href="#cb50-5"></a><span class="co">##  2 ENSG00000… -3.54     9.92 81.9   3.99e-17 2.10e-16 TRUE        FALSE         </span></span>
<span id="cb50-6"><a href="#cb50-6"></a><span class="co">##  3 ENSG00000… -0.154   10.6   1.20  2.74e- 1 2.96e- 1 FALSE       FALSE         </span></span>
<span id="cb50-7"><a href="#cb50-7"></a><span class="co">##  4 ENSG00000… -0.669   10.7  30.5   8.14e- 8 1.75e- 7 TRUE        FALSE         </span></span>
<span id="cb50-8"><a href="#cb50-8"></a><span class="co">##  5 ENSG00000…  0.826   11.3  78.4   1.49e-16 7.45e-16 TRUE        FALSE         </span></span>
<span id="cb50-9"><a href="#cb50-9"></a><span class="co">##  6 ENSG00000…  0.412   11.3  32.1   3.92e- 8 8.72e- 8 TRUE        FALSE         </span></span>
<span id="cb50-10"><a href="#cb50-10"></a><span class="co">##  7 ENSG00000…  0.0782  10.5   0.569 4.51e- 1 4.73e- 1 FALSE       FALSE         </span></span>
<span id="cb50-11"><a href="#cb50-11"></a><span class="co">##  8 ENSG00000… -1.13    11.1  34.5   1.33e- 8 3.19e- 8 TRUE        FALSE         </span></span>
<span id="cb50-12"><a href="#cb50-12"></a><span class="co">##  9 ENSG00000…  0.238   11.8   0.256 6.13e- 1 6.31e- 1 FALSE       FALSE         </span></span>
<span id="cb50-13"><a href="#cb50-13"></a><span class="co">## 10 ENSG00000… -1.68    10.5  92.0   8.66e-19 5.15e-18 TRUE        FALSE         </span></span>
<span id="cb50-14"><a href="#cb50-14"></a><span class="co">## # … with 490 more rows</span></span></code></pre></div>
</div>
<div id="drop-redundant-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#drop-redundant-transcripts" class="anchor"></a>Drop <code>redundant</code> transcripts</h2>
<p>We may want to remove redundant elements from the original data set (e.g., samples or transcripts), for example if we want to define cell-type specific signatures with low sample redundancy. <code>remove_redundancy</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and returns a tibble with redundant elements removed (e.g., samples). Two redundancy estimation approaches are supported:</p>
<ul>
<li>removal of highly correlated clusters of elements (keeping a representative) with method=“correlation”</li>
<li>removal of most proximal element pairs in a reduced dimensional space.</li>
</ul>
<p><strong>Approach 1</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb51"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">=</span>
    <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/remove_redundancy.html">remove_redundancy</a></span>(    method = <span class="st">"correlation"</span> )
</pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb52"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://github.com/dgrtwo/widyr">widyr</a></span>)

<span class="kw">.data.correlated</span> <span class="op">=</span>
    <span class="fu"><a href="https://rdrr.io/pkg/widyr/man/pairwise_cor.html">pairwise_cor</a></span>(
        <span class="kw">counts</span>,
        <span class="kw">sample</span>,
        <span class="kw">transcript</span>,
        <span class="kw">rc</span>,
        sort = <span class="fl">TRUE</span>,
        diag = <span class="fl">FALSE</span>,
        upper = <span class="fl">FALSE</span>
    ) <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/filter.html">filter</a></span>(<span class="kw">correlation</span> <span class="op">&gt;</span> <span class="kw">correlation_threshold</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/distinct.html">distinct</a></span>(<span class="kw">item1</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/rename.html">rename</a></span>(<span class="op">!</span><span class="op">!</span><span class="kw">.element</span> <span class="op">:=</span> <span class="kw">item1</span>)

<span class="co"># Return non redudant data frame</span>
<span class="kw">counts</span> <span class="op">%&gt;%</span> <span class="fu">anti_join</span>(<span class="kw">.data.correlated</span>) <span class="op">%&gt;%</span>
    <span class="fu">spread</span>(<span class="kw">sample</span>, <span class="kw">rc</span>, <span class="op">-</span> <span class="kw">transcript</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/left_join.html">left_join</a></span>(<span class="kw">annotation</span>)
</pre></div>
</div>
<div style="clear:both;">
</div>
<p>We can visualise how the reduced redundancy with the reduced dimentions look like</p>
<div class="sourceCode" id="cb53"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot_drop-1.png"><!-- --></p>
<p><strong>Approach 2</strong></p>
<div class="sourceCode" id="cb54"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">=</span>
    <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/remove_redundancy.html">remove_redundancy</a></span>(
    method = <span class="st">"reduced_dimensions"</span>,
    Dim_a_column = <span class="kw">`Dim1`</span>,
    Dim_b_column = <span class="kw">`Dim2`</span>
  )
</pre></div>
<p>We can visualise MDS reduced dimensions of the samples with the closest pair removed.</p>
<div class="sourceCode" id="cb55"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span>(<span class="fu">aes</span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
  <span class="fu">geom_point</span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="reference/figures/plot_drop2-1.png"><!-- --></p>
</div>
<div id="other-useful-wrappers" class="section level2">
<h2 class="hasAnchor">
<a href="#other-useful-wrappers" class="anchor"></a>Other useful wrappers</h2>
<p>The above wrapper streamline the most common processing of bulk RNA sequencing data. Other useful wrappers are listed above.</p>
</div>
<div id="from-bamsam-to-tibble-of-gene-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#from-bamsam-to-tibble-of-gene-counts" class="anchor"></a>From BAM/SAM to tibble of gene counts</h2>
<p>We can calculate gene counts (using FeatureCounts; Liao Y et al., 10.1093/nar/gkz114) from a list of BAM/SAM files and format them into a tidy structure (similar to counts).</p>
<div class="sourceCode" id="cb56"><pre class="downlit">
<span class="kw">counts</span> <span class="op">=</span> <span class="fu"><a href="reference/tidybulk_SAM_BAM.html">tidybulk_SAM_BAM</a></span>(
    <span class="kw">file_names</span>,
    genome = <span class="st">"hg38"</span>,
    isPairedEnd = <span class="fl">TRUE</span>,
    requireBothEndsMapped = <span class="fl">TRUE</span>,
    checkFragLength = <span class="fl">FALSE</span>,
    useMetaFeatures = <span class="fl">TRUE</span>
)
</pre></div>
</div>
<div id="from-ensembl-ids-to-gene-symbol-ids" class="section level2">
<h2 class="hasAnchor">
<a href="#from-ensembl-ids-to-gene-symbol-ids" class="anchor"></a>From ensembl IDs to gene symbol IDs</h2>
<p>We can add gene symbols from ensembl identifiers. This is useful since different resources use ensembl IDs while others use gene symbol IDs. This currently works for human and mouse.</p>
<div class="sourceCode" id="cb57"><pre class="downlit">
<span class="kw">counts_ensembl</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/ensembl_to_symbol.html">ensembl_to_symbol</a></span>(<span class="kw">ens</span>)
</pre></div>
<div class="sourceCode" id="cb58"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1"></a><span class="co">## # A tibble: 119 x 8</span></span>
<span id="cb58-2"><a href="#cb58-2"></a><span class="co">##    ens   iso   `read count` sample cases_0_project… cases_0_samples… transcript</span></span>
<span id="cb58-3"><a href="#cb58-3"></a><span class="co">##    &lt;chr&gt; &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;     </span></span>
<span id="cb58-4"><a href="#cb58-4"></a><span class="co">##  1 ENSG… 13             144 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-5"><a href="#cb58-5"></a><span class="co">##  2 ENSG… 13              72 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-6"><a href="#cb58-6"></a><span class="co">##  3 ENSG… 13               0 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-7"><a href="#cb58-7"></a><span class="co">##  4 ENSG… 13            1099 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-8"><a href="#cb58-8"></a><span class="co">##  5 ENSG… 13              11 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-9"><a href="#cb58-9"></a><span class="co">##  6 ENSG… 13               2 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-10"><a href="#cb58-10"></a><span class="co">##  7 ENSG… 13               3 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-11"><a href="#cb58-11"></a><span class="co">##  8 ENSG… 13            2678 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-12"><a href="#cb58-12"></a><span class="co">##  9 ENSG… 13             751 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-13"><a href="#cb58-13"></a><span class="co">## 10 ENSG… 13               1 TARGE… Acute Myeloid L… Primary Blood D… TSPAN6    </span></span>
<span id="cb58-14"><a href="#cb58-14"></a><span class="co">## # … with 109 more rows, and 1 more variable: ref_genome &lt;chr&gt;</span></span></code></pre></div>
</div>
<div id="from-gene-symbol-to-gene-description-gene-name-in-full" class="section level2">
<h2 class="hasAnchor">
<a href="#from-gene-symbol-to-gene-description-gene-name-in-full" class="anchor"></a>From gene symbol to gene description (gene name in full)</h2>
<p>We can add gene full name (and in future description) from symbol identifiers. This currently works for human and mouse.</p>
<div class="sourceCode" id="cb59"><pre class="downlit">
<span class="kw">tt</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/describe_transcript.html">describe_transcript</a></span>() <span class="op">%&gt;%</span> <span class="fu">select</span>(<span class="kw">transcript</span>, <span class="kw">description</span>, <span class="fu">everything</span>())
</pre></div>
<div class="sourceCode" id="cb60"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1"></a><span class="co">## # A tibble: 938,112 x 9</span></span>
<span id="cb60-2"><a href="#cb60-2"></a><span class="co">##    transcript description sample `Cell type` count time  condition batch</span></span>
<span id="cb60-3"><a href="#cb60-3"></a><span class="co">##    &lt;chr&gt;      &lt;chr&gt;       &lt;fct&gt;  &lt;fct&gt;       &lt;dbl&gt; &lt;fct&gt; &lt;lgl&gt;     &lt;int&gt;</span></span>
<span id="cb60-4"><a href="#cb60-4"></a><span class="co">##  1 DDX11L1    DEAD/H-box… SRR17… b_cell         17 0 d   TRUE          0</span></span>
<span id="cb60-5"><a href="#cb60-5"></a><span class="co">##  2 WASH7P     WASP famil… SRR17… b_cell       3568 0 d   TRUE          0</span></span>
<span id="cb60-6"><a href="#cb60-6"></a><span class="co">##  3 MIR6859-1  microRNA 6… SRR17… b_cell         57 0 d   TRUE          0</span></span>
<span id="cb60-7"><a href="#cb60-7"></a><span class="co">##  4 MIR1302-2  microRNA 1… SRR17… b_cell          1 0 d   TRUE          0</span></span>
<span id="cb60-8"><a href="#cb60-8"></a><span class="co">##  5 FAM138A    family wit… SRR17… b_cell          0 0 d   TRUE          0</span></span>
<span id="cb60-9"><a href="#cb60-9"></a><span class="co">##  6 OR4F5      olfactory … SRR17… b_cell          0 0 d   TRUE          0</span></span>
<span id="cb60-10"><a href="#cb60-10"></a><span class="co">##  7 LOC729737  uncharacte… SRR17… b_cell       1764 0 d   TRUE          0</span></span>
<span id="cb60-11"><a href="#cb60-11"></a><span class="co">##  8 LOC102725… DEAD/H-box… SRR17… b_cell         11 0 d   TRUE          0</span></span>
<span id="cb60-12"><a href="#cb60-12"></a><span class="co">##  9 MIR6859-2  microRNA 6… SRR17… b_cell         40 0 d   TRUE          0</span></span>
<span id="cb60-13"><a href="#cb60-13"></a><span class="co">## 10 OR4F29     olfactory … SRR17… b_cell          0 0 d   TRUE          0</span></span>
<span id="cb60-14"><a href="#cb60-14"></a><span class="co">## # … with 938,102 more rows, and 1 more variable: factor_of_interest &lt;lgl&gt;</span></span></code></pre></div>
</div>
<div id="add-versus-get-versus-only-modes" class="section level2">
<h2 class="hasAnchor">
<a href="#add-versus-get-versus-only-modes" class="anchor"></a>ADD versus GET versus ONLY modes</h2>
<p>Every function takes a tidytranscriptomics structured data as input, and</p>
<p>(i) with action=“add” outputs the new information joint to the original input data frame (default), (ii) with action=“get” the new information with the sample or transcript relative informatin depending on what the analysis is about, or (iii) with action=“only” just the new information. For example, from this data set</p>
<div class="sourceCode" id="cb61"><pre class="downlit">
  <span class="kw">tt.norm</span>
</pre></div>
<div class="sourceCode" id="cb62"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1"></a><span class="co">## # A tibble: 938,112 x 13</span></span>
<span id="cb62-2"><a href="#cb62-2"></a><span class="co">##    sample transcript `Cell type` count time  condition batch factor_of_inter…</span></span>
<span id="cb62-3"><a href="#cb62-3"></a><span class="co">##    &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;           </span></span>
<span id="cb62-4"><a href="#cb62-4"></a><span class="co">##  1 SRR17… A1BG       b_cell        153 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-5"><a href="#cb62-5"></a><span class="co">##  2 SRR17… A1BG-AS1   b_cell         83 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-6"><a href="#cb62-6"></a><span class="co">##  3 SRR17… A2M-AS1    b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-7"><a href="#cb62-7"></a><span class="co">##  4 SRR17… A2ML1      b_cell          3 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-8"><a href="#cb62-8"></a><span class="co">##  5 SRR17… A2MP1      b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-9"><a href="#cb62-9"></a><span class="co">##  6 SRR17… A3GALT2    b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-10"><a href="#cb62-10"></a><span class="co">##  7 SRR17… A4GALT     b_cell          4 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-11"><a href="#cb62-11"></a><span class="co">##  8 SRR17… A4GNT      b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-12"><a href="#cb62-12"></a><span class="co">##  9 SRR17… AA06       b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-13"><a href="#cb62-13"></a><span class="co">## 10 SRR17… AAAS       b_cell        868 0 d   TRUE          0 TRUE            </span></span>
<span id="cb62-14"><a href="#cb62-14"></a><span class="co">## # … with 938,102 more rows, and 5 more variables: `merged transcripts` &lt;dbl&gt;,</span></span>
<span id="cb62-15"><a href="#cb62-15"></a><span class="co">## #   count_scaled &lt;dbl&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;</span></span></code></pre></div>
<p><strong>action=“add”</strong> (Default) We can add the MDS dimensions to the original data set</p>
<div class="sourceCode" id="cb63"><pre class="downlit">
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method=<span class="st">"MDS"</span> ,
        .element = <span class="kw">sample</span>,
        .feature = <span class="kw">transcript</span>,
        .dims = <span class="fl">3</span>,
        action=<span class="st">"add"</span>
    )
</pre></div>
<div class="sourceCode" id="cb64"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1"></a><span class="co">## # A tibble: 938,112 x 16</span></span>
<span id="cb64-2"><a href="#cb64-2"></a><span class="co">##    sample transcript `Cell type` count time  condition batch factor_of_inter…</span></span>
<span id="cb64-3"><a href="#cb64-3"></a><span class="co">##    &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;           </span></span>
<span id="cb64-4"><a href="#cb64-4"></a><span class="co">##  1 SRR17… A1BG       b_cell        153 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-5"><a href="#cb64-5"></a><span class="co">##  2 SRR17… A1BG-AS1   b_cell         83 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-6"><a href="#cb64-6"></a><span class="co">##  3 SRR17… A2M-AS1    b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-7"><a href="#cb64-7"></a><span class="co">##  4 SRR17… A2ML1      b_cell          3 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-8"><a href="#cb64-8"></a><span class="co">##  5 SRR17… A2MP1      b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-9"><a href="#cb64-9"></a><span class="co">##  6 SRR17… A3GALT2    b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-10"><a href="#cb64-10"></a><span class="co">##  7 SRR17… A4GALT     b_cell          4 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-11"><a href="#cb64-11"></a><span class="co">##  8 SRR17… A4GNT      b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-12"><a href="#cb64-12"></a><span class="co">##  9 SRR17… AA06       b_cell          0 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-13"><a href="#cb64-13"></a><span class="co">## 10 SRR17… AAAS       b_cell        868 0 d   TRUE          0 TRUE            </span></span>
<span id="cb64-14"><a href="#cb64-14"></a><span class="co">## # … with 938,102 more rows, and 8 more variables: `merged transcripts` &lt;dbl&gt;,</span></span>
<span id="cb64-15"><a href="#cb64-15"></a><span class="co">## #   count_scaled &lt;dbl&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;,</span></span>
<span id="cb64-16"><a href="#cb64-16"></a><span class="co">## #   Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;, Dim3 &lt;dbl&gt;</span></span></code></pre></div>
<p><strong>action=“get”</strong> We can add the MDS dimensions to the original data set selecting just the sample-wise column</p>
<div class="sourceCode" id="cb65"><pre class="downlit">
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method=<span class="st">"MDS"</span> ,
        .element = <span class="kw">sample</span>,
        .feature = <span class="kw">transcript</span>,
        .dims = <span class="fl">3</span>,
        action=<span class="st">"get"</span>
    )
</pre></div>
<div class="sourceCode" id="cb66"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1"></a><span class="co">## # A tibble: 48 x 12</span></span>
<span id="cb66-2"><a href="#cb66-2"></a><span class="co">##    sample `Cell type` time  condition batch factor_of_inter… `merged transcr…</span></span>
<span id="cb66-3"><a href="#cb66-3"></a><span class="co">##    &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;                       &lt;dbl&gt;</span></span>
<span id="cb66-4"><a href="#cb66-4"></a><span class="co">##  1 SRR17… b_cell      0 d   TRUE          0 TRUE                            1</span></span>
<span id="cb66-5"><a href="#cb66-5"></a><span class="co">##  2 SRR17… b_cell      1 d   TRUE          1 TRUE                            1</span></span>
<span id="cb66-6"><a href="#cb66-6"></a><span class="co">##  3 SRR17… b_cell      3 d   TRUE          1 TRUE                            1</span></span>
<span id="cb66-7"><a href="#cb66-7"></a><span class="co">##  4 SRR17… b_cell      7 d   TRUE          1 TRUE                            1</span></span>
<span id="cb66-8"><a href="#cb66-8"></a><span class="co">##  5 SRR17… dendritic_… 0 d   FALSE         0 FALSE                           1</span></span>
<span id="cb66-9"><a href="#cb66-9"></a><span class="co">##  6 SRR17… dendritic_… 1 d   FALSE         0 FALSE                           1</span></span>
<span id="cb66-10"><a href="#cb66-10"></a><span class="co">##  7 SRR17… dendritic_… 3 d   FALSE         1 FALSE                           1</span></span>
<span id="cb66-11"><a href="#cb66-11"></a><span class="co">##  8 SRR17… dendritic_… 7 d   FALSE         0 FALSE                           1</span></span>
<span id="cb66-12"><a href="#cb66-12"></a><span class="co">##  9 SRR17… monocyte    0 d   FALSE         1 FALSE                           1</span></span>
<span id="cb66-13"><a href="#cb66-13"></a><span class="co">## 10 SRR17… monocyte    1 d   FALSE         1 FALSE                           1</span></span>
<span id="cb66-14"><a href="#cb66-14"></a><span class="co">## # … with 38 more rows, and 5 more variables: TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;,</span></span>
<span id="cb66-15"><a href="#cb66-15"></a><span class="co">## #   Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;, Dim3 &lt;dbl&gt;</span></span></code></pre></div>
<p><strong>action=“only”</strong> We can get just the MDS dimensions relative to each sample</p>
<div class="sourceCode" id="cb67"><pre class="downlit">
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method=<span class="st">"MDS"</span> ,
        .element = <span class="kw">sample</span>,
        .feature = <span class="kw">transcript</span>,
        .dims = <span class="fl">3</span>,
        action=<span class="st">"only"</span>
    )
</pre></div>
<div class="sourceCode" id="cb68"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a><span class="co">## # A tibble: 48 x 4</span></span>
<span id="cb68-2"><a href="#cb68-2"></a><span class="co">##    sample        Dim1   Dim2    Dim3</span></span>
<span id="cb68-3"><a href="#cb68-3"></a><span class="co">##    &lt;chr&gt;        &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span id="cb68-4"><a href="#cb68-4"></a><span class="co">##  1 SRR1740034 -1.52    0.559 -2.06  </span></span>
<span id="cb68-5"><a href="#cb68-5"></a><span class="co">##  2 SRR1740035 -1.51    0.548 -2.08  </span></span>
<span id="cb68-6"><a href="#cb68-6"></a><span class="co">##  3 SRR1740036 -1.49    0.519 -2.00  </span></span>
<span id="cb68-7"><a href="#cb68-7"></a><span class="co">##  4 SRR1740037 -1.50    0.527 -2.04  </span></span>
<span id="cb68-8"><a href="#cb68-8"></a><span class="co">##  5 SRR1740038  0.150  -1.84   0.0708</span></span>
<span id="cb68-9"><a href="#cb68-9"></a><span class="co">##  6 SRR1740039  0.0907 -1.82   0.0922</span></span>
<span id="cb68-10"><a href="#cb68-10"></a><span class="co">##  7 SRR1740040  0.161  -1.82   0.0301</span></span>
<span id="cb68-11"><a href="#cb68-11"></a><span class="co">##  8 SRR1740041  0.0450 -1.84   0.0818</span></span>
<span id="cb68-12"><a href="#cb68-12"></a><span class="co">##  9 SRR1740042  0.407  -1.95   0.200 </span></span>
<span id="cb68-13"><a href="#cb68-13"></a><span class="co">## 10 SRR1740043  0.412  -1.90   0.198 </span></span>
<span id="cb68-14"><a href="#cb68-14"></a><span class="co">## # … with 38 more rows</span></span></code></pre></div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/tidybulk">https://​www.bioconductor.org/​packages/​tidybulk</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidybulk/actions"><img src="https://github.com/stemangiola/tidybulk/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
