<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings transcriptomics to the tidyverse • tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings transcriptomics to the tidyverse">
<meta property="og:description" content="This is a collection of utility functions that allow to perform exploration of and calculations to RNA sequencing data, in a modular, pipe-friendly and tidy fashion.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/comparison_coding.html">Side-by-side comparison with standard interfaces</a>
    </li>
    <li>
      <a href="articles/introduction.html">tidybulk: An R tidy framework for modular transcriptomic data analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="tidybulk-an-r-tidy-framework-for-modular-transcriptomic-data-analysis">tidybulk: An R tidy framework for modular transcriptomic data analysis<a class="anchor" aria-label="anchor" href="#tidybulk-an-r-tidy-framework-for-modular-transcriptomic-data-analysis"></a>
</h1></div>
<p>Stefano Mangiola 2025-08-20</p>
<!-- badges: start -->

</div>
<div class="section level1">
<h1 id="section">
<img src="inst/logo.svg" height="139px" width="120px"><a class="anchor" aria-label="anchor" href="#section"></a>
</h1>
<p><strong>tidybulk</strong> is a powerful R package designed for modular transcriptomic data analysis that brings transcriptomics to the tidyverse.</p>
<div class="section level2">
<h2 id="why-tidybulk">Why tidybulk?<a class="anchor" aria-label="anchor" href="#why-tidybulk"></a>
</h2>
<p>Tidybulk provides a unified interface for comprehensive transcriptomic data analysis with seamless integration of SummarizedExperiment objects and tidyverse principles. It streamlines the entire workflow from raw data to biological insights.</p>
</div>
<div class="section level2">
<h2 id="functionsutilities-available">Functions/utilities available<a class="anchor" aria-label="anchor" href="#functionsutilities-available"></a>
</h2>
<div class="section level3">
<h3 id="abundance-normalization-functions">Abundance Normalization Functions<a class="anchor" aria-label="anchor" href="#abundance-normalization-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="45%">
<col width="54%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/scale_abundance-methods.html">scale_abundance()</a></code></td>
<td>Scale abundance data</td>
</tr>
<tr>
<td><code><a href="reference/quantile_normalise_abundance-methods.html">quantile_normalise_abundance()</a></code></td>
<td>Quantile normalization</td>
</tr>
<tr>
<td><code><a href="reference/adjust_abundance-methods.html">adjust_abundance()</a></code></td>
<td>Adjust abundance for unwanted variation</td>
</tr>
<tr>
<td><code><a href="reference/fill_missing_abundance-methods.html">fill_missing_abundance()</a></code></td>
<td>Fill missing abundance values</td>
</tr>
<tr>
<td><code><a href="reference/impute_missing_abundance-methods.html">impute_missing_abundance()</a></code></td>
<td>Impute missing abundance values</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="filtering-and-selection-functions">Filtering and Selection Functions<a class="anchor" aria-label="anchor" href="#filtering-and-selection-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="30%">
<col width="69%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/identify_abundant-methods.html">identify_abundant()</a></code></td>
<td>Identify abundant transcripts without removing them</td>
</tr>
<tr>
<td><code><a href="reference/keep_abundant.html">keep_abundant()</a></code></td>
<td>Keep abundant transcripts</td>
</tr>
<tr>
<td><code><a href="reference/keep_variable-methods.html">keep_variable()</a></code></td>
<td>Keep variable transcripts</td>
</tr>
<tr>
<td><code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr()</a></code></td>
<td>Filter by expression</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="dimensionality-reduction-functions">Dimensionality Reduction Functions<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/reduce_dimensions-methods.html">reduce_dimensions()</a></code></td>
<td>Reduce dimensions with PCA/MDS/tSNE/UMAP</td>
</tr>
<tr>
<td><code><a href="reference/rotate_dimensions-methods.html">rotate_dimensions()</a></code></td>
<td>Rotate dimensions</td>
</tr>
<tr>
<td><code><a href="reference/remove_redundancy-methods.html">remove_redundancy()</a></code></td>
<td>Remove redundant features</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="clustering-functions">Clustering Functions<a class="anchor" aria-label="anchor" href="#clustering-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/cluster_elements-methods.html">cluster_elements()</a></code></td>
<td>Cluster elements with various methods</td>
</tr>
<tr>
<td><code>kmeans clustering</code></td>
<td>K-means clustering</td>
</tr>
<tr>
<td><code>SNN clustering</code></td>
<td>Shared nearest neighbor clustering</td>
</tr>
<tr>
<td><code>hierarchical clustering</code></td>
<td>Hierarchical clustering</td>
</tr>
<tr>
<td><code>DBSCAN clustering</code></td>
<td>Density-based clustering</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="differential-analysis-functions">Differential Analysis Functions<a class="anchor" aria-label="anchor" href="#differential-analysis-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody><tr>
<td><code><a href="reference/test_differential_expression-methods.html">test_differential_expression()</a></code></td>
<td>Test differential expression with various methods</td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="cellularity-analysis-functions">Cellularity Analysis Functions<a class="anchor" aria-label="anchor" href="#cellularity-analysis-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="38%">
<col width="61%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity()</a></code></td>
<td>Deconvolve cellularity with various methods</td>
</tr>
<tr>
<td><code>cibersort()</code></td>
<td>CIBERSORT analysis</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="gene-enrichment-functions">Gene Enrichment Functions<a class="anchor" aria-label="anchor" href="#gene-enrichment-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/test_gene_enrichment-methods.html">test_gene_enrichment()</a></code></td>
<td>Test gene enrichment</td>
</tr>
<tr>
<td><code><a href="reference/test_gene_overrepresentation-methods.html">test_gene_overrepresentation()</a></code></td>
<td>Test gene overrepresentation</td>
</tr>
<tr>
<td><code><a href="reference/test_gene_rank-methods.html">test_gene_rank()</a></code></td>
<td>Test gene rank</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="utility-functions">Utility Functions<a class="anchor" aria-label="anchor" href="#utility-functions"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/describe_transcript-methods.html">describe_transcript()</a></code></td>
<td>Describe transcript characteristics</td>
</tr>
<tr>
<td><code><a href="reference/get_bibliography-methods.html">get_bibliography()</a></code></td>
<td>Get bibliography</td>
</tr>
<tr>
<td><code><a href="reference/resolve_complete_confounders_of_non_interest.html">resolve_complete_confounders_of_non_interest()</a></code></td>
<td>Resolve confounders</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="validation-and-utility-functions">Validation and Utility Functions<a class="anchor" aria-label="anchor" href="#validation-and-utility-functions"></a>
</h3>
<table class="table">
<thead><tr>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr>
<td><code><a href="reference/check_if_counts_is_na.html">check_if_counts_is_na()</a></code></td>
<td>Check if counts contain NA values</td>
</tr>
<tr>
<td><code><a href="reference/check_if_duplicated_genes.html">check_if_duplicated_genes()</a></code></td>
<td>Check for duplicated genes</td>
</tr>
<tr>
<td><code>check_if_wrong_input()</code></td>
<td>Validate input data</td>
</tr>
<tr>
<td><code><a href="reference/log10_reverse_trans.html">log10_reverse_trans()</a></code></td>
<td>Log10 reverse transformation</td>
</tr>
<tr>
<td><code><a href="reference/logit_trans.html">logit_trans()</a></code></td>
<td>Logit transformation</td>
</tr>
</tbody>
</table>
<p>All functions are directly compatible with <code>SummarizedExperiment</code> objects and follow tidyverse principles for seamless integration with the tidyverse ecosystem.</p>
</div>
<div class="section level3">
<h3 id="scientific-citation">Scientific Citation<a class="anchor" aria-label="anchor" href="#scientific-citation"></a>
</h3>
<p>Mangiola, Stefano, Ramyar Molania, Ruining Dong, Maria A. Doyle, and Anthony T. Papenfuss. 2021. “Tidybulk: An R tidy framework for modular transcriptomic data analysis.” Genome Biology 22 (42). <a href="https://doi.org/10.1186/s13059-020-02233-7" class="external-link uri">https://doi.org/10.1186/s13059-020-02233-7</a></p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7" class="external-link">Genome Biology - tidybulk: an R tidy framework for modular transcriptomic data analysis</a></p>
<p>In this vignette we will use the <code>airway</code> dataset, a <code>SummarizedExperiment</code> object containing RNA-seq data from an experiment studying the effect of dexamethasone treatment on airway smooth muscle cells. This dataset is available in the <a href="https://bioconductor.org/packages/airway/" class="external-link">airway</a> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
<p>This workflow, will use the <a href="https://bioconductor.org/packages/tidySummarizedExperiment/" class="external-link">tidySummarizedExperiment</a> package to manipulate the data in a <code>tidyverse</code> fashion. This approach streamlines the data manipulation and analysis process, making it more efficient and easier to understand.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySummarizedExperiment" class="external-link">tidySummarizedExperiment</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">## tidySummarizedExperiment says: Printing is now handled externally. If you want to visualize the data in a tidy way, do library(tidyprint). See https://github.com/tidyomics/tidyprint for more information.</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="do">## Attaching package: 'tidySummarizedExperiment'</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="do">## The following object is masked from 'package:generics':</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="do">##     tidy</span></span></code></pre></div>
<p>Here we will add a gene symbol column to the <code>airway</code> object. This will be used to interpret the differential expression analysis, and to deconvolve the cellularity.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="do">## Loading required package: AnnotationDbi</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="do">## Attaching package: 'AnnotationDbi'</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="do">## The following object is masked from 'package:dplyr':</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="do">##     select</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="do">## </span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/AnnotationDbi" class="external-link">AnnotationDbi</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add gene symbol and entrez</span></span>
<span><span class="va">airway</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>entrezid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>,</span>
<span>                                      keys <span class="op">=</span> <span class="va">.feature</span>,</span>
<span>                                      keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span>,</span>
<span>                                      column <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>                                      multiVals <span class="op">=</span> <span class="st">"first"</span></span>
<span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:org.Hs.eg.db"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:AnnotationDbi"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="installation-guide">Installation Guide<a class="anchor" aria-label="anchor" href="#installation-guide"></a>
</h1>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"tidybulk"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/tidybulk"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<h1 id="comprehensive-example-pipeline">Comprehensive Example Pipeline<a class="anchor" aria-label="anchor" href="#comprehensive-example-pipeline"></a>
</h1>
<p>This vignette demonstrates a complete transcriptomic analysis workflow using tidybulk, with special emphasis on differential expression analysis.</p>
<div class="section level2">
<h2 id="data-overview">Data Overview<a class="anchor" aria-label="anchor" href="#data-overview"></a>
</h2>
<p>We will use the <code>airway</code> dataset, a <code>SummarizedExperiment</code> object containing RNA-seq data from an experiment studying the effect of dexamethasone treatment on airway smooth muscle cells:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="do">## class: RangedSummarizedExperiment </span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="do">## dim: 63677 8 </span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="do">## metadata(1): ''</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="do">## assays(1): counts</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="do">## rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="do">##   ENSG00000273493</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="do">## rowData names(10): gene_id gene_name ... seq_coord_system symbol</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="do">## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="do">## colData names(9): SampleName cell ... Sample BioSample</span></span></code></pre></div>
<p>Loading <code>tidySummarizedExperiment</code> automatically abstracts this object as a <code>tibble</code>, making it compatible with tidyverse tools while maintaining its <code>SummarizedExperiment</code> nature:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="do">## [1] "RangedSummarizedExperiment"</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="do">## attr(,"package")</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="do">## [1] "SummarizedExperiment"</span></span></code></pre></div>
<div class="section level3">
<h3 id="prepare-data-for-analysis">Prepare Data for Analysis<a class="anchor" aria-label="anchor" href="#prepare-data-for-analysis"></a>
</h3>
<p>Before analysis, we need to ensure our variables are in the correct format:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert dex to factor for proper differential expression analysis</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">dex</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-raw-counts">Visualize Raw Counts<a class="anchor" aria-label="anchor" href="#visualize-raw-counts"></a>
</h3>
<p>Visualize the distribution of raw counts before any filtering:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">as_tibble</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">`dex`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Raw counts by treatment (before any filtering)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/plot-raw-counts-1.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="step-1-data-preprocessing">Step 1: Data Preprocessing<a class="anchor" aria-label="anchor" href="#step-1-data-preprocessing"></a>
</h2>
<div class="section level3">
<h3 id="aggregate-duplicated-transcripts-optional">Aggregate Duplicated Transcripts (optional)<a class="anchor" aria-label="anchor" href="#aggregate-duplicated-transcripts-optional"></a>
</h3>
<p>Aggregate duplicated transcripts (e.g., isoforms, ensembl IDs):</p>
<blockquote>
<p>Transcript aggregation is a standard bioinformatics approach for gene-level summarization.</p>
</blockquote>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add gene names to rowData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Aggregate duplicates</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>.transcript <span class="op">=</span> <span class="va">gene_name</span>, aggregation_function <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="do">## tidybulk says: your object does not have duplicates along the gene_name column. The input dataset is returned.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="abundance-filtering-tidybulk-approaches-only">Abundance Filtering: tidybulk approaches only<a class="anchor" aria-label="anchor" href="#abundance-filtering-tidybulk-approaches-only"></a>
</h3>
<p>Abundance filtering can be performed using tidybulk’s built-in methods (<a href="#ref-robinson2010edger">Robinson, McCarthy, and Smyth 2010</a>; <a href="#ref-chen2016edgeR">Chen, Lun, and Smyth 2016</a>).</p>
<div class="section level4">
<h4 id="id_1-tidybulk-default-formula_design-and-cpm-threshold">1. tidybulk: Default, formula_design, and CPM threshold<a class="anchor" aria-label="anchor" href="#id_1-tidybulk-default-formula_design-and-cpm-threshold"></a>
</h4>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Default (simple filtering)</span></span>
<span><span class="va">airway_abundant_default</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="do">## Warning in filterByExpr.DGEList(y, design = design, group = group, lib.size =</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="do">## lib.size, : All samples appear to belong to the same group.</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># With factor_of_interest (recommended for complex designs)</span></span>
<span><span class="va">airway_abundant_formula</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span>minimum_counts <span class="op">=</span> <span class="fl">10</span>, minimum_proportion <span class="op">=</span> <span class="fl">0.5</span>, factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="do">## Warning: The `factor_of_interest` argument of `keep_abundant()` is deprecated as of</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="do">## tidybulk 2.0.0.</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="do">## ℹ Please use the `formula_design` argument instead.</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="do">## ℹ The argument 'factor_of_interest' is deprecated and will be removed in a</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="do">##   future release. Please use the 'design' or 'formula_design' argument instead.</span></span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a><span class="do">## This warning is displayed once every 8 hours.</span></span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a><span class="do">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a><span class="do">## generated.</span></span>
<span id="cb22-9"><a href="#cb22-9" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" tabindex="-1"></a><span class="do">## Warning: The `factor_of_interest` argument of `identify_abundant()` is deprecated as of</span></span>
<span id="cb22-11"><a href="#cb22-11" tabindex="-1"></a><span class="do">## tidybulk 2.0.0.</span></span>
<span id="cb22-12"><a href="#cb22-12" tabindex="-1"></a><span class="do">## ℹ Please use the `formula_design` argument instead.</span></span>
<span id="cb22-13"><a href="#cb22-13" tabindex="-1"></a><span class="do">## ℹ The argument 'factor_of_interest' is deprecated and will be removed in a</span></span>
<span id="cb22-14"><a href="#cb22-14" tabindex="-1"></a><span class="do">##   future release. Please use the 'design' or 'formula_design' argument instead.</span></span>
<span id="cb22-15"><a href="#cb22-15" tabindex="-1"></a><span class="do">## ℹ The deprecated feature was likely used in the tidybulk package.</span></span>
<span id="cb22-16"><a href="#cb22-16" tabindex="-1"></a><span class="do">##   Please report the issue at &lt;https://github.com/stemangiola/tidybulk/issues&gt;.</span></span>
<span id="cb22-17"><a href="#cb22-17" tabindex="-1"></a><span class="do">## This warning is displayed once every 8 hours.</span></span>
<span id="cb22-18"><a href="#cb22-18" tabindex="-1"></a><span class="do">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span id="cb22-19"><a href="#cb22-19" tabindex="-1"></a><span class="do">## generated.</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># With CPM threshold (using design parameter)</span></span>
<span><span class="va">airway_abundant_cpm</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span>minimum_counts <span class="op">=</span> <span class="fl">10</span>, minimum_proportion <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="do">## Warning in filterByExpr.DGEList(y, design = design, group = group, lib.size =</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="do">## lib.size, : All samples appear to belong to the same group.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_2-summary-statistics-and-density-plots">2. Summary statistics and density plots<a class="anchor" aria-label="anchor" href="#id_2-summary-statistics-and-density-plots"></a>
</h4>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example: summary for default tidybulk filtering</span></span>
<span><span class="co"># Before filtering</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a><span class="do">## 1      63677         0            0    513766</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># After filtering</span></span>
<span><span class="va">airway_abundant_default</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a><span class="do">## 1      14224         8          394    513766</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_abundant_formula</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a><span class="do">## 1      15926         0          310    513766</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_abundant_cpm</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a><span class="do">## 1      14224         8          394    513766</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge all methods into a single tibble</span></span>
<span><span class="va">airway_abundant_all</span> <span class="op">=</span> </span>
<span>  <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"no filter"</span><span class="op">)</span>,</span>
<span>    <span class="va">airway_abundant_default</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"default"</span><span class="op">)</span>,</span>
<span>    <span class="va">airway_abundant_formula</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"formula"</span><span class="op">)</span>,</span>
<span>    <span class="va">airway_abundant_cpm</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"cpm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Density plot across methods</span></span>
<span><span class="va">airway_abundant_all</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Counts after abundance filtering (tidybulk default)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/filtering-density-plot-comparison-1.png"><!-- --></p>
<p>Update the <code>airway</code> object with the filtered data:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway_abundant_formula</span></span></code></pre></div>
<blockquote>
<p><strong>Tip:</strong> Use <code>formula_design</code> for complex designs, and use the CPM threshold for library-size-aware filtering.</p>
</blockquote>
</div>
</div>
<div class="section level3">
<h3 id="remove-redundant-transcripts">Remove Redundant Transcripts<a class="anchor" aria-label="anchor" href="#remove-redundant-transcripts"></a>
</h3>
<p>Redundancy removal is a standard approach for reducing highly correlated features.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_non_redundant</span> <span class="op">=</span> </span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/remove_redundancy-methods.html">remove_redundancy</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"correlation"</span>, top <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> </span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="do">## Getting the 100 most variable genes</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Make  </span></span>
<span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb38-4"><a href="#cb38-4" tabindex="-1"></a><span class="do">## 1      15926         0          310    513766</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summary statistics</span></span>
<span><span class="va">airway_non_redundant</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb40-4"><a href="#cb40-4" tabindex="-1"></a><span class="do">## 1      15926         0          357    401539</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot before and after</span></span>
<span><span class="co"># Merge before and after into a single tibble</span></span>
<span><span class="va">airway_all</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"before"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway_non_redundant</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"after"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Density plot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">as_tibble</span><span class="op">(</span><span class="va">airway_all</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Counts after removing redundant transcripts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/preprocessing-remove-redundancy-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="filter-variable-transcripts">Filter Variable Transcripts<a class="anchor" aria-label="anchor" href="#filter-variable-transcripts"></a>
</h3>
<p>Keep only the most variable transcripts for downstream analysis.</p>
<p>Variance-based feature selection using edgeR methodology (<a href="#ref-robinson2010edger">Robinson, McCarthy, and Smyth 2010</a>) is used for selecting informative features.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_variable</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="do">## Getting the 500 most variable genes</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-after-variable-filtering-variable-transcripts-optional">Visualize After Variable Filtering Variable Transcripts (optional)<a class="anchor" aria-label="anchor" href="#visualize-after-variable-filtering-variable-transcripts-optional"></a>
</h3>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Before filtering</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb45-3"><a href="#cb45-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb45-4"><a href="#cb45-4" tabindex="-1"></a><span class="do">## 1      15926         0          310    513766</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># After filtering</span></span>
<span><span class="va">airway_variable</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="do">## # A tibble: 1 × 4</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a><span class="do">##   n_features min_count median_count max_count</span></span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a><span class="do">##        &lt;int&gt;     &lt;int&gt;        &lt;dbl&gt;     &lt;int&gt;</span></span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a><span class="do">## 1        500         0           46     76824</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Density plot</span></span>
<span><span class="co"># Merge before and after into a single tibble</span></span>
<span><span class="va">airway_all</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"before"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway_variable</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"after"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Density plot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">as_tibble</span><span class="op">(</span><span class="va">airway_all</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Counts after variable filtering"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/filtering-variable-summary-and-plot-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="scale-abundance">Scale Abundance<a class="anchor" aria-label="anchor" href="#scale-abundance"></a>
</h3>
<p>Scale for sequencing depth using TMM (<a href="#ref-robinson2010edger">Robinson, McCarthy, and Smyth 2010</a>), upper quartile (<a href="#ref-bullard2010uq">Bullard et al. 2010</a>), and RLE (<a href="#ref-anders2010rle">Anders and Huber 2010</a>) normalization.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> </span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"TMM"</span>, suffix <span class="op">=</span> <span class="st">"_tmm"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"upperquartile"</span>, suffix <span class="op">=</span> <span class="st">"_upperquartile"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"RLE"</span>, suffix <span class="op">=</span> <span class="st">"_RLE"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="do">## tidybulk says: the sample with largest library size SRR1039517 was chosen as reference for scaling</span></span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a><span class="do">## tidybulk says: the sample with largest library size SRR1039517 was chosen as reference for scaling</span></span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a><span class="do">## tidybulk says: the sample with largest library size SRR1039517 was chosen as reference for scaling</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-after-scaling">Visualize After Scaling<a class="anchor" aria-label="anchor" href="#visualize-after-scaling"></a>
</h3>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Before scaling</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="do">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span id="cb52-2"><a href="#cb52-2" tabindex="-1"></a><span class="do">##      7.38     71.38    325.56   1375.62    989.38 328812.62</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_tmm"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="do">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="do">##     10.50     97.19    444.13   1884.49   1350.66 457443.80</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_upperquartile"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="do">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a><span class="do">##     10.51     99.10    452.57   1918.62   1376.18 465420.93</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_RLE"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="do">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a><span class="do">##     10.45     97.02    443.51   1881.64   1348.31 456680.75</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge all methods into a single tibble</span></span>
<span><span class="va">airway_scaled_all</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"no_scaling"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_tmm"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_upperquartile"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"upperquartile"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_RLE"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"RLE"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Density plot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">as_tibble</span><span class="op">(</span><span class="va">airway_scaled_all</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scaled counts by method (after scaling)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/normalization-visualize-scaling-1.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="step-2-exploratory-data-analysis">Step 2: Exploratory Data Analysis<a class="anchor" aria-label="anchor" href="#step-2-exploratory-data-analysis"></a>
</h2>
<div class="section level3">
<h3 id="remove-zero-variance-features-required-for-pca">Remove Zero-Variance Features (required for PCA)<a class="anchor" aria-label="anchor" href="#remove-zero-variance-features-required-for-pca"></a>
</h3>
<p>Variance filtering is a standard preprocessing step for dimensionality reduction.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span></span>
<span><span class="co"># Remove features with zero variance across samples</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowVars.html" class="external-link">rowVars</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dimensionality-reduction">Dimensionality Reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h3>
<p>MDS (<a href="#ref-kruskal1964mds">Kruskal 1964</a>) using limma::plotMDS (<a href="#ref-ritchie2015limma"><strong>ritchie2015limma?</strong></a>) and PCA (<a href="#ref-hotelling1933pca">Hotelling 1933</a>) are used for dimensionality reduction.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"MDS"</span>, .dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb62"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a><span class="do">## Getting the 500 most variable genes</span></span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" tabindex="-1"></a><span class="do">## [1] "MDS result_df colnames: sample, 1, 2"</span></span>
<span id="cb62-4"><a href="#cb62-4" tabindex="-1"></a></span>
<span id="cb62-5"><a href="#cb62-5" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results do `metadata(.)$tidybulk$MDS`</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span>, .dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" tabindex="-1"></a><span class="do">## Getting the 500 most variable genes</span></span>
<span id="cb64-2"><a href="#cb64-2" tabindex="-1"></a></span>
<span id="cb64-3"><a href="#cb64-3" tabindex="-1"></a><span class="do">## Fraction of variance explained by the selected principal components</span></span>
<span id="cb64-4"><a href="#cb64-4" tabindex="-1"></a></span>
<span id="cb64-5"><a href="#cb64-5" tabindex="-1"></a><span class="do">## # A tibble: 2 × 2</span></span>
<span id="cb64-6"><a href="#cb64-6" tabindex="-1"></a><span class="do">##   `Fraction of variance`    PC</span></span>
<span id="cb64-7"><a href="#cb64-7" tabindex="-1"></a><span class="do">##                    &lt;dbl&gt; &lt;int&gt;</span></span>
<span id="cb64-8"><a href="#cb64-8" tabindex="-1"></a><span class="do">## 1                 0.0652     1</span></span>
<span id="cb64-9"><a href="#cb64-9" tabindex="-1"></a><span class="do">## 2                 0.0553     2</span></span>
<span id="cb64-10"><a href="#cb64-10" tabindex="-1"></a></span>
<span id="cb64-11"><a href="#cb64-11" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results do `metadata(.)$tidybulk$PCA`</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-dimensionality-reduction-results">Visualize Dimensionality Reduction Results<a class="anchor" aria-label="anchor" href="#visualize-dimensionality-reduction-results"></a>
</h3>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># MDS plot</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"MDS Analysis"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/eda-plot-dimensionality-reduction-1.png"><!-- --></p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCA plot</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">`PC1`</span>, y<span class="op">=</span><span class="va">`PC2`</span>, color<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PCA Analysis"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/eda-plot-dimensionality-reduction-2.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="clustering-analysis">Clustering Analysis<a class="anchor" aria-label="anchor" href="#clustering-analysis"></a>
</h3>
<p>K-means clustering (<a href="#ref-macqueen1967kmeans">MacQueen 1967</a>) is used for unsupervised grouping.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/cluster_elements-methods.html">cluster_elements</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"kmeans"</span>, centers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize clustering</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`cluster_kmeans`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"K-means Clustering"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/eda-clustering-analysis-1.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="step-3-differential-expression-analysis">Step 3: Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#step-3-differential-expression-analysis"></a>
</h2>
<p>This workflow uses</p>
<p><strong>edgeR quasi-likelihood</strong> (<a href="#ref-robinson2010edger">Robinson, McCarthy, and Smyth 2010</a>; <a href="#ref-chen2016edgeR">Chen, Lun, and Smyth 2016</a>) <strong>edgeR robust likelihood ratio</strong> (<a href="#ref-chen2016edgeR">Chen, Lun, and Smyth 2016</a>), <strong>DESeq2</strong> (<a href="#ref-love2014deseq2">Love, Huber, and Anders 2014</a>), <strong>limma-voom</strong> (<a href="#ref-law2014voom">Law et al. 2014</a>), and <strong>limma-voom with sample weights</strong> (<a href="#ref-liu2015voomweights">Liu et al. 2015</a>) for differential expression analysis.</p>
<div class="section level3">
<h3 id="basic-differential-expression">Basic Differential Expression<a class="anchor" aria-label="anchor" href="#basic-differential-expression"></a>
</h3>
<p><strong>Methods:</strong></p>
<ul>
<li><p><strong>edgeR quasi-likelihood:</strong> Quasi-likelihood F-tests for differential expression</p></li>
<li><p><strong>edgeR robust likelihood ratio:</strong> Robust likelihood ratio tests</p></li>
<li><p><strong>DESeq2:</strong> Negative binomial distribution with dispersion estimation</p></li>
<li><p><strong>limma-voom:</strong> Linear modeling with empirical Bayes moderation</p></li>
<li><p><strong>limma-voom with sample weights:</strong> Enhanced voom with quality weights <strong>References:</strong></p></li>
<li><p>Robinson et al. (2010) edgeR: a Bioconductor package for differential expression analysis</p></li>
<li><p>Chen et al. (2016) From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline</p></li>
<li><p>Love et al. (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2</p></li>
<li><p>Law et al. (2014) voom: precision weights unlock linear model analysis tools for RNA-seq read counts</p></li>
<li><p>Liu et al. (2015) Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses</p></li>
</ul>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Standard differential expression analysis</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span><span class="co"># Use QL method</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>, prefix <span class="op">=</span> <span class="st">"ql__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Use edger_robust_likelihood_ratio</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"edger_robust_likelihood_ratio"</span>, prefix <span class="op">=</span> <span class="st">"lr_robust__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span><span class="co"># Use DESeq2 method</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"DESeq2"</span>, prefix <span class="op">=</span> <span class="st">"deseq2__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Use limma_voom</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"limma_voom"</span>, prefix <span class="op">=</span> <span class="st">"voom__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span><span class="co"># Use limma_voom_sample_weights</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"limma_voom_sample_weights"</span>, prefix <span class="op">=</span> <span class="st">"voom_weights__"</span><span class="op">)</span> </span></code></pre></div>
<div class="sourceCode" id="cb69"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" tabindex="-1"></a><span class="do">## Warning: The `.abundance` argument of `test_differential_abundance()` is deprecated as</span></span>
<span id="cb69-2"><a href="#cb69-2" tabindex="-1"></a><span class="do">## of tidybulk 2.0.0.</span></span>
<span id="cb69-3"><a href="#cb69-3" tabindex="-1"></a><span class="do">## ℹ Please use the `abundance` argument instead.</span></span>
<span id="cb69-4"><a href="#cb69-4" tabindex="-1"></a><span class="do">## ℹ The deprecated feature was likely used in the tidybulk package.</span></span>
<span id="cb69-5"><a href="#cb69-5" tabindex="-1"></a><span class="do">##   Please report the issue at &lt;https://github.com/stemangiola/tidybulk/issues&gt;.</span></span>
<span id="cb69-6"><a href="#cb69-6" tabindex="-1"></a><span class="do">## This warning is displayed once every 8 hours.</span></span>
<span id="cb69-7"><a href="#cb69-7" tabindex="-1"></a><span class="do">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span id="cb69-8"><a href="#cb69-8" tabindex="-1"></a><span class="do">## generated.</span></span>
<span id="cb69-9"><a href="#cb69-9" tabindex="-1"></a></span>
<span id="cb69-10"><a href="#cb69-10" tabindex="-1"></a><span class="do">## =====================================</span></span>
<span id="cb69-11"><a href="#cb69-11" tabindex="-1"></a><span class="do">## tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance</span></span>
<span id="cb69-12"><a href="#cb69-12" tabindex="-1"></a><span class="do">## or adjust_abundance have been calculated. Therefore, it is essential to add covariates</span></span>
<span id="cb69-13"><a href="#cb69-13" tabindex="-1"></a><span class="do">## such as batch effects (if applicable) in the formula.</span></span>
<span id="cb69-14"><a href="#cb69-14" tabindex="-1"></a><span class="do">## =====================================</span></span>
<span id="cb69-15"><a href="#cb69-15" tabindex="-1"></a><span class="do">## tidybulk says: The design column names are "(Intercept), dexuntrt"</span></span>
<span id="cb69-16"><a href="#cb69-16" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-17"><a href="#cb69-17" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$edgeR_quasi_likelihood_object`</span></span>
<span id="cb69-18"><a href="#cb69-18" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$edgeR_quasi_likelihood_fit`</span></span>
<span id="cb69-19"><a href="#cb69-19" tabindex="-1"></a><span class="do">## tidybulk says: The design column names are "(Intercept), dexuntrt"</span></span>
<span id="cb69-20"><a href="#cb69-20" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-21"><a href="#cb69-21" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$edger_robust_likelihood_ratio_object`</span></span>
<span id="cb69-22"><a href="#cb69-22" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$edger_robust_likelihood_ratio_fit`</span></span>
<span id="cb69-23"><a href="#cb69-23" tabindex="-1"></a><span class="do">## estimating size factors</span></span>
<span id="cb69-24"><a href="#cb69-24" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-25"><a href="#cb69-25" tabindex="-1"></a><span class="do">## estimating dispersions</span></span>
<span id="cb69-26"><a href="#cb69-26" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-27"><a href="#cb69-27" tabindex="-1"></a><span class="do">## gene-wise dispersion estimates</span></span>
<span id="cb69-28"><a href="#cb69-28" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-29"><a href="#cb69-29" tabindex="-1"></a><span class="do">## mean-dispersion relationship</span></span>
<span id="cb69-30"><a href="#cb69-30" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-31"><a href="#cb69-31" tabindex="-1"></a><span class="do">## final dispersion estimates</span></span>
<span id="cb69-32"><a href="#cb69-32" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-33"><a href="#cb69-33" tabindex="-1"></a><span class="do">## fitting model and testing</span></span>
<span id="cb69-34"><a href="#cb69-34" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-35"><a href="#cb69-35" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$DESeq2_object`</span></span>
<span id="cb69-36"><a href="#cb69-36" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$DESeq2_fit`</span></span>
<span id="cb69-37"><a href="#cb69-37" tabindex="-1"></a><span class="do">## tidybulk says: The design column names are "(Intercept), dexuntrt"</span></span>
<span id="cb69-38"><a href="#cb69-38" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-39"><a href="#cb69-39" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$limma_voom_object`</span></span>
<span id="cb69-40"><a href="#cb69-40" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$limma_voom_fit`</span></span>
<span id="cb69-41"><a href="#cb69-41" tabindex="-1"></a><span class="do">## tidybulk says: The design column names are "(Intercept), dexuntrt"</span></span>
<span id="cb69-42"><a href="#cb69-42" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb69-43"><a href="#cb69-43" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$limma_voom_sample_weights_object`</span></span>
<span id="cb69-44"><a href="#cb69-44" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$limma_voom_sample_weights_fit`</span></span>
<span id="cb69-45"><a href="#cb69-45" tabindex="-1"></a><span class="do">## This message is displayed once per session.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quality-control-of-the-fit">Quality Control of the Fit<a class="anchor" aria-label="anchor" href="#quality-control-of-the-fit"></a>
</h3>
<p>It is important to check the quality of the fit. All methods produce a fit object that can be used for quality control. The fit object produced by each underlying method are stored in as attributes of the <code>airway_mini</code> object. We can use them for example to perform quality control of the fit.</p>
<div class="section level4">
<h4 id="for-edger">For edgeR<a class="anchor" aria-label="anchor" href="#for-edger"></a>
</h4>
<p>Plot the biological coefficient of variation (BCV) trend. This plot is helpful to understant the dispersion of the data.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb71"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" tabindex="-1"></a><span class="do">## Warning: package 'edgeR' was built under R version 4.5.1</span></span>
<span id="cb71-2"><a href="#cb71-2" tabindex="-1"></a></span>
<span id="cb71-3"><a href="#cb71-3" tabindex="-1"></a><span class="do">## Loading required package: limma</span></span>
<span id="cb71-4"><a href="#cb71-4" tabindex="-1"></a></span>
<span id="cb71-5"><a href="#cb71-5" tabindex="-1"></a><span class="do">## Warning: package 'limma' was built under R version 4.5.1</span></span>
<span id="cb71-6"><a href="#cb71-6" tabindex="-1"></a></span>
<span id="cb71-7"><a href="#cb71-7" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb71-8"><a href="#cb71-8" tabindex="-1"></a><span class="do">## Attaching package: 'limma'</span></span>
<span id="cb71-9"><a href="#cb71-9" tabindex="-1"></a></span>
<span id="cb71-10"><a href="#cb71-10" tabindex="-1"></a><span class="do">## The following object is masked from 'package:BiocGenerics':</span></span>
<span id="cb71-11"><a href="#cb71-11" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb71-12"><a href="#cb71-12" tabindex="-1"></a><span class="do">##     plotMA</span></span></code></pre></div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">edgeR_quasi_likelihood_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-edgeR-object-1.png"><!-- --></p>
<p>Plot the log-fold change vs mean plot.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">edgeR_quasi_likelihood_fit</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-edgeR-fit-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="for-deseq2">For DESeq2<a class="anchor" aria-label="anchor" href="#for-deseq2"></a>
</h4>
<p>Plot the mean-variance trend.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">DESeq2_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">plotDispEsts</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-DESeq2-object-1.png"><!-- --></p>
<p>Plot the log-fold change vs mean plot.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">DESeq2_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotMA.html" class="external-link">plotMA</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-DESeq2-fit-1.png"><!-- --></p>
</div>
</div>
<div class="section level3">
<h3 id="histograms-of-p-values-across-methods">Histograms of p-values across methods<a class="anchor" aria-label="anchor" href="#histograms-of-p-values-across-methods"></a>
</h3>
<p>Inspection of the raw p-value histogram provides a rapid check of differential-expression results. When no gene is truly differentially expressed, the p-values follow a uniform U(0,1) distribution across the interval 0–1, so the histogram appears flat <a href="https://bioconductor.org/help/course-materials/2014/useR2014/Workflows.html" class="external-link">Source</a>. In a more realistic scenario where only a subset of genes changes, this uniform background is still present but an obvious spike emerges close to zero, created by the genuine signals.</p>
<p>Thanks to the modularity of the <code>tidybulk</code> workflow, that can multiplex different methods, we can easily compare the p-values across methods.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">ql__PValue</span>, </span>
<span>    <span class="va">lr_robust__PValue</span>, </span>
<span>    <span class="va">voom__P.Value</span>, </span>
<span>    <span class="va">voom_weights__P.Value</span>, </span>
<span>    <span class="va">deseq2__pvalue</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu">everything</span><span class="op">(</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"method"</span>, values_to <span class="op">=</span> <span class="st">"pvalue"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pvalue</span>, fill <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Histogram of p-values across methods"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb77"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing non-finite outside the scale range</span></span>
<span id="cb77-2"><a href="#cb77-2" tabindex="-1"></a><span class="do">## (`stat_bin()`).</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-pvalue-histograms-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="compare-results-across-methods">Compare Results Across Methods<a class="anchor" aria-label="anchor" href="#compare-results-across-methods"></a>
</h3>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summary statistics</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"ql|lr_robust|voom|voom_weights|deseq2"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"logFC"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span><span class="fu">summarise</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">everything</span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min <span class="op">=</span> <span class="va">min</span>, median <span class="op">=</span> <span class="va">median</span>, max <span class="op">=</span> <span class="va">max</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb79"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" tabindex="-1"></a><span class="do">## Warning: There was 1 warning in `summarise()`.</span></span>
<span id="cb79-2"><a href="#cb79-2" tabindex="-1"></a><span class="do">## ℹ In argument: `across(...)`.</span></span>
<span id="cb79-3"><a href="#cb79-3" tabindex="-1"></a><span class="do">## Caused by warning:</span></span>
<span id="cb79-4"><a href="#cb79-4" tabindex="-1"></a><span class="do">## ! The `...` argument of `across()` is deprecated as of dplyr 1.1.0.</span></span>
<span id="cb79-5"><a href="#cb79-5" tabindex="-1"></a><span class="do">## Supply arguments directly to `.fns` through an anonymous function instead.</span></span>
<span id="cb79-6"><a href="#cb79-6" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb79-7"><a href="#cb79-7" tabindex="-1"></a><span class="do">##   # Previously</span></span>
<span id="cb79-8"><a href="#cb79-8" tabindex="-1"></a><span class="do">##   across(a:b, mean, na.rm = TRUE)</span></span>
<span id="cb79-9"><a href="#cb79-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb79-10"><a href="#cb79-10" tabindex="-1"></a><span class="do">##   # Now</span></span>
<span id="cb79-11"><a href="#cb79-11" tabindex="-1"></a><span class="do">##   across(a:b, \(x) mean(x, na.rm = TRUE))</span></span>
<span id="cb79-12"><a href="#cb79-12" tabindex="-1"></a></span>
<span id="cb79-13"><a href="#cb79-13" tabindex="-1"></a><span class="do">## # A tibble: 1 × 0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pairplot-of-pvalues-across-methods-ggpairs">Pairplot of pvalues across methods (GGpairs)<a class="anchor" aria-label="anchor" href="#pairplot-of-pvalues-across-methods-ggpairs"></a>
</h3>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/" class="external-link">GGally</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" tabindex="-1"></a><span class="do">## Warning: package 'GGally' was built under R version 4.5.1</span></span></code></pre></div>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">ql__PValue</span>, <span class="va">lr_robust__PValue</span>, <span class="va">voom__P.Value</span>, <span class="va">voom_weights__P.Value</span>, <span class="va">deseq2__pvalue</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="reference/log10_reverse_trans.html">log10_reverse_trans</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="reference/log10_reverse_trans.html">log10_reverse_trans</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Pairplot of p-values across methods"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb83"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" tabindex="-1"></a><span class="do">## Warning in ggally_statistic(data = data, mapping = mapping, na.rm = na.rm, :</span></span>
<span id="cb83-2"><a href="#cb83-2" tabindex="-1"></a><span class="do">## Removed 30 rows containing missing values</span></span>
<span id="cb83-3"><a href="#cb83-3" tabindex="-1"></a></span>
<span id="cb83-4"><a href="#cb83-4" tabindex="-1"></a><span class="do">## Warning in ggally_statistic(data = data, mapping = mapping, na.rm = na.rm, :</span></span>
<span id="cb83-5"><a href="#cb83-5" tabindex="-1"></a><span class="do">## Removed 30 rows containing missing values</span></span>
<span id="cb83-6"><a href="#cb83-6" tabindex="-1"></a><span class="do">## Warning in ggally_statistic(data = data, mapping = mapping, na.rm = na.rm, :</span></span>
<span id="cb83-7"><a href="#cb83-7" tabindex="-1"></a><span class="do">## Removed 30 rows containing missing values</span></span>
<span id="cb83-8"><a href="#cb83-8" tabindex="-1"></a><span class="do">## Warning in ggally_statistic(data = data, mapping = mapping, na.rm = na.rm, :</span></span>
<span id="cb83-9"><a href="#cb83-9" tabindex="-1"></a><span class="do">## Removed 30 rows containing missing values</span></span>
<span id="cb83-10"><a href="#cb83-10" tabindex="-1"></a></span>
<span id="cb83-11"><a href="#cb83-11" tabindex="-1"></a><span class="do">## Warning in scale_y_continuous(trans = tidybulk::log10_reverse_trans()):</span></span>
<span id="cb83-12"><a href="#cb83-12" tabindex="-1"></a><span class="do">## log10_reverse transformation introduced infinite values.</span></span>
<span id="cb83-13"><a href="#cb83-13" tabindex="-1"></a></span>
<span id="cb83-14"><a href="#cb83-14" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb83-15"><a href="#cb83-15" tabindex="-1"></a><span class="do">## (`geom_point()`).</span></span>
<span id="cb83-16"><a href="#cb83-16" tabindex="-1"></a><span class="do">## Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb83-17"><a href="#cb83-17" tabindex="-1"></a><span class="do">## (`geom_point()`).</span></span>
<span id="cb83-18"><a href="#cb83-18" tabindex="-1"></a><span class="do">## Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb83-19"><a href="#cb83-19" tabindex="-1"></a><span class="do">## (`geom_point()`).</span></span>
<span id="cb83-20"><a href="#cb83-20" tabindex="-1"></a><span class="do">## Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb83-21"><a href="#cb83-21" tabindex="-1"></a><span class="do">## (`geom_point()`).</span></span>
<span id="cb83-22"><a href="#cb83-22" tabindex="-1"></a></span>
<span id="cb83-23"><a href="#cb83-23" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing non-finite outside the scale range</span></span>
<span id="cb83-24"><a href="#cb83-24" tabindex="-1"></a><span class="do">## (`stat_density()`).</span></span>
<span id="cb83-25"><a href="#cb83-25" tabindex="-1"></a></span>
<span id="cb83-26"><a href="#cb83-26" tabindex="-1"></a><span class="do">## Warning in scale_y_continuous(trans = tidybulk::log10_reverse_trans()):</span></span>
<span id="cb83-27"><a href="#cb83-27" tabindex="-1"></a><span class="do">## log10_reverse transformation introduced infinite values.</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-pvalue-pairplot-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="pairplot-of-effect-sizes-across-methods-ggpairs">Pairplot of effect sizes across methods (GGpairs)<a class="anchor" aria-label="anchor" href="#pairplot-of-effect-sizes-across-methods-ggpairs"></a>
</h3>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/" class="external-link">GGally</a></span><span class="op">)</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">ql__logFC</span>, <span class="va">lr_robust__logFC</span>, <span class="va">voom__logFC</span>, <span class="va">voom_weights__logFC</span>, <span class="va">deseq2__log2FoldChange</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Pairplot of effect sizes across methods"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-effectsize-pairplot-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="volcano-plots-for-each-method">Volcano Plots for Each Method<a class="anchor" aria-label="anchor" href="#volcano-plots-for-each-method"></a>
</h3>
<p>Visualising the significance and effect size of the differential expression results as a volcano plots we appreciate that some methods have much lower p-values distributions than other methods, for the same model and data.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create volcano plots</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Select the columns we want to plot</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>            <span class="va">.feature</span>,</span>
<span>      <span class="va">ql__logFC</span>, <span class="va">ql__PValue</span>,</span>
<span>      <span class="va">lr_robust__logFC</span>, <span class="va">lr_robust__PValue</span>,</span>
<span>      <span class="va">voom__logFC</span>, <span class="va">voom__P.Value</span>,</span>
<span>      <span class="va">voom_weights__logFC</span>, <span class="va">voom_weights__P.Value</span>,</span>
<span>      <span class="va">deseq2__log2FoldChange</span>, <span class="va">deseq2__pvalue</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Pivot longer to get a tidy data frame</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>      <span class="op">-</span> <span class="va">.feature</span>,</span>
<span>      names_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"method"</span>, <span class="st">"stat"</span><span class="op">)</span>,</span>
<span>      values_to <span class="op">=</span> <span class="st">"value"</span>, names_sep <span class="op">=</span> <span class="st">"__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Harmonize column names</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>stat  <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"log2FoldChange"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"logFC"</span>,</span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PValue"</span>, <span class="st">"pvalue"</span>, <span class="st">"P.Value"</span>, <span class="st">"p.value"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"PValue"</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"stat"</span>, values_from <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, size <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="reference/log10_reverse_trans.html">log10_reverse_trans</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="fl">0.5</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Volcano Plots by Method"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb86"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb86-2"><a href="#cb86-2" tabindex="-1"></a><span class="do">## (`geom_point()`).</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-volcano-plots-1-1.png"><!-- --></p>
<p>Plotting independent y-axis scales for the p-values and effect sizes allows us to compare the top genes across methods.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create volcano plots</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Select the columns we want to plot</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">as_tibble</span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>            <span class="va">symbol</span>,</span>
<span>      <span class="va">ql__logFC</span>, <span class="va">ql__PValue</span>,</span>
<span>      <span class="va">lr_robust__logFC</span>, <span class="va">lr_robust__PValue</span>,</span>
<span>      <span class="va">voom__logFC</span>, <span class="va">voom__P.Value</span>,</span>
<span>      <span class="va">voom_weights__logFC</span>, <span class="va">voom_weights__P.Value</span>,</span>
<span>      <span class="va">deseq2__log2FoldChange</span>, <span class="va">deseq2__pvalue</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Pivot longer to get a tidy data frame</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>      <span class="op">-</span> <span class="va">symbol</span>,</span>
<span>      names_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"method"</span>, <span class="st">"stat"</span><span class="op">)</span>,</span>
<span>      values_to <span class="op">=</span> <span class="st">"value"</span>, names_sep <span class="op">=</span> <span class="st">"__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Harmonize column names</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>stat  <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"log2FoldChange"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"logFC"</span>,</span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PValue"</span>, <span class="st">"pvalue"</span>, <span class="st">"P.Value"</span>, <span class="st">"p.value"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"PValue"</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"stat"</span>, values_from <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, size <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">symbol</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="fu"><a href="reference/log10_reverse_trans.html">log10_reverse_trans</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="fl">0.5</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Volcano Plots by Method"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb88"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" tabindex="-1"></a><span class="do">## Warning: Values from `value` are not uniquely identified; output will contain list-cols.</span></span>
<span id="cb88-2"><a href="#cb88-2" tabindex="-1"></a><span class="do">## • Use `values_fn = list` to suppress this warning.</span></span>
<span id="cb88-3"><a href="#cb88-3" tabindex="-1"></a><span class="do">## • Use `values_fn = {summary_fun}` to summarise duplicates.</span></span>
<span id="cb88-4"><a href="#cb88-4" tabindex="-1"></a><span class="do">## • Use the following dplyr code to identify duplicates.</span></span>
<span id="cb88-5"><a href="#cb88-5" tabindex="-1"></a><span class="do">##   {data} |&gt;</span></span>
<span id="cb88-6"><a href="#cb88-6" tabindex="-1"></a><span class="do">##   dplyr::summarise(n = dplyr::n(), .by = c(symbol, method, stat)) |&gt;</span></span>
<span id="cb88-7"><a href="#cb88-7" tabindex="-1"></a><span class="do">##   dplyr::filter(n &gt; 1L)</span></span>
<span id="cb88-8"><a href="#cb88-8" tabindex="-1"></a></span>
<span id="cb88-9"><a href="#cb88-9" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb88-10"><a href="#cb88-10" tabindex="-1"></a><span class="do">## (`geom_point()`).</span></span>
<span id="cb88-11"><a href="#cb88-11" tabindex="-1"></a></span>
<span id="cb88-12"><a href="#cb88-12" tabindex="-1"></a><span class="do">## Warning: Removed 30 rows containing missing values or values outside the scale range</span></span>
<span id="cb88-13"><a href="#cb88-13" tabindex="-1"></a><span class="do">## (`geom_text_repel()`).</span></span>
<span id="cb88-14"><a href="#cb88-14" tabindex="-1"></a></span>
<span id="cb88-15"><a href="#cb88-15" tabindex="-1"></a><span class="do">## Warning: ggrepel: 15885 unlabeled data points (too many overlaps). Consider</span></span>
<span id="cb88-16"><a href="#cb88-16" tabindex="-1"></a><span class="do">## increasing max.overlaps</span></span>
<span id="cb88-17"><a href="#cb88-17" tabindex="-1"></a></span>
<span id="cb88-18"><a href="#cb88-18" tabindex="-1"></a><span class="do">## Warning: ggrepel: 15926 unlabeled data points (too many overlaps). Consider</span></span>
<span id="cb88-19"><a href="#cb88-19" tabindex="-1"></a><span class="do">## increasing max.overlaps</span></span>
<span id="cb88-20"><a href="#cb88-20" tabindex="-1"></a></span>
<span id="cb88-21"><a href="#cb88-21" tabindex="-1"></a><span class="do">## Warning: ggrepel: 15916 unlabeled data points (too many overlaps). Consider</span></span>
<span id="cb88-22"><a href="#cb88-22" tabindex="-1"></a><span class="do">## increasing max.overlaps</span></span>
<span id="cb88-23"><a href="#cb88-23" tabindex="-1"></a></span>
<span id="cb88-24"><a href="#cb88-24" tabindex="-1"></a><span class="do">## Warning: ggrepel: 15925 unlabeled data points (too many overlaps). Consider increasing max.overlaps</span></span>
<span id="cb88-25"><a href="#cb88-25" tabindex="-1"></a><span class="do">## ggrepel: 15925 unlabeled data points (too many overlaps). Consider increasing max.overlaps</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/differential-expression-volcano-plots-2-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="differential-expression-with-contrasts">Differential Expression with Contrasts<a class="anchor" aria-label="anchor" href="#differential-expression-with-contrasts"></a>
</h3>
<p>Contrast-based differential expression analysis using <strong>edgeR</strong> (<a href="#ref-robinson2010edger">Robinson, McCarthy, and Smyth 2010</a>; <a href="#ref-chen2016edgeR">Chen, Lun, and Smyth 2016</a>) is a standard statistical approach for testing specific comparisons in complex designs.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using contrasts for more complex comparisons</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,                  </span>
<span>        .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dextrt - dexuntrt"</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>, </span>
<span>        prefix <span class="op">=</span> <span class="st">"contrasts__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Print the gene statistics</span></span>
<span>  <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"contrasts"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb90"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" tabindex="-1"></a><span class="do">## Warning: The `.contrasts` argument of `test_differential_abundance()` is deprecated as</span></span>
<span id="cb90-2"><a href="#cb90-2" tabindex="-1"></a><span class="do">## of tidybulk 1.7.4.</span></span>
<span id="cb90-3"><a href="#cb90-3" tabindex="-1"></a><span class="do">## ℹ The argument .contrasts is now deprecated please use contrasts (without the</span></span>
<span id="cb90-4"><a href="#cb90-4" tabindex="-1"></a><span class="do">##   dot).</span></span>
<span id="cb90-5"><a href="#cb90-5" tabindex="-1"></a><span class="do">## ℹ The deprecated feature was likely used in the tidybulk package.</span></span>
<span id="cb90-6"><a href="#cb90-6" tabindex="-1"></a><span class="do">##   Please report the issue at &lt;https://github.com/stemangiola/tidybulk/issues&gt;.</span></span>
<span id="cb90-7"><a href="#cb90-7" tabindex="-1"></a><span class="do">## This warning is displayed once every 8 hours.</span></span>
<span id="cb90-8"><a href="#cb90-8" tabindex="-1"></a><span class="do">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span id="cb90-9"><a href="#cb90-9" tabindex="-1"></a><span class="do">## generated.</span></span>
<span id="cb90-10"><a href="#cb90-10" tabindex="-1"></a></span>
<span id="cb90-11"><a href="#cb90-11" tabindex="-1"></a><span class="do">## tidybulk says: The design column names are "dextrt, dexuntrt"</span></span>
<span id="cb90-12"><a href="#cb90-12" tabindex="-1"></a></span>
<span id="cb90-13"><a href="#cb90-13" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$edgeR_quasi_likelihood_object`</span></span>
<span id="cb90-14"><a href="#cb90-14" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$edgeR_quasi_likelihood_fit`</span></span>
<span id="cb90-15"><a href="#cb90-15" tabindex="-1"></a></span>
<span id="cb90-16"><a href="#cb90-16" tabindex="-1"></a><span class="do">## # A tibble: 15,926 × 5</span></span>
<span id="cb90-17"><a href="#cb90-17" tabindex="-1"></a><span class="do">##    contrasts__logFC___dextrt...d…¹ contrasts__logCPM___…² contrasts__F___dextr…³</span></span>
<span id="cb90-18"><a href="#cb90-18" tabindex="-1"></a><span class="do">##                              &lt;dbl&gt;                  &lt;dbl&gt;                  &lt;dbl&gt;</span></span>
<span id="cb90-19"><a href="#cb90-19" tabindex="-1"></a><span class="do">##  1                         -0.388                    5.06                 6.82  </span></span>
<span id="cb90-20"><a href="#cb90-20" tabindex="-1"></a><span class="do">##  2                          0.195                    4.61                 5.98  </span></span>
<span id="cb90-21"><a href="#cb90-21" tabindex="-1"></a><span class="do">##  3                          0.0239                   3.48                 0.0566</span></span>
<span id="cb90-22"><a href="#cb90-22" tabindex="-1"></a><span class="do">##  4                         -0.124                    1.48                 0.214 </span></span>
<span id="cb90-23"><a href="#cb90-23" tabindex="-1"></a><span class="do">##  5                          0.432                    8.09                 2.94  </span></span>
<span id="cb90-24"><a href="#cb90-24" tabindex="-1"></a><span class="do">##  6                         -0.250                    5.91                 6.08  </span></span>
<span id="cb90-25"><a href="#cb90-25" tabindex="-1"></a><span class="do">##  7                         -0.0387                   4.84                 0.0327</span></span>
<span id="cb90-26"><a href="#cb90-26" tabindex="-1"></a><span class="do">##  8                         -0.497                    4.12                 7.07  </span></span>
<span id="cb90-27"><a href="#cb90-27" tabindex="-1"></a><span class="do">##  9                         -0.140                    3.12                 1.03  </span></span>
<span id="cb90-28"><a href="#cb90-28" tabindex="-1"></a><span class="do">## 10                         -0.0527                   7.04                 0.0331</span></span>
<span id="cb90-29"><a href="#cb90-29" tabindex="-1"></a><span class="do">## # ℹ 15,916 more rows</span></span>
<span id="cb90-30"><a href="#cb90-30" tabindex="-1"></a><span class="do">## # ℹ abbreviated names: ¹​contrasts__logFC___dextrt...dexuntrt,</span></span>
<span id="cb90-31"><a href="#cb90-31" tabindex="-1"></a><span class="do">## #   ²​contrasts__logCPM___dextrt...dexuntrt, ³​contrasts__F___dextrt...dexuntrt</span></span>
<span id="cb90-32"><a href="#cb90-32" tabindex="-1"></a><span class="do">## # ℹ 2 more variables: contrasts__PValue___dextrt...dexuntrt &lt;dbl&gt;,</span></span>
<span id="cb90-33"><a href="#cb90-33" tabindex="-1"></a><span class="do">## #   contrasts__FDR___dextrt...dexuntrt &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-expression-with-minimum-fold-change-treat-method">Differential Expression with minimum fold change (TREAT method)<a class="anchor" aria-label="anchor" href="#differential-expression-with-minimum-fold-change-treat-method"></a>
</h3>
<p>TREAT method (<a href="#ref-mccarthy2009treat">McCarthy and Smyth 2009</a>) is used for testing significance relative to a fold-change threshold.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using contrasts for more complex comparisons</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,                  </span>
<span>        .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dextrt - dexuntrt"</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>, </span>
<span>        test_above_log2_fold_change <span class="op">=</span> <span class="fl">2</span>, </span>
<span>        prefix <span class="op">=</span> <span class="st">"treat__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Print the gene statistics</span></span>
<span>  <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"treat"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb92"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" tabindex="-1"></a><span class="do">## tidybulk says: The design column names are "dextrt, dexuntrt"</span></span>
<span id="cb92-2"><a href="#cb92-2" tabindex="-1"></a></span>
<span id="cb92-3"><a href="#cb92-3" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$edgeR_quasi_likelihood_object`</span></span>
<span id="cb92-4"><a href="#cb92-4" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$edgeR_quasi_likelihood_fit`</span></span>
<span id="cb92-5"><a href="#cb92-5" tabindex="-1"></a></span>
<span id="cb92-6"><a href="#cb92-6" tabindex="-1"></a><span class="do">## # A tibble: 15,926 × 5</span></span>
<span id="cb92-7"><a href="#cb92-7" tabindex="-1"></a><span class="do">##    treat__logFC___dextrt...dexun…¹ treat__unshrunk.logF…² treat__logCPM___dext…³</span></span>
<span id="cb92-8"><a href="#cb92-8" tabindex="-1"></a><span class="do">##                              &lt;dbl&gt;                  &lt;dbl&gt;                  &lt;dbl&gt;</span></span>
<span id="cb92-9"><a href="#cb92-9" tabindex="-1"></a><span class="do">##  1                         -0.388                 -0.388                    5.06</span></span>
<span id="cb92-10"><a href="#cb92-10" tabindex="-1"></a><span class="do">##  2                          0.195                  0.195                    4.61</span></span>
<span id="cb92-11"><a href="#cb92-11" tabindex="-1"></a><span class="do">##  3                          0.0239                 0.0240                   3.48</span></span>
<span id="cb92-12"><a href="#cb92-12" tabindex="-1"></a><span class="do">##  4                         -0.124                 -0.124                    1.48</span></span>
<span id="cb92-13"><a href="#cb92-13" tabindex="-1"></a><span class="do">##  5                          0.432                  0.433                    8.09</span></span>
<span id="cb92-14"><a href="#cb92-14" tabindex="-1"></a><span class="do">##  6                         -0.250                 -0.250                    5.91</span></span>
<span id="cb92-15"><a href="#cb92-15" tabindex="-1"></a><span class="do">##  7                         -0.0387                -0.0387                   4.84</span></span>
<span id="cb92-16"><a href="#cb92-16" tabindex="-1"></a><span class="do">##  8                         -0.497                 -0.497                    4.12</span></span>
<span id="cb92-17"><a href="#cb92-17" tabindex="-1"></a><span class="do">##  9                         -0.140                 -0.140                    3.12</span></span>
<span id="cb92-18"><a href="#cb92-18" tabindex="-1"></a><span class="do">## 10                         -0.0527                -0.0527                   7.04</span></span>
<span id="cb92-19"><a href="#cb92-19" tabindex="-1"></a><span class="do">## # ℹ 15,916 more rows</span></span>
<span id="cb92-20"><a href="#cb92-20" tabindex="-1"></a><span class="do">## # ℹ abbreviated names: ¹​treat__logFC___dextrt...dexuntrt,</span></span>
<span id="cb92-21"><a href="#cb92-21" tabindex="-1"></a><span class="do">## #   ²​treat__unshrunk.logFC___dextrt...dexuntrt,</span></span>
<span id="cb92-22"><a href="#cb92-22" tabindex="-1"></a><span class="do">## #   ³​treat__logCPM___dextrt...dexuntrt</span></span>
<span id="cb92-23"><a href="#cb92-23" tabindex="-1"></a><span class="do">## # ℹ 2 more variables: treat__PValue___dextrt...dexuntrt &lt;dbl&gt;,</span></span>
<span id="cb92-24"><a href="#cb92-24" tabindex="-1"></a><span class="do">## #   treat__FDR___dextrt...dexuntrt &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="mixed-models-for-complex-designs">Mixed Models for Complex Designs<a class="anchor" aria-label="anchor" href="#mixed-models-for-complex-designs"></a>
</h3>
<p>glmmSeq (<a href="#ref-ma2020glmmseq">Ma et al. 2020</a>) is used for generalized linear mixed models for RNA-seq data.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using glmmSeq for mixed models</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span>formula_design <span class="op">=</span> <span class="op">~</span> <span class="va">dex</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Select 100 genes in the interest of execution time</span></span>
<span>  <span class="va">_</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span> <span class="op">|&gt;</span></span>
<span></span>
<span>  <span class="co"># Fit model</span></span>
<span>  <span class="fu"><a href="reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">cell</span><span class="op">)</span>, </span>
<span>    method <span class="op">=</span> <span class="st">"glmmseq_lme4"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    prefix <span class="op">=</span> <span class="st">"glmmseq__"</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<div class="sourceCode" id="cb94"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb94-2"><a href="#cb94-2" tabindex="-1"></a><span class="do">## n = 8 samples, 4 individuals</span></span>
<span id="cb94-3"><a href="#cb94-3" tabindex="-1"></a></span>
<span id="cb94-4"><a href="#cb94-4" tabindex="-1"></a><span class="do">## Time difference of 33.12868 secs</span></span>
<span id="cb94-5"><a href="#cb94-5" tabindex="-1"></a></span>
<span id="cb94-6"><a href="#cb94-6" tabindex="-1"></a><span class="do">## tidybulk says: to access the DE object do</span></span>
<span id="cb94-7"><a href="#cb94-7" tabindex="-1"></a><span class="do">## `metadata(.)$tidybulk$glmmseq_lme4_object`</span></span>
<span id="cb94-8"><a href="#cb94-8" tabindex="-1"></a></span>
<span id="cb94-9"><a href="#cb94-9" tabindex="-1"></a><span class="do">## tidybulk says: to access the raw results (fitted GLM) do</span></span>
<span id="cb94-10"><a href="#cb94-10" tabindex="-1"></a><span class="do">## `metadata(.)$tidybulk$glmmseq_lme4_fit`</span></span>
<span id="cb94-11"><a href="#cb94-11" tabindex="-1"></a></span>
<span id="cb94-12"><a href="#cb94-12" tabindex="-1"></a><span class="do">## class: RangedSummarizedExperiment </span></span>
<span id="cb94-13"><a href="#cb94-13" tabindex="-1"></a><span class="do">## dim: 100 8 </span></span>
<span id="cb94-14"><a href="#cb94-14" tabindex="-1"></a><span class="do">## metadata(2): '' tidybulk</span></span>
<span id="cb94-15"><a href="#cb94-15" tabindex="-1"></a><span class="do">## assays(4): counts counts_tmm counts_upperquartile counts_RLE</span></span>
<span id="cb94-16"><a href="#cb94-16" tabindex="-1"></a><span class="do">## rownames(100): ENSG00000000003 ENSG00000000419 ... ENSG00000006114</span></span>
<span id="cb94-17"><a href="#cb94-17" tabindex="-1"></a><span class="do">##   ENSG00000006118</span></span>
<span id="cb94-18"><a href="#cb94-18" tabindex="-1"></a><span class="do">## rowData names(63): gene_id gene_name ... glmmseq__P_dex</span></span>
<span id="cb94-19"><a href="#cb94-19" tabindex="-1"></a><span class="do">##   glmmseq__P_dex_adjusted</span></span>
<span id="cb94-20"><a href="#cb94-20" tabindex="-1"></a><span class="do">## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></span>
<span id="cb94-21"><a href="#cb94-21" tabindex="-1"></a><span class="do">## colData names(16): SampleName cell ... PC2 cluster_kmeans</span></span></code></pre></div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<div class="sourceCode" id="cb96"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" tabindex="-1"></a><span class="do">## # A tibble: 15,926 × 41</span></span>
<span id="cb96-2"><a href="#cb96-2" tabindex="-1"></a><span class="do">##    .feature  gene_id gene_name entrezid gene_biotype gene_seq_start gene_seq_end</span></span>
<span id="cb96-3"><a href="#cb96-3" tabindex="-1"></a><span class="do">##    &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;                 &lt;int&gt;        &lt;int&gt;</span></span>
<span id="cb96-4"><a href="#cb96-4" tabindex="-1"></a><span class="do">##  1 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…       99883667     99894988</span></span>
<span id="cb96-5"><a href="#cb96-5" tabindex="-1"></a><span class="do">##  2 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…       49551404     49575092</span></span>
<span id="cb96-6"><a href="#cb96-6" tabindex="-1"></a><span class="do">##  3 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…      169818772    169863408</span></span>
<span id="cb96-7"><a href="#cb96-7" tabindex="-1"></a><span class="do">##  4 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…      169631245    169823221</span></span>
<span id="cb96-8"><a href="#cb96-8" tabindex="-1"></a><span class="do">##  5 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…      196621008    196716634</span></span>
<span id="cb96-9"><a href="#cb96-9" tabindex="-1"></a><span class="do">##  6 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…      143815948    143832827</span></span>
<span id="cb96-10"><a href="#cb96-10" tabindex="-1"></a><span class="do">##  7 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…       53362139     53481768</span></span>
<span id="cb96-11"><a href="#cb96-11" tabindex="-1"></a><span class="do">##  8 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…       41040684     41067715</span></span>
<span id="cb96-12"><a href="#cb96-12" tabindex="-1"></a><span class="do">##  9 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…       24683489     24743424</span></span>
<span id="cb96-13"><a href="#cb96-13" tabindex="-1"></a><span class="do">## 10 ENSG0000… ENSG00… ENSG0000…       NA protein_cod…       24742284     24799466</span></span>
<span id="cb96-14"><a href="#cb96-14" tabindex="-1"></a><span class="do">## # ℹ 15,916 more rows</span></span>
<span id="cb96-15"><a href="#cb96-15" tabindex="-1"></a><span class="do">## # ℹ 34 more variables: seq_name &lt;chr&gt;, seq_strand &lt;int&gt;,</span></span>
<span id="cb96-16"><a href="#cb96-16" tabindex="-1"></a><span class="do">## #   seq_coord_system &lt;int&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, ql__logFC &lt;dbl&gt;,</span></span>
<span id="cb96-17"><a href="#cb96-17" tabindex="-1"></a><span class="do">## #   ql__logCPM &lt;dbl&gt;, ql__F &lt;dbl&gt;, ql__PValue &lt;dbl&gt;, ql__FDR &lt;dbl&gt;,</span></span>
<span id="cb96-18"><a href="#cb96-18" tabindex="-1"></a><span class="do">## #   lr_robust__logFC &lt;dbl&gt;, lr_robust__logCPM &lt;dbl&gt;, lr_robust__LR &lt;dbl&gt;,</span></span>
<span id="cb96-19"><a href="#cb96-19" tabindex="-1"></a><span class="do">## #   lr_robust__PValue &lt;dbl&gt;, lr_robust__FDR &lt;dbl&gt;, deseq2__baseMean &lt;dbl&gt;,</span></span>
<span id="cb96-20"><a href="#cb96-20" tabindex="-1"></a><span class="do">## #   deseq2__log2FoldChange &lt;dbl&gt;, deseq2__lfcSE &lt;dbl&gt;, deseq2__stat &lt;dbl&gt;, …</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gene-description">Gene Description<a class="anchor" aria-label="anchor" href="#gene-description"></a>
</h3>
<p>With tidybulk, retrieving gene descriptions is straightforward, making it easy to enhance the interpretability of your differential expression results.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add gene descriptions using the original SummarizedExperiment</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="fu"><a href="reference/describe_transcript-methods.html">describe_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Filter top significant genes</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ql__FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Print the gene statistics</span></span>
<span>    <span class="fu"><a href="reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.feature</span>, <span class="va">description</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"ql"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb98"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb98-2"><a href="#cb98-2" tabindex="-1"></a></span>
<span id="cb98-3"><a href="#cb98-3" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb98-4"><a href="#cb98-4" tabindex="-1"></a></span>
<span id="cb98-5"><a href="#cb98-5" tabindex="-1"></a><span class="do">## # A tibble: 6 × 7</span></span>
<span id="cb98-6"><a href="#cb98-6" tabindex="-1"></a><span class="do">##   .feature        description ql__logFC ql__logCPM ql__F ql__PValue ql__FDR</span></span>
<span id="cb98-7"><a href="#cb98-7" tabindex="-1"></a><span class="do">##   &lt;chr&gt;           &lt;lgl&gt;           &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span id="cb98-8"><a href="#cb98-8" tabindex="-1"></a><span class="do">## 1 ENSG00000002834 NA             -0.389      8.39   16.3  0.00301   0.0319 </span></span>
<span id="cb98-9"><a href="#cb98-9" tabindex="-1"></a><span class="do">## 2 ENSG00000003096 NA              0.929      4.16   31.2  0.000358  0.00883</span></span>
<span id="cb98-10"><a href="#cb98-10" tabindex="-1"></a><span class="do">## 3 ENSG00000003402 NA             -1.17       6.90   65.0  0.0000224 0.00226</span></span>
<span id="cb98-11"><a href="#cb98-11" tabindex="-1"></a><span class="do">## 4 ENSG00000003987 NA             -0.967      0.347  17.3  0.00255   0.0290 </span></span>
<span id="cb98-12"><a href="#cb98-12" tabindex="-1"></a><span class="do">## 5 ENSG00000004059 NA             -0.359      5.84   17.1  0.00260   0.0292 </span></span>
<span id="cb98-13"><a href="#cb98-13" tabindex="-1"></a><span class="do">## 6 ENSG00000004487 NA              0.309      5.86   19.8  0.00164   0.0222</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-4-batch-effect-correction">Step 4: Batch Effect Correction<a class="anchor" aria-label="anchor" href="#step-4-batch-effect-correction"></a>
</h2>
<p>ComBat-seq (<a href="#ref-zhang2017combatseq">Zhang, Parmigiani, and Johnson 2020</a>) is used for batch effect correction in RNA-seq data.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Adjust for batch effects</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/adjust_abundance-methods.html">adjust_abundance</a></span><span class="op">(</span></span>
<span>      .factor_unwanted <span class="op">=</span> <span class="va">cell</span>, </span>
<span>      .factor_of_interest <span class="op">=</span> <span class="va">dex</span>, </span>
<span>    method <span class="op">=</span> <span class="st">"combat_seq"</span>, </span>
<span>      abundance <span class="op">=</span> <span class="st">"counts_tmm"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb100"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" tabindex="-1"></a><span class="do">## Found 4 batches</span></span>
<span id="cb100-2"><a href="#cb100-2" tabindex="-1"></a><span class="do">## Using null model in ComBat-seq.</span></span>
<span id="cb100-3"><a href="#cb100-3" tabindex="-1"></a><span class="do">## Adjusting for 1 covariate(s) or covariate level(s)</span></span>
<span id="cb100-4"><a href="#cb100-4" tabindex="-1"></a><span class="do">## Estimating dispersions</span></span>
<span id="cb100-5"><a href="#cb100-5" tabindex="-1"></a><span class="do">## Fitting the GLM model</span></span>
<span id="cb100-6"><a href="#cb100-6" tabindex="-1"></a><span class="do">## Shrinkage off - using GLM estimates for parameters</span></span>
<span id="cb100-7"><a href="#cb100-7" tabindex="-1"></a><span class="do">## Adjusting the data</span></span></code></pre></div>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Scatter plot of adjusted vs unadjusted</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Subset genes to speed up plotting</span></span>
<span>  <span class="va">_</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">.sample</span>, <span class="va">counts_tmm</span>, <span class="va">counts_tmm_adjusted</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">counts_tmm</span> <span class="op">+</span> <span class="fl">1</span>, y <span class="op">=</span> <span class="va">counts_tmm_adjusted</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">.sample</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">symbol</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scatter plot of adjusted vs unadjusted"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb102"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" tabindex="-1"></a><span class="do">## tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span></span>
<span id="cb102-2"><a href="#cb102-2" tabindex="-1"></a></span>
<span id="cb102-3"><a href="#cb102-3" tabindex="-1"></a><span class="do">## Warning: ggrepel: 784 unlabeled data points (too many overlaps). Consider</span></span>
<span id="cb102-4"><a href="#cb102-4" tabindex="-1"></a><span class="do">## increasing max.overlaps</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/batch-correction-adjust-abundance-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="step-5-gene-enrichment-analysis">Step 5: Gene Enrichment Analysis<a class="anchor" aria-label="anchor" href="#step-5-gene-enrichment-analysis"></a>
</h2>
<p>Gene Set Enrichment Analysis (GSEA) (<a href="#ref-subramanian2005gsea">Subramanian et al. 2005</a>) is used for gene set enrichment.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run gene rank enrichment (GSEA style)</span></span>
<span><span class="va">gene_rank_res</span> <span class="op">=</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Filter for genes with entrez IDs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">entrezid</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>.transcript <span class="op">=</span> <span class="va">entrezid</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>  <span class="co"># Test gene rank</span></span>
<span>  <span class="fu"><a href="reference/test_gene_rank-methods.html">test_gene_rank</a></span><span class="op">(</span></span>
<span>    .entrez <span class="op">=</span> <span class="va">entrezid</span>,</span>
<span>    .arrange_desc <span class="op">=</span> <span class="va">lr_robust__logFC</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>    gene_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>, <span class="st">"C2"</span>, <span class="st">"C5"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb104"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb104-2"><a href="#cb104-2" tabindex="-1"></a></span>
<span id="cb104-3"><a href="#cb104-3" tabindex="-1"></a><span class="do">## using 'fgsea' for GSEA analysis, please cite Korotkevich et al (2019).</span></span>
<span id="cb104-4"><a href="#cb104-4" tabindex="-1"></a></span>
<span id="cb104-5"><a href="#cb104-5" tabindex="-1"></a><span class="do">## preparing geneSet collections...</span></span>
<span id="cb104-6"><a href="#cb104-6" tabindex="-1"></a></span>
<span id="cb104-7"><a href="#cb104-7" tabindex="-1"></a><span class="do">## GSEA analysis...</span></span>
<span id="cb104-8"><a href="#cb104-8" tabindex="-1"></a></span>
<span id="cb104-9"><a href="#cb104-9" tabindex="-1"></a><span class="do">## leading edge analysis...</span></span>
<span id="cb104-10"><a href="#cb104-10" tabindex="-1"></a></span>
<span id="cb104-11"><a href="#cb104-11" tabindex="-1"></a><span class="do">## done...</span></span>
<span id="cb104-12"><a href="#cb104-12" tabindex="-1"></a></span>
<span id="cb104-13"><a href="#cb104-13" tabindex="-1"></a><span class="do">## using 'fgsea' for GSEA analysis, please cite Korotkevich et al (2019).</span></span>
<span id="cb104-14"><a href="#cb104-14" tabindex="-1"></a></span>
<span id="cb104-15"><a href="#cb104-15" tabindex="-1"></a><span class="do">## preparing geneSet collections...</span></span>
<span id="cb104-16"><a href="#cb104-16" tabindex="-1"></a></span>
<span id="cb104-17"><a href="#cb104-17" tabindex="-1"></a><span class="do">## GSEA analysis...</span></span>
<span id="cb104-18"><a href="#cb104-18" tabindex="-1"></a></span>
<span id="cb104-19"><a href="#cb104-19" tabindex="-1"></a><span class="do">## leading edge analysis...</span></span>
<span id="cb104-20"><a href="#cb104-20" tabindex="-1"></a></span>
<span id="cb104-21"><a href="#cb104-21" tabindex="-1"></a><span class="do">## done...</span></span>
<span id="cb104-22"><a href="#cb104-22" tabindex="-1"></a></span>
<span id="cb104-23"><a href="#cb104-23" tabindex="-1"></a><span class="do">## using 'fgsea' for GSEA analysis, please cite Korotkevich et al (2019).</span></span>
<span id="cb104-24"><a href="#cb104-24" tabindex="-1"></a></span>
<span id="cb104-25"><a href="#cb104-25" tabindex="-1"></a><span class="do">## preparing geneSet collections...</span></span>
<span id="cb104-26"><a href="#cb104-26" tabindex="-1"></a></span>
<span id="cb104-27"><a href="#cb104-27" tabindex="-1"></a><span class="do">## GSEA analysis...</span></span>
<span id="cb104-28"><a href="#cb104-28" tabindex="-1"></a></span>
<span id="cb104-29"><a href="#cb104-29" tabindex="-1"></a><span class="do">## leading edge analysis...</span></span>
<span id="cb104-30"><a href="#cb104-30" tabindex="-1"></a></span>
<span id="cb104-31"><a href="#cb104-31" tabindex="-1"></a><span class="do">## done...</span></span>
<span id="cb104-32"><a href="#cb104-32" tabindex="-1"></a></span>
<span id="cb104-33"><a href="#cb104-33" tabindex="-1"></a><span class="do">## Warning: There were 2 warnings in `mutate()`.</span></span>
<span id="cb104-34"><a href="#cb104-34" tabindex="-1"></a><span class="do">## The first warning was:</span></span>
<span id="cb104-35"><a href="#cb104-35" tabindex="-1"></a><span class="do">## ℹ In argument: `fit = map(...)`.</span></span>
<span id="cb104-36"><a href="#cb104-36" tabindex="-1"></a><span class="do">## Caused by warning in `fgseaMultilevel()`:</span></span>
<span id="cb104-37"><a href="#cb104-37" tabindex="-1"></a><span class="do">## ! For some of the pathways the P-values were likely overestimated. For such pathways log2err is set to NA.</span></span>
<span id="cb104-38"><a href="#cb104-38" tabindex="-1"></a><span class="do">## ℹ Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.</span></span></code></pre></div>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect significant gene sets (example for C2 collection)</span></span>
<span><span class="va">gene_rank_res</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gs_collection</span> <span class="op">==</span> <span class="st">"C2"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">fit</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb106"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" tabindex="-1"></a><span class="do">## # A tibble: 143 × 13</span></span>
<span id="cb106-2"><a href="#cb106-2" tabindex="-1"></a><span class="do">##    gs_collection idx_for_plotting ID         Description setSize enrichmentScore</span></span>
<span id="cb106-3"><a href="#cb106-3" tabindex="-1"></a><span class="do">##    &lt;chr&gt;                    &lt;int&gt; &lt;chr&gt;      &lt;chr&gt;         &lt;int&gt;           &lt;dbl&gt;</span></span>
<span id="cb106-4"><a href="#cb106-4" tabindex="-1"></a><span class="do">##  1 C2                           1 CHEN_LVAD… CHEN_LVAD_…      87          -0.738</span></span>
<span id="cb106-5"><a href="#cb106-5" tabindex="-1"></a><span class="do">##  2 C2                           2 VECCHI_GA… VECCHI_GAS…     221          -0.552</span></span>
<span id="cb106-6"><a href="#cb106-6" tabindex="-1"></a><span class="do">##  3 C2                           3 FEKIR_HEP… FEKIR_HEPA…      32          -0.825</span></span>
<span id="cb106-7"><a href="#cb106-7" tabindex="-1"></a><span class="do">##  4 C2                           4 BOQUEST_S… BOQUEST_ST…     389          -0.437</span></span>
<span id="cb106-8"><a href="#cb106-8" tabindex="-1"></a><span class="do">##  5 C2                           5 SHEDDEN_L… SHEDDEN_LU…     157          -0.552</span></span>
<span id="cb106-9"><a href="#cb106-9" tabindex="-1"></a><span class="do">##  6 C2                           6 CHARAFE_B… CHARAFE_BR…     405          -0.424</span></span>
<span id="cb106-10"><a href="#cb106-10" tabindex="-1"></a><span class="do">##  7 C2                           7 WP_OVERVI… WP_OVERVIE…      29           0.778</span></span>
<span id="cb106-11"><a href="#cb106-11" tabindex="-1"></a><span class="do">##  8 C2                           8 HOSHIDA_L… HOSHIDA_LI…     194          -0.489</span></span>
<span id="cb106-12"><a href="#cb106-12" tabindex="-1"></a><span class="do">##  9 C2                           9 BLANCO_ME… BLANCO_MEL…     145           0.505</span></span>
<span id="cb106-13"><a href="#cb106-13" tabindex="-1"></a><span class="do">## 10 C2                          10 ZWANG_CLA… ZWANG_CLAS…     189          -0.478</span></span>
<span id="cb106-14"><a href="#cb106-14" tabindex="-1"></a><span class="do">## # ℹ 133 more rows</span></span>
<span id="cb106-15"><a href="#cb106-15" tabindex="-1"></a><span class="do">## # ℹ 7 more variables: NES &lt;dbl&gt;, pvalue &lt;dbl&gt;, p.adjust &lt;dbl&gt;, qvalue &lt;dbl&gt;,</span></span>
<span id="cb106-16"><a href="#cb106-16" tabindex="-1"></a><span class="do">## #   rank &lt;dbl&gt;, leading_edge &lt;chr&gt;, core_enrichment &lt;chr&gt;</span></span></code></pre></div>
<div class="section level4">
<h4 id="visualize-enrichment">Visualize enrichment<a class="anchor" aria-label="anchor" href="#visualize-enrichment"></a>
</h4>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/contribution-knowledge-mining/" class="external-link">enrichplot</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb108"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" tabindex="-1"></a><span class="do">## Warning: package 'enrichplot' was built under R version 4.5.1</span></span>
<span id="cb108-2"><a href="#cb108-2" tabindex="-1"></a></span>
<span id="cb108-3"><a href="#cb108-3" tabindex="-1"></a><span class="do">## enrichplot v1.28.4 Learn more at https://yulab-smu.top/contribution-knowledge-mining/</span></span>
<span id="cb108-4"><a href="#cb108-4" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb108-5"><a href="#cb108-5" tabindex="-1"></a><span class="do">## Please cite:</span></span>
<span id="cb108-6"><a href="#cb108-6" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb108-7"><a href="#cb108-7" tabindex="-1"></a><span class="do">## T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan,</span></span>
<span id="cb108-8"><a href="#cb108-8" tabindex="-1"></a><span class="do">## X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal</span></span>
<span id="cb108-9"><a href="#cb108-9" tabindex="-1"></a><span class="do">## enrichment tool for interpreting omics data. The Innovation. 2021,</span></span>
<span id="cb108-10"><a href="#cb108-10" tabindex="-1"></a><span class="do">## 2(3):100141</span></span>
<span id="cb108-11"><a href="#cb108-11" tabindex="-1"></a></span>
<span id="cb108-12"><a href="#cb108-12" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb108-13"><a href="#cb108-13" tabindex="-1"></a><span class="do">## Attaching package: 'enrichplot'</span></span>
<span id="cb108-14"><a href="#cb108-14" tabindex="-1"></a></span>
<span id="cb108-15"><a href="#cb108-15" tabindex="-1"></a><span class="do">## The following object is masked from 'package:GGally':</span></span>
<span id="cb108-16"><a href="#cb108-16" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb108-17"><a href="#cb108-17" tabindex="-1"></a><span class="do">##     ggtable</span></span></code></pre></div>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span>  <span class="va">gene_rank_res</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">unnest</span><span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>plot <span class="op">=</span> <span class="fu">pmap</span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">ID</span>, <span class="va">idx_for_plotting</span>, <span class="va">p.adjust</span><span class="op">)</span>,</span>
<span>      <span class="op">~</span> <span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/gseaplot2.html" class="external-link">gseaplot2</a></span><span class="op">(</span></span>
<span>        <span class="va">..1</span>, geneSetID <span class="op">=</span> <span class="va">..3</span>,</span>
<span>        title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s \nadj pvalue %s"</span>, <span class="va">..2</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">..4</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        base_size <span class="op">=</span> <span class="fl">6</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">0.5</span><span class="op">)</span>, subplots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">plot</span><span class="op">)</span> </span></code></pre></div>
<div class="sourceCode" id="cb110"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" tabindex="-1"></a><span class="do">## [[1]]</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/enrichment-visualize-gsea-plots-1.png"><!-- --></p>
<div class="sourceCode" id="cb111"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb111-2"><a href="#cb111-2" tabindex="-1"></a><span class="do">## [[2]]</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/enrichment-visualize-gsea-plots-2.png"><!-- --></p>
<div class="sourceCode" id="cb112"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb112-2"><a href="#cb112-2" tabindex="-1"></a><span class="do">## [[3]]</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/enrichment-visualize-gsea-plots-3.png"><!-- --></p>
<div class="sourceCode" id="cb113"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb113-2"><a href="#cb113-2" tabindex="-1"></a><span class="do">## [[4]]</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/enrichment-visualize-gsea-plots-4.png"><!-- --></p>
<div class="sourceCode" id="cb114"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb114-2"><a href="#cb114-2" tabindex="-1"></a><span class="do">## [[5]]</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/enrichment-visualize-gsea-plots-5.png"><!-- --></p>
<div class="sourceCode" id="cb115"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb115-2"><a href="#cb115-2" tabindex="-1"></a><span class="do">## [[6]]</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/enrichment-visualize-gsea-plots-6.png"><!-- --></p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:enrichplot"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb117"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" tabindex="-1"></a><span class="do">## Warning: 'enrichplot' namespace cannot be unloaded:</span></span>
<span id="cb117-2"><a href="#cb117-2" tabindex="-1"></a><span class="do">##   namespace 'enrichplot' is imported by 'clusterProfiler' so cannot be unloaded</span></span></code></pre></div>
<p>Gene Ontology overrepresentation analysis (<a href="#ref-ashburner2000go">Ashburner et al. 2000</a>) is used for functional enrichment.</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test gene overrepresentation</span></span>
<span><span class="va">airway_overrep</span> <span class="op">=</span> </span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Label genes to test overrepresentation of</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>genes_to_test <span class="op">=</span> <span class="va">ql__FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>    <span class="co"># Filter for genes with entrez IDs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">entrezid</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="fu"><a href="reference/test_gene_overrepresentation-methods.html">test_gene_overrepresentation</a></span><span class="op">(</span></span>
<span>    .entrez <span class="op">=</span> <span class="va">entrezid</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>    .do_test <span class="op">=</span> <span class="va">genes_to_test</span>,</span>
<span>    gene_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>, <span class="st">"C2"</span>, <span class="st">"C5"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway_overrep</span></span></code></pre></div>
<div class="sourceCode" id="cb119"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" tabindex="-1"></a><span class="do">## # A tibble: 5,291 × 13</span></span>
<span id="cb119-2"><a href="#cb119-2" tabindex="-1"></a><span class="do">##    gs_collection ID      Description GeneRatio BgRatio RichFactor FoldEnrichment</span></span>
<span id="cb119-3"><a href="#cb119-3" tabindex="-1"></a><span class="do">##    &lt;chr&gt;         &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;          &lt;dbl&gt;</span></span>
<span id="cb119-4"><a href="#cb119-4" tabindex="-1"></a><span class="do">##  1 C2            PASINI… PASINI_SUZ… 88/1848   316/22…      0.278           3.37</span></span>
<span id="cb119-5"><a href="#cb119-5" tabindex="-1"></a><span class="do">##  2 C2            CHARAF… CHARAFE_BR… 108/1848  465/22…      0.232           2.81</span></span>
<span id="cb119-6"><a href="#cb119-6" tabindex="-1"></a><span class="do">##  3 C2            REN_AL… REN_ALVEOL… 98/1848   407/22…      0.241           2.92</span></span>
<span id="cb119-7"><a href="#cb119-7" tabindex="-1"></a><span class="do">##  4 C2            CHARAF… CHARAFE_BR… 105/1848  456/22…      0.230           2.79</span></span>
<span id="cb119-8"><a href="#cb119-8" tabindex="-1"></a><span class="do">##  5 C2            CHEN_L… CHEN_LVAD_… 45/1848   102/22…      0.441           5.34</span></span>
<span id="cb119-9"><a href="#cb119-9" tabindex="-1"></a><span class="do">##  6 C2            BOQUES… BOQUEST_ST… 96/1848   428/22…      0.224           2.72</span></span>
<span id="cb119-10"><a href="#cb119-10" tabindex="-1"></a><span class="do">##  7 C2            LIM_MA… LIM_MAMMAR… 103/1848  479/22…      0.215           2.60</span></span>
<span id="cb119-11"><a href="#cb119-11" tabindex="-1"></a><span class="do">##  8 C2            LIU_PR… LIU_PROSTA… 104/1848  495/22…      0.210           2.54</span></span>
<span id="cb119-12"><a href="#cb119-12" tabindex="-1"></a><span class="do">##  9 C2            LU_AGI… LU_AGING_B… 70/1848   264/22…      0.265           3.21</span></span>
<span id="cb119-13"><a href="#cb119-13" tabindex="-1"></a><span class="do">## 10 C2            ONDER_… ONDER_CDH1… 67/1848   257/22…      0.261           3.16</span></span>
<span id="cb119-14"><a href="#cb119-14" tabindex="-1"></a><span class="do">## # ℹ 5,281 more rows</span></span>
<span id="cb119-15"><a href="#cb119-15" tabindex="-1"></a><span class="do">## # ℹ 6 more variables: zScore &lt;dbl&gt;, pvalue &lt;dbl&gt;, p.adjust &lt;dbl&gt;, qvalue &lt;dbl&gt;,</span></span>
<span id="cb119-16"><a href="#cb119-16" tabindex="-1"></a><span class="do">## #   Count &lt;int&gt;, entrez &lt;list&gt;</span></span></code></pre></div>
<p>EGSEA (<a href="#ref-alhamdoosh2017egsea">Alhamdoosh et al. 2017</a>) is used for ensemble gene set enrichment analysis. EGSEA is a method that combines multiple gene set enrichment analysis methods to provide a more robust and comprehensive analysis of gene set enrichment. It creates a web-based interactive tool that allows you to explore the results of the gene set enrichment analysis.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EGSEA</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb121"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" tabindex="-1"></a><span class="do">## Loading required package: gage</span></span>
<span id="cb121-2"><a href="#cb121-2" tabindex="-1"></a></span>
<span id="cb121-3"><a href="#cb121-3" tabindex="-1"></a><span class="do">## Loading required package: AnnotationDbi</span></span>
<span id="cb121-4"><a href="#cb121-4" tabindex="-1"></a></span>
<span id="cb121-5"><a href="#cb121-5" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-6"><a href="#cb121-6" tabindex="-1"></a><span class="do">## Attaching package: 'AnnotationDbi'</span></span>
<span id="cb121-7"><a href="#cb121-7" tabindex="-1"></a></span>
<span id="cb121-8"><a href="#cb121-8" tabindex="-1"></a><span class="do">## The following object is masked from 'package:dplyr':</span></span>
<span id="cb121-9"><a href="#cb121-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-10"><a href="#cb121-10" tabindex="-1"></a><span class="do">##     select</span></span>
<span id="cb121-11"><a href="#cb121-11" tabindex="-1"></a></span>
<span id="cb121-12"><a href="#cb121-12" tabindex="-1"></a><span class="do">## Loading required package: topGO</span></span>
<span id="cb121-13"><a href="#cb121-13" tabindex="-1"></a></span>
<span id="cb121-14"><a href="#cb121-14" tabindex="-1"></a><span class="do">## Warning: package 'topGO' was built under R version 4.5.1</span></span>
<span id="cb121-15"><a href="#cb121-15" tabindex="-1"></a></span>
<span id="cb121-16"><a href="#cb121-16" tabindex="-1"></a><span class="do">## Loading required package: graph</span></span>
<span id="cb121-17"><a href="#cb121-17" tabindex="-1"></a></span>
<span id="cb121-18"><a href="#cb121-18" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-19"><a href="#cb121-19" tabindex="-1"></a><span class="do">## Attaching package: 'graph'</span></span>
<span id="cb121-20"><a href="#cb121-20" tabindex="-1"></a></span>
<span id="cb121-21"><a href="#cb121-21" tabindex="-1"></a><span class="do">## The following object is masked from 'package:stringr':</span></span>
<span id="cb121-22"><a href="#cb121-22" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-23"><a href="#cb121-23" tabindex="-1"></a><span class="do">##     boundary</span></span>
<span id="cb121-24"><a href="#cb121-24" tabindex="-1"></a></span>
<span id="cb121-25"><a href="#cb121-25" tabindex="-1"></a><span class="do">## Loading required package: GO.db</span></span>
<span id="cb121-26"><a href="#cb121-26" tabindex="-1"></a></span>
<span id="cb121-27"><a href="#cb121-27" tabindex="-1"></a><span class="do">## Loading required package: SparseM</span></span>
<span id="cb121-28"><a href="#cb121-28" tabindex="-1"></a></span>
<span id="cb121-29"><a href="#cb121-29" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-30"><a href="#cb121-30" tabindex="-1"></a><span class="do">## groupGOTerms:    GOBPTerm, GOMFTerm, GOCCTerm environments built.</span></span>
<span id="cb121-31"><a href="#cb121-31" tabindex="-1"></a></span>
<span id="cb121-32"><a href="#cb121-32" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-33"><a href="#cb121-33" tabindex="-1"></a><span class="do">## Attaching package: 'topGO'</span></span>
<span id="cb121-34"><a href="#cb121-34" tabindex="-1"></a></span>
<span id="cb121-35"><a href="#cb121-35" tabindex="-1"></a><span class="do">## The following object is masked from 'package:gage':</span></span>
<span id="cb121-36"><a href="#cb121-36" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-37"><a href="#cb121-37" tabindex="-1"></a><span class="do">##     geneData</span></span>
<span id="cb121-38"><a href="#cb121-38" tabindex="-1"></a></span>
<span id="cb121-39"><a href="#cb121-39" tabindex="-1"></a><span class="do">## The following object is masked from 'package:IRanges':</span></span>
<span id="cb121-40"><a href="#cb121-40" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-41"><a href="#cb121-41" tabindex="-1"></a><span class="do">##     members</span></span>
<span id="cb121-42"><a href="#cb121-42" tabindex="-1"></a></span>
<span id="cb121-43"><a href="#cb121-43" tabindex="-1"></a><span class="do">## Loading required package: pathview</span></span>
<span id="cb121-44"><a href="#cb121-44" tabindex="-1"></a></span>
<span id="cb121-45"><a href="#cb121-45" tabindex="-1"></a><span class="do">## ##############################################################################</span></span>
<span id="cb121-46"><a href="#cb121-46" tabindex="-1"></a><span class="do">## Pathview is an open source software package distributed under GNU General</span></span>
<span id="cb121-47"><a href="#cb121-47" tabindex="-1"></a><span class="do">## Public License version 3 (GPLv3). Details of GPLv3 is available at</span></span>
<span id="cb121-48"><a href="#cb121-48" tabindex="-1"></a><span class="do">## http://www.gnu.org/licenses/gpl-3.0.html. Particullary, users are required to</span></span>
<span id="cb121-49"><a href="#cb121-49" tabindex="-1"></a><span class="do">## formally cite the original Pathview paper (not just mention it) in publications</span></span>
<span id="cb121-50"><a href="#cb121-50" tabindex="-1"></a><span class="do">## or products. For details, do citation("pathview") within R.</span></span>
<span id="cb121-51"><a href="#cb121-51" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-52"><a href="#cb121-52" tabindex="-1"></a><span class="do">## The pathview downloads and uses KEGG data. Non-academic uses may require a KEGG</span></span>
<span id="cb121-53"><a href="#cb121-53" tabindex="-1"></a><span class="do">## license agreement (details at http://www.kegg.jp/kegg/legal.html).</span></span>
<span id="cb121-54"><a href="#cb121-54" tabindex="-1"></a><span class="do">## ##############################################################################</span></span>
<span id="cb121-55"><a href="#cb121-55" tabindex="-1"></a></span>
<span id="cb121-56"><a href="#cb121-56" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-57"><a href="#cb121-57" tabindex="-1"></a></span>
<span id="cb121-58"><a href="#cb121-58" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb121-59"><a href="#cb121-59" tabindex="-1"></a></span>
<span id="cb121-60"><a href="#cb121-60" tabindex="-1"></a><span class="do">## </span></span></code></pre></div>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test gene enrichment</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>  <span class="co"># Filter for genes with entrez IDs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">entrezid</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>.transcript <span class="op">=</span> <span class="va">entrezid</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>  <span class="co"># Test gene enrichment</span></span>
<span>  <span class="fu"><a href="reference/test_gene_enrichment-methods.html">test_gene_enrichment</a></span><span class="op">(</span></span>
<span>    .formula <span class="op">=</span> <span class="op">~</span><span class="va">dex</span>,</span>
<span>    .entrez <span class="op">=</span> <span class="va">entrezid</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"human"</span>, </span>
<span>    gene_sets <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>    methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"roast"</span><span class="op">)</span>,  <span class="co"># Use a more robust method</span></span>
<span>    cores <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb123"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" tabindex="-1"></a><span class="do">## tidybulk says: The design column names are "(Intercept), dexuntrt"</span></span>
<span id="cb123-2"><a href="#cb123-2" tabindex="-1"></a></span>
<span id="cb123-3"><a href="#cb123-3" tabindex="-1"></a><span class="do">## [1] "Loading MSigDB Gene Sets ... "</span></span>
<span id="cb123-4"><a href="#cb123-4" tabindex="-1"></a><span class="do">## [1] "Loaded gene sets for the collection h ..."</span></span>
<span id="cb123-5"><a href="#cb123-5" tabindex="-1"></a><span class="do">## [1] "Indexed the collection h ..."</span></span>
<span id="cb123-6"><a href="#cb123-6" tabindex="-1"></a><span class="do">## [1] "Created annotation for the collection h ..."</span></span>
<span id="cb123-7"><a href="#cb123-7" tabindex="-1"></a></span>
<span id="cb123-8"><a href="#cb123-8" tabindex="-1"></a><span class="do">## EGSEA analysis has started</span></span>
<span id="cb123-9"><a href="#cb123-9" tabindex="-1"></a></span>
<span id="cb123-10"><a href="#cb123-10" tabindex="-1"></a><span class="do">## ##------ Wed Aug 20 14:41:38 2025 ------##</span></span>
<span id="cb123-11"><a href="#cb123-11" tabindex="-1"></a></span>
<span id="cb123-12"><a href="#cb123-12" tabindex="-1"></a><span class="do">## The argument 'contrast' is recommended to be a matrix object.</span></span>
<span id="cb123-13"><a href="#cb123-13" tabindex="-1"></a><span class="do">## See Vignette or Help.</span></span>
<span id="cb123-14"><a href="#cb123-14" tabindex="-1"></a></span>
<span id="cb123-15"><a href="#cb123-15" tabindex="-1"></a><span class="do">## Log fold changes are estimated using limma package ...</span></span>
<span id="cb123-16"><a href="#cb123-16" tabindex="-1"></a></span>
<span id="cb123-17"><a href="#cb123-17" tabindex="-1"></a><span class="do">## limma DE analysis is carried out ...</span></span>
<span id="cb123-18"><a href="#cb123-18" tabindex="-1"></a></span>
<span id="cb123-19"><a href="#cb123-19" tabindex="-1"></a><span class="do">## EGSEA is running on the provided data and h collection</span></span>
<span id="cb123-20"><a href="#cb123-20" tabindex="-1"></a></span>
<span id="cb123-21"><a href="#cb123-21" tabindex="-1"></a><span class="do">## ..roast*</span></span>
<span id="cb123-22"><a href="#cb123-22" tabindex="-1"></a></span>
<span id="cb123-23"><a href="#cb123-23" tabindex="-1"></a><span class="do">## ##------ Wed Aug 20 14:41:42 2025 ------##</span></span>
<span id="cb123-24"><a href="#cb123-24" tabindex="-1"></a></span>
<span id="cb123-25"><a href="#cb123-25" tabindex="-1"></a><span class="do">## EGSEA analysis took 3.66199999999998 seconds.</span></span>
<span id="cb123-26"><a href="#cb123-26" tabindex="-1"></a><span class="do">## EGSEA analysis has completed</span></span>
<span id="cb123-27"><a href="#cb123-27" tabindex="-1"></a><span class="do">## EGSEA HTML report is being generated ...</span></span>
<span id="cb123-28"><a href="#cb123-28" tabindex="-1"></a></span>
<span id="cb123-29"><a href="#cb123-29" tabindex="-1"></a><span class="do">## ##------ Wed Aug 20 14:41:42 2025 ------##</span></span>
<span id="cb123-30"><a href="#cb123-30" tabindex="-1"></a></span>
<span id="cb123-31"><a href="#cb123-31" tabindex="-1"></a><span class="do">## Report pages and figures are being generated for the h collection ...</span></span>
<span id="cb123-32"><a href="#cb123-32" tabindex="-1"></a><span class="do">##    Heat maps are being generated for top-ranked gene sets </span></span>
<span id="cb123-33"><a href="#cb123-33" tabindex="-1"></a><span class="do">## based on logFC ... </span></span>
<span id="cb123-34"><a href="#cb123-34" tabindex="-1"></a><span class="do">##    Summary plots are being generated ... </span></span>
<span id="cb123-35"><a href="#cb123-35" tabindex="-1"></a><span class="do">##    Comparison summary plots are being generated  ...</span></span>
<span id="cb123-36"><a href="#cb123-36" tabindex="-1"></a></span>
<span id="cb123-37"><a href="#cb123-37" tabindex="-1"></a><span class="do">## ##------ Wed Aug 20 14:42:40 2025 ------##</span></span>
<span id="cb123-38"><a href="#cb123-38" tabindex="-1"></a></span>
<span id="cb123-39"><a href="#cb123-39" tabindex="-1"></a><span class="do">## EGSEA report generation took 58.602 seconds.</span></span>
<span id="cb123-40"><a href="#cb123-40" tabindex="-1"></a><span class="do">## EGSEA report has been generated.</span></span>
<span id="cb123-41"><a href="#cb123-41" tabindex="-1"></a></span>
<span id="cb123-42"><a href="#cb123-42" tabindex="-1"></a><span class="do">## # A tibble: 0 × 0</span></span></code></pre></div>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:EGSEA"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="step-6-cellularity-analysis">Step 6: Cellularity Analysis<a class="anchor" aria-label="anchor" href="#step-6-cellularity-analysis"></a>
</h2>
<p>CIBERSORT (<a href="#ref-newman2015cibersort">Newman et al. 2015</a>) is used for cell type deconvolution.</p>
<p>Cellularity deconvolution is a standard approach for estimating the cellular composition of a sample.</p>
<div class="section level3">
<h3 id="available-deconvolution-methods">Available Deconvolution Methods<a class="anchor" aria-label="anchor" href="#available-deconvolution-methods"></a>
</h3>
<p>The <code>tidybulk</code> package provides several methods for deconvolution:</p>
<ul>
<li>
<strong>CIBERSORT</strong> (<a href="#ref-newman2015cibersort">Newman et al. 2015</a>): Uses support vector regression to deconvolve cell type proportions. Requires the <code>class</code>, <code>e1071</code>, and <code>preprocessCore</code> packages.</li>
<li>
<strong>LLSR</strong> (<a href="#ref-newman2015cibersort">Newman et al. 2015</a>): Linear Least Squares Regression for deconvolution.</li>
<li>
<strong>EPIC</strong> (<a href="#ref-racle2017epic"><strong>racle2017epic?</strong></a>): Uses a reference-based approach to estimate cell fractions.</li>
<li>
<strong>MCP-counter</strong> (<a href="#ref-becht2016mcp"><strong>becht2016mcp?</strong></a>): Quantifies the abundance of immune and stromal cell populations.</li>
<li>
<strong>quanTIseq</strong> (<a href="#ref-finotello2019quantiseq"><strong>finotello2019quantiseq?</strong></a>): A computational framework for inferring the immune contexture of tumors.</li>
<li>
<strong>xCell</strong> (<a href="#ref-aran2017xcell"><strong>aran2017xcell?</strong></a>): Performs cell type enrichment analysis.</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-usage">Example Usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h3>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> </span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">symbol</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"cibersort"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"cibersort__"</span>, feature_column <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span> </span></code></pre></div>
<p>For the rest of the methods, you need to install the <code>immunedeconv</code> package.</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"immunedeconv"</span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"immunedeconv"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> </span>
<span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span><span class="co"># Example using LLSR</span></span>
<span><span class="fu"><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"llsr"</span>, prefix <span class="op">=</span> <span class="st">"llsr__"</span>, feature_column <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using EPIC</span></span>
<span><span class="fu"><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"epic"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"epic__"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using MCP-counter</span></span>
<span><span class="fu"><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"mcp_counter"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"mcp__"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using quanTIseq</span></span>
<span><span class="fu"><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"quantiseq"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"quantiseq__"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using xCell</span></span>
<span><span class="fu"><a href="reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"xcell"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"xcell__"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plotting-results">Plotting Results<a class="anchor" aria-label="anchor" href="#plotting-results"></a>
</h3>
<p>Visualize the cell type proportions as a stacked barplot for each method:</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize CIBERSORT results</span></span>
<span><span class="va">airway</span>   <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"cibersort__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CIBERSORT Cell Type Proportions"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/deconvolution-plotting-1.png"><!-- --></p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="co"># Repeat similar plotting for LLSR, EPIC, MCP-counter, quanTIseq, and xCell</span></span>
<span><span class="va">airway</span>   <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"llsr__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"LLSR Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"epic__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"EPIC Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"mcp__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"MCP-counter Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"quantiseq__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"quanTIseq Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"xcell__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"xCell Cell Type Proportions"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p><code>tidybulk</code> allows you to get the bibliography of all methods used in our workflow.</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get bibliography of all methods used in our workflow</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="reference/get_bibliography-methods.html">get_bibliography</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb131"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" tabindex="-1"></a><span class="do">##  @Article{tidybulk,</span></span>
<span id="cb131-2"><a href="#cb131-2" tabindex="-1"></a><span class="do">##   title = {tidybulk: an R tidy framework for modular transcriptomic data analysis},</span></span>
<span id="cb131-3"><a href="#cb131-3" tabindex="-1"></a><span class="do">##   author = {Stefano Mangiola and Ramyar Molania and Ruining Dong and Maria A. Doyle &amp; Anthony T. Papenfuss},</span></span>
<span id="cb131-4"><a href="#cb131-4" tabindex="-1"></a><span class="do">##   journal = {Genome Biology},</span></span>
<span id="cb131-5"><a href="#cb131-5" tabindex="-1"></a><span class="do">##   year = {2021},</span></span>
<span id="cb131-6"><a href="#cb131-6" tabindex="-1"></a><span class="do">##   volume = {22},</span></span>
<span id="cb131-7"><a href="#cb131-7" tabindex="-1"></a><span class="do">##   number = {42},</span></span>
<span id="cb131-8"><a href="#cb131-8" tabindex="-1"></a><span class="do">##   url = {https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7},</span></span>
<span id="cb131-9"><a href="#cb131-9" tabindex="-1"></a><span class="do">##   }</span></span>
<span id="cb131-10"><a href="#cb131-10" tabindex="-1"></a><span class="do">## @article{wickham2019welcome,</span></span>
<span id="cb131-11"><a href="#cb131-11" tabindex="-1"></a><span class="do">##   title={Welcome to the Tidyverse},</span></span>
<span id="cb131-12"><a href="#cb131-12" tabindex="-1"></a><span class="do">##   author={Wickham, Hadley and Averick, Mara and Bryan, Jennifer and Chang, Winston and McGowan, Lucy D'Agostino and Francois, Romain and Grolemund, Garrett and Hayes, Alex and Henry, Lionel and Hester, Jim and others},</span></span>
<span id="cb131-13"><a href="#cb131-13" tabindex="-1"></a><span class="do">##   journal={Journal of Open Source Software},</span></span>
<span id="cb131-14"><a href="#cb131-14" tabindex="-1"></a><span class="do">##   volume={4},</span></span>
<span id="cb131-15"><a href="#cb131-15" tabindex="-1"></a><span class="do">##   number={43},</span></span>
<span id="cb131-16"><a href="#cb131-16" tabindex="-1"></a><span class="do">##   pages={1686},</span></span>
<span id="cb131-17"><a href="#cb131-17" tabindex="-1"></a><span class="do">##   year={2019}</span></span>
<span id="cb131-18"><a href="#cb131-18" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb131-19"><a href="#cb131-19" tabindex="-1"></a><span class="do">## @article{robinson2010edger,</span></span>
<span id="cb131-20"><a href="#cb131-20" tabindex="-1"></a><span class="do">##   title={edgeR: a Bioconductor package for differential expression analysis of digital gene expression data},</span></span>
<span id="cb131-21"><a href="#cb131-21" tabindex="-1"></a><span class="do">##   author={Robinson, Mark D and McCarthy, Davis J and Smyth, Gordon K},</span></span>
<span id="cb131-22"><a href="#cb131-22" tabindex="-1"></a><span class="do">##   journal={Bioinformatics},</span></span>
<span id="cb131-23"><a href="#cb131-23" tabindex="-1"></a><span class="do">##   volume={26},</span></span>
<span id="cb131-24"><a href="#cb131-24" tabindex="-1"></a><span class="do">##   number={1},</span></span>
<span id="cb131-25"><a href="#cb131-25" tabindex="-1"></a><span class="do">##   pages={139--140},</span></span>
<span id="cb131-26"><a href="#cb131-26" tabindex="-1"></a><span class="do">##   year={2010},</span></span>
<span id="cb131-27"><a href="#cb131-27" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb131-28"><a href="#cb131-28" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb131-29"><a href="#cb131-29" tabindex="-1"></a><span class="do">## @article{robinson2010scaling,</span></span>
<span id="cb131-30"><a href="#cb131-30" tabindex="-1"></a><span class="do">##   title={A scaling normalization method for differential expression analysis of RNA-seq data},</span></span>
<span id="cb131-31"><a href="#cb131-31" tabindex="-1"></a><span class="do">##   author={Robinson, Mark D and Oshlack, Alicia},</span></span>
<span id="cb131-32"><a href="#cb131-32" tabindex="-1"></a><span class="do">##   journal={Genome biology},</span></span>
<span id="cb131-33"><a href="#cb131-33" tabindex="-1"></a><span class="do">##   volume={11},</span></span>
<span id="cb131-34"><a href="#cb131-34" tabindex="-1"></a><span class="do">##   number={3},</span></span>
<span id="cb131-35"><a href="#cb131-35" tabindex="-1"></a><span class="do">##   pages={1--9},</span></span>
<span id="cb131-36"><a href="#cb131-36" tabindex="-1"></a><span class="do">##   year={2010},</span></span>
<span id="cb131-37"><a href="#cb131-37" tabindex="-1"></a><span class="do">##   publisher={BioMed Central}</span></span>
<span id="cb131-38"><a href="#cb131-38" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb131-39"><a href="#cb131-39" tabindex="-1"></a><span class="do">## @incollection{smyth2005limma,</span></span>
<span id="cb131-40"><a href="#cb131-40" tabindex="-1"></a><span class="do">##   title={Limma: linear models for microarray data},</span></span>
<span id="cb131-41"><a href="#cb131-41" tabindex="-1"></a><span class="do">##   author={Smyth, Gordon K},</span></span>
<span id="cb131-42"><a href="#cb131-42" tabindex="-1"></a><span class="do">##   booktitle={Bioinformatics and computational biology solutions using R and Bioconductor},</span></span>
<span id="cb131-43"><a href="#cb131-43" tabindex="-1"></a><span class="do">##   pages={397--420},</span></span>
<span id="cb131-44"><a href="#cb131-44" tabindex="-1"></a><span class="do">##   year={2005},</span></span>
<span id="cb131-45"><a href="#cb131-45" tabindex="-1"></a><span class="do">##   publisher={Springer}</span></span>
<span id="cb131-46"><a href="#cb131-46" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb131-47"><a href="#cb131-47" tabindex="-1"></a><span class="do">## @Manual{,</span></span>
<span id="cb131-48"><a href="#cb131-48" tabindex="-1"></a><span class="do">##     title = {R: A Language and Environment for Statistical Computing},</span></span>
<span id="cb131-49"><a href="#cb131-49" tabindex="-1"></a><span class="do">##     author = {{R Core Team}},</span></span>
<span id="cb131-50"><a href="#cb131-50" tabindex="-1"></a><span class="do">##     organization = {R Foundation for Statistical Computing},</span></span>
<span id="cb131-51"><a href="#cb131-51" tabindex="-1"></a><span class="do">##     address = {Vienna, Austria},</span></span>
<span id="cb131-52"><a href="#cb131-52" tabindex="-1"></a><span class="do">##     year = {2020},</span></span>
<span id="cb131-53"><a href="#cb131-53" tabindex="-1"></a><span class="do">##     url = {https://www.R-project.org/},</span></span>
<span id="cb131-54"><a href="#cb131-54" tabindex="-1"></a><span class="do">##   }</span></span>
<span id="cb131-55"><a href="#cb131-55" tabindex="-1"></a><span class="do">## @article{lund2012detecting,</span></span>
<span id="cb131-56"><a href="#cb131-56" tabindex="-1"></a><span class="do">##   title={Detecting differential expression in RNA-sequence data using quasi-likelihood with shrunken dispersion estimates},</span></span>
<span id="cb131-57"><a href="#cb131-57" tabindex="-1"></a><span class="do">##   author={Lund, Steven P and Nettleton, Dan and McCarthy, Davis J and Smyth, Gordon K},</span></span>
<span id="cb131-58"><a href="#cb131-58" tabindex="-1"></a><span class="do">##   journal={Statistical applications in genetics and molecular biology},</span></span>
<span id="cb131-59"><a href="#cb131-59" tabindex="-1"></a><span class="do">##   volume={11},</span></span>
<span id="cb131-60"><a href="#cb131-60" tabindex="-1"></a><span class="do">##   number={5},</span></span>
<span id="cb131-61"><a href="#cb131-61" tabindex="-1"></a><span class="do">##   year={2012},</span></span>
<span id="cb131-62"><a href="#cb131-62" tabindex="-1"></a><span class="do">##   publisher={De Gruyter}</span></span>
<span id="cb131-63"><a href="#cb131-63" tabindex="-1"></a><span class="do">##     }</span></span>
<span id="cb131-64"><a href="#cb131-64" tabindex="-1"></a><span class="do">## @article{zhou2014robustly,</span></span>
<span id="cb131-65"><a href="#cb131-65" tabindex="-1"></a><span class="do">##   title={Robustly detecting differential expression in RNA sequencing data using observation weights},</span></span>
<span id="cb131-66"><a href="#cb131-66" tabindex="-1"></a><span class="do">##   author={Zhou, Xiaobei and Lindsay, Helen and Robinson, Mark D},</span></span>
<span id="cb131-67"><a href="#cb131-67" tabindex="-1"></a><span class="do">##   journal={Nucleic acids research},</span></span>
<span id="cb131-68"><a href="#cb131-68" tabindex="-1"></a><span class="do">##   volume={42},</span></span>
<span id="cb131-69"><a href="#cb131-69" tabindex="-1"></a><span class="do">##   number={11},</span></span>
<span id="cb131-70"><a href="#cb131-70" tabindex="-1"></a><span class="do">##   pages={e91--e91},</span></span>
<span id="cb131-71"><a href="#cb131-71" tabindex="-1"></a><span class="do">##   year={2014},</span></span>
<span id="cb131-72"><a href="#cb131-72" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb131-73"><a href="#cb131-73" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb131-74"><a href="#cb131-74" tabindex="-1"></a><span class="do">## @article{love2014moderated,</span></span>
<span id="cb131-75"><a href="#cb131-75" tabindex="-1"></a><span class="do">##   title={Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2},</span></span>
<span id="cb131-76"><a href="#cb131-76" tabindex="-1"></a><span class="do">##   author={Love, Michael I and Huber, Wolfgang and Anders, Simon},</span></span>
<span id="cb131-77"><a href="#cb131-77" tabindex="-1"></a><span class="do">##   journal={Genome biology},</span></span>
<span id="cb131-78"><a href="#cb131-78" tabindex="-1"></a><span class="do">##   volume={15},</span></span>
<span id="cb131-79"><a href="#cb131-79" tabindex="-1"></a><span class="do">##   number={12},</span></span>
<span id="cb131-80"><a href="#cb131-80" tabindex="-1"></a><span class="do">##   pages={550},</span></span>
<span id="cb131-81"><a href="#cb131-81" tabindex="-1"></a><span class="do">##   year={2014},</span></span>
<span id="cb131-82"><a href="#cb131-82" tabindex="-1"></a><span class="do">##   publisher={Springer}</span></span>
<span id="cb131-83"><a href="#cb131-83" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb131-84"><a href="#cb131-84" tabindex="-1"></a><span class="do">## @article{law2014voom,</span></span>
<span id="cb131-85"><a href="#cb131-85" tabindex="-1"></a><span class="do">##   title={voom: Precision weights unlock linear model analysis tools for RNA-seq read counts},</span></span>
<span id="cb131-86"><a href="#cb131-86" tabindex="-1"></a><span class="do">##   author={Law, Charity W and Chen, Yunshun and Shi, Wei and Smyth, Gordon K},</span></span>
<span id="cb131-87"><a href="#cb131-87" tabindex="-1"></a><span class="do">##   journal={Genome biology},</span></span>
<span id="cb131-88"><a href="#cb131-88" tabindex="-1"></a><span class="do">##   volume={15},</span></span>
<span id="cb131-89"><a href="#cb131-89" tabindex="-1"></a><span class="do">##   number={2},</span></span>
<span id="cb131-90"><a href="#cb131-90" tabindex="-1"></a><span class="do">##   pages={R29},</span></span>
<span id="cb131-91"><a href="#cb131-91" tabindex="-1"></a><span class="do">##   year={2014},</span></span>
<span id="cb131-92"><a href="#cb131-92" tabindex="-1"></a><span class="do">##   publisher={Springer}</span></span>
<span id="cb131-93"><a href="#cb131-93" tabindex="-1"></a><span class="do">##     }</span></span>
<span id="cb131-94"><a href="#cb131-94" tabindex="-1"></a><span class="do">## @article{liu2015weight,</span></span>
<span id="cb131-95"><a href="#cb131-95" tabindex="-1"></a><span class="do">##   title={Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses},</span></span>
<span id="cb131-96"><a href="#cb131-96" tabindex="-1"></a><span class="do">##   author={Liu, Ruijie and Holik, Aliaksei Z and Su, Shian and Jansz, Natasha and Chen, Kelan and Leong, Huei San and Blewitt, Marnie E and Asselin-Labat, Marie-Liesse and Smyth, Gordon K and Ritchie, Matthew E},</span></span>
<span id="cb131-97"><a href="#cb131-97" tabindex="-1"></a><span class="do">##   journal={Nucleic acids research},</span></span>
<span id="cb131-98"><a href="#cb131-98" tabindex="-1"></a><span class="do">##   volume={43},</span></span>
<span id="cb131-99"><a href="#cb131-99" tabindex="-1"></a><span class="do">##   number={15},</span></span>
<span id="cb131-100"><a href="#cb131-100" tabindex="-1"></a><span class="do">##   pages={e97--e97},</span></span>
<span id="cb131-101"><a href="#cb131-101" tabindex="-1"></a><span class="do">##   year={2015},</span></span>
<span id="cb131-102"><a href="#cb131-102" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb131-103"><a href="#cb131-103" tabindex="-1"></a><span class="do">##     }</span></span>
<span id="cb131-104"><a href="#cb131-104" tabindex="-1"></a><span class="do">## @article{leek2012sva,</span></span>
<span id="cb131-105"><a href="#cb131-105" tabindex="-1"></a><span class="do">##   title={The sva package for removing batch effects and other unwanted variation in high-throughput experiments},</span></span>
<span id="cb131-106"><a href="#cb131-106" tabindex="-1"></a><span class="do">##   author={Leek, Jeffrey T and Johnson, W Evan and Parker, Hilary S and Jaffe, Andrew E and Storey, John D},</span></span>
<span id="cb131-107"><a href="#cb131-107" tabindex="-1"></a><span class="do">##   journal={Bioinformatics},</span></span>
<span id="cb131-108"><a href="#cb131-108" tabindex="-1"></a><span class="do">##   volume={28},</span></span>
<span id="cb131-109"><a href="#cb131-109" tabindex="-1"></a><span class="do">##   number={6},</span></span>
<span id="cb131-110"><a href="#cb131-110" tabindex="-1"></a><span class="do">##   pages={882--883},</span></span>
<span id="cb131-111"><a href="#cb131-111" tabindex="-1"></a><span class="do">##   year={2012},</span></span>
<span id="cb131-112"><a href="#cb131-112" tabindex="-1"></a><span class="do">##   publisher={Oxford University Press}</span></span>
<span id="cb131-113"><a href="#cb131-113" tabindex="-1"></a><span class="do">##  }</span></span>
<span id="cb131-114"><a href="#cb131-114" tabindex="-1"></a><span class="do">## @article{newman2015robust,</span></span>
<span id="cb131-115"><a href="#cb131-115" tabindex="-1"></a><span class="do">##   title={Robust enumeration of cell subsets from tissue expression profiles},</span></span>
<span id="cb131-116"><a href="#cb131-116" tabindex="-1"></a><span class="do">##   author={Newman, Aaron M and Liu, Chih Long and Green, Michael R and Gentles, Andrew J and Feng, Weiguo and Xu, Yue and Hoang, Chuong D and Diehn, Maximilian and Alizadeh, Ash A},</span></span>
<span id="cb131-117"><a href="#cb131-117" tabindex="-1"></a><span class="do">##   journal={Nature methods},</span></span>
<span id="cb131-118"><a href="#cb131-118" tabindex="-1"></a><span class="do">##   volume={12},</span></span>
<span id="cb131-119"><a href="#cb131-119" tabindex="-1"></a><span class="do">##   number={5},</span></span>
<span id="cb131-120"><a href="#cb131-120" tabindex="-1"></a><span class="do">##   pages={453--457},</span></span>
<span id="cb131-121"><a href="#cb131-121" tabindex="-1"></a><span class="do">##   year={2015},</span></span>
<span id="cb131-122"><a href="#cb131-122" tabindex="-1"></a><span class="do">##   publisher={Nature Publishing Group}</span></span>
<span id="cb131-123"><a href="#cb131-123" tabindex="-1"></a><span class="do">##  }</span></span></code></pre></div>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb133"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" tabindex="-1"></a><span class="do">## R version 4.5.0 (2025-04-11)</span></span>
<span id="cb133-2"><a href="#cb133-2" tabindex="-1"></a><span class="do">## Platform: x86_64-apple-darwin20</span></span>
<span id="cb133-3"><a href="#cb133-3" tabindex="-1"></a><span class="do">## Running under: macOS Sonoma 14.6.1</span></span>
<span id="cb133-4"><a href="#cb133-4" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-5"><a href="#cb133-5" tabindex="-1"></a><span class="do">## Matrix products: default</span></span>
<span id="cb133-6"><a href="#cb133-6" tabindex="-1"></a><span class="do">## BLAS:   /Library/Frameworks/R.framework/Versions/4.5-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span id="cb133-7"><a href="#cb133-7" tabindex="-1"></a><span class="do">## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span id="cb133-8"><a href="#cb133-8" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-9"><a href="#cb133-9" tabindex="-1"></a><span class="do">## locale:</span></span>
<span id="cb133-10"><a href="#cb133-10" tabindex="-1"></a><span class="do">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span id="cb133-11"><a href="#cb133-11" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-12"><a href="#cb133-12" tabindex="-1"></a><span class="do">## time zone: Australia/Adelaide</span></span>
<span id="cb133-13"><a href="#cb133-13" tabindex="-1"></a><span class="do">## tzcode source: internal</span></span>
<span id="cb133-14"><a href="#cb133-14" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-15"><a href="#cb133-15" tabindex="-1"></a><span class="do">## attached base packages:</span></span>
<span id="cb133-16"><a href="#cb133-16" tabindex="-1"></a><span class="do">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span id="cb133-17"><a href="#cb133-17" tabindex="-1"></a><span class="do">## [8] base     </span></span>
<span id="cb133-18"><a href="#cb133-18" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-19"><a href="#cb133-19" tabindex="-1"></a><span class="do">## other attached packages:</span></span>
<span id="cb133-20"><a href="#cb133-20" tabindex="-1"></a><span class="do">##  [1] pathview_1.48.0                 topGO_2.60.1                   </span></span>
<span id="cb133-21"><a href="#cb133-21" tabindex="-1"></a><span class="do">##  [3] SparseM_1.84-2                  GO.db_3.21.0                   </span></span>
<span id="cb133-22"><a href="#cb133-22" tabindex="-1"></a><span class="do">##  [5] graph_1.86.0                    AnnotationDbi_1.70.0           </span></span>
<span id="cb133-23"><a href="#cb133-23" tabindex="-1"></a><span class="do">##  [7] gage_2.58.0                     patchwork_1.3.1                </span></span>
<span id="cb133-24"><a href="#cb133-24" tabindex="-1"></a><span class="do">##  [9] GGally_2.3.0                    DESeq2_1.48.1                  </span></span>
<span id="cb133-25"><a href="#cb133-25" tabindex="-1"></a><span class="do">## [11] edgeR_4.6.3                     limma_3.64.3                   </span></span>
<span id="cb133-26"><a href="#cb133-26" tabindex="-1"></a><span class="do">## [13] tidySummarizedExperiment_1.19.4 airway_1.28.0                  </span></span>
<span id="cb133-27"><a href="#cb133-27" tabindex="-1"></a><span class="do">## [15] SummarizedExperiment_1.38.1     Biobase_2.68.0                 </span></span>
<span id="cb133-28"><a href="#cb133-28" tabindex="-1"></a><span class="do">## [17] GenomicRanges_1.60.0            GenomeInfoDb_1.44.1            </span></span>
<span id="cb133-29"><a href="#cb133-29" tabindex="-1"></a><span class="do">## [19] IRanges_2.42.0                  S4Vectors_0.46.0               </span></span>
<span id="cb133-30"><a href="#cb133-30" tabindex="-1"></a><span class="do">## [21] BiocGenerics_0.54.0             generics_0.1.4                 </span></span>
<span id="cb133-31"><a href="#cb133-31" tabindex="-1"></a><span class="do">## [23] MatrixGenerics_1.20.0           matrixStats_1.5.0              </span></span>
<span id="cb133-32"><a href="#cb133-32" tabindex="-1"></a><span class="do">## [25] tidybulk_1.99.2                 ttservice_0.5.3                </span></span>
<span id="cb133-33"><a href="#cb133-33" tabindex="-1"></a><span class="do">## [27] ggrepel_0.9.6                   magrittr_2.0.3                 </span></span>
<span id="cb133-34"><a href="#cb133-34" tabindex="-1"></a><span class="do">## [29] lubridate_1.9.4                 forcats_1.0.0                  </span></span>
<span id="cb133-35"><a href="#cb133-35" tabindex="-1"></a><span class="do">## [31] stringr_1.5.1                   dplyr_1.1.4                    </span></span>
<span id="cb133-36"><a href="#cb133-36" tabindex="-1"></a><span class="do">## [33] purrr_1.1.0                     readr_2.1.5                    </span></span>
<span id="cb133-37"><a href="#cb133-37" tabindex="-1"></a><span class="do">## [35] tidyr_1.3.1                     tibble_3.3.0                   </span></span>
<span id="cb133-38"><a href="#cb133-38" tabindex="-1"></a><span class="do">## [37] ggplot2_3.5.2                   tidyverse_2.0.0                </span></span>
<span id="cb133-39"><a href="#cb133-39" tabindex="-1"></a><span class="do">## [39] knitr_1.50                     </span></span>
<span id="cb133-40"><a href="#cb133-40" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb133-41"><a href="#cb133-41" tabindex="-1"></a><span class="do">## loaded via a namespace (and not attached):</span></span>
<span id="cb133-42"><a href="#cb133-42" tabindex="-1"></a><span class="do">##   [1] fs_1.6.6                    GSVA_2.2.0                 </span></span>
<span id="cb133-43"><a href="#cb133-43" tabindex="-1"></a><span class="do">##   [3] bitops_1.0-9                R2HTML_2.3.4               </span></span>
<span id="cb133-44"><a href="#cb133-44" tabindex="-1"></a><span class="do">##   [5] enrichplot_1.28.4           httr_1.4.7                 </span></span>
<span id="cb133-45"><a href="#cb133-45" tabindex="-1"></a><span class="do">##   [7] RColorBrewer_1.1-3          numDeriv_2016.8-1.1        </span></span>
<span id="cb133-46"><a href="#cb133-46" tabindex="-1"></a><span class="do">##   [9] Rgraphviz_2.52.0            doRNG_1.8.6.2              </span></span>
<span id="cb133-47"><a href="#cb133-47" tabindex="-1"></a><span class="do">##  [11] tools_4.5.0                 backports_1.5.0            </span></span>
<span id="cb133-48"><a href="#cb133-48" tabindex="-1"></a><span class="do">##  [13] utf8_1.2.6                  R6_2.6.1                   </span></span>
<span id="cb133-49"><a href="#cb133-49" tabindex="-1"></a><span class="do">##  [15] DT_0.33                     HDF5Array_1.36.0           </span></span>
<span id="cb133-50"><a href="#cb133-50" tabindex="-1"></a><span class="do">##  [17] sn_2.1.1                    lazyeval_0.2.2             </span></span>
<span id="cb133-51"><a href="#cb133-51" tabindex="-1"></a><span class="do">##  [19] mgcv_1.9-3                  rhdf5filters_1.20.0        </span></span>
<span id="cb133-52"><a href="#cb133-52" tabindex="-1"></a><span class="do">##  [21] withr_3.0.2                 preprocessCore_1.70.0      </span></span>
<span id="cb133-53"><a href="#cb133-53" tabindex="-1"></a><span class="do">##  [23] cli_3.6.5                   sandwich_3.1-1             </span></span>
<span id="cb133-54"><a href="#cb133-54" tabindex="-1"></a><span class="do">##  [25] labeling_0.4.3              KEGGgraph_1.68.0           </span></span>
<span id="cb133-55"><a href="#cb133-55" tabindex="-1"></a><span class="do">##  [27] mvtnorm_1.3-3               S7_0.2.0                   </span></span>
<span id="cb133-56"><a href="#cb133-56" tabindex="-1"></a><span class="do">##  [29] genefilter_1.90.0           proxy_0.4-27               </span></span>
<span id="cb133-57"><a href="#cb133-57" tabindex="-1"></a><span class="do">##  [31] PADOG_1.50.0                yulab.utils_0.2.0          </span></span>
<span id="cb133-58"><a href="#cb133-58" tabindex="-1"></a><span class="do">##  [33] gson_0.1.0                  DOSE_4.2.0                 </span></span>
<span id="cb133-59"><a href="#cb133-59" tabindex="-1"></a><span class="do">##  [35] R.utils_2.13.0              HTMLUtils_0.1.9            </span></span>
<span id="cb133-60"><a href="#cb133-60" tabindex="-1"></a><span class="do">##  [37] plotrix_3.8-4               rstudioapi_0.17.1          </span></span>
<span id="cb133-61"><a href="#cb133-61" tabindex="-1"></a><span class="do">##  [39] RSQLite_2.4.2               gridGraphics_0.5-1         </span></span>
<span id="cb133-62"><a href="#cb133-62" tabindex="-1"></a><span class="do">##  [41] hwriter_1.3.2.1             gtools_3.9.5               </span></span>
<span id="cb133-63"><a href="#cb133-63" tabindex="-1"></a><span class="do">##  [43] Matrix_1.7-3                abind_1.4-8                </span></span>
<span id="cb133-64"><a href="#cb133-64" tabindex="-1"></a><span class="do">##  [45] R.methodsS3_1.8.2           lifecycle_1.0.4            </span></span>
<span id="cb133-65"><a href="#cb133-65" tabindex="-1"></a><span class="do">##  [47] multcomp_1.4-28             yaml_2.3.10                </span></span>
<span id="cb133-66"><a href="#cb133-66" tabindex="-1"></a><span class="do">##  [49] mathjaxr_1.8-0              KEGGdzPathwaysGEO_1.46.0   </span></span>
<span id="cb133-67"><a href="#cb133-67" tabindex="-1"></a><span class="do">##  [51] gplots_3.2.0                rhdf5_2.52.1               </span></span>
<span id="cb133-68"><a href="#cb133-68" tabindex="-1"></a><span class="do">##  [53] qvalue_2.40.0               SparseArray_1.8.1          </span></span>
<span id="cb133-69"><a href="#cb133-69" tabindex="-1"></a><span class="do">##  [55] grid_4.5.0                  blob_1.2.4                 </span></span>
<span id="cb133-70"><a href="#cb133-70" tabindex="-1"></a><span class="do">##  [57] crayon_1.5.3                ggtangle_0.0.7             </span></span>
<span id="cb133-71"><a href="#cb133-71" tabindex="-1"></a><span class="do">##  [59] lattice_0.22-7              beachmat_2.24.0            </span></span>
<span id="cb133-72"><a href="#cb133-72" tabindex="-1"></a><span class="do">##  [61] msigdbr_25.1.1              cowplot_1.2.0              </span></span>
<span id="cb133-73"><a href="#cb133-73" tabindex="-1"></a><span class="do">##  [63] annotate_1.86.1             KEGGREST_1.48.1            </span></span>
<span id="cb133-74"><a href="#cb133-74" tabindex="-1"></a><span class="do">##  [65] magick_2.8.7                pillar_1.11.0              </span></span>
<span id="cb133-75"><a href="#cb133-75" tabindex="-1"></a><span class="do">##  [67] fgsea_1.34.2                hgu133plus2.db_3.13.0      </span></span>
<span id="cb133-76"><a href="#cb133-76" tabindex="-1"></a><span class="do">##  [69] hgu133a.db_3.13.0           rjson_0.2.23               </span></span>
<span id="cb133-77"><a href="#cb133-77" tabindex="-1"></a><span class="do">##  [71] widyr_0.1.5                 codetools_0.2-20           </span></span>
<span id="cb133-78"><a href="#cb133-78" tabindex="-1"></a><span class="do">##  [73] fastmatch_1.1-6             mutoss_0.1-13              </span></span>
<span id="cb133-79"><a href="#cb133-79" tabindex="-1"></a><span class="do">##  [75] glue_1.8.0                  ggfun_0.2.0                </span></span>
<span id="cb133-80"><a href="#cb133-80" tabindex="-1"></a><span class="do">##  [77] data.table_1.17.8           Rdpack_2.6.4               </span></span>
<span id="cb133-81"><a href="#cb133-81" tabindex="-1"></a><span class="do">##  [79] vctrs_0.6.5                 png_0.1-8                  </span></span>
<span id="cb133-82"><a href="#cb133-82" tabindex="-1"></a><span class="do">##  [81] treeio_1.32.0               org.Mm.eg.db_3.21.0        </span></span>
<span id="cb133-83"><a href="#cb133-83" tabindex="-1"></a><span class="do">##  [83] gtable_0.3.6                org.Rn.eg.db_3.21.0        </span></span>
<span id="cb133-84"><a href="#cb133-84" tabindex="-1"></a><span class="do">##  [85] assertthat_0.2.1            cachem_1.1.0               </span></span>
<span id="cb133-85"><a href="#cb133-85" tabindex="-1"></a><span class="do">##  [87] xfun_0.53                   rbibutils_2.3              </span></span>
<span id="cb133-86"><a href="#cb133-86" tabindex="-1"></a><span class="do">##  [89] S4Arrays_1.8.1              survival_3.8-3             </span></span>
<span id="cb133-87"><a href="#cb133-87" tabindex="-1"></a><span class="do">##  [91] SingleCellExperiment_1.30.1 iterators_1.0.14           </span></span>
<span id="cb133-88"><a href="#cb133-88" tabindex="-1"></a><span class="do">##  [93] statmod_1.5.0               TH.data_1.1-3              </span></span>
<span id="cb133-89"><a href="#cb133-89" tabindex="-1"></a><span class="do">##  [95] ellipsis_0.3.2              nlme_3.1-168               </span></span>
<span id="cb133-90"><a href="#cb133-90" tabindex="-1"></a><span class="do">##  [97] ggtree_3.16.3               bit64_4.6.0-1              </span></span>
<span id="cb133-91"><a href="#cb133-91" tabindex="-1"></a><span class="do">##  [99] rprojroot_2.1.0             SnowballC_0.7.1            </span></span>
<span id="cb133-92"><a href="#cb133-92" tabindex="-1"></a><span class="do">## [101] irlba_2.3.5.1               KernSmooth_2.23-26         </span></span>
<span id="cb133-93"><a href="#cb133-93" tabindex="-1"></a><span class="do">## [103] colorspace_2.1-1            DBI_1.2.3                  </span></span>
<span id="cb133-94"><a href="#cb133-94" tabindex="-1"></a><span class="do">## [105] mnormt_2.1.1                tidyselect_1.2.1           </span></span>
<span id="cb133-95"><a href="#cb133-95" tabindex="-1"></a><span class="do">## [107] bit_4.6.0                   compiler_4.5.0             </span></span>
<span id="cb133-96"><a href="#cb133-96" tabindex="-1"></a><span class="do">## [109] curl_6.4.0                  h5mread_1.0.1              </span></span>
<span id="cb133-97"><a href="#cb133-97" tabindex="-1"></a><span class="do">## [111] TFisher_0.2.0               DelayedArray_0.34.1        </span></span>
<span id="cb133-98"><a href="#cb133-98" tabindex="-1"></a><span class="do">## [113] plotly_4.11.0               scales_1.4.0               </span></span>
<span id="cb133-99"><a href="#cb133-99" tabindex="-1"></a><span class="do">## [115] caTools_1.18.3              SpatialExperiment_1.18.1   </span></span>
<span id="cb133-100"><a href="#cb133-100" tabindex="-1"></a><span class="do">## [117] digest_0.6.37               rmarkdown_2.29             </span></span>
<span id="cb133-101"><a href="#cb133-101" tabindex="-1"></a><span class="do">## [119] XVector_0.48.0              htmltools_0.5.8.1          </span></span>
<span id="cb133-102"><a href="#cb133-102" tabindex="-1"></a><span class="do">## [121] pkgconfig_2.0.3             sparseMatrixStats_1.20.0   </span></span>
<span id="cb133-103"><a href="#cb133-103" tabindex="-1"></a><span class="do">## [123] fastmap_1.2.0               rlang_1.1.6                </span></span>
<span id="cb133-104"><a href="#cb133-104" tabindex="-1"></a><span class="do">## [125] htmlwidgets_1.6.4           UCSC.utils_1.4.0           </span></span>
<span id="cb133-105"><a href="#cb133-105" tabindex="-1"></a><span class="do">## [127] farver_2.1.2                zoo_1.8-14                 </span></span>
<span id="cb133-106"><a href="#cb133-106" tabindex="-1"></a><span class="do">## [129] jsonlite_2.0.0              BiocParallel_1.42.1        </span></span>
<span id="cb133-107"><a href="#cb133-107" tabindex="-1"></a><span class="do">## [131] GOSemSim_2.34.0             tokenizers_0.3.0           </span></span>
<span id="cb133-108"><a href="#cb133-108" tabindex="-1"></a><span class="do">## [133] R.oo_1.27.1                 BiocSingular_1.24.0        </span></span>
<span id="cb133-109"><a href="#cb133-109" tabindex="-1"></a><span class="do">## [135] RCurl_1.98-1.17             GenomeInfoDbData_1.2.14    </span></span>
<span id="cb133-110"><a href="#cb133-110" tabindex="-1"></a><span class="do">## [137] ggplotify_0.1.2             Rhdf5lib_1.30.0            </span></span>
<span id="cb133-111"><a href="#cb133-111" tabindex="-1"></a><span class="do">## [139] Rcpp_1.1.0                  ape_5.8-1                  </span></span>
<span id="cb133-112"><a href="#cb133-112" tabindex="-1"></a><span class="do">## [141] babelgene_22.9              stringi_1.8.7              </span></span>
<span id="cb133-113"><a href="#cb133-113" tabindex="-1"></a><span class="do">## [143] MASS_7.3-65                 globaltest_5.62.0          </span></span>
<span id="cb133-114"><a href="#cb133-114" tabindex="-1"></a><span class="do">## [145] plyr_1.8.9                  org.Hs.eg.db_3.21.0        </span></span>
<span id="cb133-115"><a href="#cb133-115" tabindex="-1"></a><span class="do">## [147] ggstats_0.10.0              parallel_4.5.0             </span></span>
<span id="cb133-116"><a href="#cb133-116" tabindex="-1"></a><span class="do">## [149] Biostrings_2.76.0           splines_4.5.0              </span></span>
<span id="cb133-117"><a href="#cb133-117" tabindex="-1"></a><span class="do">## [151] multtest_2.64.0             hms_1.1.3                  </span></span>
<span id="cb133-118"><a href="#cb133-118" tabindex="-1"></a><span class="do">## [153] qqconf_1.3.2                locfit_1.5-9.12            </span></span>
<span id="cb133-119"><a href="#cb133-119" tabindex="-1"></a><span class="do">## [155] igraph_2.1.4                rngtools_1.5.2             </span></span>
<span id="cb133-120"><a href="#cb133-120" tabindex="-1"></a><span class="do">## [157] EGSEAdata_1.36.0            reshape2_1.4.4             </span></span>
<span id="cb133-121"><a href="#cb133-121" tabindex="-1"></a><span class="do">## [159] ScaledMatrix_1.16.0         XML_3.99-0.18              </span></span>
<span id="cb133-122"><a href="#cb133-122" tabindex="-1"></a><span class="do">## [161] GSA_1.03.3                  evaluate_1.0.4             </span></span>
<span id="cb133-123"><a href="#cb133-123" tabindex="-1"></a><span class="do">## [163] metap_1.12                  tidytext_0.4.2             </span></span>
<span id="cb133-124"><a href="#cb133-124" tabindex="-1"></a><span class="do">## [165] foreach_1.5.2               tzdb_0.5.0                 </span></span>
<span id="cb133-125"><a href="#cb133-125" tabindex="-1"></a><span class="do">## [167] rsvd_1.0.5                  broom_1.0.8                </span></span>
<span id="cb133-126"><a href="#cb133-126" tabindex="-1"></a><span class="do">## [169] xtable_1.8-4                e1071_1.7-16               </span></span>
<span id="cb133-127"><a href="#cb133-127" tabindex="-1"></a><span class="do">## [171] tidytree_0.4.6              janeaustenr_1.0.0          </span></span>
<span id="cb133-128"><a href="#cb133-128" tabindex="-1"></a><span class="do">## [173] class_7.3-23                viridisLite_0.4.2          </span></span>
<span id="cb133-129"><a href="#cb133-129" tabindex="-1"></a><span class="do">## [175] clusterProfiler_4.16.0      aplot_0.2.8                </span></span>
<span id="cb133-130"><a href="#cb133-130" tabindex="-1"></a><span class="do">## [177] safe_3.48.0                 memoise_2.0.1              </span></span>
<span id="cb133-131"><a href="#cb133-131" tabindex="-1"></a><span class="do">## [179] timechange_0.3.0            sva_3.56.0                 </span></span>
<span id="cb133-132"><a href="#cb133-132" tabindex="-1"></a><span class="do">## [181] GSEABase_1.70.0</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-alhamdoosh2017egsea" class="csl-entry">
Alhamdoosh, M, M Ng, NJ Wilson, JM Sheridan, H Huynh, MJ Wilson, and ME Ritchie. 2017. “Combining Multiple Tools Outperforms Individual Methods in Gene Set Enrichment Analyses.” <em>Bioinformatics</em> 33 (3): 414–24.
</div>
<div id="ref-anders2010rle" class="csl-entry">
Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biology</em> 11 (10): R106.
</div>
<div id="ref-ashburner2000go" class="csl-entry">
Ashburner, Michael, Catherine A Ball, Judith A Blake, David Botstein, Heather Butler, J Michael Cherry, J Michael Davis, et al. 2000. “Gene Ontology: Tool for the Unification of Biology.” <em>Nature Genetics</em> 25 (1): 25–29.
</div>
<div id="ref-bullard2010uq" class="csl-entry">
Bullard, James H, Elizabeth Purdom, Kasper D Hansen, and Sandrine Dudoit. 2010. “Evaluation of Statistical Methods for Normalization and Differential Expression in mRNA-Seq Experiments.” <em>BMC Bioinformatics</em> 11 (1): 94.
</div>
<div id="ref-chen2016edgeR" class="csl-entry">
Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. “From Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.” <em>F1000Research</em> 5.
</div>
<div id="ref-hotelling1933pca" class="csl-entry">
Hotelling, Harold. 1933. “Analysis of a Complex of Statistical Variables into Principal Components.” <em>Journal of Educational Psychology</em> 24 (6): 417.
</div>
<div id="ref-kruskal1964mds" class="csl-entry">
Kruskal, Joseph B. 1964. “Multidimensional Scaling by Optimizing Goodness of Fit to a Nonmetric Hypothesis.” <em>Psychometrika</em> 29 (1): 1–27.
</div>
<div id="ref-law2014voom" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. “Voom: Precision Weights Unlock Linear Model Analysis Tools for RNA-Seq Read Counts.” <em>Genome Biology</em> 15 (2): R29.
</div>
<div id="ref-liu2015voomweights" class="csl-entry">
Liu, Rui, Anna Z Holik, Shian Su, Nikolas Jansz, Yunshun Chen, Hsiu S Leong, Marnie E Blewitt, Marie-Liesse Asselin-Labat, Gordon K Smyth, and Matthew E Ritchie. 2015. “Why Weight? Modelling Sample and Observational Level Variability Improves Power in RNA-Seq Analyses.” <em>Nucleic Acids Research</em> 43 (15): e97–97.
</div>
<div id="ref-love2014deseq2" class="csl-entry">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.” <em>Genome Biology</em> 15 (12): 550.
</div>
<div id="ref-ma2020glmmseq" class="csl-entry">
Ma, Shuang, Bing Zhang, Laura M LaFave, Amanda S Earl, Zhicheng Chiang, Yuling Hu, Jun Ding, et al. 2020. “A General Linear Mixed Model for Differential Expression Analysis of Single-Cell RNA-Seq Data.” <em>Nature Communications</em> 11 (1): 1–11.
</div>
<div id="ref-macqueen1967kmeans" class="csl-entry">
MacQueen, J. 1967. “Some Methods for Classification and Analysis of Multivariate Observations.” In <em>Proceedings of the Fifth Berkeley Symposium on Mathematical Statistics and Probability</em>, 1:281–97. 14. Oakland, CA, USA.
</div>
<div id="ref-mccarthy2009treat" class="csl-entry">
McCarthy, Davis J, and Gordon K Smyth. 2009. “Testing Significance Relative to a Fold-Change Threshold Is a TREAT.” <em>Bioinformatics</em> 25 (6): 765–71.
</div>
<div id="ref-newman2015cibersort" class="csl-entry">
Newman, Aaron M, Chih Long Liu, Michael R Green, Andrew J Gentles, Weiguo Feng, Yue Xu, Chuong D Hoang, Maximilian Diehn, and Ash A Alizadeh. 2015. “Robust Enumeration of Cell Subsets from Tissue Expression Profiles.” <em>Nature Methods</em> 12 (5): 453–57.
</div>
<div id="ref-robinson2010edger" class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. “edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40.
</div>
<div id="ref-subramanian2005gsea" class="csl-entry">
Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005. “Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.” <em>Proceedings of the National Academy of Sciences</em> 102 (43): 15545–50.
</div>
<div id="ref-zhang2017combatseq" class="csl-entry">
Zhang, Yu, Giovanni Parmigiani, and W Evan Johnson. 2020. “ComBat-Seq: Batch Effect Adjustment for RNA-Seq Count Data.” <em>NAR Genomics and Bioinformatics</em> 2 (3): lqaa078.
</div>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/tidybulk" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/stemangiola/tidybulk/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/stemangiola/tidybulk/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidybulk</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stefano Mangiola <br><small class="roles"> Author, maintainer </small>   </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a></li>
<li><a href="https://github.com/stemangiola/tidybulk/actions/" class="external-link"><img src="https://github.com/stemangiola/tidybulk/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/tidybulk/" class="external-link"><img src="https://bioconductor.org/shields/build/release/bioc/tidybulk.svg" alt="Bioconductor status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
