---
title: "Side-by-side comparison with standard interfaces"
author: "Stefano Mangiola"
date: "`r Sys.Date()`"
package: tidybulk
output:
  BiocStyle::html_document:
    toc_float: true
abstract: >
  Tidybulk is a comprehensive R package for modular transcriptomic data analysis that brings transcriptomics 
  to the tidyverse. It provides a unified interface for data transformation, normalization, filtering, 
  dimensionality reduction, clustering, differential analysis, cellularity analysis, and gene enrichment 
  with seamless integration of SummarizedExperiment objects and tidyverse principles.
keywords: "transcriptomics, RNA-seq, differential expression, data analysis, tidyverse, SummarizedExperiment, 
  bioinformatics, genomics, gene expression, clustering, dimensionality reduction, cellularity analysis, 
  gene enrichment, R package"
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{Side-by-side comparison with standard interfaces}
  %\usepackage[UTF-8]{inputenc}

---


<!-- <style type="text/css"> -->
<!-- .main-container { -->
<!--   max-width: 2800px;!important -->
<!--   margin-left: auto; -->
<!--   margin-right: auto; -->
<!-- } -->
<!-- </style> -->

<style>
.column-left{
  float: left;
  width: 50%;
  text-align: left;
}
.column-right{
  float: right;
  width: 50%;
  text-align: right;
}
</style>

<!-- badges: start -->
[![Lifecycle:maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing) 
[![R build status](https://github.com/stemangiola/tidybulk/workflows/R-CMD-check/badge.svg)](https://github.com/stemangiola/tidybulk/actions/)
[![Bioconductor status](https://bioconductor.org/shields/build/release/bioc/tidybulk.svg)](https://bioconductor.org/checkResults/release/bioc-LATEST/tidybulk/)
<!-- badges: end -->

```{r, echo=FALSE, out.height='139px', out.width='120px'}
knitr::include_graphics("../inst/logo.svg")
```

**tidybulk** is a powerful R package designed for modular transcriptomic data analysis that brings transcriptomics to the tidyverse.

## Why tidybulk?

Tidybulk provides a unified interface for comprehensive transcriptomic data analysis with seamless integration of SummarizedExperiment objects and tidyverse principles. It streamlines the entire workflow from raw data to biological insights.






### Scientific Citation

Mangiola, Stefano, Ramyar Molania, Ruining Dong, Maria A. Doyle, and Anthony T. Papenfuss. 2021. "Tidybulk: An R tidy framework for modular transcriptomic data analysis." Genome Biology 22 (42). https://doi.org/10.1186/s13059-020-02233-7

[Genome Biology - tidybulk: an R tidy framework for modular transcriptomic data analysis](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7)


```{r, echo=FALSE, include=FALSE, }
library(knitr)
# knitr::opts_chunk$set(cache = TRUE, warning = FALSE,
#                       message = FALSE, cache.lazy = FALSE)

library(dplyr)
library(tidyr)
library(tibble)
library(magrittr)
library(ggplot2)
library(ggrepel)
library(tidybulk)
library(tidySummarizedExperiment)
library(airway)

my_theme = 	
	theme_bw() +
	theme(
		panel.border = element_blank(),
		axis.line = element_line(),
		panel.grid.major = element_line(size = 0.2),
		panel.grid.minor = element_line(size = 0.1),
		text = element_text(size=12),
		legend.position="bottom",
		aspect.ratio=1,
		strip.background = element_blank(),
		axis.title.x  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
		axis.title.y  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10))
	)

# Load airway dataset
data(airway)

# Add gene symbol and entrez for better annotation
airway <-
  airway |> 
  mutate(symbol = AnnotationDbi::mapIds(org.Hs.eg.db::org.Hs.eg.db,
                                        keys = .feature,
                                        keytype = "ENSEMBL",
                                        column = "SYMBOL",
                                        multiVals = "first"
  )) |> 
  
  mutate(entrez = AnnotationDbi::mapIds(org.Hs.eg.db::org.Hs.eg.db,
                                      keys = .feature,
                                      keytype = "ENSEMBL",
                                      column = "ENTREZID",
                                      multiVals = "first"
)) 

# Convert dex to factor for proper differential expression analysis
colData(airway)$dex = as.factor(colData(airway)$dex)

```

## Aggregate duplicated `transcripts`

tidybulk provide the `aggregate_duplicates` function to aggregate duplicated transcripts (e.g., isoforms, ensembl). For example, we often have to convert ensembl symbols to gene/transcript symbol, but in doing so we have to deal with duplicates. `aggregate_duplicates` takes a tibble and column names (as symbols; for `sample`, `transcript` and `count`) as arguments and returns a tibble with transcripts with the same name aggregated. All the rest of the columns are appended, and factors and boolean are appended as characters.

<div class="column-left">
TidyTranscriptomics
```{r aggregate, message=FALSE, warning=FALSE, results='hide', class.source='yellow', eval=FALSE}
rowData(airway)$gene_name = rownames(airway)
airway.aggr = airway |> aggregate_duplicates(.transcript = gene_name)
```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r aggregate long, eval=FALSE}
temp = data.frame(
	symbol = dge_list$genes$symbol,
	dge_list$counts
)
dge_list.nr <- by(temp,	temp$symbol,
	function(df)
		if(length(df[1,1])>0)
			matrixStats:::colSums(as.matrix(df[,-1]))
)
dge_list.nr <- do.call("rbind", dge_list.nr)
colnames(dge_list.nr) <- colnames(dge_list)
```
</div>
<div style="clear:both;"></div>

## Scale `counts`

We may want to compensate for sequencing depth, scaling the transcript abundance (e.g., with TMM algorithm, Robinson and Oshlack doi.org/10.1186/gb-2010-11-3-r25). `scale_abundance` takes a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and a method as arguments and returns a tibble with additional columns with scaled data as `<NAME OF COUNT COLUMN>_scaled`.

<div class="column-left">
TidyTranscriptomics
```{r normalise}
airway.norm = airway |> identify_abundant(factor_of_interest = dex) |> scale_abundance()
```

</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r normalise long, eval=FALSE}
library(edgeR)

dgList <- DGEList(count_m=x,group=group)
keep <- filterByExpr(dgList)
dgList <- dgList[keep,,keep.lib.sizes=FALSE]
# ... additional processing steps ...
dgList <- calcNormFactors(dgList, method="TMM")
norm_counts.table <- cpm(dgList)
```
</div>
<div style="clear:both;"></div>

```{r, include=FALSE}
airway.norm |> select(`counts`, counts_scaled, .abundant, everything())
```

We can easily plot the scaled density to check the scaling outcome. On the x axis we have the log scaled counts, on the y axes we have the density, data is grouped by sample and coloured by treatment.


```{r plot_normalise, fig.width = 10, fig.height = 10}
airway.norm |>
	ggplot(aes(counts_scaled + 1, group=.sample, color=`dex`)) +
	geom_density() +
	scale_x_log10() +
	my_theme
```

## Filter `variable transcripts`

We may want to identify and filter variable transcripts.

<div class="column-left">
TidyTranscriptomics
```{r filter variable}
airway.norm.variable = airway.norm |> keep_variable()
```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r filter variable long, eval=FALSE}
library(edgeR)

x = norm_counts.table

s <- rowMeans((x-rowMeans(x))^2)
o <- order(s,decreasing=TRUE)
x <- x[o[1L:top],,drop=FALSE]

norm_counts.table = norm_counts.table[rownames(x)]

norm_counts.table$cell_type = tibble_counts[
	match(
		tibble_counts$sample,
		rownames(norm_counts.table)
	),
	"cell"
]
```

</div>
<div style="clear:both;"></div>


## Reduce `dimensions`

We may want to reduce the dimensions of our data, for example using PCA or MDS algorithms. `reduce_dimensions` takes a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and a method (e.g., MDS or PCA) as arguments and returns a tibble with additional columns for the reduced dimensions.

**MDS** (Robinson et al., 10.1093/bioinformatics/btp616)

<div class="column-left">
TidyTranscriptomics
```{r mds}
airway.norm.MDS =
  airway.norm |>
  reduce_dimensions(method="MDS", .dims = 2)

```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r, eval = FALSE}
library(limma)

count_m_log = log(count_m + 1)
cmds = limma::plotMDS(count_m_log, ndim = 3, plot = FALSE)

cmds = cmds %$%	
	cmdscale.out |>
	setNames(sprintf("Dim%s", 1:6))

cmds$cell_type = tibble_counts[
	match(tibble_counts$sample, rownames(cmds)),
	"cell"
]
```
</div>
<div style="clear:both;"></div>

On the x and y axes axis we have the reduced dimensions 1 to 3, data is coloured by treatment.

```{r plot_mds, fig.width = 10, fig.height = 10, eval=FALSE}
airway.norm.MDS |> pivot_sample()  |> select(contains("Dim"), everything())

airway.norm.MDS |>
	pivot_sample() |>
  GGally::ggpairs(columns = 9:11, ggplot2::aes(colour=`dex`))


```

**PCA**

<div class="column-left">
TidyTranscriptomics
```{r pca, message=FALSE, warning=FALSE, results='hide'}
airway.norm.PCA =
  airway.norm |>
  reduce_dimensions(method="PCA", .dims = 2)
```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r,eval=FALSE}
count_m_log = log(count_m + 1)
pc = count_m_log |> prcomp(scale = TRUE)
variance = pc$sdev^2
variance = (variance / sum(variance))[1:6]
pc$cell_type = counts[
	match(counts$sample, rownames(pc)),
	"cell"
]
```
</div>
<div style="clear:both;"></div>

On the x and y axes axis we have the reduced dimensions 1 to 3, data is coloured by treatment.

```{r plot_pca, fig.width = 10, fig.height = 10, eval=FALSE}

airway.norm.PCA |> pivot_sample() |> select(contains("PC"), everything())

airway.norm.PCA |>
	 pivot_sample() |>
  GGally::ggpairs(columns = 11:13, ggplot2::aes(colour=`dex`))
```



## Rotate `dimensions`

We may want to rotate the reduced dimensions (or any two numeric columns really) of our data, of a set angle. `rotate_dimensions` takes a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and an angle as arguments and returns a tibble with additional columns for the rotated dimensions. The rotated dimensions will be added to the original data set as `<NAME OF DIMENSION> rotated <ANGLE>` by default, or as specified in the input arguments.
<div class="column-left">
TidyTranscriptomics
```{r rotate}
airway.norm.MDS.rotated =
  airway.norm.MDS |>
	rotate_dimensions(`Dim1`, `Dim2`, rotation_degrees = 45)
```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r, eval=FALSE}
rotation = function(m, d) {
	r = d * pi / 180
	((bind_rows(
		c(`1` = cos(r), `2` = -sin(r)),
		c(`1` = sin(r), `2` = cos(r))
	) |> as_matrix()) %*% m)
}
mds_r = pca |> rotation(rotation_degrees)
mds_r$cell_type = counts[
	match(counts$sample, rownames(mds_r)),
	"cell"
]
```
</div>
<div style="clear:both;"></div>

**Original**
On the x and y axes axis we have the first two reduced dimensions, data is coloured by treatment.

```{r plot_rotate_1, fig.width = 10, fig.height = 10}
airway.norm.MDS.rotated |>
	ggplot(aes(x=`Dim1`, y=`Dim2`, color=`dex` )) +
  geom_point() +
  my_theme
```

**Rotated**
On the x and y axes axis we have the first two reduced dimensions rotated of 45 degrees, data is coloured by treatment.

```{r plot_rotate_2, fig.width = 10, fig.height = 10}
airway.norm.MDS.rotated |>
	pivot_sample() |>
	ggplot(aes(x=`Dim1_rotated_45`, y=`Dim2_rotated_45`, color=`dex` )) +
  geom_point() +
  my_theme
```

## Test `differential abundance`

We may want to test for differential transcription between sample-wise factors of interest (e.g., with edgeR). `test_differential_expression` takes a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and a formula representing the desired linear model as arguments and returns a tibble with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).
<div class="column-left">
TidyTranscriptomics
```{r de, message=FALSE, warning=FALSE, results='hide'}
airway.de =
	airway |>
	test_differential_expression( ~ dex, method = "edgeR_quasi_likelihood") |>
  pivot_transcript()
airway.de
```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r, eval=FALSE}
library(edgeR)

dgList <- DGEList(counts=counts_m,group=group)
keep <- filterByExpr(dgList)
dgList <- dgList[keep,,keep.lib.sizes=FALSE]
dgList <- calcNormFactors(dgList)
design <- model.matrix(~group)
dgList <- estimateDisp(dgList,design)
fit <- glmQLFit(dgList,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf, n=Inf)
```
</div>
<div style="clear:both;"></div>

The functon `test_differential_expression` operated with contrasts too. The constrasts hve the name of the design matrix (generally <NAME_COLUMN_COVARIATE><VALUES_OF_COVARIATE>)
```{r de contrast, message=FALSE, warning=FALSE, results='hide', eval=FALSE}
airway.de =
	airway |>
	identify_abundant(factor_of_interest = dex) |>
	test_differential_expression(
		~ 0 + dex,                  
		.contrasts = c( "dexuntrt - dextrt"),
		method = "edgeR_quasi_likelihood"
	) |> 
  pivot_transcript()
```

## Adjust `counts`

We may want to adjust `counts` for (known) unwanted variation. `adjust_abundance` takes as arguments a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and a formula representing the desired linear model where the first covariate is the factor of interest and the second covariate is the unwanted variation, and returns a tibble with additional columns for the adjusted counts as `<COUNT COLUMN>_adjusted`. At the moment just an unwanted covariates is allowed at a time.

<div class="column-left">
TidyTranscriptomics
```{r adjust, message=FALSE, warning=FALSE, results='hide'}
airway.norm.adj =
	airway.norm 	|> adjust_abundance(	.factor_unwanted = cell, .factor_of_interest = dex, method="combat")


```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r, eval=FALSE}
library(sva)

count_m_log = log(count_m + 1)

design =
		model.matrix(
			object = ~ factor_of_interest + batch,
			data = annotation
		)

count_m_log.sva =
	ComBat(
			batch =	design[,2],
			mod = design,
			...
		)

count_m_log.sva = ceiling(exp(count_m_log.sva) -1)
count_m_log.sva$cell_type = counts[
	match(counts$sample, rownames(count_m_log.sva)),
	"cell"
]

```
</div>
<div style="clear:both;"></div>

## Deconvolve `Cell type composition`

We may want to infer the cell type composition of our samples (with the algorithm Cibersort; Newman et al., 10.1038/nmeth.3337). `deconvolve_cellularity` takes as arguments a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and returns a tibble with additional columns for the adjusted  cell type proportions.

*Note: Cellularity deconvolution requires gene symbols that match the reference data. The airway dataset uses Ensembl IDs, so this example is commented out.*

<div class="column-left">
TidyTranscriptomics
```{r cibersort, eval=FALSE}
# Requires gene symbols that match reference data
# airway.cibersort =
# 	airway |>
# 	deconvolve_cellularity( cores=1, prefix = "cibersort__") |> 
#   pivot_sample()

```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r, eval=FALSE}

source("CIBERSORT.R")
count_m |> write.table("mixture_file.txt")
results <- CIBERSORT(
	"sig_matrix_file.txt",
	"mixture_file.txt",
	perm=100, QN=TRUE
)
results$cell_type = tibble_counts[
	match(tibble_counts$sample, rownames(results)),
	"cell"
]

```
</div>
<div style="clear:both;"></div>

*Note: The plotting code is commented out as the deconvolution step is not executed.*

```{r plot_cibersort, eval=FALSE}
# airway.cibersort |>
# 	pivot_longer(
# 		names_to= "Cell_type_inferred", 
# 		values_to = "proportion", 
# 		names_prefix ="cibersort__", 
# 		cols=contains("cibersort__")
# 	) |>
#   ggplot(aes(x=Cell_type_inferred, y=proportion, fill=cell)) +
#   geom_boxplot() +
#   facet_wrap(~cell) +
#   my_theme +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), aspect.ratio=1/5)

```

## Cluster samples

We may want to cluster our samples based on the transcriptomic profiles. `cluster_elements` takes as arguments a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and returns a tibble with additional columns for the cluster labels.

<div class="column-left">
TidyTranscriptomics
```{r cluster}
airway.norm.cluster =
	airway.norm |>
	cluster_elements(method="kmeans", centers = 2)
```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r, eval=FALSE}
library(stats)

count_m_log = log(count_m + 1)
count_m_log_centered = count_m_log - rowMeans(count_m_log)

kmeans_result = kmeans(t(count_m_log_centered), centers = 2)

cluster_labels = kmeans_result$cluster
```
</div>
<div style="clear:both;"></div>

## Test differential abundance

We may want to test for differential abundance between conditions. `test_differential_abundance` takes as arguments a tibble, column names (as symbols; for `sample`, `transcript` and `count`) and a formula representing the desired linear model, and returns a tibble with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).

<div class="column-left">
TidyTranscriptomics
```{r differential}
airway.norm.de =
	airway.norm |>
	test_differential_abundance(~ dex, method="edgeR_quasi_likelihood")
```
</div>
<div class="column-right">
Standard procedure (comparative purpose)
```{r, eval=FALSE}
library(edgeR)

count_m_log = log(count_m + 1)

design = model.matrix(~ dex, data = annotation)

dge = DGEList(counts = count_m)
dge = calcNormFactors(dge)
dge = estimateDisp(dge, design)

fit = glmQLFit(dge, design)
qlf = glmQLFTest(fit, coef=2)

results = topTags(qlf, n = Inf)
```
</div>
<div style="clear:both;"></div>

## Conclusion

Tidybulk provides a unified interface for comprehensive transcriptomic data analysis with seamless integration of SummarizedExperiment objects and tidyverse principles. It streamlines the entire workflow from raw data to biological insights while maintaining compatibility with standard Bioconductor practices.

```{r session-info}
sessionInfo()
```