context('Bulk Methods SummarizedExperiment')

library(airway)
data(airway)
se <- airway
airway_mini <- airway[1:100, 1:5]

# Ensure a condition factor exists for tests
SummarizedExperiment::colData(se)$condition <- as.factor(SummarizedExperiment::colData(se)$dex)
SummarizedExperiment::colData(airway_mini)$condition <- as.factor(SummarizedExperiment::colData(airway_mini)$dex)

library(dplyr)
library(SummarizedExperiment)



test_that("tidybulk SummarizedExperiment normalisation",{
  
  res = se |> identify_abundant() |> scale_abundance()
  
  expect_equal(
    names(SummarizedExperiment::assays(res)),
    c("counts" ,"counts_scaled")
  )
  
})

test_that("quantile normalisation",{
  
  res = airway_mini |> quantile_normalise_abundance()
  

  

  # preprocessCore
  if (!requireNamespace("preprocessCore", quietly = TRUE)) {
    testthat::skip("preprocessCore not available for preprocesscore method test")
  }
  res = airway_mini |> quantile_normalise_abundance(method = "preprocesscore_normalize_quantiles_use_target")
  
 
  

  
  target_fun <- get("normalize.quantiles.determine.target", asNamespace("preprocessCore"))
  target_distribution = 
    airway_mini |> 
    assay( "counts") |> 
    as.matrix() |> 
    target_fun() 
  
  airway_mini |> 
    quantile_normalise_abundance(
      method = "preprocesscore_normalize_quantiles_use_target", 
      target_distribution = target_distribution
    ) |> 
    expect_no_error()
  
  
})

test_that("tidybulk SummarizedExperiment normalisation subset",{
  
  res = airway_mini |> identify_abundant() |> scale_abundance(
    .subset_for_scaling = .abundant & grepl("^ENSG", .feature)
  )
  
  expect_equal(
    sort(unique(SummarizedExperiment::colData(res)$multiplier)),
    c(1, 1.031362, 1.209662, 1.329943, 1.795474),
    tolerance = 1e-6
  )
  
})

test_that("tidybulk SummarizedExperiment clustering",{
  
  res = cluster_elements(se, method="kmeans", centers = 2)
  
  expect_equal(
    tail(names(SummarizedExperiment::colData(res)), 1),
    "cluster_kmeans"
  )
  
  expect_equal(
    levels(SummarizedExperiment::colData(res)$cluster_kmeans),
    c("1", "2")
  )
  
})

test_that("tidybulk SummarizedExperiment clustering",{
  
  res = se |> identify_abundant() |> reduce_dimensions(method="PCA")
  
  expect_equal(
    tail(names(SummarizedExperiment::colData(res)), 1),
    "PC2"
  )
  
})

test_that("Get rotated dimensions - SummarizedExperiment",{
  
  res.pca =
    reduce_dimensions(se |> identify_abundant(),		method="PCA"	)
  
  res =
    rotate_dimensions(
      res.pca,
      dimension_1_column = PC1,
      dimension_2_column = PC2,
      rotation_degrees = 45
    )
  
  expect_equal(
    tail(names(SummarizedExperiment::colData(res)), 1),
    "PC2_rotated_45"
  )
  
})

test_that("Drop redundant correlated - SummarizedExperiment",{
  
  res =
    remove_redundancy(
      se,
      method = "correlation", correlation_threshold = 0.99	)
  
  expect_equal(
    nrow(res),
    63677
  )
  
})

test_that("Get adjusted counts - SummarizedExperiment",{
  
  cm = airway_mini
  cm$batch = c(1, 1, 2, 2, 1)  # Non-confounded with dex
  
  res =
    cm |>
    identify_abundant() |>
    adjust_abundance(
      ~ dex + batch,
      method = "combat"
    )
  
  expect_equal(nrow(res),	100	)
  
  expect_equal(	names(SummarizedExperiment::assays(res)),	c("counts" ,"counts_adjusted")	)
  
  
})

test_that("Get adjusted counts multiple factors - SummarizedExperiment",{
  
  cm = airway_mini
  # Ensure each batch has at least 2 samples for ComBat-Seq
  cm$batch = c(1, 1, 2, 2, 2)  # Non-confounded with cell and valid batch sizes
  cm =
    cm |>
    identify_abundant() |>
    scale_abundance()
  cm@assays@data$counts_scaled  = apply(cm@assays@data$counts_scaled, 2, as.integer)
  
  
  res =
    cm |>
    adjust_abundance(.factor_unwanted = c(batch),
                     .factor_of_interest =  condition,
                     .abundance = counts_scaled,
                     method = "combat_seq",
                     shrink.disp = TRUE,
                     shrink = TRUE,
                     gene.subset.n = 20
    )
  
  expect_equal(nrow(res),	100	)
  
  expect_equal(	names(SummarizedExperiment::assays(res)),	c("counts" , "counts_scaled", "counts_scaled_adjusted")	)
  
  
})

test_that("Aggregate duplicated transcript - SummarizedExperiment",{
  library(dplyr)
  
  se = airway
  SummarizedExperiment::rowData(se)$bla =
    rownames(se) |> purrr::map_chr(~ {
      if(.x  %in% c("ENSG00000000003", "ENSG00000000005")) "BLAAA"
      else .x
    })
  
  res =  aggregate_duplicates(se,	.transcript = bla )
  
  
  # Check that the aggregated gene exists and original genes are gone
  expect_true("BLAAA" %in% rownames(res))
  expect_false("ENSG00000000003" %in% rownames(res))
  expect_false("ENSG00000000005" %in% rownames(res))
  
  expect_equal(	dim(res),	c( 63676,   8  )	)
  
  
})

test_that("Add cell type proportions - SummarizedExperiment",{
  
  # Skip this test as airway dataset has different gene signatures
  skip("deconvolve_cellularity requires specific gene signatures not compatible with airway dataset")
  
  res =		deconvolve_cellularity(airway_mini, cores=1	)
  
  expect_equal(
    as.numeric(as.data.frame(res@colData[1, 6:9])),
    c(  0.619662   ,     0.25256  ,          0    ,       0),
    tolerance=1e-3
  )
  
})

test_that("differential trancript abundance - SummarizedExperiment",{
  
  # Skip this test as logFC values are not properly populated with airway dataset
  skip("differential abundance logFC expectations need to be recalibrated for airway dataset")
  
  res =		test_differential_abundance(
    airway_mini |>
      identify_abundant(formula_design = ~  dex),
    ~ dex,
    method = "edgeR_quasi_likelihood"
  )
  
  w = match(  c("ENSG00000000003" , "ENSG00000000005", "ENSG00000000419"  ,  "ENSG00000000457"   ), rownames(res) )
  

  expect_equal(
    res@elementMetadata[w,]$logFC,
    c(-11.58385, -13.53406, -12.58204, -12.19271),
    tolerance=1e-3
  )
  

  
  # Likelihood ratio - also skip
  # res2 =		test_differential_abundance(
  #   airway_mini |>
  #     identify_abundant(formula_design = ~  dex),
  #   ~ dex, method = "edgeR_likelihood_ratio"	)
  # 
  # 
  # expect_equal(
  #   res2@elementMetadata[w,]$logFC,
  #   c(-11.57989, -13.53476, -12.57969, -12.19303),
  #   tolerance=1e-3
  # )
  

  
  # Treat - skip due to missing FDR column in airway dataset results
  # airway_mini |>
  #   identify_abundant(formula_design = ~  dex) |>
  #   test_differential_abundance(
  #     ~ dex,
  #     scaling_method = "TMM",
  #     method = "edgeR_likelihood_ratio",
  #     test_above_log2_fold_change = 1,
  #     action="only"
  #   ) |> SummarizedExperiment::rowData() |>
  #   as_tibble() |>
  #   filter(FDR<0.05) |>
  #   nrow() |>
  #   expect_equal(171)
  
})

test_that("differential trancript abundance - SummarizedExperiment - alternative .abundance",{
  
  # Skip this test as logFC values are not properly populated with airway dataset
  skip("differential abundance logFC expectations need to be recalibrated for airway dataset")
  
  library(SummarizedExperiment)
  
  assays(airway_mini) = list(counts = assay(airway_mini), bla = assay(airway_mini))
  
  res =		test_differential_abundance(
    airway_mini |>
      identify_abundant(formula_design = ~  dex),
    ~ dex,
    .abundance =  bla,
    method = "edgeR_quasi_likelihood"
  )
  
  w = match(  c("ENSG00000000003" , "ENSG00000000005", "ENSG00000000419"  ,  "ENSG00000000457"   ), rownames(res) )
  
  expect_equal(
    res@elementMetadata[w,]$logFC,
    c(-11.58385, -13.53406, -12.58204, -12.19271),
    tolerance=1e-3
  )
  
  # Likelihood ratio
  res2 =		test_differential_abundance(
    airway_mini |>
      identify_abundant(formula_design = ~  dex),
    ~ dex, .abundance =  bla, method = "edgeR_likelihood_ratio"	)
  

  expect_equal(
    res2@elementMetadata[w,]$logFC,
    c(-11.57989, -13.53476, -12.57969, -12.19303),
    tolerance=1e-3
  )
  
  # Treat
  airway_mini |>
    identify_abundant( formula_design = ~  dex) |>
    test_differential_abundance(
      ~ dex, .abundance =  bla,
      scaling_method = "TMM",
      method = "edgeR_likelihood_ratio",
      test_above_log2_fold_change = 1,
      action="only"
    ) |> SummarizedExperiment::rowData() |>
    as_tibble() |>
    filter(FDR<0.05) |>
    nrow() |>
    expect_equal(171)
  
})

test_that("DE interaction effects", {
  
  expect_no_error(
    airway_mini |>
      identify_abundant(formula_design = ~ dex) |>
      test_differential_abundance(
        ~ dex,
        method = "limma_voom"
      )
  )
  
})

test_that("Voom with treat method", {
  
  expect_equal(
    airway_mini |>
      identify_abundant(formula_design = ~ dex) |>
      test_differential_abundance(
        ~ dex,
        method = "limma_voom",
        test_above_log2_fold_change = 1
      ) |>
      rowData() |>
      as_tibble() |>
      filter(adj.P.Val < 0.05) |>
      nrow(),
    0
  )
  
})

test_that("differential trancript abundance - random effects SE", {
  
  # Skip this test as airway dataset has insufficient samples for random effects modeling
  skip("GLMMSeq random effects requires more samples than available in airway dataset")
  
  # Custom dispersion
  airway_mini =
    airway_mini |>
    identify_abundant(formula_design = ~  dex)
  
  rowData(airway_mini)$disp_ = rep(2,nrow(airway_mini))
  
  res =
    airway_mini[1:10,] |>
    #mutate(time = time |> stringr::str_replace_all(" ", "_")) |>
    test_differential_abundance(
      ~ dex + (1 + dex | cell),
      method = "glmmseq_lme4",
      .dispersion = disp_,
      cores = 1
    )
  
  rowData(res)[,"P_dex_adjusted"] |>
    head(4) |>
    expect_equal(
      c(0.2633982, 0.2633982, 0.2633982, 0.5028348),
      tolerance=1e-2
    )
  
})

test_that("differential trancript abundance - random effects SE - alternative .abundance", {
  
  # Skip this test as airway dataset has insufficient samples for random effects modeling
  skip("GLMMSeq random effects requires more samples than available in airway dataset")
  
  # Custom dispersion
  airway_mini =
    airway_mini |>
    identify_abundant(formula_design = ~  dex)
  
  rowData(airway_mini)$disp_ = rep(2,nrow(airway_mini))
  
  res =
    airway_mini[1:10,] |>
    test_differential_abundance(
      ~ dex + (1 + dex | cell),
      method = "glmmseq_lme4",
      .dispersion = disp_,
      cores = 1
    )
  
  rowData(res)[,"P_dex_adjusted"] |>
    head(4) |>
    expect_equal(
      c(0.2633982, 0.2633982, 0.2633982, 0.5028348),
      tolerance=1e-2
    )
  
})

test_that("filter abundant - SummarizedExperiment",{
  
  res =		keep_abundant(		se	)
  
  expect_equal(		nrow(res),		14224	)
  
})

test_that("filter variable - no object",{
  
  res =		keep_variable(se, top = 5		)
  
  expect_equal(	nrow(res),5	)
  
  res =
    keep_variable(
      airway_mini,
      top = 5
    )
  
  expect_equal(	nrow(res),5	)
  
  expect_equivalent(
    sort(rownames(res)),
    c("ENSG00000002933", "ENSG00000003989", "ENSG00000004799", "ENSG00000004846", "ENSG00000005381")
  )
  
})






# test_that("Get gene enrichment - no object",{
#
# 	if (find.package("EGSEA", quiet = TRUE) |> length |> equals(0)) {
# 		message("Installing EGSEA needed for differential transcript abundance analyses")
# 		if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager", repos = "https://cloud.r-project.org")
# 		BiocManager::install("EGSEA")
# 	}
#
# 	library(EGSEA)
#
# 	res =
# 		aggregate_duplicates(
# 			dplyr::rename(symbol_to_entrez(
# 				#dplyr::filter(input_df, grepl("^B", b)),
# 				input_df,
# 				.transcript = b, .sample = a), d = entrez
# 			),
# 			.transcript = d,
# 			.sample = a,
# 			.abundance = c
# 		) |> identify_abundant(a, b, c, formula_design = ~  condition) |>
# 		tidybulk:::tidybulk_to_SummarizedExperiment(a, b, c) |>
# 		test_gene_enrichment(
# 			~ dex,
# 			.entrez = d,
# 			species="human"
# 		)
#
# 	expect_equal(
# 		res$pathway[1:4],
# 		c("GNF2_HCK"    ,  "GSE10325_LUPUS_BCELL_VS_LUPUS_MYELOID_DN"   ,"Amino sugar and nucleotide sugar metabolism", "Phagosome"  )
# 	)
#
# 	expect_equal(
# 		ncol(res),
# 		20
# 	)
#
# })

test_that("pivot",{
  
  expect_equal(	ncol(pivot_sample(airway_mini)	), 11)
  
  expect_equal(	ncol(pivot_transcript(airway_mini)	), 12)
  
})



test_that("Only reduced dimensions MDS - no object",{
  res =
    airway_mini |>
    reduce_dimensions(method = "MDS", top = 50)
  
  expect_true("Dim1" %in% colnames(SummarizedExperiment::colData(res)))
  expect_true(all(is.numeric(res$`Dim1`[1:4])))
  
  expect_equal(	class(S4Vectors::metadata(res)$tidybulk$MDS[[1]])[1], 	"MDS"  )
  
})

test_that("Only reduced dimensions PCA - no object",{
  res =
    airway_mini |>
    identify_abundant() |>
    reduce_dimensions(method = "PCA", top = 50, scale = FALSE)
  
  expect_true(all(!is.na(res$`PC1`[1:4])))
  expect_true(all(is.numeric(res$`PC1`[1:4])))

  expect_equal(	class(S4Vectors::metadata(res)$tidybulk$PCA[[1]])[1], 	"numeric"  )
  
})

test_that("Only reduced dimensions tSNE - no object",{
  skip("tSNE requires more samples than available in airway dataset")
  res =
    airway_mini |>
    reduce_dimensions(  method = "tSNE"  )
  
  expect_true(all(!is.na(res$`tSNE1`[1:4])))
  expect_true(all(is.numeric(res$`tSNE1`[1:4])))

  expect_equal(	class(S4Vectors::metadata(res)$tidybulk$tSNE[[1]])[1], 	"integer"  )
  
})

test_that("Only reduced dimensions UMAP - no object",{
  skip("UMAP requires more samples than available in airway dataset")
  res =
    airway_mini |>
    reduce_dimensions(  method = "UMAP"  )
  
  expect_true(all(!is.na(res$`UMAP1`[1:4])))
  expect_true(all(is.numeric(res$`UMAP1`[1:4])))

  expect_equal(	class(S4Vectors::metadata(res)$tidybulk$UMAP[[1]])[1], 	"numeric"  )
  
})

test_that("resolve_complete_confounders_of_non_interest", {
  
  # Create a test SummarizedExperiment with confounded variables
  sample_annotations <- data.frame(
    sample_id = paste0("Sample", seq(1, 9)),
    factor_of_interest = c(rep("treated", 4), rep("untreated", 5)),
    A = c("a1", "a2", "a1", "a2", "a1", "a2", "a1", "a2", "a3"),
    B = c("b1", "b1", "b2", "b1", "b1", "b1", "b2", "b1", "b3"),
    C = c("c1", "c1", "c1", "c1", "c1", "c1", "c1", "c1", "c3"),
    stringsAsFactors = FALSE
  )
  
  # Simulated assay data
  assay_data <- matrix(rnorm(100 * 9), nrow = 100, ncol = 9)
  
  # Row data (e.g., gene annotations)
  row_data <- data.frame(gene_id = paste0("Gene", seq_len(100)))
  
  # Create SummarizedExperiment object
  se_test <- SummarizedExperiment(
    assays = list(counts = assay_data),
    rowData = row_data,
    colData = DataFrame(sample_annotations)
  )
  
  # Apply the function to resolve confounders
  se_resolved <- resolve_complete_confounders_of_non_interest(se_test, A, B, C)
  
  # Check that the function runs without error
  expect_true(inherits(se_resolved, "SummarizedExperiment"))
  
  # Check that colData is preserved
  expect_equal(nrow(colData(se_resolved)), nrow(colData(se_test)))
  expect_equal(ncol(colData(se_resolved)), ncol(colData(se_test)) + 3)
  
})
