<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Adjust transcript abundance for unwanted variation — adjust_abundance • tidybulk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Adjust transcript abundance for unwanted variation — adjust_abundance"><meta property="og:description" content="adjust_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a consistent object (to the input) with an additional adjusted abundance column. This method uses scaled counts if present."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/comparison_coding.html">Side-by-side comparison with standard interfaces</a>
    </li>
    <li>
      <a href="../articles/introduction.html">tidybulk: An R tidy framework for modular transcriptomic data analysis</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Adjust transcript abundance for unwanted variation</h1>
    <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/master/R/adjust_abundance.R" class="external-link"><code>R/adjust_abundance.R</code></a></small>
    <div class="hidden name"><code>adjust_abundance-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>adjust_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a consistent object (to the input) with an additional adjusted abundance column. This method uses scaled counts if present.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">adjust_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .formula <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .factor_unwanted <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  abundance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">.data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"combat_seq"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  inverse_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">adjust_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .formula <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .factor_unwanted <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  abundance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">.data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"combat_seq"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  inverse_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'RangedSummarizedExperiment'</span></span>
<span><span class="fu">adjust_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .formula <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .factor_unwanted <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  abundance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">.data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"combat_seq"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  inverse_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg--data">.data<a class="anchor" aria-label="anchor" href="#arg--data"></a></dt>
<dd><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></dd>


<dt id="arg--formula">.formula<a class="anchor" aria-label="anchor" href="#arg--formula"></a></dt>
<dd><p>DEPRECATED - A formula with no response variable, representing the desired linear model where the first covariate is the factor of interest and the second covariate is the unwanted variation (of the kind ~ factor_of_interest + batch)</p></dd>


<dt id="arg--factor-unwanted">.factor_unwanted<a class="anchor" aria-label="anchor" href="#arg--factor-unwanted"></a></dt>
<dd><p>A tidy select, e.g. column names without double quotation. c(batch, country) These are the factor that we want to adjust for, including unwanted batcheffect, and unwanted biological effects.</p></dd>


<dt id="arg--factor-of-interest">.factor_of_interest<a class="anchor" aria-label="anchor" href="#arg--factor-of-interest"></a></dt>
<dd><p>A tidy select, e.g. column names without double quotation. c(treatment) These are the factor that we want to preserve.</p></dd>


<dt id="arg-abundance">abundance<a class="anchor" aria-label="anchor" href="#arg-abundance"></a></dt>
<dd><p>The name of the transcript/gene abundance column (character, preferred)</p></dd>


<dt id="arg--abundance">.abundance<a class="anchor" aria-label="anchor" href="#arg--abundance"></a></dt>
<dd><p>DEPRECATED. The name of the transcript/gene abundance column (symbolic, for backward compatibility)</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string. Methods include combat_seq (default), combat and limma_remove_batch_effect.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further parameters passed to the function sva::ComBat</p></dd>


<dt id="arg-log-transform">log_transform<a class="anchor" aria-label="anchor" href="#arg-log-transform"></a></dt>
<dd><p>DEPRECATED - A boolean, whether the value should be log-transformed (e.g., TRUE for RNA sequencing data)</p></dd>


<dt id="arg-transform">transform<a class="anchor" aria-label="anchor" href="#arg-transform"></a></dt>
<dd><p>DEPRECATED - A function that will tranform the counts, by default it is log1p for RNA sequencing data, but for avoinding tranformation you can use identity</p></dd>


<dt id="arg-inverse-transform">inverse_transform<a class="anchor" aria-label="anchor" href="#arg-inverse-transform"></a></dt>
<dd><p>DEPRECATED - A function that is the inverse of transform (e.g. expm1 is inverse of log1p). This is needed to tranform back the counts after analysis.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A consistent object (to the input) with additional columns for the adjusted counts as `&lt;COUNT COLUMN&gt;_adjusted`</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>`r lifecycle::badge("maturing")`</p>
<p>This function adjusts the abundance for (known) unwanted variation.
At the moment just an unwanted covariate is allowed at a time using Combat (DOI: 10.1093/bioinformatics/bts034)</p>
<p>Underlying method:
	sva::ComBat(data, batch = my_batch,	mod = design,	prior.plots = FALSE, ...)</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mangiola, S., Molania, R., Dong, R., Doyle, M. A., &amp; Papenfuss, A. T. (2021). tidybulk: an R tidy framework for modular transcriptomic data analysis. Genome Biology, 22(1), 42. doi:10.1186/s13059-020-02233-7</p>
<p>Zhang, Y., Parmigiani, G., &amp; Johnson, W. E. (2020). ComBat-seq: batch effect adjustment for RNA-seq count data. NAR Genomics and Bioinformatics, 2(3), lqaa078. doi:10.1093/nargab/lqaa078</p>
<p>Johnson, W. E., Li, C., &amp; Rabinovic, A. (2007). Adjusting batch effects in microarray expression data using empirical Bayes methods. Biostatistics, 8(1), 118–127. doi:10.1093/biostatistics/kxj037</p>
<p>Ritchie, M. E., Phipson, B., Wu, D., Hu, Y., Law, C. W., Shi, W., &amp; Smyth, G. K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research, 43(7), e47. doi:10.1093/nar/gkv007</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load airway dataset for examples</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'airway'</span>, package <span class="op">=</span> <span class="st">'airway'</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="co"># Ensure a 'condition' column exists for examples expecting it</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cm</span> <span class="op">=</span> <span class="va">airway</span></span></span>
<span class="r-in"><span><span class="co"># Create a balanced two-level batch within each condition to avoid confounding</span></span></span>
<span class="r-in"><span><span class="va">cond</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span></span></span>
<span class="r-in"><span><span class="va">cm</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA_character_</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">for</span> <span class="op">(</span><span class="va">lev</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cond</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">cond</span> <span class="op">==</span> <span class="va">lev</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">cm</span><span class="op">$</span><span class="va">batch</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'B'</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span><span class="va">cm</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">cm</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cm</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span><span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span><span class="fu">adjust_abundance</span><span class="op">(</span>  .factor_unwanted <span class="op">=</span> <span class="va">batch</span>, .factor_of_interest <span class="op">=</span>  <span class="va">condition</span>, method<span class="op">=</span><span class="st">"combat_seq"</span>  <span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All samples appear to belong to the same group.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Found 2 batches</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Using null model in ComBat-seq.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Adjusting for 1 covariate(s) or covariate level(s)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Estimating dispersions</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fitting the GLM model</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Shrinkage off - using GLM estimates for parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Adjusting the data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 63677 8 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(2): '' tidybulk</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(2): counts counts_adjusted</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ENSG00000273493</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(11): gene_id gene_name ... symbol .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(11): SampleName cell ... condition batch</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

