<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Adjust transcript abundance for unwanted variation — adjust_abundance • tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Adjust transcript abundance for unwanted variation — adjust_abundance" />
<meta property="og:description" content="adjust_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a consistent object (to the input) with an additional adjusted abundance column. This method uses scaled counts if present." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential feature abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stemangiola/tidybulk/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Adjust transcript abundance for unwanted variation</h1>
    <small class="dont-index">Source: <a href='https://github.com/stemangiola/tidybulk/blob/master/R/methods.R'><code>R/methods.R</code></a>, <a href='https://github.com/stemangiola/tidybulk/blob/master/R/methods_SE.R'><code>R/methods_SE.R</code></a></small>
    <div class="hidden name"><code>adjust_abundance-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>adjust_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a consistent object (to the input) with an additional adjusted abundance column. This method uses scaled counts if present.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>adjust_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for spec_tbl_df</span>
<span class='fu'>adjust_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for tbl_df</span>
<span class='fu'>adjust_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for tidybulk</span>
<span class='fu'>adjust_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>adjust_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for RangedSummarizedExperiment</span>
<span class='fu'>adjust_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></td>
    </tr>
    <tr>
      <th>.formula</th>
      <td><p>A formula with no response variable, representing the desired linear model where the first covariate is the factor of interest and the second covariate is the unwanted variation (of the kind ~ factor_of_interest + batch)</p></td>
    </tr>
    <tr>
      <th>.sample</th>
      <td><p>The name of the sample column</p></td>
    </tr>
    <tr>
      <th>.transcript</th>
      <td><p>The name of the transcript/gene column</p></td>
    </tr>
    <tr>
      <th>.abundance</th>
      <td><p>The name of the transcript/gene abundance column</p></td>
    </tr>
    <tr>
      <th>log_transform</th>
      <td><p>A boolean, whether the value should be log-transformed (e.g., TRUE for RNA sequencing data)</p></td>
    </tr>
    <tr>
      <th>action</th>
      <td><p>A character string. Whether to join the new information to the input tbl (add), or just get the non-redundant tbl with the new information (get).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Further parameters passed to the function sva::ComBat</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A consistent object (to the input) with additional columns for the adjusted counts as `&lt;COUNT COLUMN&gt;_adjusted`</p>
<p>A consistent object (to the input) with additional columns for the adjusted counts as `&lt;COUNT COLUMN&gt;_adjusted`</p>
<p>A consistent object (to the input) with additional columns for the adjusted counts as `&lt;COUNT COLUMN&gt;_adjusted`</p>
<p>A consistent object (to the input) with additional columns for the adjusted counts as `&lt;COUNT COLUMN&gt;_adjusted`</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>`r lifecycle::badge("maturing")`</p>
<p>This function adjusts the abundance for (known) unwanted variation.
At the moment just an unwanted covariate is allowed at a time using Combat (DOI: 10.1093/bioinformatics/bts034)</p>
<p>Underlying method:
	sva::ComBat(data, batch = my_batch,	mod = design,	prior.plots = FALSE, ...)</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'></span>
<span class='r-in'></span>
<span class='r-in'></span>
<span class='r-in'><span class='va'>cm</span> <span class='op'>=</span> <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='se_mini.html'>se_mini</a></span></span>
<span class='r-in'><span class='va'>cm</span><span class='op'>$</span><span class='va'>batch</span> <span class='op'>=</span> <span class='fl'>0</span></span>
<span class='r-in'><span class='va'>cm</span><span class='op'>$</span><span class='va'>batch</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>cm</span><span class='op'>)</span> <span class='op'><a href='https://rdrr.io/r/base/match.html'>%in%</a></span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"SRR1740035"</span>, <span class='st'>"SRR1740043"</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>=</span> <span class='fl'>1</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='va'>res</span> <span class='op'>=</span></span>
<span class='r-in'> <span class='va'>cm</span> <span class='op'><a href='pipe.html'>%&gt;%</a></span></span>
<span class='r-in'> <span class='fu'><a href='tidybulk-methods.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'><a href='pipe.html'>%&gt;%</a></span></span>
<span class='r-in'> <span class='fu'><a href='identify_abundant-methods.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'><a href='pipe.html'>%&gt;%</a></span></span>
<span class='r-in'>  <span class='fu'>adjust_abundance</span><span class='op'>(</span>  <span class='op'>~</span> <span class='va'>condition</span> <span class='op'>+</span> <span class='va'>batch</span>  <span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Found2batches</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Adjusting for1covariate(s) or covariate level(s)</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Standardizing Data across genes</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Fitting L/S model and finding priors</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Finding parametric adjustments</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Adjusting the Data</span>
<span class='r-in'></span>
<span class='r-in'></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Stefano Mangiola.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


