<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Identify abundant transcripts/genes — identify_abundant • tidybulk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Identify abundant transcripts/genes — identify_abundant"><meta property="og:description" content="Identifies transcripts/genes that are consistently expressed above a threshold across samples.
This function adds a logical column `.abundant` to indicate which features pass the filtering criteria."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.21.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identify abundant transcripts/genes</h1>
    <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/master/R/methods.R" class="external-link"><code>R/methods.R</code></a>, <a href="https://github.com/stemangiola/tidybulk/blob/master/R/methods_SE.R" class="external-link"><code>R/methods_SE.R</code></a></small>
    <div class="hidden name"><code>identify_abundant-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Identifies transcripts/genes that are consistently expressed above a threshold across samples.
This function adds a logical column `.abundant` to indicate which features pass the filtering criteria.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">identify_abundant</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minimum_counts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minimum_proportion <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'spec_tbl_df'</span></span>
<span><span class="fu">identify_abundant</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minimum_counts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minimum_proportion <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'tbl_df'</span></span>
<span><span class="fu">identify_abundant</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minimum_counts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minimum_proportion <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'tidybulk'</span></span>
<span><span class="fu">identify_abundant</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minimum_counts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minimum_proportion <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">identify_abundant</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minimum_counts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minimum_proportion <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'RangedSummarizedExperiment'</span></span>
<span><span class="fu">identify_abundant</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  factor_of_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  design <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minimum_counts <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minimum_proportion <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg--data">.data<a class="anchor" aria-label="anchor" href="#arg--data"></a></dt>
<dd><p>A `tbl` or `SummarizedExperiment` object containing transcript/gene abundance data</p></dd>


<dt id="arg--sample">.sample<a class="anchor" aria-label="anchor" href="#arg--sample"></a></dt>
<dd><p>The name of the sample column</p></dd>


<dt id="arg--transcript">.transcript<a class="anchor" aria-label="anchor" href="#arg--transcript"></a></dt>
<dd><p>The name of the transcript/gene column</p></dd>


<dt id="arg--abundance">.abundance<a class="anchor" aria-label="anchor" href="#arg--abundance"></a></dt>
<dd><p>The name of the transcript/gene abundance column</p></dd>


<dt id="arg-factor-of-interest">factor_of_interest<a class="anchor" aria-label="anchor" href="#arg-factor-of-interest"></a></dt>
<dd><p>The name of the column containing groups/conditions for filtering.
Used by edgeR's filterByExpr to define sample groups.</p></dd>


<dt id="arg-design">design<a class="anchor" aria-label="anchor" href="#arg-design"></a></dt>
<dd><p>A design matrix for more complex experimental designs. If provided,
this is passed to filterByExpr instead of factor_of_interest.</p></dd>


<dt id="arg-minimum-counts">minimum_counts<a class="anchor" aria-label="anchor" href="#arg-minimum-counts"></a></dt>
<dd><p>A positive number specifying the minimum counts per million (CPM) threshold
for a transcript to be considered abundant (default = 10)</p></dd>


<dt id="arg-minimum-proportion">minimum_proportion<a class="anchor" aria-label="anchor" href="#arg-minimum-proportion"></a></dt>
<dd><p>A number between 0 and 1 specifying the minimum proportion of samples
that must exceed the minimum_counts threshold (default = 0.7)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Returns the input object with an additional logical column `.abundant` indicating which
features passed the abundance threshold criteria.</p>
<p>A consistent object (to the input) with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A consistent object (to the input) with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A consistent object (to the input) with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>`r lifecycle::badge("maturing")`</p>

<p>This function uses edgeR's filterByExpr() function to identify consistently expressed features.
A feature is considered abundant if it has CPM &gt; minimum_counts in at least minimum_proportion
of samples in at least one experimental group (defined by factor_of_interest or design).</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>McCarthy, D. J., Chen, Y., &amp; Smyth, G. K. (2012). Differential expression analysis of
multifactor RNA-Seq experiments with respect to biological variation. Nucleic Acids Research,
40(10), 4288-4297. DOI: 10.1093/bioinformatics/btp616</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Basic usage</span></span></span>
<span class="r-in"><span><span class="va">se_mini</span> <span class="op">|&gt;</span> <span class="fu">identify_abundant</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 527 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): entrez .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># With custom thresholds</span></span></span>
<span class="r-in"><span><span class="va">se_mini</span> <span class="op">|&gt;</span> <span class="fu">identify_abundant</span><span class="op">(</span></span></span>
<span class="r-in"><span>  minimum_counts <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>  minimum_proportion <span class="op">=</span> <span class="fl">0.5</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 527 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): entrez .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Using a factor of interest</span></span></span>
<span class="r-in"><span><span class="va">se_mini</span> <span class="op">|&gt;</span> <span class="fu">identify_abundant</span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 527 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): entrez .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

