<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="X_cibersort.html">X_cibersort</a></code> </p>
        </td>
        <td><p>Cibersort reference</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="adjust_abundance-methods.html">adjust_abundance()</a></code> </p>
        </td>
        <td><p>Adjust transcript abundance for unwanted variation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aggregate_duplicates-methods.html">aggregate_duplicates()</a></code> </p>
        </td>
        <td><p>Aggregates multiple counts from the same samples (e.g., from isoforms), concatenates other character columns, and averages other numeric columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="as_matrix.html">as_matrix()</a></code> </p>
        </td>
        <td><p>Get matrix from tibble</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bind.html">bind</a></code> </p>
        </td>
        <td><p>Efficiently bind multiple data frames by row and column</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="breast_tcga_mini.html">breast_tcga_mini</a></code> </p>
        </td>
        <td><p>Data set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cluster_elements-methods.html">cluster_elements()</a></code> </p>
        </td>
        <td><p>Get clusters of elements (e.g., samples or transcripts)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="counts.html">counts</a></code> </p>
        </td>
        <td><p>Example data set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="counts_ensembl.html">counts_ensembl</a></code> </p>
        </td>
        <td><p>Counts with ensembl annotation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="counts_mini.html">counts_mini</a></code> </p>
        </td>
        <td><p>Example data set reduced</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deconvolve_cellularity-methods.html">deconvolve_cellularity()</a></code> </p>
        </td>
        <td><p>Get cell type proportions from samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="describe_transcript.html">describe_transcript()</a></code> </p>
        </td>
        <td><p>Get DESCRIPTION from gene SYMBOL for Human and Mouse</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="distinct.html">distinct()</a></code> </p>
        </td>
        <td><p>distinct</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dplyr-methods.html">arrange()</a></code> <code><a href="dplyr-methods.html">bind_rows()</a></code> <code><a href="dplyr-methods.html">bind_cols()</a></code> <code><a href="dplyr-methods.html">ungroup()</a></code> </p>
        </td>
        <td><p>drplyr-methods</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ensembl_symbol_mapping.html">ensembl_symbol_mapping</a></code> </p>
        </td>
        <td><p>Data set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ensembl_to_symbol-methods.html">ensembl_to_symbol()</a></code> </p>
        </td>
        <td><p>Add transcript symbol column from ensembl id for human and mouse data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fill_missing_abundance-methods.html">fill_missing_abundance()</a></code> </p>
        </td>
        <td><p>Fill transcript abundance if missing from sample-transcript pairs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter.html">filter()</a></code> </p>
        </td>
        <td><p>Subset rows using column values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flybaseIDs.html">flybaseIDs</a></code> </p>
        </td>
        <td><p>flybaseIDs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="full_join.html">full_join()</a></code> </p>
        </td>
        <td><p>Full join datasets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_bibliography-methods.html">get_bibliography()</a></code> </p>
        </td>
        <td><p>Produces the bibliography list of your workflow</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group_by.html">group_by()</a></code> </p>
        </td>
        <td><p>Group by one or more variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="identify_abundant-methods.html">identify_abundant()</a></code> </p>
        </td>
        <td><p>find abundant transcripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="impute_missing_abundance-methods.html">impute_missing_abundance()</a></code> </p>
        </td>
        <td><p>impute transcript abundance if missing from sample-transcript pairs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="inner_join.html">inner_join()</a></code> </p>
        </td>
        <td><p>Inner join datasets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="keep_abundant-methods.html">keep_abundant()</a></code> </p>
        </td>
        <td><p>Keep abundant transcripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="keep_variable-methods.html">keep_variable()</a></code> </p>
        </td>
        <td><p>Keep variable transcripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="left_join.html">left_join()</a></code> </p>
        </td>
        <td><p>Left join datasets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="log10_reverse_trans.html">log10_reverse_trans()</a></code> </p>
        </td>
        <td><p>log10_reverse_trans</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mutate.html">mutate()</a></code> </p>
        </td>
        <td><p>Create, modify, and delete columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_formula_survival.html">parse_formula_survival()</a></code> </p>
        </td>
        <td><p>Formula parser with survival</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pivot_sample-methods.html">pivot_sample()</a></code> </p>
        </td>
        <td><p>Extract sample-wise information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pivot_transcript-methods.html">pivot_transcript()</a></code> </p>
        </td>
        <td><p>Extract transcript-wise information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reduce_dimensions-methods.html">reduce_dimensions()</a></code> </p>
        </td>
        <td><p>Dimension reduction of the transcript abundance data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="remove_redundancy-methods.html">remove_redundancy()</a></code> </p>
        </td>
        <td><p>Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rename.html">rename()</a></code> </p>
        </td>
        <td><p>Rename columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="right_join.html">right_join()</a></code> </p>
        </td>
        <td><p>Right join datasets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rotate_dimensions-methods.html">rotate_dimensions()</a></code> </p>
        </td>
        <td><p>Rotate two dimensions (e.g., principal components) of an arbitrary angle</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rowwise.html">rowwise()</a></code> </p>
        </td>
        <td><p>Group input by rows</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scale_abundance-methods.html">scale_abundance()</a></code> </p>
        </td>
        <td><p>Scale the counts of transcripts/genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="se.html">se</a></code> </p>
        </td>
        <td><p>SummarizedExperiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="se_mini.html">se_mini</a></code> </p>
        </td>
        <td><p>SummarizedExperiment mini for vignette</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarise.html">summarise()</a></code> </p>
        </td>
        <td><p>Summarise each group to fewer rows</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="symbol_to_entrez.html">symbol_to_entrez()</a></code> </p>
        </td>
        <td><p>Get ENTREZ id from gene SYMBOL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_deseq2_df.html">test_deseq2_df</a></code> </p>
        </td>
        <td><p>SummarizedExperiment mini for vignette</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_differential_abundance-methods.html">test_differential_abundance()</a></code> </p>
        </td>
        <td><p>Perform differential transcription testing using edgeR, limma-voom or DESeq2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_differential_cellularity-methods.html">test_differential_cellularity()</a></code> </p>
        </td>
        <td><p>Add differential tissue composition information to a tbl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_gene_enrichment-methods.html">test_gene_enrichment()</a></code> </p>
        </td>
        <td><p>analyse gene enrichment with EGSEA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_gene_overrepresentation-methods.html">test_gene_overrepresentation()</a></code> </p>
        </td>
        <td><p>analyse gene over-representation with GSEA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tidybulk-methods.html">tidybulk()</a></code> </p>
        </td>
        <td><p>Creates a `tt` object from a `tbl` or `SummarizedExperiment` object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tidybulk_SAM_BAM-methods.html">tidybulk_SAM_BAM()</a></code> </p>
        </td>
        <td><p>Creates a `tt` object from a list of file names of BAM/SAM</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tidyr-methods.html">unnest()</a></code> <code><a href="tidyr-methods.html">nest()</a></code> </p>
        </td>
        <td><p>unnest</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


