<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Dimension reduction of the transcript abundance data — reduce_dimensions • tidybulk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Dimension reduction of the transcript abundance data — reduce_dimensions"><meta property="og:description" content="reduce_dimensions() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and calculates the reduced dimensional space of the transcript abundance."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/comparison_coding.html">Side-by-side comparison with standard interfaces</a>
    </li>
    <li>
      <a href="../articles/introduction.html">tidybulk: An R tidy framework for modular transcriptomic data analysis</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Dimension reduction of the transcript abundance data</h1>
    <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/master/R/reduce_dimensions.R" class="external-link"><code>R/reduce_dimensions.R</code></a></small>
    <div class="hidden name"><code>reduce_dimensions-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>reduce_dimensions() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and calculates the reduced dimensional space of the transcript abundance.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">reduce_dimensions</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  .dims <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">log1p</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">reduce_dimensions</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  .dims <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">log1p</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'RangedSummarizedExperiment'</span></span>
<span><span class="fu">reduce_dimensions</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  .dims <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">log1p</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg--data">.data<a class="anchor" aria-label="anchor" href="#arg--data"></a></dt>
<dd><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></dd>


<dt id="arg--element">.element<a class="anchor" aria-label="anchor" href="#arg--element"></a></dt>
<dd><p>The name of the element column (normally samples).</p></dd>


<dt id="arg--feature">.feature<a class="anchor" aria-label="anchor" href="#arg--feature"></a></dt>
<dd><p>The name of the feature column (normally transcripts/genes)</p></dd>


<dt id="arg--abundance">.abundance<a class="anchor" aria-label="anchor" href="#arg--abundance"></a></dt>
<dd><p>The name of the column including the numerical value the clustering is based on (normally transcript abundance)</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string. The dimension reduction algorithm to use (PCA, MDS, tSNE).</p></dd>


<dt id="arg--dims">.dims<a class="anchor" aria-label="anchor" href="#arg--dims"></a></dt>
<dd><p>An integer. The number of dimensions your are interested in (e.g., 4 for returning the first four principal components).</p></dd>


<dt id="arg-top">top<a class="anchor" aria-label="anchor" href="#arg-top"></a></dt>
<dd><p>An integer. How many top genes to select for dimensionality reduction</p></dd>


<dt id="arg-of-samples">of_samples<a class="anchor" aria-label="anchor" href="#arg-of-samples"></a></dt>
<dd><p>A boolean. In case the input is a tidybulk object, it indicates Whether the element column will be sample or transcript column</p></dd>


<dt id="arg-transform">transform<a class="anchor" aria-label="anchor" href="#arg-transform"></a></dt>
<dd><p>A function that will tranform the counts, by default it is log1p for RNA sequencing data, but for avoinding tranformation you can use identity</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>A boolean for method="PCA", this will be passed to the `prcomp` function. It is not included in the ... argument because although the default for `prcomp` if FALSE, it is advisable to set it as TRUE.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further parameters passed to the function prcomp if you choose method="PCA" or Rtsne if you choose method="tSNE", or uwot::tumap if you choose method="umap"</p></dd>


<dt id="arg-log-transform">log_transform<a class="anchor" aria-label="anchor" href="#arg-log-transform"></a></dt>
<dd><p>DEPRECATED - A boolean, whether the value should be log-transformed (e.g., TRUE for RNA sequencing data)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A tbl object with additional columns for the reduced dimensions</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>`r lifecycle::badge("maturing")`</p>
<p>This function reduces the dimensions of the transcript abundances.
It can use multi-dimensional scaling (MDS; DOI.org/10.1186/gb-2010-11-3-r25),
principal component analysis (PCA), or tSNE (Jesse Krijthe et al. 2018)</p>
<p>Underlying method for PCA:
prcomp(scale = scale, ...)</p>
<p>Underlying method for MDS:
limma::plotMDS(ndim = .dims, plot = FALSE, top = top)</p>
<p>Underlying method for tSNE:
Rtsne::Rtsne(data, ...)</p>
<p>Underlying method for UMAP:</p>
<p>df_source =
.data |&gt;</p>
<p># Filter NA symbol
  filter(!!.feature |&gt; is.na() |&gt; not()) |&gt;</p>
<p># Prepare data frame
  distinct(!!.feature,!!.element,!!.abundance) |&gt;</p>
<p># Filter most variable genes
  keep_variable_transcripts(top) |&gt;
  reduce_dimensions(method="PCA",  .dims = calculate_for_pca_dimensions ) |&gt;
  as_matrix(rownames = quo_name(.element)) |&gt;
  uwot::tumap(...)</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mangiola, S., Molania, R., Dong, R., Doyle, M. A., &amp; Papenfuss, A. T. (2021). tidybulk: an R tidy framework for modular transcriptomic data analysis. Genome Biology, 22(1), 42. doi:10.1186/s13059-020-02233-7</p>
<p>Krijthe, J. H. (2015). Rtsne: T-Distributed Stochastic Neighbor Embedding using a Barnes-Hut Implementation. R package version 0.15.</p>
<p>McInnes, L., Healy, J., &amp; Melville, J. (2018). UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction. arXiv preprint arXiv:1802.03426.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load airway dataset for examples</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'airway'</span>, package <span class="op">=</span> <span class="st">'airway'</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="co"># Ensure a 'condition' column exists for examples expecting it</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">dex</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">counts.MDS</span> <span class="op">=</span></span></span>
<span class="r-in"><span> <span class="va">airway</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu">reduce_dimensions</span><span class="op">(</span> method<span class="op">=</span><span class="st">"MDS"</span>, .dims <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All samples appear to belong to the same group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting the 500 most variable genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "MDS result_df colnames: sample, 1, 2, 3"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results do `metadata(.)$tidybulk$MDS`</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">counts.PCA</span> <span class="op">=</span></span></span>
<span class="r-in"><span> <span class="va">airway</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu">reduce_dimensions</span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span>, .dims <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All samples appear to belong to the same group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting the 500 most variable genes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fraction of variance explained by the selected principal components</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3 × 2</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   `Fraction of variance`    PC</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>                 0.065<span style="text-decoration: underline;">0</span>     1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span>                 0.045<span style="text-decoration: underline;">3</span>     2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span>                 0.023<span style="text-decoration: underline;">3</span>     3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results do `metadata(.)$tidybulk$PCA`</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

