<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated — remove_redundancy • tidybulk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated — remove_redundancy"><meta property="og:description" content="remove_redundancy() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) for correlation method or | &amp;lt;DIMENSION 1&amp;gt; | &amp;lt;DIMENSION 2&amp;gt; | &amp;lt;...&amp;gt; | for reduced_dimensions method, and returns a consistent object (to the input) with dropped elements (e.g., samples)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.6.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential feature abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated</h1>
    <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/HEAD/R/methods.R" class="external-link"><code>R/methods.R</code></a>, <a href="https://github.com/stemangiola/tidybulk/blob/HEAD/R/methods_SE.R" class="external-link"><code>R/methods_SE.R</code></a></small>
    <div class="hidden name"><code>remove_redundancy-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>remove_redundancy() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) for correlation method or | &lt;DIMENSION 1&gt; | &lt;DIMENSION 2&gt; | &lt;...&gt; | for reduced_dimensions method, and returns a consistent object (to the input) with dropped elements (e.g., samples).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">remove_redundancy</span><span class="op">(</span>
  <span class="va">.data</span>,
  .element <span class="op">=</span> <span class="cn">NULL</span>,
  .feature <span class="op">=</span> <span class="cn">NULL</span>,
  .abundance <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">method</span>,
  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,
  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,
  top <span class="op">=</span> <span class="cn">Inf</span>,
  log_transform <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">Dim_a_column</span>,
  <span class="va">Dim_b_column</span>
<span class="op">)</span>

<span class="co"># S4 method for spec_tbl_df</span>
<span class="fu">remove_redundancy</span><span class="op">(</span>
  <span class="va">.data</span>,
  .element <span class="op">=</span> <span class="cn">NULL</span>,
  .feature <span class="op">=</span> <span class="cn">NULL</span>,
  .abundance <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">method</span>,
  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,
  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,
  top <span class="op">=</span> <span class="cn">Inf</span>,
  log_transform <span class="op">=</span> <span class="cn">FALSE</span>,
  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,
  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span>

<span class="co"># S4 method for tbl_df</span>
<span class="fu">remove_redundancy</span><span class="op">(</span>
  <span class="va">.data</span>,
  .element <span class="op">=</span> <span class="cn">NULL</span>,
  .feature <span class="op">=</span> <span class="cn">NULL</span>,
  .abundance <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">method</span>,
  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,
  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,
  top <span class="op">=</span> <span class="cn">Inf</span>,
  log_transform <span class="op">=</span> <span class="cn">FALSE</span>,
  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,
  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span>

<span class="co"># S4 method for tidybulk</span>
<span class="fu">remove_redundancy</span><span class="op">(</span>
  <span class="va">.data</span>,
  .element <span class="op">=</span> <span class="cn">NULL</span>,
  .feature <span class="op">=</span> <span class="cn">NULL</span>,
  .abundance <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">method</span>,
  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,
  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,
  top <span class="op">=</span> <span class="cn">Inf</span>,
  log_transform <span class="op">=</span> <span class="cn">FALSE</span>,
  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,
  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span>

<span class="co"># S4 method for SummarizedExperiment</span>
<span class="fu">remove_redundancy</span><span class="op">(</span>
  <span class="va">.data</span>,
  .element <span class="op">=</span> <span class="cn">NULL</span>,
  .feature <span class="op">=</span> <span class="cn">NULL</span>,
  .abundance <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">method</span>,
  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,
  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,
  top <span class="op">=</span> <span class="cn">Inf</span>,
  log_transform <span class="op">=</span> <span class="cn">FALSE</span>,
  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,
  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span>

<span class="co"># S4 method for RangedSummarizedExperiment</span>
<span class="fu">remove_redundancy</span><span class="op">(</span>
  <span class="va">.data</span>,
  .element <span class="op">=</span> <span class="cn">NULL</span>,
  .feature <span class="op">=</span> <span class="cn">NULL</span>,
  .abundance <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">method</span>,
  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,
  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,
  top <span class="op">=</span> <span class="cn">Inf</span>,
  log_transform <span class="op">=</span> <span class="cn">FALSE</span>,
  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,
  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>.data</dt>
<dd><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></dd>
<dt>.element</dt>
<dd><p>The name of the element column (normally samples).</p></dd>
<dt>.feature</dt>
<dd><p>The name of the feature column (normally transcripts/genes)</p></dd>
<dt>.abundance</dt>
<dd><p>The name of the column including the numerical value the clustering is based on (normally transcript abundance)</p></dd>
<dt>method</dt>
<dd><p>A character string. The method to use, correlation and reduced_dimensions are available. The latter eliminates one of the most proximar pairs of samples in PCA reduced dimensions.</p></dd>
<dt>of_samples</dt>
<dd><p>A boolean. In case the input is a tidybulk object, it indicates Whether the element column will be sample or transcript column</p></dd>
<dt>correlation_threshold</dt>
<dd><p>A real number between 0 and 1. For correlation based calculation.</p></dd>
<dt>top</dt>
<dd><p>An integer. How many top genes to select for correlation based method</p></dd>
<dt>log_transform</dt>
<dd><p>A boolean, whether the value should be log-transformed (e.g., TRUE for RNA sequencing data)</p></dd>
<dt>Dim_a_column</dt>
<dd><p>A character string. For reduced_dimension based calculation. The column of one principal component</p></dd>
<dt>Dim_b_column</dt>
<dd><p>A character string. For reduced_dimension based calculation. The column of another principal component</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A tbl object with with dropped redundant elements (e.g., samples).
A tbl object with with dropped redundant elements (e.g., samples).
A tbl object with with dropped redundant elements (e.g., samples).
A tbl object with with dropped redundant elements (e.g., samples).
A `SummarizedExperiment` object
A `SummarizedExperiment` object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    


    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"> <span class="fu">tidybulk</span><span class="fu">::</span><span class="va"><a href="se_mini.html">se_mini</a></span> |&gt;</span>
<span class="r-in"> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> |&gt;</span>
<span class="r-in">   <span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span class="r-in">     .element <span class="op">=</span> <span class="va">sample</span>,</span>
<span class="r-in">     .feature <span class="op">=</span> <span class="va">transcript</span>,</span>
<span class="r-in">       .abundance <span class="op">=</span>  <span class="va">count</span>,</span>
<span class="r-in">       method <span class="op">=</span> <span class="st">"correlation"</span></span>
<span class="r-in">       <span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting the 182 most variable genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 527 4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(1): .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(4): SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(3): Cell.type time condition</span>
<span class="r-in"></span>
<span class="r-in"><span class="va">counts.MDS</span> <span class="op">=</span></span>
<span class="r-in"> <span class="fu">tidybulk</span><span class="fu">::</span><span class="va"><a href="se_mini.html">se_mini</a></span> |&gt;</span>
<span class="r-in"> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> |&gt;</span>
<span class="r-in">  <span class="fu"><a href="reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span> method<span class="op">=</span><span class="st">"MDS"</span>, .dims <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting the 182 most variable genes</span>
<span class="r-in"></span>
<span class="r-in"><span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">counts.MDS</span>,</span>
<span class="r-in">  Dim_a_column <span class="op">=</span> <span class="va">`Dim1`</span>,</span>
<span class="r-in">  Dim_b_column <span class="op">=</span> <span class="va">`Dim2`</span>,</span>
<span class="r-in">  .element <span class="op">=</span> <span class="va">sample</span>,</span>
<span class="r-in">  method <span class="op">=</span> <span class="st">"reduced_dimensions"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 527 3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(1): .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(3): SRR1740035 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(6): Cell.type time ... Dim2 Dim3</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

