<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated — remove_redundancy • tidybulk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated — remove_redundancy"><meta property="og:description" content="remove_redundancy() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) for correlation method or | &amp;lt;DIMENSION 1&amp;gt; | &amp;lt;DIMENSION 2&amp;gt; | &amp;lt;...&amp;gt; | for reduced_dimensions method, and returns a consistent object (to the input) with dropped elements (e.g., samples)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated</h1>
    <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/resolve-confounders-without-altering-original-columns/R/methods.R" class="external-link"><code>R/methods.R</code></a>, <a href="https://github.com/stemangiola/tidybulk/blob/resolve-confounders-without-altering-original-columns/R/methods_SE.R" class="external-link"><code>R/methods_SE.R</code></a></small>
    <div class="hidden name"><code>remove_redundancy-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>remove_redundancy() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) for correlation method or | &lt;DIMENSION 1&gt; | &lt;DIMENSION 2&gt; | &lt;...&gt; | for reduced_dimensions method, and returns a consistent object (to the input) with dropped elements (e.g., samples).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">identity</span>,</span>
<span>  <span class="va">Dim_a_column</span>,</span>
<span>  <span class="va">Dim_b_column</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'spec_tbl_df'</span></span>
<span><span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">identity</span>,</span>
<span>  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'tbl_df'</span></span>
<span><span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">identity</span>,</span>
<span>  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'tidybulk'</span></span>
<span><span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">identity</span>,</span>
<span>  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">identity</span>,</span>
<span>  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'RangedSummarizedExperiment'</span></span>
<span><span class="fu">remove_redundancy</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  .element <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .feature <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">method</span>,</span>
<span>  of_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  correlation_threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  transform <span class="op">=</span> <span class="va">identity</span>,</span>
<span>  Dim_a_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Dim_b_column <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log_transform <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg--data">.data<a class="anchor" aria-label="anchor" href="#arg--data"></a></dt>
<dd><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></dd>


<dt id="arg--element">.element<a class="anchor" aria-label="anchor" href="#arg--element"></a></dt>
<dd><p>The name of the element column (normally samples).</p></dd>


<dt id="arg--feature">.feature<a class="anchor" aria-label="anchor" href="#arg--feature"></a></dt>
<dd><p>The name of the feature column (normally transcripts/genes)</p></dd>


<dt id="arg--abundance">.abundance<a class="anchor" aria-label="anchor" href="#arg--abundance"></a></dt>
<dd><p>The name of the column including the numerical value the clustering is based on (normally transcript abundance)</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string. The method to use, correlation and reduced_dimensions are available. The latter eliminates one of the most proximar pairs of samples in PCA reduced dimensions.</p></dd>


<dt id="arg-of-samples">of_samples<a class="anchor" aria-label="anchor" href="#arg-of-samples"></a></dt>
<dd><p>A boolean. In case the input is a tidybulk object, it indicates Whether the element column will be sample or transcript column</p></dd>


<dt id="arg-correlation-threshold">correlation_threshold<a class="anchor" aria-label="anchor" href="#arg-correlation-threshold"></a></dt>
<dd><p>A real number between 0 and 1. For correlation based calculation.</p></dd>


<dt id="arg-top">top<a class="anchor" aria-label="anchor" href="#arg-top"></a></dt>
<dd><p>An integer. How many top genes to select for correlation based method</p></dd>


<dt id="arg-transform">transform<a class="anchor" aria-label="anchor" href="#arg-transform"></a></dt>
<dd><p>A function that will tranform the counts, by default it is log1p for RNA sequencing data, but for avoinding tranformation you can use identity</p></dd>


<dt id="arg-dim-a-column">Dim_a_column<a class="anchor" aria-label="anchor" href="#arg-dim-a-column"></a></dt>
<dd><p>A character string. For reduced_dimension based calculation. The column of one principal component</p></dd>


<dt id="arg-dim-b-column">Dim_b_column<a class="anchor" aria-label="anchor" href="#arg-dim-b-column"></a></dt>
<dd><p>A character string. For reduced_dimension based calculation. The column of another principal component</p></dd>


<dt id="arg-log-transform">log_transform<a class="anchor" aria-label="anchor" href="#arg-log-transform"></a></dt>
<dd><p>DEPRECATED - A boolean, whether the value should be log-transformed (e.g., TRUE for RNA sequencing data)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>`r lifecycle::badge("maturing")`</p>
<p>This function removes redundant elements from the original data set (e.g., samples or transcripts).
For example, if we want to define cell-type specific signatures with low sample redundancy.
This function returns a tibble with dropped redundant elements (e.g., samples).
Two redundancy estimation approaches are supported:
(i) removal of highly correlated clusters of elements (keeping a representative) with method="correlation";
(ii) removal of most proximal element pairs in a reduced dimensional space.</p>
<p>Underlying method for correlation:
widyr::pairwise_cor(sample, transcript,count, sort = TRUE, diag = FALSE, upper = FALSE)</p>
<p>Underlying custom method for reduced dimensions:
select_closest_pairs = function(df)
		couples &lt;- df |&gt; head(n = 0)		while (df |&gt; nrow() &gt; 0)
			pair &lt;- df |&gt;
			arrange(dist) |&gt;
			head(n = 1)
			couples &lt;- couples |&gt; bind_rows(pair)
			df &lt;- df |&gt;
				filter(
					!`sample 1` <!-- %in% (pair |&gt; select(1:2) |&gt; as.character()) &amp; -->
						!`sample 2` <!-- %in% (pair |&gt; select(1:2) |&gt; as.character()) -->
				)		couples</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="va"><a href="se_mini.html">se_mini</a></span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>   <span class="fu">remove_redundancy</span><span class="op">(</span></span></span>
<span class="r-in"><span>     .element <span class="op">=</span> <span class="va">sample</span>,</span></span>
<span class="r-in"><span>     .feature <span class="op">=</span> <span class="va">transcript</span>,</span></span>
<span class="r-in"><span>       .abundance <span class="op">=</span>  <span class="va">count</span>,</span></span>
<span class="r-in"><span>       method <span class="op">=</span> <span class="st">"correlation"</span></span></span>
<span class="r-in"><span>       <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting the 182 most variable genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 527 4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): entrez .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(4): SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">counts.MDS</span> <span class="op">=</span></span></span>
<span class="r-in"><span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="va"><a href="se_mini.html">se_mini</a></span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span> method<span class="op">=</span><span class="st">"MDS"</span>, .dims <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting the 182 most variable genes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results do `attr(..., "internals")$MDS`</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">remove_redundancy</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">counts.MDS</span>,</span></span>
<span class="r-in"><span>  Dim_a_column <span class="op">=</span> <span class="va">`Dim1`</span>,</span></span>
<span class="r-in"><span>  Dim_b_column <span class="op">=</span> <span class="va">`Dim2`</span>,</span></span>
<span class="r-in"><span>  .element <span class="op">=</span> <span class="va">sample</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"reduced_dimensions"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 527 3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): entrez .abundant</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(3): SRR1740035 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(8): Cell.type time ... Dim2 Dim3</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

