<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2 — test_differential_abundance • tidybulk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2 — test_differential_abundance"><meta property="og:description" content="test_differential_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a consistent object (to the input) with additional columns for the statistics from the hypothesis test."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.19.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2</h1>
    <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/master/R/methods.R" class="external-link"><code>R/methods.R</code></a>, <a href="https://github.com/stemangiola/tidybulk/blob/master/R/methods_SE.R" class="external-link"><code>R/methods_SE.R</code></a></small>
    <div class="hidden name"><code>test_differential_abundance-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>test_differential_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a consistent object (to the input) with additional columns for the statistics from the hypothesis test.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">test_differential_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.formula</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,</span>
<span>  test_above_log2_fold_change <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  omit_contrast_in_colnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  action <span class="op">=</span> <span class="st">"add"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fill_missing_values <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .contrasts <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'spec_tbl_df'</span></span>
<span><span class="fu">test_differential_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.formula</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,</span>
<span>  test_above_log2_fold_change <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  omit_contrast_in_colnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  action <span class="op">=</span> <span class="st">"add"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fill_missing_values <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .contrasts <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'tbl_df'</span></span>
<span><span class="fu">test_differential_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.formula</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,</span>
<span>  test_above_log2_fold_change <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  omit_contrast_in_colnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  action <span class="op">=</span> <span class="st">"add"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fill_missing_values <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .contrasts <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'tidybulk'</span></span>
<span><span class="fu">test_differential_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.formula</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,</span>
<span>  test_above_log2_fold_change <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  omit_contrast_in_colnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  action <span class="op">=</span> <span class="st">"add"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fill_missing_values <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .contrasts <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">test_differential_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.formula</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,</span>
<span>  test_above_log2_fold_change <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  omit_contrast_in_colnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  action <span class="op">=</span> <span class="st">"add"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fill_missing_values <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .contrasts <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'RangedSummarizedExperiment'</span></span>
<span><span class="fu">test_differential_abundance</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.formula</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .abundance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,</span>
<span>  test_above_log2_fold_change <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  omit_contrast_in_colnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  action <span class="op">=</span> <span class="st">"add"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  significance_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fill_missing_values <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  .contrasts <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg--data">.data<a class="anchor" aria-label="anchor" href="#arg--data"></a></dt>
<dd><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></dd>


<dt id="arg--formula">.formula<a class="anchor" aria-label="anchor" href="#arg--formula"></a></dt>
<dd><p>A formula representing the desired linear model. If there is more than one factor, they should be in the order factor of interest + additional factors.</p></dd>


<dt id="arg--sample">.sample<a class="anchor" aria-label="anchor" href="#arg--sample"></a></dt>
<dd><p>The name of the sample column</p></dd>


<dt id="arg--transcript">.transcript<a class="anchor" aria-label="anchor" href="#arg--transcript"></a></dt>
<dd><p>The name of the transcript/gene column</p></dd>


<dt id="arg--abundance">.abundance<a class="anchor" aria-label="anchor" href="#arg--abundance"></a></dt>
<dd><p>The name of the transcript/gene abundance column</p></dd>


<dt id="arg-contrasts">contrasts<a class="anchor" aria-label="anchor" href="#arg-contrasts"></a></dt>
<dd><p>This parameter takes the format of the contrast parameter of the method of choice. For edgeR and limma-voom is a character vector. For DESeq2 is a list including a character vector of length three. The first covariate is the one the model is tested against (e.g., ~ factor_of_interest)</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A string character. Either "edgeR_quasi_likelihood" (i.e., QLF), "edgeR_likelihood_ratio" (i.e., LRT), "edger_robust_likelihood_ratio", "DESeq2", "limma_voom", "limma_voom_sample_weights", "glmmseq_lme4", "glmmseq_glmmtmb"</p></dd>


<dt id="arg-test-above-log-fold-change">test_above_log2_fold_change<a class="anchor" aria-label="anchor" href="#arg-test-above-log-fold-change"></a></dt>
<dd><p>A positive real value. This works for edgeR and limma_voom methods. It uses the `treat` function, which tests that the difference in abundance is bigger than this threshold rather than zero <a href="https://pubmed.ncbi.nlm.nih.gov/19176553" class="external-link">https://pubmed.ncbi.nlm.nih.gov/19176553</a>.</p></dd>


<dt id="arg-scaling-method">scaling_method<a class="anchor" aria-label="anchor" href="#arg-scaling-method"></a></dt>
<dd><p>A character string. The scaling method passed to the back-end functions: edgeR and limma-voom (i.e., edgeR::calcNormFactors; "TMM","TMMwsp","RLE","upperquartile"). Setting the parameter to \"none\" will skip the compensation for sequencing-depth for the method edgeR or limma-voom.</p></dd>


<dt id="arg-omit-contrast-in-colnames">omit_contrast_in_colnames<a class="anchor" aria-label="anchor" href="#arg-omit-contrast-in-colnames"></a></dt>
<dd><p>If just one contrast is specified you can choose to omit the contrast label in the colnames.</p></dd>


<dt id="arg-prefix">prefix<a class="anchor" aria-label="anchor" href="#arg-prefix"></a></dt>
<dd><p>A character string. The prefix you would like to add to the result columns. It is useful if you want to compare several methods.</p></dd>


<dt id="arg-action">action<a class="anchor" aria-label="anchor" href="#arg-action"></a></dt>
<dd><p>A character string. Whether to join the new information to the input tbl (add), or just get the non-redundant tbl with the new information (get).</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to some of the internal experimental functions. For example for glmmSeq, it is possible to pass .dispersion, and .scaling_factor column tidyeval to skip the caluclation of dispersion and scaling and use precalculated values. This is helpful is you want to calculate those quantities on many genes and do DE testing on fewer genes. .scaling_factor is the TMM value that can be obtained with tidybulk::scale_abundance.</p></dd>


<dt id="arg-significance-threshold">significance_threshold<a class="anchor" aria-label="anchor" href="#arg-significance-threshold"></a></dt>
<dd><p>DEPRECATED - A real between 0 and 1 (usually 0.05).</p></dd>


<dt id="arg-fill-missing-values">fill_missing_values<a class="anchor" aria-label="anchor" href="#arg-fill-missing-values"></a></dt>
<dd><p>DEPRECATED - A boolean. Whether to fill missing sample/transcript values with the median of the transcript. This is rarely needed.</p></dd>


<dt id="arg--contrasts">.contrasts<a class="anchor" aria-label="anchor" href="#arg--contrasts"></a></dt>
<dd><p>DEPRECATED - This parameter takes the format of the contrast parameter of the method of choice. For edgeR and limma-voom is a character vector. For DESeq2 is a list including a character vector of length three. The first covariate is the one the model is tested against (e.g., ~ factor_of_interest)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A consistent object (to the input) with additional columns for the statistics from the test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A consistent object (to the input) with additional columns for the statistics from the test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A consistent object (to the input) with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A consistent object (to the input) with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>`r lifecycle::badge("maturing")`</p>
<p>This function provides the option to use edgeR <a href="https://doi.org/10.1093/bioinformatics/btp616" class="external-link">https://doi.org/10.1093/bioinformatics/btp616</a>, limma-voom <a href="https://doi.org/10.1186/gb-2014-15-2-r29" class="external-link">https://doi.org/10.1186/gb-2014-15-2-r29</a>, limma_voom_sample_weights <a href="https://doi.org/10.1093/nar/gkv412" class="external-link">https://doi.org/10.1093/nar/gkv412</a> or  DESeq2 <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">https://doi.org/10.1186/s13059-014-0550-8</a> to perform the testing.
All methods use raw counts, irrespective of if scale_abundance or adjust_abundance have been calculated, therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</p>
<p>Underlying method for edgeR framework:</p>
<p>.data |&gt;</p>
<p># Filter
keep_abundant(
		factor_of_interest = !!(as.symbol(parse_formula(.formula)[1])),
		minimum_counts = minimum_counts,
		minimum_proportion = minimum_proportion
	) |&gt;</p>
<p># Format
		select(!!.transcript,!!.sample,!!.abundance) |&gt;
		spread(!!.sample,!!.abundance) |&gt;
		as_matrix(rownames = !!.transcript) <!-- %&gt;% --></p>
<p># edgeR
		edgeR::DGEList(counts = .) |&gt;
		edgeR::calcNormFactors(method = scaling_method) |&gt;
		edgeR::estimateDisp(design) |&gt;</p>
<p># Fit
		edgeR::glmQLFit(design) |&gt; // or glmFit according to choice
		edgeR::glmQLFTest(coef = 2, contrast = my_contrasts) // or glmLRT according to choice</p>

<p>Underlying method for DESeq2 framework:</p>
<p>keep_abundant(
		factor_of_interest = !!as.symbol(parse_formula(.formula)[[1]]),
		minimum_counts = minimum_counts,
		minimum_proportion = minimum_proportion
) |&gt;</p>
<p># DESeq2
DESeq2::DESeqDataSet(design = .formula) |&gt;
DESeq2::DESeq() |&gt;
DESeq2::results()</p>


<p>Underlying method for glmmSeq framework:</p>
<p>counts =
.data <!-- %&gt;% -->
  assay(my_assay)</p>
<p># Create design matrix for dispersion, removing random effects
design =
  model.matrix(
    object = .formula |&gt; lme4::nobars(),
    data = metadata
  )</p>
<p>dispersion = counts |&gt; edgeR::estimateDisp(design = design) <!-- %$% tagwise.dispersion |&gt; setNames(rownames(counts)) --></p>
<p>glmmSeq( .formula,
           countdata = counts ,
           metadata =   metadata |&gt; as.data.frame(),
           dispersion = dispersion,
           progress = TRUE,
           method = method |&gt; str_remove("(?i)^glmmSeq_" ),
  )</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="co"># edgeR</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="va"><a href="se_mini.html">se_mini</a></span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu">test_differential_abundance</span><span class="op">(</span> <span class="op">~</span> <span class="va">condition</span> <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> =====================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> or adjust_abundance have been calculated. Therefore, it is essential to add covariates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> such as batch effects (if applicable) in the formula.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> =====================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">This message is displayed once per session.</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: The design column names are "(Intercept), conditionTRUE"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR_quasi_likelihood`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 182 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(182): ACAP1 ACP5 ... ZNF286A ZNF324</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(7): entrez .abundant ... PValue FDR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="co"># The function `test_differential_abundance` operates with contrasts too</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="va"><a href="se_mini.html">se_mini</a></span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu">test_differential_abundance</span><span class="op">(</span></span></span>
<span class="r-in"><span>      <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">condition</span>,</span></span>
<span class="r-in"><span>      contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"conditionTRUE - conditionFALSE"</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: The design column names are "conditionFALSE, conditionTRUE"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR_quasi_likelihood`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 394 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(394): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(7): entrez .abundant ... PValue___conditionTRUE -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   conditionFALSE FDR___conditionTRUE - conditionFALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="co"># DESeq2 - equivalent for limma-voom</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">my_se_mini</span> <span class="op">=</span> <span class="fu">tidybulk</span><span class="fu">::</span><span class="va"><a href="se_mini.html">se_mini</a></span></span></span>
<span class="r-in"><span><span class="va">my_se_mini</span><span class="op">$</span><span class="va">condition</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">my_se_mini</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># demontrating with `fitType` that you can access any arguments to DESeq()</span></span></span>
<span class="r-in"><span><span class="va">my_se_mini</span>  <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>   <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>       <span class="fu">test_differential_abundance</span><span class="op">(</span> <span class="op">~</span> <span class="va">condition</span>, method<span class="op">=</span><span class="st">"deseq2"</span>, fitType<span class="op">=</span><span class="st">"local"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$deseq2`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 394 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(394): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(8): entrez .abundant ... pvalue padj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># testing above a log2 threshold, passes along value to lfcThreshold of results()</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">my_se_mini</span>  <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>   <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>        <span class="fu">test_differential_abundance</span><span class="op">(</span> <span class="op">~</span> <span class="va">condition</span>, method<span class="op">=</span><span class="st">"deseq2"</span>,</span></span>
<span class="r-in"><span>            fitType<span class="op">=</span><span class="st">"local"</span>,</span></span>
<span class="r-in"><span>            test_above_log2_fold_change<span class="op">=</span><span class="fl">4</span> <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$deseq2`</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Use random intercept and random effect models</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="va">se_mini</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span>,<span class="op">]</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu">test_differential_abundance</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="op">~</span> <span class="va">condition</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">condition</span> <span class="op">|</span> <span class="va">time</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    method <span class="op">=</span> <span class="st">"glmmseq_lme4"</span>, cores <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> n = 5 samples, 2 individuals</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 32.07547 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$glmmseq_lme4`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 40 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(40): ABCB4 ABCB9 ... C5AR1 C5AR2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(26): entrez .abundant ... P_condition</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   P_condition_adjusted</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># confirm that lfcThreshold was used</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span>    <span class="va">res</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>        <span class="fu">mcols</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>        <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqResults.html" class="external-link">DESeqResults</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>        <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotMA.html" class="external-link">plotMA</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The function `test_differential_abundance` operates with contrasts too</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="va">my_se_mini</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span> <span class="fu">test_differential_abundance</span><span class="op">(</span></span></span>
<span class="r-in"><span>      <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">condition</span>,</span></span>
<span class="r-in"><span>      contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"TRUE"</span>, <span class="st">"FALSE"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>      method<span class="op">=</span><span class="st">"deseq2"</span>,</span></span>
<span class="r-in"><span>         fitType<span class="op">=</span><span class="st">"local"</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$deseq2`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 182 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(182): ACAP1 ACP5 ... ZNF286A ZNF324</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(8): entrez .abundant ... pvalue___condition TRUE-FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   padj___condition TRUE-FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(5): Cell.type time condition days dead</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

