<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>analyse gene over-representation with GSEA — test_gene_overrepresentation • tidybulk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="analyse gene over-representation with GSEA — test_gene_overrepresentation"><meta property="og:description" content="test_gene_overrepresentation() takes as input a `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a `tbl` with the GSEA statistics"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/comparison_coding.html">Side-by-side comparison with standard interfaces</a>
    </li>
    <li>
      <a href="../articles/introduction.html">tidybulk: An R tidy framework for modular transcriptomic data analysis</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>analyse gene over-representation with GSEA</h1>
    <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/master/R/test_gene_overrepresentation.R" class="external-link"><code>R/test_gene_overrepresentation.R</code></a></small>
    <div class="hidden name"><code>test_gene_overrepresentation-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>test_gene_overrepresentation() takes as input a `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a `tbl` with the GSEA statistics</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">test_gene_overrepresentation</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.entrez</span>,</span>
<span>  <span class="va">.do_test</span>,</span>
<span>  <span class="va">species</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_sets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_set <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">test_gene_overrepresentation</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.entrez</span>,</span>
<span>  <span class="va">.do_test</span>,</span>
<span>  <span class="va">species</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_sets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_set <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'RangedSummarizedExperiment'</span></span>
<span><span class="fu">test_gene_overrepresentation</span><span class="op">(</span></span>
<span>  <span class="va">.data</span>,</span>
<span>  <span class="va">.entrez</span>,</span>
<span>  <span class="va">.do_test</span>,</span>
<span>  <span class="va">species</span>,</span>
<span>  .sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_sets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_set <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg--data">.data<a class="anchor" aria-label="anchor" href="#arg--data"></a></dt>
<dd><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></dd>


<dt id="arg--entrez">.entrez<a class="anchor" aria-label="anchor" href="#arg--entrez"></a></dt>
<dd><p>The ENTREZ ID of the transcripts/genes</p></dd>


<dt id="arg--do-test">.do_test<a class="anchor" aria-label="anchor" href="#arg--do-test"></a></dt>
<dd><p>A boolean column name symbol. It indicates the transcript to check</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>A character. For example, human or mouse. MSigDB uses the latin species names (e.g., \"Mus musculus\", \"Homo sapiens\")</p></dd>


<dt id="arg--sample">.sample<a class="anchor" aria-label="anchor" href="#arg--sample"></a></dt>
<dd><p>The name of the sample column</p></dd>


<dt id="arg-gene-sets">gene_sets<a class="anchor" aria-label="anchor" href="#arg-gene-sets"></a></dt>
<dd><p>A character vector. The subset of MSigDB datasets you want to test against (e.g. \"C2\"). If NULL all gene sets are used (suggested). This argument was added to avoid time overflow of the examples.</p></dd>


<dt id="arg-gene-set">gene_set<a class="anchor" aria-label="anchor" href="#arg-gene-set"></a></dt>
<dd><p>DEPRECATED. Use gene_sets instead.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A consistent object (to the input)</p>
<p>A `SummarizedExperiment` object</p>
<p>A `RangedSummarizedExperiment` object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>`r lifecycle::badge("maturing")`</p>
<p>This wrapper execute gene enrichment analyses of the dataset using a list of transcripts and GSEA.
This wrapper uses clusterProfiler (DOI: doi.org/10.1089/omi.2011.0118) on the back-end.</p>
<p># Get MSigDB data
msigdb_data = msigdbr::msigdbr(species = species)</p>
<p># Filter for specific gene collections if provided</p>
<p>msigdb_data = filter(msigdb_data, gs_collection <!-- %in% gene_collections) --></p>

<p># Process the data
msigdb_data |&gt;
  nest(data = -gs_collection) |&gt;
  mutate(test =
           map(
             data,
             ~ clusterProfiler::enricher(
               my_entrez_rank,
               TERM2GENE=.x |&gt; select(gs_name, ncbi_gene),
               pvalueCutoff = 1
             ) |&gt;
               as_tibble()
           ))</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mangiola, S., Molania, R., Dong, R., Doyle, M. A., &amp; Papenfuss, A. T. (2021). tidybulk: an R tidy framework for modular transcriptomic data analysis. Genome Biology, 22(1), 42. doi:10.1186/s13059-020-02233-7</p>
<p>Yu, G., Wang, L. G., Han, Y., &amp; He, Q. Y. (2012). clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology, 16(5), 284-287. doi:10.1089/omi.2011.0118</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load airway dataset for examples</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'airway'</span>, package <span class="op">=</span> <span class="st">'airway'</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="co"># Ensure a 'condition' column exists for examples expecting it</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">dex</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Not run for build time."</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Not run for build time."</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># airway = airway[!rowData(airway)$entrez |&gt; is.na(),] |&gt; aggregate_duplicates(.transcript = entrez)</span></span></span>
<span class="r-in"><span><span class="co"># df_entrez = mutate(df_entrez, do_test = feature %in% c("TNFRSF4", "PLCH2", "PADI4", "PAX7"))</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span>  <span class="fu">test_gene_overrepresentation</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">df_entrez</span>,</span></span>
<span class="r-in"><span>    .sample <span class="op">=</span> <span class="va">sample</span>,</span></span>
<span class="r-in"><span>    .entrez <span class="op">=</span> <span class="va">entrez</span>,</span></span>
<span class="r-in"><span>    .do_test <span class="op">=</span> <span class="va">do_test</span>,</span></span>
<span class="r-in"><span>    species<span class="op">=</span><span class="st">"Homo sapiens"</span>,</span></span>
<span class="r-in"><span>   gene_sets <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C2"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

