<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Side-by-side comparison with standard interfaces • tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Side-by-side comparison with standard interfaces">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparison_coding.html">Side-by-side comparison with standard interfaces</a>
    </li>
    <li>
      <a href="../articles/introduction.html">tidybulk: An R tidy framework for modular transcriptomic data analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Side-by-side comparison with standard
interfaces</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2025-09-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/master/vignettes/comparison_coding.Rmd" class="external-link"><code>vignettes/comparison_coding.Rmd</code></a></small>
      <div class="hidden name"><code>comparison_coding.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Tidybulk is a comprehensive R package for modular
      transcriptomic data analysis that brings transcriptomics to the
      tidyverse. It provides a unified interface for data
      transformation, normalization, filtering, dimensionality
      reduction, clustering, differential analysis, cellularity
      analysis, and gene enrichment with seamless integration of
      SummarizedExperiment objects and tidyverse principles.</p>
    </div>
    
<!-- <style type="text/css"> -->
<!-- .main-container { -->
<!--   max-width: 2800px;!important -->
<!--   margin-left: auto; -->
<!--   margin-right: auto; -->
<!-- } -->
<!-- </style> -->
<style>
.column-left{
  float: left;
  width: 50%;
  text-align: left;
}
.column-right{
  float: right;
  width: 50%;
  text-align: right;
}
</style>
<!-- badges: start --><p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidybulk/actions/" class="external-link"><img src="https://github.com/stemangiola/tidybulk/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <a href="https://bioconductor.org/checkResults/release/bioc-LATEST/tidybulk/" class="external-link"><img src="https://bioconductor.org/shields/build/release/bioc/tidybulk.svg" alt="Bioconductor status"></a> <!-- badges: end --></p>
<p><strong>tidybulk</strong> is a powerful R package designed for
modular transcriptomic data analysis that brings transcriptomics to the
tidyverse.</p>
<div class="section level2">
<h2 id="why-tidybulk">Why tidybulk?<a class="anchor" aria-label="anchor" href="#why-tidybulk"></a>
</h2>
<p>Tidybulk provides a unified interface for comprehensive
transcriptomic data analysis with seamless integration of
SummarizedExperiment objects and tidyverse principles. It streamlines
the entire workflow from raw data to biological insights.</p>
<div class="section level3">
<h3 id="scientific-citation">Scientific Citation<a class="anchor" aria-label="anchor" href="#scientific-citation"></a>
</h3>
<p>Mangiola, Stefano, Ramyar Molania, Ruining Dong, Maria A. Doyle, and
Anthony T. Papenfuss. 2021. “Tidybulk: An R tidy framework for modular
transcriptomic data analysis.” Genome Biology 22 (42). <a href="https://doi.org/10.1186/s13059-020-02233-7" class="external-link uri">https://doi.org/10.1186/s13059-020-02233-7</a></p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7" class="external-link">Genome
Biology - tidybulk: an R tidy framework for modular transcriptomic data
analysis</a></p>
</div>
</div>
<div class="section level2">
<h2 id="aggregate-duplicated-transcripts">Aggregate duplicated <code>transcripts</code><a class="anchor" aria-label="anchor" href="#aggregate-duplicated-transcripts"></a>
</h2>
<p>tidybulk provide the <code>aggregate_duplicates</code> function to
aggregate duplicated transcripts (e.g., isoforms, ensembl). For example,
we often have to convert ensembl symbols to gene/transcript symbol, but
in doing so we have to deal with duplicates.
<code>aggregate_duplicates</code> takes a tibble and column names (as
symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) as arguments and returns a tibble with transcripts
with the same name aggregated. All the rest of the columns are appended,
and factors and boolean are appended as characters.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="va">airway.aggr</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>.transcript <span class="op">=</span> <span class="va">gene_name</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    symbol <span class="op">=</span> <span class="va">dge_list</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">symbol</span>,</span>
<span>    <span class="va">dge_list</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/by.html" class="external-link">by</a></span><span class="op">(</span><span class="va">temp</span>, <span class="va">temp</span><span class="op">$</span><span class="va">symbol</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span></span>
<span>            <span class="fu">matrixStats</span><span class="fu">:::</span><span class="fu">colSums</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="va">dge_list.nr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dge_list.nr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dge_list</span><span class="op">)</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level2">
<h2 id="scale-counts">Scale <code>counts</code><a class="anchor" aria-label="anchor" href="#scale-counts"></a>
</h2>
<p>We may want to compensate for sequencing depth, scaling the
transcript abundance (e.g., with TMM algorithm, Robinson and Oshlack
doi.org/10.1186/gb-2010-11-3-r25). <code>scale_abundance</code> takes a
tibble, column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and a method as
arguments and returns a tibble with additional columns with scaled data
as <code>&lt;NAME OF COUNT COLUMN&gt;_scaled</code>.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `factor_of_interest` argument of `identify_abundant()` is deprecated as of</span></span>
<span><span class="co">## tidybulk 2.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `formula_design` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The argument 'factor_of_interest' is deprecated and will be removed in a</span></span>
<span><span class="co">##   future release. Please use the 'design' or 'formula_design' argument instead.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: the sample with largest library size SRR1039517 was chosen as reference for scaling</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span>count_m<span class="op">=</span><span class="va">x</span>,group<span class="op">=</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep</span>,,keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="co"># ... additional processing steps ...</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span>, method<span class="op">=</span><span class="st">"TMM"</span><span class="op">)</span></span>
<span><span class="va">norm_counts.table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>We can easily plot the scaled density to check the scaling outcome.
On the x axis we have the log scaled counts, on the y axes we have the
density, data is grouped by sample and coloured by treatment.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">counts_scaled</span> <span class="op">+</span> <span class="fl">1</span>, group<span class="op">=</span><span class="va">.sample</span>, color<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span></span></code></pre></div>
<p><img src="comparison_coding_files/figure-html/plot_normalise-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="filter-variable-transcripts">Filter <code>variable transcripts</code><a class="anchor" aria-label="anchor" href="#filter-variable-transcripts"></a>
</h2>
<p>We may want to identify and filter variable transcripts.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.variable</span> <span class="op">=</span> <span class="va">airway.norm</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 500 most variable genes</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="va">norm_counts.table</span></span>
<span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">(</span><span class="va">x</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">s</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1L</span><span class="op">:</span><span class="va">top</span><span class="op">]</span>,,drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="va">norm_counts.table</span> <span class="op">=</span> <span class="va">norm_counts.table</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">norm_counts.table</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span></span>
<span>        <span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">norm_counts.table</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="st">"cell"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level2">
<h2 id="reduce-dimensions">Reduce <code>dimensions</code><a class="anchor" aria-label="anchor" href="#reduce-dimensions"></a>
</h2>
<p>We may want to reduce the dimensions of our data, for example using
PCA or MDS algorithms. <code>reduce_dimensions</code> takes a tibble,
column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and a method (e.g., MDS
or PCA) as arguments and returns a tibble with additional columns for
the reduced dimensions.</p>
<p><strong>MDS</strong> (Robinson et al.,
10.1093/bioinformatics/btp616)</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.MDS</span> <span class="op">=</span></span>
<span>  <span class="va">airway.norm</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"MDS"</span>, .dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 500 most variable genes</span></span></code></pre>
<pre><code><span><span class="co">## [1] "MDS result_df colnames: sample, 1, 2"</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the raw results do `metadata(.)$tidybulk$MDS`</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cmds</span> <span class="op">=</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html" class="external-link">plotMDS</a></span><span class="op">(</span><span class="va">count_m_log</span>, ndim <span class="op">=</span> <span class="fl">3</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cmds</span> <span class="op">=</span> <span class="va">cmds</span> <span class="op">%$%</span> </span>
<span>    <span class="va">cmdscale.out</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Dim%s"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cmds</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cmds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"cell"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data
is coloured by treatment.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.MDS</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"Dim"</span><span class="op">)</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">airway.norm.MDS</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">9</span><span class="op">:</span><span class="fl">11</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>PCA</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.PCA</span> <span class="op">=</span></span>
<span>  <span class="va">airway.norm</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span>, .dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pc</span> <span class="op">=</span> <span class="va">count_m_log</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span>scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">variance</span> <span class="op">=</span> <span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span></span>
<span><span class="va">variance</span> <span class="op">=</span> <span class="op">(</span><span class="va">variance</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">variance</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"cell"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data
is coloured by treatment.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.PCA</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"PC"</span><span class="op">)</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">airway.norm.PCA</span> <span class="op">|&gt;</span></span>
<span>     <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">11</span><span class="op">:</span><span class="fl">13</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="rotate-dimensions">Rotate <code>dimensions</code><a class="anchor" aria-label="anchor" href="#rotate-dimensions"></a>
</h2>
We may want to rotate the reduced dimensions (or any two numeric columns
really) of our data, of a set angle. <code>rotate_dimensions</code>
takes a tibble, column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and an angle as
arguments and returns a tibble with additional columns for the rotated
dimensions. The rotated dimensions will be added to the original data
set as <code>&lt;NAME OF DIMENSION&gt; rotated &lt;ANGLE&gt;</code> by
default, or as specified in the input arguments.
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.MDS.rotated</span> <span class="op">=</span></span>
<span>  <span class="va">airway.norm.MDS</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/rotate_dimensions-methods.html">rotate_dimensions</a></span><span class="op">(</span><span class="va">`Dim1`</span>, <span class="va">`Dim2`</span>, rotation_degrees <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rotation</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">m</span>, <span class="va">d</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">r</span> <span class="op">=</span> <span class="va">d</span> <span class="op">*</span> <span class="va">pi</span> <span class="op">/</span> <span class="fl">180</span></span>
<span>    <span class="op">(</span><span class="op">(</span><span class="fu">bind_rows</span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span>, `2` <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span>, `2` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/as_matrix.html">as_matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">m</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">mds_r</span> <span class="op">=</span> <span class="va">pca</span> <span class="op">|&gt;</span> <span class="fu">rotation</span><span class="op">(</span><span class="va">rotation_degrees</span><span class="op">)</span></span>
<span><span class="va">mds_r</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mds_r</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"cell"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p><strong>Original</strong> On the x and y axes axis we have the first
two reduced dimensions, data is coloured by treatment.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.MDS.rotated</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`dex`</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span></span></code></pre></div>
<p><img src="comparison_coding_files/figure-html/plot_rotate_1-1.png" width="960"></p>
<p><strong>Rotated</strong> On the x and y axes axis we have the first
two reduced dimensions rotated of 45 degrees, data is coloured by
treatment.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.MDS.rotated</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1_rotated_45`</span>, y<span class="op">=</span><span class="va">`Dim2_rotated_45`</span>, color<span class="op">=</span><span class="va">`dex`</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span></span></code></pre></div>
<p><img src="comparison_coding_files/figure-html/plot_rotate_2-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="test-differential-abundance">Test <code>differential abundance</code><a class="anchor" aria-label="anchor" href="#test-differential-abundance"></a>
</h2>
We may want to test for differential transcription between sample-wise
factors of interest (e.g., with edgeR).
<code>test_differential_expression</code> takes a tibble, column names
(as symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) and a formula representing the desired linear model
as arguments and returns a tibble with additional columns for the
statistics from the hypothesis test (e.g., log fold change, p-value and
false discovery rate).
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.de</span> <span class="op">=</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">airway.de</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">counts_m</span>,group<span class="op">=</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep</span>,,keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">dgList</span>,<span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFit.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">dgList</span>,<span class="va">design</span><span class="op">)</span></span>
<span><span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>,coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">qlf</span>, n<span class="op">=</span><span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>The functon <code>test_differential_expression</code> operated with
contrasts too. The constrasts hve the name of the design matrix
(generally <name_column_covariate><values_of_covariate>)</values_of_covariate></name_column_covariate></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.de</span> <span class="op">=</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,                  </span>
<span>        .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"dexuntrt - dextrt"</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="adjust-counts">Adjust <code>counts</code><a class="anchor" aria-label="anchor" href="#adjust-counts"></a>
</h2>
<p>We may want to adjust <code>counts</code> for (known) unwanted
variation. <code>adjust_abundance</code> takes as arguments a tibble,
column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and a formula
representing the desired linear model where the first covariate is the
factor of interest and the second covariate is the unwanted variation,
and returns a tibble with additional columns for the adjusted counts as
<code>&lt;COUNT COLUMN&gt;_adjusted</code>. At the moment just an
unwanted covariates is allowed at a time.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.adj</span> <span class="op">=</span></span>
<span>    <span class="va">airway.norm</span>     <span class="op">|&gt;</span> <span class="fu"><a href="../reference/adjust_abundance-methods.html">adjust_abundance</a></span><span class="op">(</span>    .factor_unwanted <span class="op">=</span> <span class="va">cell</span>, .factor_of_interest <span class="op">=</span> <span class="va">dex</span>, method<span class="op">=</span><span class="st">"combat"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sva</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">=</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span></span>
<span>            object <span class="op">=</span> <span class="op">~</span> <span class="va">factor_of_interest</span> <span class="op">+</span> <span class="va">batch</span>,</span>
<span>            data <span class="op">=</span> <span class="va">annotation</span></span>
<span>        <span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log.sva</span> <span class="op">=</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat.html" class="external-link">ComBat</a></span><span class="op">(</span></span>
<span>            batch <span class="op">=</span> <span class="va">design</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>            mod <span class="op">=</span> <span class="va">design</span>,</span>
<span>            <span class="va">...</span></span>
<span>        <span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log.sva</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">count_m_log.sva</span><span class="op">)</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">count_m_log.sva</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">count_m_log.sva</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"cell"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level2">
<h2 id="deconvolve-cell-type-composition">Deconvolve <code>Cell type composition</code><a class="anchor" aria-label="anchor" href="#deconvolve-cell-type-composition"></a>
</h2>
<p>We may want to infer the cell type composition of our samples (with
the algorithm Cibersort; Newman et al., 10.1038/nmeth.3337).
<code>deconvolve_cellularity</code> takes as arguments a tibble, column
names (as symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) and returns a tibble with additional columns for the
adjusted cell type proportions.</p>
<p><em>Note: Cellularity deconvolution requires gene symbols that match
the reference data. The airway dataset uses Ensembl IDs, so this example
is commented out.</em></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Requires gene symbols that match reference data</span></span>
<span><span class="co"># airway.cibersort =</span></span>
<span><span class="co">#   airway |&gt;</span></span>
<span><span class="co">#   deconvolve_cellularity( cores=1, prefix = "cibersort__") |&gt; </span></span>
<span><span class="co">#   pivot_sample()</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"CIBERSORT.R"</span><span class="op">)</span></span>
<span><span class="va">count_m</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="st">"mixture_file.txt"</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">CIBERSORT</span><span class="op">(</span></span>
<span>    <span class="st">"sig_matrix_file.txt"</span>,</span>
<span>    <span class="st">"mixture_file.txt"</span>,</span>
<span>    perm<span class="op">=</span><span class="fl">100</span>, QN<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"cell"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p><em>Note: The plotting code is commented out as the deconvolution
step is not executed.</em></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># airway.cibersort |&gt;</span></span>
<span><span class="co">#   pivot_longer(</span></span>
<span><span class="co">#       names_to= "Cell_type_inferred", </span></span>
<span><span class="co">#       values_to = "proportion", </span></span>
<span><span class="co">#       names_prefix ="cibersort__", </span></span>
<span><span class="co">#       cols=contains("cibersort__")</span></span>
<span><span class="co">#   ) |&gt;</span></span>
<span><span class="co">#   ggplot(aes(x=Cell_type_inferred, y=proportion, fill=cell)) +</span></span>
<span><span class="co">#   geom_boxplot() +</span></span>
<span><span class="co">#   facet_wrap(~cell) +</span></span>
<span><span class="co">#   my_theme +</span></span>
<span><span class="co">#   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), aspect.ratio=1/5)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cluster-samples">Cluster samples<a class="anchor" aria-label="anchor" href="#cluster-samples"></a>
</h2>
<p>We may want to cluster our samples based on the transcriptomic
profiles. <code>cluster_elements</code> takes as arguments a tibble,
column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and returns a tibble
with additional columns for the cluster labels.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.cluster</span> <span class="op">=</span></span>
<span>    <span class="va">airway.norm</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/cluster_elements-methods.html">cluster_elements</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"kmeans"</span>, centers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">count_m_log_centered</span> <span class="op">=</span> <span class="va">count_m_log</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">count_m_log</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kmeans_result</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">count_m_log_centered</span><span class="op">)</span>, centers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster_labels</span> <span class="op">=</span> <span class="va">kmeans_result</span><span class="op">$</span><span class="va">cluster</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level2">
<h2 id="test-differential-abundance-1">Test differential abundance<a class="anchor" aria-label="anchor" href="#test-differential-abundance-1"></a>
</h2>
<p>We may want to test for differential abundance between conditions.
<code>test_differential_abundance</code> takes as arguments a tibble,
column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and a formula
representing the desired linear model, and returns a tibble with
additional columns for the statistics from the hypothesis test (e.g.,
log fold change, p-value and false discovery rate).</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway.norm.de</span> <span class="op">=</span></span>
<span>    <span class="va">airway.norm</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method<span class="op">=</span><span class="st">"edgeR_quasi_likelihood"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidybulk says: The design column names are "(Intercept), dexuntrt"</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$edgeR_quasi_likelihood_object`</span></span>
<span><span class="co">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$edgeR_quasi_likelihood_fit`</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, data <span class="op">=</span> <span class="va">annotation</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dge</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_m</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFit.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">qlf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">qlf</span>, n <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Tidybulk provides a unified interface for comprehensive
transcriptomic data analysis with seamless integration of
SummarizedExperiment objects and tidyverse principles. It streamlines
the entire workflow from raw data to biological insights while
maintaining compatibility with standard Bioconductor practices.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] airway_1.29.0                   tidySummarizedExperiment_1.19.0</span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.39.1     Biobase_2.69.0                 </span></span>
<span><span class="co">##  [5] GenomicRanges_1.61.2            Seqinfo_0.99.2                 </span></span>
<span><span class="co">##  [7] IRanges_2.43.0                  S4Vectors_0.47.0               </span></span>
<span><span class="co">##  [9] BiocGenerics_0.55.1             generics_0.1.4                 </span></span>
<span><span class="co">## [11] MatrixGenerics_1.21.0           matrixStats_1.5.0              </span></span>
<span><span class="co">## [13] tidybulk_1.99.2                 ttservice_0.5.3                </span></span>
<span><span class="co">## [15] ggrepel_0.9.6                   ggplot2_3.5.2                  </span></span>
<span><span class="co">## [17] magrittr_2.0.3                  tibble_3.3.0                   </span></span>
<span><span class="co">## [19] tidyr_1.3.1                     dplyr_1.1.4                    </span></span>
<span><span class="co">## [21] knitr_1.50                      BiocStyle_2.37.1               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] DBI_1.2.3            rlang_1.1.6          compiler_4.5.1      </span></span>
<span><span class="co">##  [4] RSQLite_2.4.3        mgcv_1.9-3           png_0.1-8           </span></span>
<span><span class="co">##  [7] systemfonts_1.2.3    vctrs_0.6.5          sva_3.57.0          </span></span>
<span><span class="co">## [10] stringr_1.5.2        pkgconfig_2.0.3      crayon_1.5.3        </span></span>
<span><span class="co">## [13] fastmap_1.2.0        XVector_0.49.0       ellipsis_0.3.2      </span></span>
<span><span class="co">## [16] labeling_0.4.3       utf8_1.2.6           rmarkdown_2.29      </span></span>
<span><span class="co">## [19] ragg_1.5.0           purrr_1.1.0          bit_4.6.0           </span></span>
<span><span class="co">## [22] xfun_0.53            cachem_1.1.0         jsonlite_2.0.0      </span></span>
<span><span class="co">## [25] blob_1.2.4           DelayedArray_0.35.2  BiocParallel_1.43.4 </span></span>
<span><span class="co">## [28] parallel_4.5.1       R6_2.6.1             bslib_0.9.0         </span></span>
<span><span class="co">## [31] stringi_1.8.7        RColorBrewer_1.1-3   limma_3.65.4        </span></span>
<span><span class="co">## [34] genefilter_1.91.0    jquerylib_0.1.4      Rcpp_1.1.0          </span></span>
<span><span class="co">## [37] bookdown_0.44        splines_4.5.1        Matrix_1.7-4        </span></span>
<span><span class="co">## [40] tidyselect_1.2.1     abind_1.4-8          yaml_2.3.10         </span></span>
<span><span class="co">## [43] codetools_0.2-20     lattice_0.22-7       withr_3.0.2         </span></span>
<span><span class="co">## [46] KEGGREST_1.49.1      evaluate_1.0.5       desc_1.4.3          </span></span>
<span><span class="co">## [49] survival_3.8-3       Biostrings_2.77.2    pillar_1.11.0       </span></span>
<span><span class="co">## [52] BiocManager_1.30.26  plotly_4.11.0        scales_1.4.0        </span></span>
<span><span class="co">## [55] xtable_1.8-4         glue_1.8.0           lazyeval_0.2.2      </span></span>
<span><span class="co">## [58] tools_4.5.1          data.table_1.17.8    annotate_1.87.0     </span></span>
<span><span class="co">## [61] locfit_1.5-9.12      fs_1.6.6             XML_3.99-0.19       </span></span>
<span><span class="co">## [64] grid_4.5.1           AnnotationDbi_1.71.1 edgeR_4.7.4         </span></span>
<span><span class="co">## [67] nlme_3.1-168         cli_3.6.5            textshaping_1.0.3   </span></span>
<span><span class="co">## [70] fansi_1.0.6          S4Arrays_1.9.1       viridisLite_0.4.2   </span></span>
<span><span class="co">## [73] gtable_0.3.6         sass_0.4.10          digest_0.6.37       </span></span>
<span><span class="co">## [76] SparseArray_1.9.1    org.Hs.eg.db_3.21.0  htmlwidgets_1.6.4   </span></span>
<span><span class="co">## [79] farver_2.1.2         memoise_2.0.1        htmltools_0.5.8.1   </span></span>
<span><span class="co">## [82] pkgdown_2.1.3        lifecycle_1.0.4      httr_1.4.7          </span></span>
<span><span class="co">## [85] statmod_1.5.0        bit64_4.6.0-1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
