<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the tidybulk package • tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the tidybulk package">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the tidybulk package</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2025-03-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/comply-with-breaking-changes-for-msigdbr/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- <style type="text/css"> -->
<!-- .main-container { -->
<!--   max-width: 2800px;!important -->
<!--   margin-left: auto; -->
<!--   margin-right: auto; -->
<!-- } -->
<!-- </style> -->
<style>
.column-left{
  float: left;
  width: 50%;
  text-align: left;
}
.column-right{
  float: right;
  width: 50%;
  text-align: right;
}
</style>
<p><strong>Brings transcriptomics to the tidyverse</strong></p>
<p><img src="../reference/figures/new_SE_usage-01.png" width="800px"></p>
<div class="section level3">
<h3 id="functionsutilities-available">Functions/utilities available<a class="anchor" aria-label="anchor" href="#functionsutilities-available"></a>
</h3>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>identify_abundant</code></td>
<td>identify the abundant genes</td>
</tr>
<tr class="even">
<td><code>aggregate_duplicates</code></td>
<td>Aggregate abundance and annotation of duplicated transcripts in a
robust way</td>
</tr>
<tr class="odd">
<td><code>scale_abundance</code></td>
<td>Scale (normalise) abundance for RNA sequencing depth</td>
</tr>
<tr class="even">
<td><code>reduce_dimensions</code></td>
<td>Perform dimensionality reduction (PCA, MDS, tSNE)</td>
</tr>
<tr class="odd">
<td><code>cluster_elements</code></td>
<td>Labels elements with cluster identity (kmeans, SNN)</td>
</tr>
<tr class="even">
<td><code>remove_redundancy</code></td>
<td>Filter out elements with highly correlated features</td>
</tr>
<tr class="odd">
<td><code>adjust_abundance</code></td>
<td>Remove known unwanted variation (Combat)</td>
</tr>
<tr class="even">
<td><code>test_differential_abundance</code></td>
<td>Differential transcript abundance testing (DE)</td>
</tr>
<tr class="odd">
<td><code>deconvolve_cellularity</code></td>
<td>Estimated tissue composition (Cibersort or llsr)</td>
</tr>
<tr class="even">
<td><code>test_differential_cellularity</code></td>
<td>Differential cell-type abundance testing</td>
</tr>
<tr class="odd">
<td><code>keep_variable</code></td>
<td>Filter for top variable features</td>
</tr>
<tr class="even">
<td><code>keep_abundant</code></td>
<td>Filter out lowly abundant transcripts</td>
</tr>
<tr class="odd">
<td><code>test_gene_enrichment</code></td>
<td>Gene enrichment analyses (EGSEA)</td>
</tr>
<tr class="even">
<td><code>test_gene_overrepresentation</code></td>
<td>Gene enrichment on list of transcript names (no rank)</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>get_bibliography</code></td>
<td>Get the bibliography of your workflow</td>
</tr>
<tr class="even">
<td><code>tidybulk</code></td>
<td>add tidybulk attributes to a tibble object</td>
</tr>
<tr class="odd">
<td><code>tidybulk_SAM_BAM</code></td>
<td>Convert SAM BAM files into tidybulk tibble</td>
</tr>
<tr class="even">
<td><code>pivot_sample</code></td>
<td>Select sample-wise columns/information</td>
</tr>
<tr class="odd">
<td><code>pivot_transcript</code></td>
<td>Select transcript-wise columns/information</td>
</tr>
<tr class="even">
<td><code>rotate_dimensions</code></td>
<td>Rotate two dimensions of a degree</td>
</tr>
<tr class="odd">
<td><code>ensembl_to_symbol</code></td>
<td>Add gene symbol from ensembl IDs</td>
</tr>
<tr class="even">
<td><code>symbol_to_entrez</code></td>
<td>Add entrez ID from gene symbol</td>
</tr>
<tr class="odd">
<td><code>describe_transcript</code></td>
<td>Add gene description from gene symbol</td>
</tr>
<tr class="even">
<td><code>impute_missing_abundance</code></td>
<td>Impute abundance for missing data points using sample groupings</td>
</tr>
<tr class="odd">
<td><code>fill_missing_abundance</code></td>
<td>Fill abundance for missing data points using an arbitrary value</td>
</tr>
</tbody>
</table>
<p>All functions are directly compatibble with
<code>SummarizedExperiment</code> object.</p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>From Bioconductor</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"tidybulk"</span><span class="op">)</span></span></code></pre></div>
<p>From Github</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/tidybulk"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>We will use a <code>SummarizedExperiment</code> object</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SummarizedExperiment-tibble abstraction: 2,635 × 9</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=527 | Samples=5 | Assays=count</span></span></span>
<span><span class="co">##    .feature .sample    count Cell.type time  condition  days  dead entrez</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ABCB4    SRR1740034  <span style="text-decoration: underline;">1</span>035 b_cell    0 d   TRUE          1     1 5244  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ABCB9    SRR1740034    45 b_cell    0 d   TRUE          1     1 23457 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ACAP1    SRR1740034  <span style="text-decoration: underline;">7</span>151 b_cell    0 d   TRUE          1     1 9744  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ACHE     SRR1740034     2 b_cell    0 d   TRUE          1     1 43    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ACP5     SRR1740034  <span style="text-decoration: underline;">2</span>278 b_cell    0 d   TRUE          1     1 54    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ADAM28   SRR1740034 <span style="text-decoration: underline;">11</span>156 b_cell    0 d   TRUE          1     1 10863 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ADAMDEC1 SRR1740034    72 b_cell    0 d   TRUE          1     1 27299 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ADAMTS3  SRR1740034     0 b_cell    0 d   TRUE          1     1 9508  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ADRB2    SRR1740034   298 b_cell    0 d   TRUE          1     1 154   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AIF1     SRR1740034     8 b_cell    0 d   TRUE          1     1 199   </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 40 more rows</span></span></span></code></pre>
<p>Loading <code>tidySummarizedExperiment</code> will automatically
abstract this object as <code>tibble</code>, so we can display it and
manipulate it with tidy tools. Although it looks different, and more
tools (tidyverse) are available to us, this object is in fact a
<code>SummarizedExperiment</code> object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">se_mini</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SummarizedExperiment"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "SummarizedExperiment"</span></span></code></pre>
<div class="section level3">
<h3 id="get-the-bibliography-of-your-workflow">Get the bibliography of your workflow<a class="anchor" aria-label="anchor" href="#get-the-bibliography-of-your-workflow"></a>
</h3>
<p>First of all, you can cite all articles utilised within your workflow
automatically from any tidybulk tibble</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini</span> <span class="op">|&gt;</span>  <span class="fu"><a href="../reference/get_bibliography-methods.html">get_bibliography</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="aggregate-duplicated-transcripts">Aggregate duplicated <code>transcripts</code><a class="anchor" aria-label="anchor" href="#aggregate-duplicated-transcripts"></a>
</h3>
<p>tidybulk provide the <code>aggregate_duplicates</code> function to
aggregate duplicated transcripts (e.g., isoforms, ensembl). For example,
we often have to convert ensembl symbols to gene/transcript symbol, but
in doing so we have to deal with duplicates.
<code>aggregate_duplicates</code> takes a tibble and column names (as
symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) as arguments and returns a tibble with transcripts
with the same name aggregated. All the rest of the columns are appended,
and factors and boolean are appended as characters.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se_mini</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se_mini</span><span class="op">)</span></span>
<span><span class="va">se_mini.aggr</span> <span class="op">=</span> <span class="va">se_mini</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>.transcript <span class="op">=</span> <span class="va">gene_name</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    symbol <span class="op">=</span> <span class="va">dge_list</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">symbol</span>,</span>
<span>    <span class="va">dge_list</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/by.html" class="external-link">by</a></span><span class="op">(</span><span class="va">temp</span>, <span class="va">temp</span><span class="op">$</span><span class="va">symbol</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span></span>
<span>            <span class="fu">matrixStats</span><span class="fu">:::</span><span class="fu">colSums</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="va">dge_list.nr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dge_list.nr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dge_list</span><span class="op">)</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level3">
<h3 id="scale-counts">Scale <code>counts</code><a class="anchor" aria-label="anchor" href="#scale-counts"></a>
</h3>
<p>We may want to compensate for sequencing depth, scaling the
transcript abundance (e.g., with TMM algorithm, Robinson and Oshlack
doi.org/10.1186/gb-2010-11-3-r25). <code>scale_abundance</code> takes a
tibble, column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and a method as
arguments and returns a tibble with additional columns with scaled data
as <code>&lt;NAME OF COUNT COLUMN&gt;_scaled</code>.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm</span> <span class="op">=</span> <span class="va">se_mini.aggr</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidybulk says: the sample with largest library size SRR1740035 was chosen as reference for scaling</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>dgList <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">count_m=</span>x,<span class="at">group=</span>group)</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dgList)</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>dgList <span class="ot">&lt;-</span> dgList[keep,,keep.lib.sizes<span class="ot">=</span><span class="cn">FALSE</span>]</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>[...]</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>dgList <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dgList, <span class="at">method=</span><span class="st">"TMM"</span>)</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>norm_counts.table <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dgList)</span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>We can easily plot the scaled density to check the scaling outcome.
On the x axis we have the log scaled counts, on the y axes we have the
density, data is grouped by sample and coloured by cell type.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">count_scaled</span> <span class="op">+</span> <span class="fl">1</span>, group<span class="op">=</span><span class="va">.sample</span>, color<span class="op">=</span><span class="va">`Cell.type`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot_normalise-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="filter-variable-transcripts">Filter <code>variable transcripts</code><a class="anchor" aria-label="anchor" href="#filter-variable-transcripts"></a>
</h3>
<p>We may want to identify and filter variable transcripts.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.variable</span> <span class="op">=</span> <span class="va">se_mini.norm</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 394 most variable genes</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="va">norm_counts.table</span></span>
<span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">(</span><span class="va">x</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">s</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1L</span><span class="op">:</span><span class="va">top</span><span class="op">]</span>,,drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="va">norm_counts.table</span> <span class="op">=</span> <span class="va">norm_counts.table</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">norm_counts.table</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span></span>
<span>        <span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">norm_counts.table</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="st">"Cell.type"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level3">
<h3 id="reduce-dimensions">Reduce <code>dimensions</code><a class="anchor" aria-label="anchor" href="#reduce-dimensions"></a>
</h3>
<p>We may want to reduce the dimensions of our data, for example using
PCA or MDS algorithms. <code>reduce_dimensions</code> takes a tibble,
column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and a method (e.g., MDS
or PCA) as arguments and returns a tibble with additional columns for
the reduced dimensions.</p>
<p><strong>MDS</strong> (Robinson et al.,
10.1093/bioinformatics/btp616)</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.MDS</span> <span class="op">=</span></span>
<span>  <span class="va">se_mini.norm</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"MDS"</span>, .dims <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 394 most variable genes</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the raw results do `attr(..., "internals")$MDS`</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cmds</span> <span class="op">=</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html" class="external-link">plotMDS</a></span><span class="op">(</span>ndim <span class="op">=</span> <span class="va">.dims</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cmds</span> <span class="op">=</span> <span class="va">cmds</span> <span class="op">%$%</span> </span>
<span>    <span class="va">cmdscale.out</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Dim%s"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cmds</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cmds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Cell.type"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data
is coloured by cell type.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.MDS</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"Dim"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_mini.norm.MDS</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">9</span><span class="op">:</span><span class="fl">11</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">`Cell.type`</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>PCA</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.PCA</span> <span class="op">=</span></span>
<span>  <span class="va">se_mini.norm</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span>, .dims <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pc</span> <span class="op">=</span> <span class="va">count_m_log</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span>scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">variance</span> <span class="op">=</span> <span class="va">pc</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span></span>
<span><span class="va">variance</span> <span class="op">=</span> <span class="op">(</span><span class="va">variance</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">variance</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Cell.type"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data
is coloured by cell type.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.PCA</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"PC"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_mini.norm.PCA</span> <span class="op">|&gt;</span></span>
<span>     <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">11</span><span class="op">:</span><span class="fl">13</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour<span class="op">=</span><span class="va">`Cell.type`</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<strong>tSNE</strong>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.tSNE</span> <span class="op">=</span></span>
<span>    <span class="va">breast_tcga_mini_SE</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span></span>
<span>        method <span class="op">=</span> <span class="st">"tSNE"</span>,</span>
<span>        perplexity<span class="op">=</span><span class="fl">10</span>,</span>
<span>        pca_scale <span class="op">=</span><span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tsne</span> <span class="op">=</span> <span class="fu">Rtsne</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">count_m_log</span><span class="op">)</span>,</span>
<span>    perplexity<span class="op">=</span><span class="fl">10</span>,</span>
<span>        pca_scale <span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">Y</span></span>
<span><span class="va">tsne</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tsne</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Cell.type"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>Plot</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.tSNE</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"tSNE"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 251 × 4</span></span></span>
<span><span class="co">##      tSNE1  tSNE2 .sample                      Call </span></span>
<span><span class="co">##      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                        <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>   0.769  -<span style="color: #BB0000;">8.96</span> TCGA-A1-A0SD-01A-11R-A115-07 LumA </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>   7.38    5.86 TCGA-A1-A0SF-01A-11R-A144-07 LumA </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>   0.313 -<span style="color: #BB0000;">15.3</span>  TCGA-A1-A0SG-01A-11R-A144-07 LumA </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>  -<span style="color: #BB0000;">5.40</span>   -<span style="color: #BB0000;">2.71</span> TCGA-A1-A0SH-01A-11R-A084-07 LumA </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>  -<span style="color: #BB0000;">1.30</span>   -<span style="color: #BB0000;">4.84</span> TCGA-A1-A0SI-01A-11R-A144-07 LumB </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>   4.18   12.4  TCGA-A1-A0SJ-01A-11R-A084-07 LumA </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>  -<span style="color: #BB0000;">5.31</span>   27.6  TCGA-A1-A0SK-01A-12R-A084-07 Basal</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> -<span style="color: #BB0000;">10.7</span>     2.40 TCGA-A1-A0SM-01A-11R-A084-07 LumA </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> -<span style="color: #BB0000;">11.4</span>     1.40 TCGA-A1-A0SN-01A-11R-A144-07 LumB </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>   2.58  -<span style="color: #BB0000;">22.0</span>  TCGA-A1-A0SQ-01A-21R-A144-07 LumA </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 241 more rows</span></span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.tSNE</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`tSNE1`</span>, y <span class="op">=</span> <span class="va">`tSNE2`</span>, color<span class="op">=</span><span class="va">Call</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">my_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="rotate-dimensions">Rotate <code>dimensions</code><a class="anchor" aria-label="anchor" href="#rotate-dimensions"></a>
</h3>
We may want to rotate the reduced dimensions (or any two numeric columns
really) of our data, of a set angle. <code>rotate_dimensions</code>
takes a tibble, column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and an angle as
arguments and returns a tibble with additional columns for the rotated
dimensions. The rotated dimensions will be added to the original data
set as <code>&lt;NAME OF DIMENSION&gt; rotated &lt;ANGLE&gt;</code> by
default, or as specified in the input arguments.
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.MDS.rotated</span> <span class="op">=</span></span>
<span>  <span class="va">se_mini.norm.MDS</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/rotate_dimensions-methods.html">rotate_dimensions</a></span><span class="op">(</span><span class="va">`Dim1`</span>, <span class="va">`Dim2`</span>, rotation_degrees <span class="op">=</span> <span class="fl">45</span>, action<span class="op">=</span><span class="st">"get"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rotation</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">m</span>, <span class="va">d</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">r</span> <span class="op">=</span> <span class="va">d</span> <span class="op">*</span> <span class="va">pi</span> <span class="op">/</span> <span class="fl">180</span></span>
<span>    <span class="op">(</span><span class="op">(</span><span class="fu"><a href="../reference/bind_rows.html">bind_rows</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span>, `2` <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">sin</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span>, `2` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/as_matrix.html">as_matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">m</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">mds_r</span> <span class="op">=</span> <span class="va">pca</span> <span class="op">|&gt;</span> <span class="fu">rotation</span><span class="op">(</span><span class="va">rotation_degrees</span><span class="op">)</span></span>
<span><span class="va">mds_r</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mds_r</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Cell.type"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p><strong>Original</strong> On the x and y axes axis we have the first
two reduced dimensions, data is coloured by cell type.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.MDS.rotated</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`Cell.type`</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot_rotate_1-1.png" width="700"></p>
<p><strong>Rotated</strong> On the x and y axes axis we have the first
two reduced dimensions rotated of 45 degrees, data is coloured by cell
type.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.MDS.rotated</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1_rotated_45`</span>, y<span class="op">=</span><span class="va">`Dim2_rotated_45`</span>, color<span class="op">=</span><span class="va">`Cell.type`</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot_rotate_2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="test-differential-abundance">Test <code>differential abundance</code><a class="anchor" aria-label="anchor" href="#test-differential-abundance"></a>
</h3>
We may want to test for differential transcription between sample-wise
factors of interest (e.g., with edgeR).
<code>test_differential_abundance</code> takes a tibble, column names
(as symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) and a formula representing the desired linear model
as arguments and returns a tibble with additional columns for the
statistics from the hypothesis test (e.g., log fold change, p-value and
false discovery rate).
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.de</span> <span class="op">=</span></span>
<span>    <span class="va">se_mini</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">condition</span>, action<span class="op">=</span><span class="st">"get"</span><span class="op">)</span></span>
<span><span class="va">se_mini.de</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">counts_m</span>,group<span class="op">=</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep</span>,,keep.lib.sizes<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">dgList</span>,<span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFit.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">dgList</span>,<span class="va">design</span><span class="op">)</span></span>
<span><span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>,coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">qlf</span>, n<span class="op">=</span><span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>The functon <code>test_differential_abundance</code> operated with
contrasts too. The constrasts hve the name of the design matrix
(generally <name_column_covariate><values_of_covariate>)</values_of_covariate></name_column_covariate></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.de</span> <span class="op">=</span></span>
<span>    <span class="va">se_mini</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">condition</span>,                  </span>
<span>        .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"conditionTRUE - conditionFALSE"</span><span class="op">)</span>,</span>
<span>        action<span class="op">=</span><span class="st">"get"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="adjust-counts">Adjust <code>counts</code><a class="anchor" aria-label="anchor" href="#adjust-counts"></a>
</h3>
<p>We may want to adjust <code>counts</code> for (known) unwanted
variation. <code>adjust_abundance</code> takes as arguments a tibble,
column names (as symbols; for <code>sample</code>,
<code>transcript</code> and <code>count</code>) and a formula
representing the desired linear model where the first covariate is the
factor of interest and the second covariate is the unwanted variation,
and returns a tibble with additional columns for the adjusted counts as
<code>&lt;COUNT COLUMN&gt;_adjusted</code>. At the moment just an
unwanted covariates is allowed at a time.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.adj</span> <span class="op">=</span></span>
<span>    <span class="va">se_mini.norm</span>    <span class="op">|&gt;</span> <span class="fu"><a href="../reference/adjust_abundance-methods.html">adjust_abundance</a></span><span class="op">(</span>    .factor_unwanted <span class="op">=</span> <span class="va">time</span>, .factor_of_interest <span class="op">=</span> <span class="va">condition</span>, method<span class="op">=</span><span class="st">"combat"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sva</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">=</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span></span>
<span>            object <span class="op">=</span> <span class="op">~</span> <span class="va">factor_of_interest</span> <span class="op">+</span> <span class="va">batch</span>,</span>
<span>            data <span class="op">=</span> <span class="va">annotation</span></span>
<span>        <span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log.sva</span> <span class="op">=</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat.html" class="external-link">ComBat</a></span><span class="op">(</span></span>
<span>            batch <span class="op">=</span> <span class="va">design</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>            mod <span class="op">=</span> <span class="va">design</span>,</span>
<span>            <span class="va">...</span></span>
<span>        <span class="op">)</span></span>
<span></span>
<span><span class="va">count_m_log.sva</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">count_m_log.sva</span><span class="op">)</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">count_m_log.sva</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">count_m_log.sva</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="st">"Cell.type"</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div class="section level3">
<h3 id="deconvolve-cell-type-composition">Deconvolve <code>Cell type composition</code><a class="anchor" aria-label="anchor" href="#deconvolve-cell-type-composition"></a>
</h3>
<p>We may want to infer the cell type composition of our samples (with
the algorithm Cibersort; Newman et al., 10.1038/nmeth.3337).
<code>deconvolve_cellularity</code> takes as arguments a tibble, column
names (as symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) and returns a tibble with additional columns for the
adjusted cell type proportions.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.cibersort</span> <span class="op">=</span></span>
<span>    <span class="va">se_mini</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>action<span class="op">=</span><span class="st">"get"</span>, cores<span class="op">=</span><span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"cibersort__"</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="fu">source</span>(‘CIBERSORT.R’)</span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a>count_m <span class="sc">|&gt;</span> <span class="fu">write.table</span>(<span class="st">"mixture_file.txt"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">CIBERSORT</span>(</span>
<span id="cb40-4"><a href="#cb40-4" tabindex="-1"></a>    <span class="st">"sig_matrix_file.txt"</span>,</span>
<span id="cb40-5"><a href="#cb40-5" tabindex="-1"></a>    <span class="st">"mixture_file.txt"</span>,</span>
<span id="cb40-6"><a href="#cb40-6" tabindex="-1"></a>    <span class="at">perm=</span><span class="dv">100</span>, <span class="at">QN=</span><span class="cn">TRUE</span></span>
<span id="cb40-7"><a href="#cb40-7" tabindex="-1"></a>)</span>
<span id="cb40-8"><a href="#cb40-8" tabindex="-1"></a>results<span class="sc">$</span>cell_type <span class="ot">=</span> tibble_counts[</span>
<span id="cb40-9"><a href="#cb40-9" tabindex="-1"></a>    <span class="fu">match</span>(tibble_counts<span class="sc">$</span>sample, <span class="fu">rownames</span>(results)),</span>
<span id="cb40-10"><a href="#cb40-10" tabindex="-1"></a>    <span class="st">"Cell.type"</span></span>
<span id="cb40-11"><a href="#cb40-11" tabindex="-1"></a>]</span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>With the new annotated data frame, we can plot the distributions of
cell types across samples, and compare them with the nominal cell type
labels to check for the purity of isolation. On the x axis we have the
cell types inferred by Cibersort, on the y axis we have the inferred
proportions. The data is facetted and coloured by nominal cell types
(annotation given by the researcher after FACS sorting).</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.cibersort</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>        names_to<span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, </span>
<span>        values_to <span class="op">=</span> <span class="st">"proportion"</span>, </span>
<span>        names_prefix <span class="op">=</span><span class="st">"cibersort__"</span>, </span>
<span>        cols<span class="op">=</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"cibersort__"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Cell_type_inferred</span>, y<span class="op">=</span><span class="va">proportion</span>, fill<span class="op">=</span><span class="va">`Cell.type`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">`Cell.type`</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, aspect.ratio<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="test-differential-cell-type-abundance">Test differential cell-type abundance<a class="anchor" aria-label="anchor" href="#test-differential-cell-type-abundance"></a>
</h3>
<p>We can also perform a statistical test on the differential cell-type
abundance across conditions</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="va">se_mini</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_cellularity-methods.html">test_differential_cellularity</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">condition</span> <span class="op">)</span></span></code></pre></div>
<p>We can also perform regression analysis with censored data
(coxph).</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="va">se_mini</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_cellularity-methods.html">test_differential_cellularity</a></span><span class="op">(</span><span class="fu">survival</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">dead</span><span class="op">)</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cluster-samples">Cluster <code>samples</code><a class="anchor" aria-label="anchor" href="#cluster-samples"></a>
</h3>
<p>We may want to cluster our data (e.g., using k-means sample-wise).
<code>cluster_elements</code> takes as arguments a tibble, column names
(as symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) and returns a tibble with additional columns for the
cluster annotation. At the moment only k-means clustering is supported,
the plan is to introduce more clustering methods.</p>
<p><strong>k-means</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.cluster</span> <span class="op">=</span> <span class="va">se_mini.norm.MDS</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/cluster_elements-methods.html">cluster_elements</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"kmeans"</span>, centers <span class="op">=</span> <span class="fl">2</span>, action<span class="op">=</span><span class="st">"get"</span> <span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">count_m</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">k</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">count_m_log</span>, iter.max <span class="op">=</span> <span class="fl">1000</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">cluster</span> <span class="op">=</span> <span class="va">k</span><span class="op">$</span><span class="va">cluster</span></span>
<span></span>
<span><span class="va">cluster</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell.type"</span>, <span class="st">"Dim1"</span>, <span class="st">"Dim2"</span><span class="op">)</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>We can add cluster annotation to the MDS dimension reduced data set
and plot.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">se_mini.norm.cluster</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`cluster_kmeans`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot_cluster-1.png" width="700"></p>
<p><strong>SNN</strong></p>
<p>Matrix package (v1.3-3) causes an error with Seurat::FindNeighbors
used in this method. We are trying to solve this issue. At the moment
this option in unaviable.</p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.SNN</span> <span class="op">=</span></span>
<span>    <span class="va">se_mini.norm.tSNE</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/cluster_elements-methods.html">cluster_elements</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"SNN"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">count_m</span><span class="op">)</span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span></span>
<span>    <span class="va">snn</span>, display.progress <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    num.cores<span class="op">=</span><span class="fl">4</span>, do.par <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">snn</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">snn</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">snn</span>, npcs <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">snn</span><span class="op">)</span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">snn</span>, method <span class="op">=</span> <span class="st">"igraph"</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">snn</span> <span class="op">=</span> <span class="va">snn</span><span class="op">[[</span><span class="st">"seurat_clusters"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">snn</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">tibble_counts</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">tibble_counts</span><span class="op">$</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">snn</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell.type"</span>, <span class="st">"Dim1"</span>, <span class="st">"Dim2"</span><span class="op">)</span></span>
<span><span class="op">]</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.SNN</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"tSNE"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">se_mini.norm.SNN</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">gather</span><span class="op">(</span><span class="va">source</span>, <span class="va">Call</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster_SNN"</span>, <span class="st">"Call"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/distinct-methods.html">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`tSNE1`</span>, y <span class="op">=</span> <span class="va">`tSNE2`</span>, color<span class="op">=</span><span class="va">Call</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span> <span class="va">my_theme</span></span>
<span></span>
<span></span>
<span><span class="co"># Do differential transcription between clusters</span></span>
<span><span class="va">se_mini.norm.SNN</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/mutate-methods.html">mutate</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">`cluster_SNN`</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span> <span class="va">factor_of_interest</span>,</span>
<span>    action<span class="op">=</span><span class="st">"get"</span></span>
<span>   <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="drop-redundant-transcripts">Drop <code>redundant</code> transcripts<a class="anchor" aria-label="anchor" href="#drop-redundant-transcripts"></a>
</h3>
<p>We may want to remove redundant elements from the original data set
(e.g., samples or transcripts), for example if we want to define
cell-type specific signatures with low sample redundancy.
<code>remove_redundancy</code> takes as arguments a tibble, column names
(as symbols; for <code>sample</code>, <code>transcript</code> and
<code>count</code>) and returns a tibble with redundant elements removed
(e.g., samples). Two redundancy estimation approaches are supported:</p>
<ul>
<li>removal of highly correlated clusters of elements (keeping a
representative) with method=“correlation”</li>
<li>removal of most proximal element pairs in a reduced dimensional
space.</li>
</ul>
<p><strong>Approach 1</strong></p>
<div class="column-left">
<p>TidyTranscriptomics</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.non_redundant</span> <span class="op">=</span></span>
<span>    <span class="va">se_mini.norm.MDS</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/remove_redundancy-methods.html">remove_redundancy</a></span><span class="op">(</span>    method <span class="op">=</span> <span class="st">"correlation"</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 394 most variable genes</span></span></code></pre>
</div>
<div class="column-right">
<p>Standard procedure (comparative purpose)</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/juliasilge/widyr" class="external-link">widyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">.data.correlated</span> <span class="op">=</span></span>
<span>    <span class="fu"><a href="https://juliasilge.github.io/widyr/reference/pairwise_cor.html" class="external-link">pairwise_cor</a></span><span class="op">(</span></span>
<span>        <span class="va">counts</span>,</span>
<span>        <span class="va">sample</span>,</span>
<span>        <span class="va">transcript</span>,</span>
<span>        <span class="va">rc</span>,</span>
<span>        sort <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        diag <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        upper <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filter-methods.html">filter</a></span><span class="op">(</span><span class="va">correlation</span> <span class="op">&gt;</span> <span class="va">correlation_threshold</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/distinct-methods.html">distinct</a></span><span class="op">(</span><span class="va">item1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/rename-methods.html">rename</a></span><span class="op">(</span><span class="op">!</span><span class="op">!</span><span class="va">.element</span> <span class="op">:=</span> <span class="va">item1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Return non redudant data frame</span></span>
<span><span class="va">counts</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">anti_join</a></span><span class="op">(</span><span class="va">.data.correlated</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">spread</span><span class="op">(</span><span class="va">sample</span>, <span class="va">rc</span>, <span class="op">-</span> <span class="va">transcript</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/dplyr-methods.html">left_join</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span></span></code></pre></div>
</div>
<div style="clear:both;">

</div>
<p>We can visualise how the reduced redundancy with the reduced
dimentions look like</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.non_redundant</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`Cell.type`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot_drop-1.png" width="700"></p>
<p><strong>Approach 2</strong></p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.non_redundant</span> <span class="op">=</span></span>
<span>    <span class="va">se_mini.norm.MDS</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/remove_redundancy-methods.html">remove_redundancy</a></span><span class="op">(</span></span>
<span>    method <span class="op">=</span> <span class="st">"reduced_dimensions"</span>,</span>
<span>    Dim_a_column <span class="op">=</span> <span class="va">`Dim1`</span>,</span>
<span>    Dim_b_column <span class="op">=</span> <span class="va">`Dim2`</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>We can visualise MDS reduced dimensions of the samples with the
closest pair removed.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini.norm.non_redundant</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`Cell.type`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot_drop2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="other-useful-wrappers">Other useful wrappers<a class="anchor" aria-label="anchor" href="#other-useful-wrappers"></a>
</h3>
<p>The above wrapper streamline the most common processing of bulk RNA
sequencing data. Other useful wrappers are listed above.</p>
</div>
<div class="section level3">
<h3 id="from-bamsam-to-tibble-of-gene-counts">From BAM/SAM to tibble of gene counts<a class="anchor" aria-label="anchor" href="#from-bamsam-to-tibble-of-gene-counts"></a>
</h3>
<p>We can calculate gene counts (using FeatureCounts; Liao Y et al.,
10.1093/nar/gkz114) from a list of BAM/SAM files and format them into a
tidy structure (similar to counts).</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">=</span> <span class="fu"><a href="../reference/tidybulk_SAM_BAM-methods.html">tidybulk_SAM_BAM</a></span><span class="op">(</span></span>
<span>    <span class="va">file_names</span>,</span>
<span>    genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>    isPairedEnd <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    requireBothEndsMapped <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    checkFragLength <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    useMetaFeatures <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="from-ensembl-ids-to-gene-symbol-ids">From ensembl IDs to gene symbol IDs<a class="anchor" aria-label="anchor" href="#from-ensembl-ids-to-gene-symbol-ids"></a>
</h3>
<p>We can add gene symbols from ensembl identifiers. This is useful
since different resources use ensembl IDs while others use gene symbol
IDs. This currently works for human and mouse.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_ensembl</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/ensembl_to_symbol-methods.html">ensembl_to_symbol</a></span><span class="op">(</span><span class="va">ens</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 119 × 8</span></span></span>
<span><span class="co">##    ens   iso   `read count` sample cases_0_project_dise…¹ cases_0_samples_0_sa…²</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ENSG… 13             144 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ENSG… 13              72 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ENSG… 13               0 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ENSG… 13            <span style="text-decoration: underline;">1</span>099 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ENSG… 13              11 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ENSG… 13               2 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ENSG… 13               3 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ENSG… 13            <span style="text-decoration: underline;">2</span>678 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ENSG… 13             751 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ENSG… 13               1 TARGE… Acute Myeloid Leukemia Primary Blood Derived…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 109 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ abbreviated names: ¹​cases_0_project_disease_type,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ²​cases_0_samples_0_sample_type</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: transcript &lt;chr&gt;, ref_genome &lt;chr&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="from-gene-symbol-to-gene-description-gene-name-in-full">From gene symbol to gene description (gene name in full)<a class="anchor" aria-label="anchor" href="#from-gene-symbol-to-gene-description-gene-name-in-full"></a>
</h3>
<p>We can add gene full name (and in future description) from symbol
identifiers. This currently works for human and mouse.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mini</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/describe_transcript-methods.html">describe_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">feature</span>, <span class="va">description</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## Warning in is_sample_feature_deprecated_used(.data, .cols):</span></span>
<span><span class="co">## tidySummarizedExperiment says: from version 1.3.1, the special columns</span></span>
<span><span class="co">## including sample/feature id (colnames(se), rownames(se)) has changed to</span></span>
<span><span class="co">## ".sample" and ".feature". This dataset is returned with the old-style</span></span>
<span><span class="co">## vocabulary (feature and sample), however we suggest to update your workflow to</span></span>
<span><span class="co">## reflect the new vocabulary (.feature, .sample)</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SummarizedExperiment-tibble abstraction: 2,635 × 11</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=527 | Samples=5 | Assays=count</span></span></span>
<span><span class="co">##    feature sample count Cell.type time  condition  days  dead description entrez</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ABCB4   SRR17…  <span style="text-decoration: underline;">1</span>035 b_cell    0 d   TRUE          1     1 ATP bindin… 5244  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ABCB9   SRR17…    45 b_cell    0 d   TRUE          1     1 ATP bindin… 23457 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ACAP1   SRR17…  <span style="text-decoration: underline;">7</span>151 b_cell    0 d   TRUE          1     1 ArfGAP wit… 9744  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ACHE    SRR17…     2 b_cell    0 d   TRUE          1     1 acetylchol… 43    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ACP5    SRR17…  <span style="text-decoration: underline;">2</span>278 b_cell    0 d   TRUE          1     1 acid phosp… 54    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ADAM28  SRR17… <span style="text-decoration: underline;">11</span>156 b_cell    0 d   TRUE          1     1 ADAM metal… 10863 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ADAMDE… SRR17…    72 b_cell    0 d   TRUE          1     1 ADAM like … 27299 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ADAMTS3 SRR17…     0 b_cell    0 d   TRUE          1     1 ADAM metal… 9508  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ADRB2   SRR17…   298 b_cell    0 d   TRUE          1     1 adrenocept… 154   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AIF1    SRR17…     8 b_cell    0 d   TRUE          1     1 allograft … 199   </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 40 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: gene_name &lt;chr&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h3>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2025-03-08 r87910)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] tidySummarizedExperiment_1.17.0 SummarizedExperiment_1.37.0    </span></span>
<span><span class="co">##  [3] Biobase_2.67.0                  GenomicRanges_1.59.1           </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.43.4             IRanges_2.41.3                 </span></span>
<span><span class="co">##  [7] S4Vectors_0.45.4                BiocGenerics_0.53.6            </span></span>
<span><span class="co">##  [9] generics_0.1.3                  MatrixGenerics_1.19.1          </span></span>
<span><span class="co">## [11] matrixStats_1.5.0               tidybulk_1.17.10               </span></span>
<span><span class="co">## [13] ttservice_0.4.1                 ggrepel_0.9.6                  </span></span>
<span><span class="co">## [15] ggplot2_3.5.1                   magrittr_2.0.3                 </span></span>
<span><span class="co">## [17] tibble_3.2.1                    tidyr_1.3.1                    </span></span>
<span><span class="co">## [19] dplyr_1.1.4                     knitr_1.49                     </span></span>
<span><span class="co">## [21] BiocStyle_2.35.0               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.9.1          farver_2.1.2            rmarkdown_2.29         </span></span>
<span><span class="co">##   [4] fs_1.6.5                ragg_1.3.3              vctrs_0.6.5            </span></span>
<span><span class="co">##   [7] memoise_2.0.1           htmltools_0.5.8.1       S4Arrays_1.7.3         </span></span>
<span><span class="co">##  [10] broom_1.0.7             janeaustenr_1.0.0       SparseArray_1.7.6      </span></span>
<span><span class="co">##  [13] sass_0.4.9              bslib_0.9.0             htmlwidgets_1.6.4      </span></span>
<span><span class="co">##  [16] tokenizers_0.3.0        desc_1.4.3              plyr_1.8.9             </span></span>
<span><span class="co">##  [19] plotly_4.10.4           cachem_1.1.0            lifecycle_1.0.4        </span></span>
<span><span class="co">##  [22] pkgconfig_2.0.3         Matrix_1.7-3            R6_2.6.1               </span></span>
<span><span class="co">##  [25] fastmap_1.2.0           GenomeInfoDbData_1.2.14 digest_0.6.37          </span></span>
<span><span class="co">##  [28] tidytext_0.4.2          colorspace_2.1-1        AnnotationDbi_1.69.0   </span></span>
<span><span class="co">##  [31] textshaping_1.0.0       SnowballC_0.7.1         RSQLite_2.3.9          </span></span>
<span><span class="co">##  [34] org.Hs.eg.db_3.21.0     labeling_0.4.3          org.Mm.eg.db_3.21.0    </span></span>
<span><span class="co">##  [37] fansi_1.0.6             httr_1.4.7              abind_1.4-8            </span></span>
<span><span class="co">##  [40] mgcv_1.9-1              compiler_4.5.0          proxy_0.4-27           </span></span>
<span><span class="co">##  [43] bit64_4.6.0-1           withr_3.0.2             backports_1.5.0        </span></span>
<span><span class="co">##  [46] BiocParallel_1.41.2     DBI_1.2.3               DelayedArray_0.33.6    </span></span>
<span><span class="co">##  [49] tools_4.5.0             glue_1.8.0              nlme_3.1-167           </span></span>
<span><span class="co">##  [52] grid_4.5.0              Rtsne_0.17              reshape2_1.4.4         </span></span>
<span><span class="co">##  [55] sva_3.55.0              gtable_0.3.6            tzdb_0.4.0             </span></span>
<span><span class="co">##  [58] class_7.3-23            preprocessCore_1.69.0   data.table_1.17.0      </span></span>
<span><span class="co">##  [61] hms_1.1.3               utf8_1.2.4              XVector_0.47.2         </span></span>
<span><span class="co">##  [64] pillar_1.10.1           stringr_1.5.1           limma_3.63.9           </span></span>
<span><span class="co">##  [67] genefilter_1.89.0       splines_4.5.0           lattice_0.22-6         </span></span>
<span><span class="co">##  [70] survival_3.8-3          bit_4.6.0               annotate_1.85.0        </span></span>
<span><span class="co">##  [73] tidyselect_1.2.1        locfit_1.5-9.12         Biostrings_2.75.4      </span></span>
<span><span class="co">##  [76] bookdown_0.42           edgeR_4.5.9             xfun_0.51              </span></span>
<span><span class="co">##  [79] statmod_1.5.0           stringi_1.8.4           UCSC.utils_1.3.1       </span></span>
<span><span class="co">##  [82] lazyeval_0.2.2          yaml_2.3.10             evaluate_1.0.3         </span></span>
<span><span class="co">##  [85] codetools_0.2-20        widyr_0.1.5             BiocManager_1.30.25    </span></span>
<span><span class="co">##  [88] cli_3.6.4               xtable_1.8-4            systemfonts_1.2.1      </span></span>
<span><span class="co">##  [91] munsell_0.5.1           jquerylib_0.1.4         Rcpp_1.0.14            </span></span>
<span><span class="co">##  [94] png_0.1-8               XML_3.99-0.18           parallel_4.5.0         </span></span>
<span><span class="co">##  [97] ellipsis_0.3.2          pkgdown_2.1.1           readr_2.1.5            </span></span>
<span><span class="co">## [100] blob_1.2.4              viridisLite_0.4.2       scales_1.3.0           </span></span>
<span><span class="co">## [103] e1071_1.7-16            purrr_1.0.4             crayon_1.5.3           </span></span>
<span><span class="co">## [106] rlang_1.1.5             KEGGREST_1.47.0</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
