<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>tidybulk: An R tidy framework for modular transcriptomic data analysis • tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="tidybulk: An R tidy framework for modular transcriptomic data analysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparison_coding.html">Side-by-side comparison with standard interfaces</a>
    </li>
    <li>
      <a href="../articles/introduction.html">tidybulk: An R tidy framework for modular transcriptomic data analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/tidybulk/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>tidybulk: An R tidy framework for modular
transcriptomic data analysis</h1>
                        <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            
            <h4 data-toc-skip class="date">2025-10-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/tidybulk/blob/master/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <a href="https://github.com/stemangiola/tidybulk/actions/" class="external-link"><img src="https://github.com/stemangiola/tidybulk/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <a href="https://bioconductor.org/checkResults/release/bioc-LATEST/tidybulk/" class="external-link"><img src="https://bioconductor.org/shields/build/release/bioc/tidybulk.svg" alt="Bioconductor status"></a> <!-- badges: end --></p>
<p><strong>tidybulk</strong> is a powerful R package designed for
modular transcriptomic data analysis that brings transcriptomics to the
tidyverse.</p>
<div class="section level3">
<h3 id="why-tidybulk">Why tidybulk?<a class="anchor" aria-label="anchor" href="#why-tidybulk"></a>
</h3>
<p>Tidybulk provides a unified interface for comprehensive
transcriptomic data analysis with seamless integration of
SummarizedExperiment objects and tidyverse principles. It streamlines
the entire workflow from raw data to biological insights.</p>
</div>
<div class="section level3">
<h3 id="functionsutilities-available">Functions/utilities available<a class="anchor" aria-label="anchor" href="#functionsutilities-available"></a>
</h3>
<div class="section level4">
<h4 id="abundance-normalization-functions">Abundance Normalization Functions<a class="anchor" aria-label="anchor" href="#abundance-normalization-functions"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/scale_abundance-methods.html">scale_abundance()</a></code></td>
<td>Scale abundance data</td>
</tr>
<tr class="even">
<td><code><a href="../reference/quantile_normalise_abundance-methods.html">quantile_normalise_abundance()</a></code></td>
<td>Quantile normalization</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/adjust_abundance-methods.html">adjust_abundance()</a></code></td>
<td>Adjust abundance for unwanted variation</td>
</tr>
<tr class="even">
<td><code><a href="../reference/fill_missing_abundance-methods.html">fill_missing_abundance()</a></code></td>
<td>Fill missing abundance values</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/impute_missing_abundance-methods.html">impute_missing_abundance()</a></code></td>
<td>Impute missing abundance values</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="filtering-and-selection-functions">Filtering and Selection Functions<a class="anchor" aria-label="anchor" href="#filtering-and-selection-functions"></a>
</h4>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/identify_abundant-methods.html">identify_abundant()</a></code></td>
<td>Identify abundant transcripts without removing them</td>
</tr>
<tr class="even">
<td><code><a href="../reference/keep_abundant.html">keep_abundant()</a></code></td>
<td>Keep abundant transcripts</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/keep_variable-methods.html">keep_variable()</a></code></td>
<td>Keep variable transcripts</td>
</tr>
<tr class="even">
<td><code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr()</a></code></td>
<td>Filter by expression</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="dimensionality-reduction-functions">Dimensionality Reduction Functions<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-functions"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions()</a></code></td>
<td>Reduce dimensions with PCA/MDS/tSNE/UMAP</td>
</tr>
<tr class="even">
<td><code><a href="../reference/rotate_dimensions-methods.html">rotate_dimensions()</a></code></td>
<td>Rotate dimensions</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/remove_redundancy-methods.html">remove_redundancy()</a></code></td>
<td>Remove redundant features</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="clustering-functions">Clustering Functions<a class="anchor" aria-label="anchor" href="#clustering-functions"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/cluster_elements-methods.html">cluster_elements()</a></code></td>
<td>Cluster elements with various methods</td>
</tr>
<tr class="even">
<td><code>kmeans clustering</code></td>
<td>K-means clustering</td>
</tr>
<tr class="odd">
<td><code>SNN clustering</code></td>
<td>Shared nearest neighbor clustering</td>
</tr>
<tr class="even">
<td><code>hierarchical clustering</code></td>
<td>Hierarchical clustering</td>
</tr>
<tr class="odd">
<td><code>DBSCAN clustering</code></td>
<td>Density-based clustering</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="differential-analysis-functions">Differential Analysis Functions<a class="anchor" aria-label="anchor" href="#differential-analysis-functions"></a>
</h4>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code><a href="../reference/test_differential_expression-methods.html">test_differential_expression()</a></code></td>
<td>Test differential expression with various methods</td>
</tr></tbody>
</table>
</div>
<div class="section level4">
<h4 id="cellularity-analysis-functions">Cellularity Analysis Functions<a class="anchor" aria-label="anchor" href="#cellularity-analysis-functions"></a>
</h4>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity()</a></code></td>
<td>Deconvolve cellularity with various methods</td>
</tr>
<tr class="even">
<td><code>cibersort()</code></td>
<td>CIBERSORT analysis</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="gene-enrichment-functions">Gene Enrichment Functions<a class="anchor" aria-label="anchor" href="#gene-enrichment-functions"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/test_gene_enrichment-methods.html">test_gene_enrichment()</a></code></td>
<td>Test gene enrichment</td>
</tr>
<tr class="even">
<td><code><a href="../reference/test_gene_overrepresentation-methods.html">test_gene_overrepresentation()</a></code></td>
<td>Test gene overrepresentation</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/test_gene_rank-methods.html">test_gene_rank()</a></code></td>
<td>Test gene rank</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="utility-functions">Utility Functions<a class="anchor" aria-label="anchor" href="#utility-functions"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/describe_transcript-methods.html">describe_transcript()</a></code></td>
<td>Describe transcript characteristics</td>
</tr>
<tr class="even">
<td><code><a href="../reference/get_bibliography-methods.html">get_bibliography()</a></code></td>
<td>Get bibliography</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/resolve_complete_confounders_of_non_interest.html">resolve_complete_confounders_of_non_interest()</a></code></td>
<td>Resolve confounders</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="validation-and-utility-functions">Validation and Utility Functions<a class="anchor" aria-label="anchor" href="#validation-and-utility-functions"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/check_if_counts_is_na.html">check_if_counts_is_na()</a></code></td>
<td>Check if counts contain NA values</td>
</tr>
<tr class="even">
<td><code><a href="../reference/check_if_duplicated_genes.html">check_if_duplicated_genes()</a></code></td>
<td>Check for duplicated genes</td>
</tr>
<tr class="odd">
<td><code>check_if_wrong_input()</code></td>
<td>Validate input data</td>
</tr>
<tr class="even">
<td><code><a href="../reference/log10_reverse_trans.html">log10_reverse_trans()</a></code></td>
<td>Log10 reverse transformation</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/logit_trans.html">logit_trans()</a></code></td>
<td>Logit transformation</td>
</tr>
</tbody>
</table>
<p>All functions are directly compatible with
<code>SummarizedExperiment</code> objects and follow tidyverse
principles for seamless integration with the tidyverse ecosystem.</p>
</div>
<div class="section level4">
<h4 id="scientific-citation">Scientific Citation<a class="anchor" aria-label="anchor" href="#scientific-citation"></a>
</h4>
<p>Mangiola, Stefano, Ramyar Molania, Ruining Dong, Maria A. Doyle, and
Anthony T. Papenfuss. 2021. “Tidybulk: An R tidy framework for modular
transcriptomic data analysis.” Genome Biology 22 (42). <a href="https://doi.org/10.1186/s13059-020-02233-7" class="external-link uri">https://doi.org/10.1186/s13059-020-02233-7</a></p>
<p><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7" class="external-link">Genome
Biology - tidybulk: an R tidy framework for modular transcriptomic data
analysis</a></p>
</div>
</div>
<div class="section level2">
<h2 id="full-vignette">Full vignette<a class="anchor" aria-label="anchor" href="#full-vignette"></a>
</h2>
<p><a href="https://stemangiola.github.io/tidybulk/" class="external-link">HERE</a></p>
</div>
<div class="section level2">
<h2 id="installation-guide">Installation Guide<a class="anchor" aria-label="anchor" href="#installation-guide"></a>
</h2>
<p><strong>Bioconductor</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"tidybulk"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Github</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/tidybulk"</span><span class="op">)</span></span></code></pre></div>
<p>In this vignette we will use the <code>airway</code> dataset, a
<code>SummarizedExperiment</code> object containing RNA-seq data from an
experiment studying the effect of dexamethasone treatment on airway
smooth muscle cells. This dataset is available in the <a href="https://bioconductor.org/packages/airway/" class="external-link">airway</a> package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
<p>This workflow, will use the <a href="https://bioconductor.org/packages/tidySummarizedExperiment/" class="external-link">tidySummarizedExperiment</a>
package to manipulate the data in a <code>tidyverse</code> fashion. This
approach streamlines the data manipulation and analysis process, making
it more efficient and easier to understand.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySummarizedExperiment" class="external-link">tidySummarizedExperiment</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'tidySummarizedExperiment'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:generics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     tidy</span></span></code></pre>
<p>Here we will add a gene symbol column to the <code>airway</code>
object. This will be used to interpret the differential expression
analysis, and to deconvolve the cellularity.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: AnnotationDbi</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'AnnotationDbi'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     select</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/AnnotationDbi" class="external-link">AnnotationDbi</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add gene symbol and entrez</span></span>
<span><span class="va">airway</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>entrezid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>,</span>
<span>                                      keys <span class="op">=</span> <span class="va">gene_name</span>,</span>
<span>                                      keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                                      column <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>                                      multiVals <span class="op">=</span> <span class="st">"first"</span></span>
<span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:org.Hs.eg.db"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:AnnotationDbi"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="comprehensive-example-pipeline">Comprehensive Example Pipeline<a class="anchor" aria-label="anchor" href="#comprehensive-example-pipeline"></a>
</h2>
<p>This vignette demonstrates a complete transcriptomic analysis
workflow using tidybulk, with special emphasis on differential
expression analysis.</p>
<div class="section level3">
<h3 id="data-overview">Data Overview<a class="anchor" aria-label="anchor" href="#data-overview"></a>
</h3>
<p>We will use the <code>airway</code> dataset, a
<code>SummarizedExperiment</code> object containing RNA-seq data from an
experiment studying the effect of dexamethasone treatment on airway
smooth muscle cells:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SummarizedExperiment-tibble abstraction: 509,416 × 23</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=63677 | Samples=8 | Assays=counts</span></span></span>
<span><span class="co">##    .feature        .sample   counts SampleName cell  dex   albut Run   avgLength</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ENSG00000000003 SRR10395…    679 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ENSG00000000005 SRR10395…      0 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ENSG00000000419 SRR10395…    467 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ENSG00000000457 SRR10395…    260 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ENSG00000000460 SRR10395…     60 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ENSG00000000938 SRR10395…      0 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ENSG00000000971 SRR10395…   <span style="text-decoration: underline;">3</span>251 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ENSG00000001036 SRR10395…   <span style="text-decoration: underline;">1</span>433 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ENSG00000001084 SRR10395…    519 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ENSG00000001167 SRR10395…    394 GSM1275862 N613… untrt untrt SRR1…       126</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 40 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 14 more variables: Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   gene_id &lt;chr&gt;, gene_name &lt;chr&gt;, entrezid &lt;chr&gt;, gene_biotype &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   gene_seq_start &lt;int&gt;, gene_seq_end &lt;int&gt;, seq_name &lt;chr&gt;, seq_strand &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   seq_coord_system &lt;int&gt;, symbol &lt;chr&gt;, GRangesList &lt;list&gt;</span></span></span></code></pre>
<p>Loading <code>tidySummarizedExperiment</code> makes the
<code>SummarizedExperiment</code> objects compatible with tidyverse
tools while maintaining its <code>SummarizedExperiment</code> nature.
This is useful because it allows us to use the <code>tidyverse</code>
tools to manipulate the data.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "RangedSummarizedExperiment"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "SummarizedExperiment"</span></span></code></pre>
<div class="section level4">
<h4 id="prepare-data-for-analysis">Prepare Data for Analysis<a class="anchor" aria-label="anchor" href="#prepare-data-for-analysis"></a>
</h4>
<p>Before analysis, we need to ensure our variables are in the correct
format:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert dex to factor for proper differential expression analysis</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>dex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">dex</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualize-raw-counts">Visualize Raw Counts<a class="anchor" aria-label="anchor" href="#visualize-raw-counts"></a>
</h4>
<p>Visualize the distribution of raw counts before any filtering:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">dex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Raw counts by treatment (before any filtering)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/plot-raw-counts-1.png" width="960"></p>
</div>
</div>
<div class="section level3">
<h3 id="step-1-data-preprocessing">Step 1: Data Preprocessing<a class="anchor" aria-label="anchor" href="#step-1-data-preprocessing"></a>
</h3>
<div class="section level4">
<h4 id="aggregate-duplicated-transcripts-optional">Aggregate Duplicated Transcripts (optional)<a class="anchor" aria-label="anchor" href="#aggregate-duplicated-transcripts-optional"></a>
</h4>
<p>Aggregate duplicated transcripts (e.g., isoforms, ensembl IDs):</p>
<blockquote>
<p>Transcript aggregation is a standard bioinformatics approach for
gene-level summarization.</p>
</blockquote>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Aggregate duplicates</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="st">"gene_name"</span>, aggregation_function <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="abundance-filtering">Abundance Filtering<a class="anchor" aria-label="anchor" href="#abundance-filtering"></a>
</h4>
<p>Abundance filtering can be performed using established methods <span class="citation">(<a href="#ref-robinson2010edger">Robinson et al.
2010</a>; <a href="#ref-chen2016edgeR">Chen et al. 2016</a>)</span>.</p>
<div class="section level5">
<h5 id="run-multiple-methods">Run multiple methods<a class="anchor" aria-label="anchor" href="#run-multiple-methods"></a>
</h5>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Default (simple filtering)</span></span>
<span><span class="va">airway_abundant_default</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span>minimum_counts <span class="op">=</span> <span class="fl">10</span>, minimum_proportion <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in filterByExpr.DGEList(y, design = design, group = group, lib.size =</span></span>
<span><span class="co">## lib.size, : All samples appear to belong to the same group.</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># With factor_of_interest (recommended for complex designs)</span></span>
<span><span class="va">airway_abundant_formula</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span>minimum_counts <span class="op">=</span> <span class="fl">10</span>, minimum_proportion <span class="op">=</span> <span class="fl">0.5</span>, factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `factor_of_interest` argument of `keep_abundant()` is deprecated as of</span></span>
<span><span class="co">## tidybulk 2.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `formula_design` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The argument 'factor_of_interest' is deprecated and will be removed in a</span></span>
<span><span class="co">##   future release. Please use the 'design' or 'formula_design' argument instead.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## Warning: The `factor_of_interest` argument of `identify_abundant()` is deprecated as of</span></span>
<span><span class="co">## tidybulk 2.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `formula_design` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The argument 'factor_of_interest' is deprecated and will be removed in a</span></span>
<span><span class="co">##   future release. Please use the 'design' or 'formula_design' argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">tidybulk</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/stemangiola/tidybulk/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># With CPM threshold (using design parameter)</span></span>
<span><span class="va">airway_abundant_cpm</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span>minimum_count_per_million  <span class="op">=</span> <span class="fl">10</span>, minimum_proportion <span class="op">=</span> <span class="fl">0.5</span>, factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="compare-methods">Compare methods<a class="anchor" aria-label="anchor" href="#compare-methods"></a>
</h5>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example: summary for default tidybulk filtering</span></span>
<span><span class="co"># Before filtering</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     <span style="text-decoration: underline;">113</span>276        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># After filtering</span></span>
<span><span class="va">airway_abundant_default</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>      <span style="text-decoration: underline;">28</span>240        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_abundant_formula</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>      <span style="text-decoration: underline;">31</span>580        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_abundant_cpm</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>      <span style="text-decoration: underline;">18</span>376        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge all methods into a single tibble</span></span>
<span><span class="va">airway_abundant_all</span> <span class="op">=</span> </span>
<span>  <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"no filter"</span><span class="op">)</span>,</span>
<span>    <span class="va">airway_abundant_default</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"default"</span><span class="op">)</span>,</span>
<span>    <span class="va">airway_abundant_formula</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"formula"</span><span class="op">)</span>,</span>
<span>    <span class="va">airway_abundant_cpm</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"cpm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Density plot across methods</span></span>
<span><span class="va">airway_abundant_all</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="fu">fct_relevel</span><span class="op">(</span><span class="va">method</span>, <span class="st">"no filter"</span>, <span class="st">"default"</span>, <span class="st">"formula"</span>, <span class="st">"cpm"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Counts after abundance filtering (tidybulk default)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/filtering-density-plot-comparison-1.png" width="960"></p>
<p>Update the <code>airway</code> object with the filtered data:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway_abundant_formula</span></span></code></pre></div>
<blockquote>
<p><strong>Tip:</strong> Use <code>formula_design</code> for complex
designs, and use the CPM threshold for library-size-aware filtering.</p>
</blockquote>
</div>
</div>
<div class="section level4">
<h4 id="remove-redundant-transcripts">Remove Redundant Transcripts<a class="anchor" aria-label="anchor" href="#remove-redundant-transcripts"></a>
</h4>
<p>Redundancy removal is a standard approach for reducing highly
correlated features.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_non_redundant</span> <span class="op">=</span> </span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/remove_redundancy-methods.html">remove_redundancy</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"correlation"</span>, top <span class="op">=</span> <span class="fl">100</span>, of_samples <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Getting the 100 most variable genes</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Make  </span></span>
<span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## Warning in check_se_dimnames(se): tidySummarizedExperiment says: the assays in</span></span>
<span><span class="co">## your SummarizedExperiment have row names, but they don't agree with the row</span></span>
<span><span class="co">## names of the SummarizedExperiment object itself. It is strongly recommended to</span></span>
<span><span class="co">## make the assays consistent, to avoid erroneous matching of features.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>      <span style="text-decoration: underline;">31</span>580        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summary statistics</span></span>
<span><span class="va">airway_non_redundant</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## Warning in check_se_dimnames(se): tidySummarizedExperiment says: the assays in</span></span>
<span><span class="co">## your SummarizedExperiment have row names, but they don't agree with the row</span></span>
<span><span class="co">## names of the SummarizedExperiment object itself. It is strongly recommended to</span></span>
<span><span class="co">## make the assays consistent, to avoid erroneous matching of features.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>      <span style="text-decoration: underline;">31</span>578        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot before and after</span></span>
<span><span class="co"># Merge before and after into a single tibble</span></span>
<span><span class="va">airway_all</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"before"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway_non_redundant</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"after"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Density plot</span></span>
<span><span class="va">airway_all</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="fu">fct_relevel</span><span class="op">(</span><span class="va">method</span>, <span class="st">"before"</span>, <span class="st">"after"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Counts after removing redundant transcripts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/preprocessing-remove-redundancy-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="filter-variable-transcripts">Filter Variable Transcripts<a class="anchor" aria-label="anchor" href="#filter-variable-transcripts"></a>
</h4>
<p>Keep only the most variable transcripts for downstream analysis.</p>
<p>Variance-based feature selection using edgeR methodology <span class="citation">(<a href="#ref-robinson2010edger">Robinson et al.
2010</a>)</span> is used for selecting informative features.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway_variable</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 500 most variable genes</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="visualize-after-variable-filtering-variable-transcripts-optional">Visualize After Variable Filtering Variable Transcripts
(optional)<a class="anchor" aria-label="anchor" href="#visualize-after-variable-filtering-variable-transcripts-optional"></a>
</h4>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Before filtering</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## Warning in check_se_dimnames(se): tidySummarizedExperiment says: the assays in</span></span>
<span><span class="co">## your SummarizedExperiment have row names, but they don't agree with the row</span></span>
<span><span class="co">## names of the SummarizedExperiment object itself. It is strongly recommended to</span></span>
<span><span class="co">## make the assays consistent, to avoid erroneous matching of features.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>      <span style="text-decoration: underline;">31</span>580        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># After filtering</span></span>
<span><span class="va">airway_variable</span> <span class="op">|&gt;</span> <span class="fu">summarise</span><span class="op">(</span></span>
<span>  n_features <span class="op">=</span> <span class="fu">n_distinct</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span>,</span>
<span>  min_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  median_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>  max_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## Warning in check_se_dimnames(se): tidySummarizedExperiment says: the assays in</span></span>
<span><span class="co">## your SummarizedExperiment have row names, but they don't agree with the row</span></span>
<span><span class="co">## names of the SummarizedExperiment object itself. It is strongly recommended to</span></span>
<span><span class="co">## make the assays consistent, to avoid erroneous matching of features.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">##   n_features min_count median_count max_count</span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>       <span style="text-decoration: underline;">1</span>000        <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>        <span style="color: #BB0000;">NA</span></span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Density plot</span></span>
<span><span class="co"># Merge before and after into a single tibble</span></span>
<span><span class="va">airway_all</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"before"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway_variable</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"after"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Density plot</span></span>
<span><span class="va">airway_all</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="fu">fct_relevel</span><span class="op">(</span><span class="va">method</span>, <span class="st">"before"</span>, <span class="st">"after"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Counts after variable filtering"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/filtering-variable-summary-and-plot-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="scale-abundance">Scale Abundance<a class="anchor" aria-label="anchor" href="#scale-abundance"></a>
</h4>
<p>Scale for sequencing depth using TMM <span class="citation">(<a href="#ref-robinson2010edger">Robinson et al. 2010</a>)</span>, upper
quartile <span class="citation">(<a href="#ref-bullard2010uq">Bullard et
al. 2010</a>)</span>, and RLE <span class="citation">(<a href="#ref-anders2010rle">Anders and Huber 2010</a>)</span>
normalization.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> </span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="../reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"TMM"</span>, suffix <span class="op">=</span> <span class="st">"_tmm"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"upperquartile"</span>, suffix <span class="op">=</span> <span class="st">"_upperquartile"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"RLE"</span>, suffix <span class="op">=</span> <span class="st">"_RLE"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidybulk says: the sample with largest library size SRR1039517 was chosen as reference for scaling</span></span>
<span><span class="co">## tidybulk says: the sample with largest library size SRR1039517 was chosen as reference for scaling</span></span>
<span><span class="co">## tidybulk says: the sample with largest library size SRR1039517 was chosen as reference for scaling</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="visualize-after-scaling">Visualize After Scaling<a class="anchor" aria-label="anchor" href="#visualize-after-scaling"></a>
</h4>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Before scaling</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">##      7.38     70.50    315.25   1349.47    961.47 328812.62</span></span></code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_tmm"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">##     10.50     95.82    432.54   1848.78   1315.63 457450.72</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_upperquartile"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">##     10.47     97.16    438.49   1871.36   1331.12 462648.85</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_RLE"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span></span>
<span><span class="co">##     10.45     95.61    431.75   1845.81   1313.26 456675.06</span></span></code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge all methods into a single tibble</span></span>
<span><span class="va">airway_scaled_all</span> <span class="op">=</span> <span class="fu">bind_rows</span><span class="op">(</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"no_scaling"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_tmm"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_upperquartile"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"upperquartile"</span><span class="op">)</span>,</span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="st">"counts_RLE"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">".feature"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span>, names_to <span class="op">=</span> <span class="st">".sample"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">mutate</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"RLE"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Density plot</span></span>
<span><span class="va">airway_scaled_all</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group <span class="op">=</span> <span class="va">.sample</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="fu">fct_relevel</span><span class="op">(</span><span class="va">method</span>, <span class="st">"no_scaling"</span>, <span class="st">"TMM"</span>, <span class="st">"upperquartile"</span>, <span class="st">"RLE"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scaled counts by method (after scaling)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/normalization-visualize-scaling-1.png" width="960"></p>
</div>
</div>
<div class="section level3">
<h3 id="step-2-exploratory-data-analysis">Step 2: Exploratory Data Analysis<a class="anchor" aria-label="anchor" href="#step-2-exploratory-data-analysis"></a>
</h3>
<div class="section level4">
<h4 id="remove-zero-variance-features-required-for-pca">Remove Zero-Variance Features (required for PCA)<a class="anchor" aria-label="anchor" href="#remove-zero-variance-features-required-for-pca"></a>
</h4>
<p>Variance filtering is a standard preprocessing step for
dimensionality reduction.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span></span>
<span><span class="co"># Remove features with zero variance across samples</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowVars.html" class="external-link">rowVars</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="dimensionality-reduction">Dimensionality Reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h4>
<p>MDS <span class="citation">(<a href="#ref-kruskal1964mds">Kruskal
1964</a>)</span> using limma::plotMDS <span class="citation">(<a href="#ref-ritchie2015limma"><strong>ritchie2015limma?</strong></a>)</span>
and PCA <span class="citation">(<a href="#ref-hotelling1933pca">Hotelling 1933</a>)</span> are used for
dimensionality reduction.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"MDS"</span>, .dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 500 most variable genes</span></span></code></pre>
<pre><code><span><span class="co">## [1] "MDS result_df colnames: sample, 1, 2"</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the raw results do `metadata(.)$tidybulk$MDS`</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"PCA"</span>, .dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting the 500 most variable genes</span></span></code></pre>
<pre><code><span><span class="co">## Fraction of variance explained by the selected principal components</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   `Fraction of variance`    PC</span></span>
<span><span class="co">##                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>                 0.065<span style="text-decoration: underline;">4</span>     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>                 0.055<span style="text-decoration: underline;">8</span>     2</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the raw results do `metadata(.)$tidybulk$PCA`</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="visualize-dimensionality-reduction-results">Visualize Dimensionality Reduction Results<a class="anchor" aria-label="anchor" href="#visualize-dimensionality-reduction-results"></a>
</h4>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># MDS plot</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"MDS Analysis"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/eda-plot-dimensionality-reduction-1.png" width="960"></p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCA plot</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`PC1`</span>, y<span class="op">=</span><span class="va">`PC2`</span>, color<span class="op">=</span><span class="va">`dex`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="va">my_theme</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PCA Analysis"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/eda-plot-dimensionality-reduction-2.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="clustering-analysis">Clustering Analysis<a class="anchor" aria-label="anchor" href="#clustering-analysis"></a>
</h4>
<p>K-means clustering <span class="citation">(<a href="#ref-macqueen1967kmeans">MacQueen 1967</a>)</span> is used for
unsupervised grouping.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/cluster_elements-methods.html">cluster_elements</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"kmeans"</span>, centers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize clustering</span></span>
<span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">`Dim1`</span>, y<span class="op">=</span><span class="va">`Dim2`</span>, color<span class="op">=</span><span class="va">`cluster_kmeans`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"K-means Clustering"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/eda-clustering-analysis-1.png" width="960"></p>
</div>
</div>
<div class="section level3">
<h3 id="step-3-differential-expression-analysis">Step 3: Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#step-3-differential-expression-analysis"></a>
</h3>
<div class="section level4">
<h4 id="basic-differential-expression">Basic Differential Expression<a class="anchor" aria-label="anchor" href="#basic-differential-expression"></a>
</h4>
<p><strong>Methods:</strong></p>
<ul>
<li><p><strong>edgeR quasi-likelihood:</strong> Quasi-likelihood F-tests
for differential expression <span class="citation">(<a href="#ref-robinson2010edger">Robinson et al. 2010</a>; <a href="#ref-chen2016edgeR">Chen et al. 2016</a>)</span></p></li>
<li><p><strong>edgeR robust likelihood ratio:</strong> Robust likelihood
ratio tests <span class="citation">(<a href="#ref-chen2016edgeR">Chen et
al. 2016</a>)</span></p></li>
<li><p><strong>DESeq2:</strong> Negative binomial distribution with
dispersion estimation <span class="citation">(<a href="#ref-love2014deseq2">Love et al. 2014</a>)</span></p></li>
<li><p><strong>limma-voom:</strong> Linear modeling with empirical Bayes
moderation <span class="citation">(<a href="#ref-law2014voom">Law et al.
2014</a>)</span></p></li>
<li><p><strong>limma-voom with sample weights:</strong> Enhanced voom
with quality weights <span class="citation">(<a href="#ref-liu2015voomweights">Liu et al. 2015</a>)</span></p></li>
</ul>
<p><strong>References:</strong></p>
<ul>
<li><p>Robinson et al. (2010) edgeR: a Bioconductor package for
differential expression analysis</p></li>
<li><p>Chen et al. (2016) From reads to genes to pathways: differential
expression analysis of RNA-Seq experiments using Rsubread and the edgeR
quasi-likelihood pipeline</p></li>
<li><p>Love et al. (2014) Moderated estimation of fold change and
dispersion for RNA-seq data with DESeq2</p></li>
<li><p>Law et al. (2014) voom: precision weights unlock linear model
analysis tools for RNA-seq read counts</p></li>
<li><p>Liu et al. (2015) Why weight? Modelling sample and observational
level variability improves power in RNA-seq analyses</p></li>
</ul>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Standard differential expression analysis</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span><span class="co"># Use QL method</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>, prefix <span class="op">=</span> <span class="st">"ql__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Use edger_robust_likelihood_ratio</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"edger_robust_likelihood_ratio"</span>, prefix <span class="op">=</span> <span class="st">"lr_robust__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span><span class="co"># Use DESeq2 method</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"DESeq2"</span>, prefix <span class="op">=</span> <span class="st">"deseq2__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Use limma_voom</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"limma_voom"</span>, prefix <span class="op">=</span> <span class="st">"voom__"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span><span class="co"># Use limma_voom_sample_weights</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span>, method <span class="op">=</span> <span class="st">"limma_voom_sample_weights"</span>, prefix <span class="op">=</span> <span class="st">"voom_weights__"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `.abundance` argument of `test_differential_abundance()` is deprecated as</span></span>
<span><span class="co">## of tidybulk 2.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `abundance` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">tidybulk</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/stemangiola/tidybulk/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="quality-control-of-the-fit">Quality Control of the Fit<a class="anchor" aria-label="anchor" href="#quality-control-of-the-fit"></a>
</h4>
<p>It is important to check the quality of the fit. All methods produce
a fit object that can be used for quality control. The fit object
produced by each underlying method are stored in as attributes of the
<code>airway_mini</code> object. We can use them for example to perform
quality control of the fit.</p>
<div class="section level5">
<h5 id="for-edger">For edgeR<a class="anchor" aria-label="anchor" href="#for-edger"></a>
</h5>
<p>Plot the biological coefficient of variation (BCV) trend. This plot
is helpful to understant the dispersion of the data.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: limma</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'limma'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">edgeR_quasi_likelihood_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/differential-expression-edgeR-object-1.png" width="700"></p>
<p>Plot the log-fold change vs mean plot.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">edgeR_quasi_likelihood_fit</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/differential-expression-edgeR-fit-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="for-deseq2">For DESeq2<a class="anchor" aria-label="anchor" href="#for-deseq2"></a>
</h5>
<p>Plot the mean-variance trend.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">DESeq2_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">plotDispEsts</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/differential-expression-DESeq2-object-1.png" width="700"></p>
<p>Plot the log-fold change vs mean plot.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">$</span><span class="va">tidybulk</span><span class="op">$</span><span class="va">DESeq2_object</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotMA.html" class="external-link">plotMA</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/differential-expression-DESeq2-fit-1.png" width="700"></p>
</div>
</div>
<div class="section level4">
<h4 id="histograms-of-p-values-across-methods">Histograms of p-values across methods<a class="anchor" aria-label="anchor" href="#histograms-of-p-values-across-methods"></a>
</h4>
<p>Inspection of the raw p-value histogram provides a rapid check of
differential-expression results. When no gene is truly differentially
expressed, the p-values follow a uniform U(0,1) distribution across the
interval 0–1, so the histogram appears flat <a href="https://bioconductor.org/help/course-materials/2014/useR2014/Workflows.html" class="external-link">Source</a>.
In a more realistic scenario where only a subset of genes changes, this
uniform background is still present but an obvious spike emerges close
to zero, created by the genuine signals.</p>
<p>Thanks to the modularity of the <code>tidybulk</code> workflow, that
can multiplex different methods, we can easily compare the p-values
across methods.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">ql__PValue</span>, </span>
<span>    <span class="va">lr_robust__PValue</span>, </span>
<span>    <span class="va">voom__P.Value</span>, </span>
<span>    <span class="va">voom_weights__P.Value</span>, </span>
<span>    <span class="va">deseq2__pvalue</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="fu">everything</span><span class="op">(</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"method"</span>, values_to <span class="op">=</span> <span class="st">"pvalue"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pvalue</span>, fill <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Histogram of p-values across methods"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 29 rows containing non-finite outside the scale range</span></span>
<span><span class="co">## (`stat_bin()`).</span></span></code></pre>
<p><img src="introduction_files/figure-html/differential-expression-pvalue-histograms-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="compare-results-across-methods">Compare Results Across Methods<a class="anchor" aria-label="anchor" href="#compare-results-across-methods"></a>
</h4>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summary statistics</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"ql|lr_robust|voom|voom_weights|deseq2"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"logFC"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">summarise</span><span class="op">(</span><span class="fu">across</span><span class="op">(</span><span class="fu">everything</span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min <span class="op">=</span> <span class="va">min</span>, median <span class="op">=</span> <span class="va">median</span>, max <span class="op">=</span> <span class="va">max</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: There was 1 warning in `summarise()`.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> In argument: `across(...)`.</span></span>
<span><span class="co">## Caused by warning:</span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> The `...` argument of `across()` is deprecated as of dplyr 1.1.0.</span></span>
<span><span class="co">## Supply arguments directly to `.fns` through an anonymous function instead.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   # Previously</span></span>
<span><span class="co">##   across(a:b, mean, na.rm = TRUE)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   # Now</span></span>
<span><span class="co">##   across(a:b, \(x) mean(x, na.rm = TRUE))</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1 × 0</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="pairplot-of-pvalues-across-methods-ggpairs">Pairplot of pvalues across methods (GGpairs)<a class="anchor" aria-label="anchor" href="#pairplot-of-pvalues-across-methods-ggpairs"></a>
</h4>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/" class="external-link">GGally</a></span><span class="op">)</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">ql__PValue</span>, <span class="va">lr_robust__PValue</span>, <span class="va">voom__P.Value</span>, <span class="va">voom_weights__P.Value</span>, <span class="va">deseq2__pvalue</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_y_log10_reverse.html">scale_y_log10_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_x_log10_reverse.html">scale_x_log10_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Pairplot of p-values across methods"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/differential-expression-pvalue-pairplot-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="pairplot-of-effect-sizes-across-methods-ggpairs">Pairplot of effect sizes across methods (GGpairs)<a class="anchor" aria-label="anchor" href="#pairplot-of-effect-sizes-across-methods-ggpairs"></a>
</h4>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/" class="external-link">GGally</a></span><span class="op">)</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">ql__logFC</span>, <span class="va">lr_robust__logFC</span>, <span class="va">voom__logFC</span>, <span class="va">voom_weights__logFC</span>, <span class="va">deseq2__log2FoldChange</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Pairplot of effect sizes across methods"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/differential-expression-effectsize-pairplot-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="volcano-plots-for-each-method">Volcano Plots for Each Method<a class="anchor" aria-label="anchor" href="#volcano-plots-for-each-method"></a>
</h4>
<p>Visualising the significance and effect size of the differential
expression results as a volcano plots we appreciate that some methods
have much lower p-values distributions than other methods, for the same
model and data.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create volcano plots</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Select the columns we want to plot</span></span>
<span>    <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>            <span class="va">.feature</span>,</span>
<span>      <span class="va">ql__logFC</span>, <span class="va">ql__PValue</span>,</span>
<span>      <span class="va">lr_robust__logFC</span>, <span class="va">lr_robust__PValue</span>,</span>
<span>      <span class="va">voom__logFC</span>, <span class="va">voom__P.Value</span>,</span>
<span>      <span class="va">voom_weights__logFC</span>, <span class="va">voom_weights__P.Value</span>,</span>
<span>      <span class="va">deseq2__log2FoldChange</span>, <span class="va">deseq2__pvalue</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Pivot longer to get a tidy data frame</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>      <span class="op">-</span> <span class="va">.feature</span>,</span>
<span>      names_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"method"</span>, <span class="st">"stat"</span><span class="op">)</span>,</span>
<span>      values_to <span class="op">=</span> <span class="st">"value"</span>, names_sep <span class="op">=</span> <span class="st">"__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Harmonize column names</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>stat  <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"log2FoldChange"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"logFC"</span>,</span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PValue"</span>, <span class="st">"pvalue"</span>, <span class="st">"P.Value"</span>, <span class="st">"p.value"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"PValue"</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"stat"</span>, values_from <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, size <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_y_log10_reverse.html">scale_y_log10_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="fl">0.5</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Volcano Plots by Method"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 29 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_point()`).</span></span></code></pre>
<p><img src="introduction_files/figure-html/differential-expression-volcano-plots-1-1.png" width="960"></p>
<p>Plotting independent y-axis scales for the p-values and effect sizes
allows us to compare the top genes across methods.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create volcano plots</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Select the columns we want to plot</span></span>
<span>    <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>            <span class="va">symbol</span>,</span>
<span>      <span class="va">ql__logFC</span>, <span class="va">ql__PValue</span>,</span>
<span>      <span class="va">lr_robust__logFC</span>, <span class="va">lr_robust__PValue</span>,</span>
<span>      <span class="va">voom__logFC</span>, <span class="va">voom__P.Value</span>,</span>
<span>      <span class="va">voom_weights__logFC</span>, <span class="va">voom_weights__P.Value</span>,</span>
<span>      <span class="va">deseq2__log2FoldChange</span>, <span class="va">deseq2__pvalue</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Pivot longer to get a tidy data frame</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span></span>
<span>      <span class="op">-</span> <span class="va">symbol</span>,</span>
<span>      names_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"method"</span>, <span class="st">"stat"</span><span class="op">)</span>,</span>
<span>      values_to <span class="op">=</span> <span class="st">"value"</span>, names_sep <span class="op">=</span> <span class="st">"__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Harmonize column names</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>stat  <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"log2FoldChange"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"logFC"</span>,</span>
<span>        <span class="va">stat</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PValue"</span>, <span class="st">"pvalue"</span>, <span class="st">"P.Value"</span>, <span class="st">"p.value"</span><span class="op">)</span> <span class="op">~</span> <span class="st">"PValue"</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"stat"</span>, values_from <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, size <span class="op">=</span> <span class="va">PValue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">symbol</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/scale_y_log10_reverse.html">scale_y_log10_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="fl">0.5</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Volcano Plots by Method"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 29 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_point()`).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 29 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_text_repel()`).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: ggrepel: 15750 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre>
<pre><code><span><span class="co">## Warning: ggrepel: 15790 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre>
<pre><code><span><span class="co">## Warning: ggrepel: 15780 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre>
<pre><code><span><span class="co">## Warning: ggrepel: 15789 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre>
<pre><code><span><span class="co">## Warning: ggrepel: 15790 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre>
<p><img src="introduction_files/figure-html/differential-expression-volcano-plots-2-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="differential-expression-with-contrasts">Differential Expression with Contrasts<a class="anchor" aria-label="anchor" href="#differential-expression-with-contrasts"></a>
</h4>
<p>Contrast-based differential expression analysis, is available for
most methods. It is a standard statistical approach for testing specific
comparisons in complex designs.</p>
<div class="section level5">
<h5 id="for-edger-1">For edgeR<a class="anchor" aria-label="anchor" href="#for-edger-1"></a>
</h5>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using contrasts for more complex comparisons</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,                  </span>
<span>        contrasts  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dextrt - dexuntrt"</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>, </span>
<span>        prefix <span class="op">=</span> <span class="st">"contrasts__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Print the gene statistics</span></span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"contrasts"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidybulk says: The design column names are "dextrt, dexuntrt"</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$edgeR_quasi_likelihood_object`</span></span>
<span><span class="co">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$edgeR_quasi_likelihood_fit`</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 15,790 × 5</span></span></span>
<span><span class="co">##    contrasts__logFC___dextrt...d…¹ contrasts__logCPM___…² contrasts__F___dextr…³</span></span>
<span><span class="co">##                              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>                         -<span style="color: #BB0000;">0.389</span>                    5.10                 6.81  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>                          0.194                    4.65                 6.00  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>                          0.023<span style="text-decoration: underline;">1</span>                   3.52                 0.054<span style="text-decoration: underline;">4</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>                         -<span style="color: #BB0000;">0.125</span>                    1.52                 0.219 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>                          0.432                    8.13                 2.92  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>                         -<span style="color: #BB0000;">0.251</span>                    5.95                 6.14  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>                         -<span style="color: #BB0000;">0.039</span><span style="color: #BB0000; text-decoration: underline;">5</span>                   4.88                 0.033<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>                         -<span style="color: #BB0000;">0.498</span>                    4.16                 7.07  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>                         -<span style="color: #BB0000;">0.141</span>                    3.16                 1.06  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>                         -<span style="color: #BB0000;">0.053</span><span style="color: #BB0000; text-decoration: underline;">3</span>                   7.08                 0.033<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 15,780 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ abbreviated names: ¹​contrasts__logFC___dextrt...dexuntrt,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ²​contrasts__logCPM___dextrt...dexuntrt, ³​contrasts__F___dextrt...dexuntrt</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: contrasts__PValue___dextrt...dexuntrt &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   contrasts__FDR___dextrt...dexuntrt &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level5">
<h5 id="for-deseq2-1">For DESeq2<a class="anchor" aria-label="anchor" href="#for-deseq2-1"></a>
</h5>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using contrasts for more complex comparisons</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,                  </span>
<span>        contrasts  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dex"</span>, <span class="st">"trt"</span>, <span class="st">"untrt"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"DESeq2"</span>,</span>
<span>        prefix <span class="op">=</span> <span class="st">"contrasts__"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"contrasts"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## estimating size factors</span></span></code></pre>
<pre><code><span><span class="co">## estimating dispersions</span></span></code></pre>
<pre><code><span><span class="co">## gene-wise dispersion estimates</span></span></code></pre>
<pre><code><span><span class="co">## mean-dispersion relationship</span></span></code></pre>
<pre><code><span><span class="co">## final dispersion estimates</span></span></code></pre>
<pre><code><span><span class="co">## fitting model and testing</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$DESeq2_object`</span></span>
<span><span class="co">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$DESeq2_fit`</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 15,790 × 6</span></span></span>
<span><span class="co">##    contrasts__baseMean___dex.trt…¹ contrasts__log2FoldC…² contrasts__lfcSE___d…³</span></span>
<span><span class="co">##                              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>                           710.                 -<span style="color: #BB0000;">0.385</span>                  0.172 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>                           521.                  0.197                  0.095<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>                           238.                  0.027<span style="text-decoration: underline;">9</span>                 0.122 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>                            58.0                -<span style="color: #BB0000;">0.124</span>                  0.308 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>                          <span style="text-decoration: underline;">5</span>826.                  0.434                  0.260 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>                          <span style="text-decoration: underline;">1</span>285.                 -<span style="color: #BB0000;">0.248</span>                  0.116 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>                           611.                 -<span style="color: #BB0000;">0.036</span><span style="color: #BB0000; text-decoration: underline;">0</span>                 0.233 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>                           370.                 -<span style="color: #BB0000;">0.496</span>                  0.211 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>                           184.                 -<span style="color: #BB0000;">0.140</span>                  0.164 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>                          <span style="text-decoration: underline;">2</span>820.                 -<span style="color: #BB0000;">0.049</span><span style="color: #BB0000; text-decoration: underline;">4</span>                 0.294 </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 15,780 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ abbreviated names: ¹​contrasts__baseMean___dex.trt.untrt,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ²​contrasts__log2FoldChange___dex.trt.untrt,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ³​contrasts__lfcSE___dex.trt.untrt</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: contrasts__stat___dex.trt.untrt &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   contrasts__pvalue___dex.trt.untrt &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   contrasts__padj___dex.trt.untrt &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level5">
<h5 id="for-limma-voom">For limma-voom<a class="anchor" aria-label="anchor" href="#for-limma-voom"></a>
</h5>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using contrasts for more complex comparisons</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,                  </span>
<span>        contrasts  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dextrt - dexuntrt"</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"limma_voom"</span>,</span>
<span>        prefix <span class="op">=</span> <span class="st">"contrasts__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"contrasts"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidybulk says: The design column names are "dextrt, dexuntrt"</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$limma_voom_object`</span></span>
<span><span class="co">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$limma_voom_fit`</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 15,790 × 6</span></span></span>
<span><span class="co">##    contrasts__logFC___dextrt...d…¹ contrasts__AveExpr__…² contrasts__t___dextr…³</span></span>
<span><span class="co">##                              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>                         -<span style="color: #BB0000;">0.383</span>                    5.07                -<span style="color: #BB0000;">2.55</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>                          0.196                    4.64                 2.49  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>                          0.031<span style="text-decoration: underline;">8</span>                   3.51                 0.324 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>                         -<span style="color: #BB0000;">0.067</span><span style="color: #BB0000; text-decoration: underline;">7</span>                   1.45                -<span style="color: #BB0000;">0.241</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>                          0.408                    8.07                 1.55  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>                         -<span style="color: #BB0000;">0.252</span>                    5.94                -<span style="color: #BB0000;">2.47</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>                         -<span style="color: #BB0000;">0.065</span><span style="color: #BB0000; text-decoration: underline;">6</span>                   4.84                -<span style="color: #BB0000;">0.305</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>                         -<span style="color: #BB0000;">0.481</span>                    4.11                -<span style="color: #BB0000;">2.58</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>                         -<span style="color: #BB0000;">0.159</span>                    3.14                -<span style="color: #BB0000;">1.14</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>                         -<span style="color: #BB0000;">0.027</span><span style="color: #BB0000; text-decoration: underline;">3</span>                   7.02                -<span style="color: #BB0000;">0.095</span><span style="color: #BB0000; text-decoration: underline;">6</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 15,780 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ abbreviated names: ¹​contrasts__logFC___dextrt...dexuntrt,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ²​contrasts__AveExpr___dextrt...dexuntrt, ³​contrasts__t___dextrt...dexuntrt</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: contrasts__P.Value___dextrt...dexuntrt &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   contrasts__adj.P.Val___dextrt...dexuntrt &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   contrasts__B___dextrt...dexuntrt &lt;dbl&gt;</span></span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="differential-expression-with-minimum-fold-change-treat-method">Differential Expression with minimum fold change (TREAT method)<a class="anchor" aria-label="anchor" href="#differential-expression-with-minimum-fold-change-treat-method"></a>
</h4>
<p>TREAT method <span class="citation">(<a href="#ref-mccarthy2009treat">McCarthy and Smyth 2009</a>)</span> is
used for testing significance relative to a fold-change threshold.</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using contrasts for more complex comparisons</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">dex</span>,                  </span>
<span>        contrasts  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dextrt - dexuntrt"</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>, </span>
<span>        test_above_log2_fold_change <span class="op">=</span> <span class="fl">2</span>, </span>
<span>        prefix <span class="op">=</span> <span class="st">"treat__"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="co"># Print the gene statistics</span></span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"treat"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidybulk says: The design column names are "dextrt, dexuntrt"</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the DE object do `metadata(.)$tidybulk$edgeR_quasi_likelihood_object`</span></span>
<span><span class="co">## tidybulk says: to access the raw results (fitted GLM) do `metadata(.)$tidybulk$edgeR_quasi_likelihood_fit`</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 15,790 × 5</span></span></span>
<span><span class="co">##    treat__logFC___dextrt...dexun…¹ treat__unshrunk.logF…² treat__logCPM___dext…³</span></span>
<span><span class="co">##                              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>                         -<span style="color: #BB0000;">0.389</span>                 -<span style="color: #BB0000;">0.389</span>                    5.10</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>                          0.194                  0.194                    4.65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>                          0.023<span style="text-decoration: underline;">1</span>                 0.023<span style="text-decoration: underline;">2</span>                   3.52</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>                         -<span style="color: #BB0000;">0.125</span>                 -<span style="color: #BB0000;">0.125</span>                    1.52</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>                          0.432                  0.432                    8.13</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>                         -<span style="color: #BB0000;">0.251</span>                 -<span style="color: #BB0000;">0.251</span>                    5.95</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>                         -<span style="color: #BB0000;">0.039</span><span style="color: #BB0000; text-decoration: underline;">5</span>                -<span style="color: #BB0000;">0.039</span><span style="color: #BB0000; text-decoration: underline;">5</span>                   4.88</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>                         -<span style="color: #BB0000;">0.498</span>                 -<span style="color: #BB0000;">0.498</span>                    4.16</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>                         -<span style="color: #BB0000;">0.141</span>                 -<span style="color: #BB0000;">0.141</span>                    3.16</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>                         -<span style="color: #BB0000;">0.053</span><span style="color: #BB0000; text-decoration: underline;">3</span>                -<span style="color: #BB0000;">0.053</span><span style="color: #BB0000; text-decoration: underline;">3</span>                   7.08</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 15,780 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ abbreviated names: ¹​treat__logFC___dextrt...dexuntrt,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ²​treat__unshrunk.logFC___dextrt...dexuntrt,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ³​treat__logCPM___dextrt...dexuntrt</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: treat__PValue___dextrt...dexuntrt &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   treat__FDR___dextrt...dexuntrt &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="mixed-models-for-complex-designs">Mixed Models for Complex Designs<a class="anchor" aria-label="anchor" href="#mixed-models-for-complex-designs"></a>
</h4>
<p>glmmSeq <span class="citation">(<a href="#ref-ma2020glmmseq"><span class="nocase">Ma et al.</span> 2020</a>)</span> is used for generalized
linear mixed models for RNA-seq data.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using glmmSeq for mixed models</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/keep_abundant.html">keep_abundant</a></span><span class="op">(</span>formula_design <span class="op">=</span> <span class="op">~</span> <span class="va">dex</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Select 100 genes in the interest of execution time</span></span>
<span>  <span class="va">_</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span> <span class="op">|&gt;</span></span>
<span></span>
<span>  <span class="co"># Fit model</span></span>
<span>  <span class="fu"><a href="../reference/test_differential_expression-methods.html">test_differential_expression</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">cell</span><span class="op">)</span>, </span>
<span>    method <span class="op">=</span> <span class="st">"glmmseq_lme4"</span>, </span>
<span>    cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    prefix <span class="op">=</span> <span class="st">"glmmseq__"</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## n = 8 samples, 4 individuals</span></span></code></pre>
<pre><code><span><span class="co">## Time difference of 1.271983 mins</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the DE object do</span></span>
<span><span class="co">## `metadata(.)$tidybulk$glmmseq_lme4_object`</span></span></code></pre>
<pre><code><span><span class="co">## tidybulk says: to access the raw results (fitted GLM) do</span></span>
<span><span class="co">## `metadata(.)$tidybulk$glmmseq_lme4_fit`</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SummarizedExperiment-tibble abstraction: 800 × 87</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=100 | Samples=8 | Assays=counts, counts_tmm, counts_upperquartile,</span></span></span>
<span><span class="co">## # <span style="color: #555555;">  counts_RLE</span></span></span>
<span><span class="co">##    .feature .sample counts counts_tmm counts_upperquartile counts_RLE SampleName</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> TSPAN6   SRR103…    679      937.                 972.       930.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> DPM1     SRR103…    467      644.                 669.       640.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> SCYL3    SRR103…    260      359.                 372.       356.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> C1orf112 SRR103…     60       82.8                 85.9       82.2 GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CFH      SRR103…   <span style="text-decoration: underline;">3</span>251     <span style="text-decoration: underline;">4</span>485.                <span style="text-decoration: underline;">4</span>654.      <span style="text-decoration: underline;">4</span>454.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> FUCA2    SRR103…   <span style="text-decoration: underline;">1</span>433     <span style="text-decoration: underline;">1</span>977.                <span style="text-decoration: underline;">2</span>052.      <span style="text-decoration: underline;">1</span>963.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> GCLC     SRR103…    519      716.                 743.       711.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> NFYA     SRR103…    394      544.                 564.       540.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> STPG1    SRR103…    172      237.                 246.       236.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> NIPAL3   SRR103…   <span style="text-decoration: underline;">2</span>112     <span style="text-decoration: underline;">2</span>914.                <span style="text-decoration: underline;">3</span>024.      <span style="text-decoration: underline;">2</span>893.  GSM1275862</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 40 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 80 more variables: cell &lt;fct&gt;, dex &lt;fct&gt;, albut &lt;fct&gt;, Run &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   avgLength &lt;int&gt;, Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   cluster_kmeans &lt;fct&gt;, gene_name &lt;chr&gt;, gene_seq_end &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   gene_seq_start &lt;dbl&gt;, seq_coord_system &lt;dbl&gt;, seq_strand &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   entrezid &lt;chr&gt;, gene_biotype &lt;chr&gt;, gene_id &lt;chr&gt;, seq_name &lt;chr&gt;, …</span></span></span></code></pre>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 15,790 × 42</span></span></span>
<span><span class="co">##    .feature gene_name gene_seq_end gene_seq_start seq_coord_system seq_strand</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> TSPAN6   TSPAN6        99<span style="text-decoration: underline;">894</span>988       99<span style="text-decoration: underline;">883</span>667               <span style="color: #BB0000;">NA</span>         -<span style="color: #BB0000;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> DPM1     DPM1          49<span style="text-decoration: underline;">575</span>092       49<span style="text-decoration: underline;">551</span>404               <span style="color: #BB0000;">NA</span>         -<span style="color: #BB0000;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> SCYL3    SCYL3        169<span style="text-decoration: underline;">863</span>408      169<span style="text-decoration: underline;">818</span>772               <span style="color: #BB0000;">NA</span>         -<span style="color: #BB0000;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> C1orf112 C1orf112     169<span style="text-decoration: underline;">823</span>221      169<span style="text-decoration: underline;">631</span>245               <span style="color: #BB0000;">NA</span>          1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> CFH      CFH          196<span style="text-decoration: underline;">716</span>634      196<span style="text-decoration: underline;">621</span>008               <span style="color: #BB0000;">NA</span>          1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> FUCA2    FUCA2        143<span style="text-decoration: underline;">832</span>827      143<span style="text-decoration: underline;">815</span>948               <span style="color: #BB0000;">NA</span>         -<span style="color: #BB0000;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> GCLC     GCLC          53<span style="text-decoration: underline;">481</span>768       53<span style="text-decoration: underline;">362</span>139               <span style="color: #BB0000;">NA</span>         -<span style="color: #BB0000;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> NFYA     NFYA          41<span style="text-decoration: underline;">067</span>715       41<span style="text-decoration: underline;">040</span>684               <span style="color: #BB0000;">NA</span>          1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> STPG1    STPG1         24<span style="text-decoration: underline;">743</span>424       24<span style="text-decoration: underline;">683</span>489               <span style="color: #BB0000;">NA</span>         -<span style="color: #BB0000;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> NIPAL3   NIPAL3        24<span style="text-decoration: underline;">799</span>466       24<span style="text-decoration: underline;">742</span>284               <span style="color: #BB0000;">NA</span>          1</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 15,780 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 36 more variables: entrezid &lt;chr&gt;, gene_biotype &lt;chr&gt;, gene_id &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   seq_name &lt;chr&gt;, symbol &lt;chr&gt;, merged_transcripts &lt;int&gt;, .abundant &lt;lgl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ql__logFC &lt;dbl&gt;, ql__logCPM &lt;dbl&gt;, ql__F &lt;dbl&gt;, ql__PValue &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ql__FDR &lt;dbl&gt;, lr_robust__logFC &lt;dbl&gt;, lr_robust__logCPM &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   lr_robust__LR &lt;dbl&gt;, lr_robust__PValue &lt;dbl&gt;, lr_robust__FDR &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   deseq2__baseMean &lt;dbl&gt;, deseq2__log2FoldChange &lt;dbl&gt;, …</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="gene-description">Gene Description<a class="anchor" aria-label="anchor" href="#gene-description"></a>
</h4>
<p>With tidybulk, retrieving gene descriptions is straightforward,
making it easy to enhance the interpretability of your differential
expression results.</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add gene descriptions using the original SummarizedExperiment</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>    <span class="fu"><a href="../reference/describe_transcript-methods.html">describe_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Filter top significant genes</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ql__FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Print the gene statistics</span></span>
<span>    <span class="fu"><a href="../reference/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">description</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">not</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.feature</span>, <span class="va">description</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"ql"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">##   .feature description             ql__logFC ql__logCPM ql__F ql__PValue ql__FDR</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> LASP1    LIM and SH3 protein 1      -<span style="color: #BB0000;">0.389</span>      8.43   16.4  0.002<span style="text-decoration: underline;">99</span>   0.031<span style="text-decoration: underline;">6</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> KLHL13   kelch like family memb…     0.930      4.20   31.3  0.000<span style="text-decoration: underline;">361</span>  0.008<span style="text-decoration: underline;">92</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> CFLAR    CASP8 and FADD like ap…    -<span style="color: #BB0000;">1.17</span>       6.94   64.7  0.000<span style="text-decoration: underline;">023</span>7 0.002<span style="text-decoration: underline;">32</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> MTMR7    myotubularin related p…    -<span style="color: #BB0000;">0.967</span>      0.387  17.7  0.002<span style="text-decoration: underline;">40</span>   0.028<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ARF5     ARF GTPase 5               -<span style="color: #BB0000;">0.358</span>      5.88   17.1  0.002<span style="text-decoration: underline;">66</span>   0.029<span style="text-decoration: underline;">5</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> KDM1A    lysine demethylase 1A       0.310      5.90   20.0  0.001<span style="text-decoration: underline;">61</span>   0.021<span style="text-decoration: underline;">7</span></span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="step-4-batch-effect-correction">Step 4: Batch Effect Correction<a class="anchor" aria-label="anchor" href="#step-4-batch-effect-correction"></a>
</h3>
<p>ComBat-seq <span class="citation">(<a href="#ref-zhang2017combatseq">Zhang et al. 2020</a>)</span> is used for
batch effect correction in RNA-seq data.</p>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Adjust for batch effects</span></span>
<span><span class="va">airway</span> <span class="op">=</span> <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/adjust_abundance-methods.html">adjust_abundance</a></span><span class="op">(</span></span>
<span>      .factor_unwanted <span class="op">=</span> <span class="va">cell</span>, </span>
<span>      .factor_of_interest <span class="op">=</span> <span class="va">dex</span>, </span>
<span>    method <span class="op">=</span> <span class="st">"combat_seq"</span>, </span>
<span>      abundance <span class="op">=</span> <span class="st">"counts_tmm"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Found 4 batches</span></span>
<span><span class="co">## Using null model in ComBat-seq.</span></span>
<span><span class="co">## Adjusting for 1 covariate(s) or covariate level(s)</span></span>
<span><span class="co">## Estimating dispersions</span></span>
<span><span class="co">## Fitting the GLM model</span></span>
<span><span class="co">## Shrinkage off - using GLM estimates for parameters</span></span>
<span><span class="co">## Adjusting the data</span></span></code></pre>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Scatter plot of adjusted vs unadjusted</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Subset genes to speed up plotting</span></span>
<span>  <span class="va">_</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">.sample</span>, <span class="va">counts_tmm</span>, <span class="va">counts_tmm_adjusted</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">counts_tmm</span> <span class="op">+</span> <span class="fl">1</span>, y <span class="op">=</span> <span class="va">counts_tmm_adjusted</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">.sample</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">symbol</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, max.overlaps <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Scatter plot of adjusted vs unadjusted"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: ggrepel: 789 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">## increasing max.overlaps</span></span></code></pre>
<p><img src="introduction_files/figure-html/batch-correction-adjust-abundance-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="step-5-gene-enrichment-analysis">Step 5: Gene Enrichment Analysis<a class="anchor" aria-label="anchor" href="#step-5-gene-enrichment-analysis"></a>
</h3>
<p>Gene Set Enrichment Analysis (GSEA) <span class="citation">(<a href="#ref-subramanian2005gsea"><span class="nocase">Subramanian et
al.</span> 2005</a>)</span> is used for gene set enrichment.</p>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run gene rank enrichment (GSEA style)</span></span>
<span><span class="va">gene_rank_res</span> <span class="op">=</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span></span>
<span>    <span class="co"># Filter for genes with entrez IDs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">entrezid</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="st">"entrezid"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>  <span class="co"># Test gene rank</span></span>
<span>  <span class="fu"><a href="../reference/test_gene_rank-methods.html">test_gene_rank</a></span><span class="op">(</span></span>
<span>    .entrez <span class="op">=</span> <span class="va">entrezid</span>,</span>
<span>    .arrange_desc <span class="op">=</span> <span class="va">lr_robust__logFC</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>    gene_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>, <span class="st">"C2"</span>, <span class="st">"C5"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## using 'fgsea' for GSEA analysis, please cite Korotkevich et al (2019).</span></span></code></pre>
<pre><code><span><span class="co">## preparing geneSet collections...</span></span></code></pre>
<pre><code><span><span class="co">## GSEA analysis...</span></span></code></pre>
<pre><code><span><span class="co">## leading edge analysis...</span></span></code></pre>
<pre><code><span><span class="co">## done...</span></span></code></pre>
<pre><code><span><span class="co">## using 'fgsea' for GSEA analysis, please cite Korotkevich et al (2019).</span></span></code></pre>
<pre><code><span><span class="co">## preparing geneSet collections...</span></span></code></pre>
<pre><code><span><span class="co">## GSEA analysis...</span></span></code></pre>
<pre><code><span><span class="co">## leading edge analysis...</span></span></code></pre>
<pre><code><span><span class="co">## done...</span></span></code></pre>
<pre><code><span><span class="co">## using 'fgsea' for GSEA analysis, please cite Korotkevich et al (2019).</span></span></code></pre>
<pre><code><span><span class="co">## preparing geneSet collections...</span></span></code></pre>
<pre><code><span><span class="co">## GSEA analysis...</span></span></code></pre>
<pre><code><span><span class="co">## leading edge analysis...</span></span></code></pre>
<pre><code><span><span class="co">## done...</span></span></code></pre>
<pre><code><span><span class="co">## Warning: There was 1 warning in `mutate()`.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> In argument: `fit = map(...)`.</span></span>
<span><span class="co">## Caused by warning in `fgseaMultilevel()`:</span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.</span></span></code></pre>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect significant gene sets (example for C2 collection)</span></span>
<span><span class="va">gene_rank_res</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gs_collection</span> <span class="op">==</span> <span class="st">"C2"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">fit</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">unnest</span><span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 36 × 13</span></span></span>
<span><span class="co">##    gs_collection idx_for_plotting ID         Description setSize enrichmentScore</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> C2                           1 CHEN_LVAD… CHEN_LVAD_…      83          -<span style="color: #BB0000;">0.726</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> C2                           2 VECCHI_GA… VECCHI_GAS…     205          -<span style="color: #BB0000;">0.574</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> C2                           3 SHEDDEN_L… SHEDDEN_LU…     154          -<span style="color: #BB0000;">0.554</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> C2                           4 BOQUEST_S… BOQUEST_ST…     374          -<span style="color: #BB0000;">0.436</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> C2                           5 FEKIR_HEP… FEKIR_HEPA…      27          -<span style="color: #BB0000;">0.841</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> C2                           6 HOSHIDA_L… HOSHIDA_LI…     187          -<span style="color: #BB0000;">0.505</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> C2                           7 CHARAFE_B… CHARAFE_BR…     387          -<span style="color: #BB0000;">0.418</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> C2                           8 ZWANG_CLA… ZWANG_CLAS…     186          -<span style="color: #BB0000;">0.483</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> C2                           9 YAO_TEMPO… YAO_TEMPOR…      55          -<span style="color: #BB0000;">0.656</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> C2                          10 SCHAEFFER… SCHAEFFER_…     277          -<span style="color: #BB0000;">0.420</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 26 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 7 more variables: NES &lt;dbl&gt;, pvalue &lt;dbl&gt;, p.adjust &lt;dbl&gt;, qvalue &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   rank &lt;dbl&gt;, leading_edge &lt;chr&gt;, core_enrichment &lt;chr&gt;</span></span></span></code></pre>
<div class="section level5">
<h5 id="visualize-enrichment">Visualize enrichment<a class="anchor" aria-label="anchor" href="#visualize-enrichment"></a>
</h5>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/contribution-knowledge-mining/" class="external-link">enrichplot</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## enrichplot v1.29.2 Learn more at https://yulab-smu.top/contribution-knowledge-mining/</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Please cite:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Guangchuang Yu, Fei Li, Yide Qin, Xiaochen Bo, Yibo Wu and Shengqi</span></span>
<span><span class="co">## Wang. GOSemSim: an R package for measuring semantic similarity among GO</span></span>
<span><span class="co">## terms and gene products. Bioinformatics. 2010, 26(7):976-978</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'enrichplot'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GGally':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     ggtable</span></span></code></pre>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span>  <span class="va">gene_rank_res</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">unnest</span><span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>plot <span class="op">=</span> <span class="fu">pmap</span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">ID</span>, <span class="va">idx_for_plotting</span>, <span class="va">p.adjust</span><span class="op">)</span>,</span>
<span>      <span class="op">~</span> <span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/gseaplot2.html" class="external-link">gseaplot2</a></span><span class="op">(</span></span>
<span>        <span class="va">..1</span>, geneSetID <span class="op">=</span> <span class="va">..3</span>,</span>
<span>        title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s \nadj pvalue %s"</span>, <span class="va">..2</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">..4</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        base_size <span class="op">=</span> <span class="fl">6</span>, rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">0.5</span><span class="op">)</span>, subplots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">plot</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Warning: There was 1 warning in `mutate()`.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> In argument: `plot = pmap(...)`.</span></span>
<span><span class="co">## Caused by warning:</span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> `aes_()` was deprecated in ggplot2 3.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use tidy evaluation idioms with `aes()`</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">enrichplot</span> package.</span></span>
<span><span class="co">##   Please report the issue at</span></span>
<span><span class="co">##   <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/GuangchuangYu/enrichplot/issues&gt;</span>.</span></span></code></pre>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="introduction_files/figure-html/enrichment-visualize-gsea-plots-1.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span></code></pre>
<p><img src="introduction_files/figure-html/enrichment-visualize-gsea-plots-2.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span></code></pre>
<p><img src="introduction_files/figure-html/enrichment-visualize-gsea-plots-3.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span></code></pre>
<p><img src="introduction_files/figure-html/enrichment-visualize-gsea-plots-4.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span></code></pre>
<p><img src="introduction_files/figure-html/enrichment-visualize-gsea-plots-5.png" width="700"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span></code></pre>
<p><img src="introduction_files/figure-html/enrichment-visualize-gsea-plots-6.png" width="700"></p>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:enrichplot"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: 'enrichplot' namespace cannot be unloaded:</span></span>
<span><span class="co">##   namespace 'enrichplot' is imported by 'clusterProfiler' so cannot be unloaded</span></span></code></pre>
<p>Gene Ontology overrepresentation analysis <span class="citation">(<a href="#ref-ashburner2000go"><span class="nocase">Ashburner et al.</span>
2000</a>)</span> is used for functional enrichment.</p>
<div class="sourceCode" id="cb187"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test gene overrepresentation</span></span>
<span><span class="va">airway_overrep</span> <span class="op">=</span> </span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Label genes to test overrepresentation of</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>genes_to_test <span class="op">=</span> <span class="va">ql__FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>    <span class="co"># Filter for genes with entrez IDs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">entrezid</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="fu"><a href="../reference/test_gene_overrepresentation-methods.html">test_gene_overrepresentation</a></span><span class="op">(</span></span>
<span>    .entrez <span class="op">=</span> <span class="va">entrezid</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>    .do_test <span class="op">=</span> <span class="va">genes_to_test</span>,</span>
<span>    gene_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>, <span class="st">"C2"</span>, <span class="st">"C5"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway_overrep</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 5,417 × 13</span></span></span>
<span><span class="co">##    gs_collection ID      Description GeneRatio BgRatio RichFactor FoldEnrichment</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> C2            PASINI… PASINI_SUZ… 85/1747   316/22…      0.269           3.45</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> C2            CHARAF… CHARAFE_BR… 102/1747  465/22…      0.219           2.81</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> C2            CHARAF… CHARAFE_BR… 100/1747  456/22…      0.219           2.81</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> C2            REN_AL… REN_ALVEOL… 93/1747   407/22…      0.229           2.93</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> C2            CHEN_L… CHEN_LVAD_… 40/1747   102/22…      0.392           5.02</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> C2            ONDER_… ONDER_CDH1… 66/1747   257/22…      0.257           3.29</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> C2            LIU_PR… LIU_PROSTA… 98/1747   495/22…      0.198           2.54</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> C2            BOQUES… BOQUEST_ST… 89/1747   428/22…      0.208           2.66</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> C2            LU_AGI… LU_AGING_B… 65/1747   264/22…      0.246           3.15</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> C2            LIM_MA… LIM_MAMMAR… 94/1747   479/22…      0.196           2.51</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 5,407 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 6 more variables: zScore &lt;dbl&gt;, pvalue &lt;dbl&gt;, p.adjust &lt;dbl&gt;, qvalue &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Count &lt;int&gt;, entrez &lt;list&gt;</span></span></span></code></pre>
<p>EGSEA <span class="citation">(<a href="#ref-alhamdoosh2017egsea">Alhamdoosh et al. 2017</a>)</span> is
used for ensemble gene set enrichment analysis. EGSEA is a method that
combines multiple gene set enrichment analysis methods to provide a more
robust and comprehensive analysis of gene set enrichment. It creates a
web-based interactive tool that allows you to explore the results of the
gene set enrichment analysis.</p>
<div class="sourceCode" id="cb189"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EGSEA</span><span class="op">)</span></span>
<span><span class="co"># Test gene enrichment</span></span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>  <span class="co"># Filter for genes with entrez IDs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">entrezid</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="st">"entrezid"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span>  <span class="co"># Test gene enrichment</span></span>
<span>  <span class="fu"><a href="../reference/test_gene_enrichment-methods.html">test_gene_enrichment</a></span><span class="op">(</span></span>
<span>    .formula <span class="op">=</span> <span class="op">~</span><span class="va">dex</span>,</span>
<span>    .entrez <span class="op">=</span> <span class="va">entrezid</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"human"</span>, </span>
<span>    gene_sets <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>    methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"roast"</span><span class="op">)</span>,  <span class="co"># Use a more robust method</span></span>
<span>    cores <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:EGSEA"</span>, unload <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="step-6-cellularity-analysis">Step 6: Cellularity Analysis<a class="anchor" aria-label="anchor" href="#step-6-cellularity-analysis"></a>
</h3>
<p>CIBERSORT <span class="citation">(<a href="#ref-newman2015cibersort">Newman et al. 2015</a>)</span> is used
for cell type deconvolution.</p>
<p>Cellularity deconvolution is a standard approach for estimating the
cellular composition of a sample.</p>
<div class="section level4">
<h4 id="available-deconvolution-methods">Available Deconvolution Methods<a class="anchor" aria-label="anchor" href="#available-deconvolution-methods"></a>
</h4>
<p>The <code>tidybulk</code> package provides several methods for
deconvolution:</p>
<ul>
<li>
<strong>CIBERSORT</strong> <span class="citation">(<a href="#ref-newman2015cibersort">Newman et al. 2015</a>)</span>: Uses
support vector regression to deconvolve cell type proportions. Requires
the <code>class</code>, <code>e1071</code>, and
<code>preprocessCore</code> packages.</li>
<li>
<strong>LLSR</strong> <span class="citation">(<a href="#ref-newman2015cibersort">Newman et al. 2015</a>)</span>: Linear
Least Squares Regression for deconvolution.</li>
<li>
<strong>EPIC</strong> <span class="citation">(<a href="#ref-racle2017epic"><strong>racle2017epic?</strong></a>)</span>:
Uses a reference-based approach to estimate cell fractions.</li>
<li>
<strong>MCP-counter</strong> <span class="citation">(<a href="#ref-becht2016mcp"><strong>becht2016mcp?</strong></a>)</span>:
Quantifies the abundance of immune and stromal cell populations.</li>
<li>
<strong>quanTIseq</strong> <span class="citation">(<a href="#ref-finotello2019quantiseq"><strong>finotello2019quantiseq?</strong></a>)</span>:
A computational framework for inferring the immune contexture of
tumors.</li>
<li>
<strong>xCell</strong> <span class="citation">(<a href="#ref-aran2017xcell"><strong>aran2017xcell?</strong></a>)</span>:
Performs cell type enrichment analysis.</li>
</ul>
</div>
<div class="section level4">
<h4 id="example-usage">Example Usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h4>
<div class="sourceCode" id="cb190"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> </span>
<span>  <span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">symbol</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"cibersort"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"cibersort__"</span>, feature_column <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span> </span></code></pre></div>
<p>For the rest of the methods, you need to install the
<code>immunedeconv</code> package.</p>
<div class="sourceCode" id="cb191"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"immunedeconv"</span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"immunedeconv"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb192"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span> <span class="op">=</span> </span>
<span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> </span>
<span><span class="co"># Example using LLSR</span></span>
<span><span class="fu"><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"llsr"</span>, prefix <span class="op">=</span> <span class="st">"llsr__"</span>, feature_column <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using EPIC</span></span>
<span><span class="fu"><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"epic"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"epic__"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using MCP-counter</span></span>
<span><span class="fu"><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"mcp_counter"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"mcp__"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using quanTIseq</span></span>
<span><span class="fu"><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"quantiseq"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"quantiseq__"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span></span>
<span><span class="co"># Example using xCell</span></span>
<span><span class="fu"><a href="../reference/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"xcell"</span>, cores <span class="op">=</span> <span class="fl">1</span>, prefix <span class="op">=</span> <span class="st">"xcell__"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plotting-results">Plotting Results<a class="anchor" aria-label="anchor" href="#plotting-results"></a>
</h4>
<p>Visualize the cell type proportions as a stacked barplot for each
method:</p>
<div class="sourceCode" id="cb193"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize CIBERSORT results</span></span>
<span><span class="va">airway</span>   <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"cibersort__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CIBERSORT Cell Type Proportions"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/deconvolution-plotting-1.png" width="960"></p>
<div class="sourceCode" id="cb194"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="co"># Repeat similar plotting for LLSR, EPIC, MCP-counter, quanTIseq, and xCell</span></span>
<span><span class="va">airway</span>   <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"llsr__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"LLSR Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"epic__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"EPIC Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"mcp__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"MCP-counter Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"quantiseq__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"quanTIseq Cell Type Proportions"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">airway</span>     <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.sample</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"xcell__"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, names_to <span class="op">=</span> <span class="st">"Cell_type_inferred"</span>, values_to <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">Cell_type_inferred</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">my_theme</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"xCell Cell Type Proportions"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h3>
<p><code>tidybulk</code> allows you to get the bibliography of all
methods used in our workflow.</p>
<div class="sourceCode" id="cb195"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get bibliography of all methods used in our workflow</span></span>
<span><span class="va">airway</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/get_bibliography-methods.html">get_bibliography</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  @Article{tidybulk,</span></span>
<span><span class="co">##   title = {tidybulk: an R tidy framework for modular transcriptomic data analysis},</span></span>
<span><span class="co">##   author = {Stefano Mangiola and Ramyar Molania and Ruining Dong and Maria A. Doyle &amp; Anthony T. Papenfuss},</span></span>
<span><span class="co">##   journal = {Genome Biology},</span></span>
<span><span class="co">##   year = {2021},</span></span>
<span><span class="co">##   volume = {22},</span></span>
<span><span class="co">##   number = {42},</span></span>
<span><span class="co">##   url = {https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7},</span></span>
<span><span class="co">##   }</span></span>
<span><span class="co">## @article{wickham2019welcome,</span></span>
<span><span class="co">##   title={Welcome to the Tidyverse},</span></span>
<span><span class="co">##   author={Wickham, Hadley and Averick, Mara and Bryan, Jennifer and Chang, Winston and McGowan, Lucy D'Agostino and Francois, Romain and Grolemund, Garrett and Hayes, Alex and Henry, Lionel and Hester, Jim and others},</span></span>
<span><span class="co">##   journal={Journal of Open Source Software},</span></span>
<span><span class="co">##   volume={4},</span></span>
<span><span class="co">##   number={43},</span></span>
<span><span class="co">##   pages={1686},</span></span>
<span><span class="co">##   year={2019}</span></span>
<span><span class="co">##  }</span></span>
<span><span class="co">## @article{robinson2010edger,</span></span>
<span><span class="co">##   title={edgeR: a Bioconductor package for differential expression analysis of digital gene expression data},</span></span>
<span><span class="co">##   author={Robinson, Mark D and McCarthy, Davis J and Smyth, Gordon K},</span></span>
<span><span class="co">##   journal={Bioinformatics},</span></span>
<span><span class="co">##   volume={26},</span></span>
<span><span class="co">##   number={1},</span></span>
<span><span class="co">##   pages={139--140},</span></span>
<span><span class="co">##   year={2010},</span></span>
<span><span class="co">##   publisher={Oxford University Press}</span></span>
<span><span class="co">##  }</span></span>
<span><span class="co">## @article{robinson2010scaling,</span></span>
<span><span class="co">##   title={A scaling normalization method for differential expression analysis of RNA-seq data},</span></span>
<span><span class="co">##   author={Robinson, Mark D and Oshlack, Alicia},</span></span>
<span><span class="co">##   journal={Genome biology},</span></span>
<span><span class="co">##   volume={11},</span></span>
<span><span class="co">##   number={3},</span></span>
<span><span class="co">##   pages={1--9},</span></span>
<span><span class="co">##   year={2010},</span></span>
<span><span class="co">##   publisher={BioMed Central}</span></span>
<span><span class="co">##  }</span></span>
<span><span class="co">## @incollection{smyth2005limma,</span></span>
<span><span class="co">##   title={Limma: linear models for microarray data},</span></span>
<span><span class="co">##   author={Smyth, Gordon K},</span></span>
<span><span class="co">##   booktitle={Bioinformatics and computational biology solutions using R and Bioconductor},</span></span>
<span><span class="co">##   pages={397--420},</span></span>
<span><span class="co">##   year={2005},</span></span>
<span><span class="co">##   publisher={Springer}</span></span>
<span><span class="co">##  }</span></span>
<span><span class="co">## @Manual{,</span></span>
<span><span class="co">##     title = {R: A Language and Environment for Statistical Computing},</span></span>
<span><span class="co">##     author = {{R Core Team}},</span></span>
<span><span class="co">##     organization = {R Foundation for Statistical Computing},</span></span>
<span><span class="co">##     address = {Vienna, Austria},</span></span>
<span><span class="co">##     year = {2020},</span></span>
<span><span class="co">##     url = {https://www.R-project.org/},</span></span>
<span><span class="co">##   }</span></span>
<span><span class="co">## @article{lund2012detecting,</span></span>
<span><span class="co">##   title={Detecting differential expression in RNA-sequence data using quasi-likelihood with shrunken dispersion estimates},</span></span>
<span><span class="co">##   author={Lund, Steven P and Nettleton, Dan and McCarthy, Davis J and Smyth, Gordon K},</span></span>
<span><span class="co">##   journal={Statistical applications in genetics and molecular biology},</span></span>
<span><span class="co">##   volume={11},</span></span>
<span><span class="co">##   number={5},</span></span>
<span><span class="co">##   year={2012},</span></span>
<span><span class="co">##   publisher={De Gruyter}</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">## @article{zhou2014robustly,</span></span>
<span><span class="co">##   title={Robustly detecting differential expression in RNA sequencing data using observation weights},</span></span>
<span><span class="co">##   author={Zhou, Xiaobei and Lindsay, Helen and Robinson, Mark D},</span></span>
<span><span class="co">##   journal={Nucleic acids research},</span></span>
<span><span class="co">##   volume={42},</span></span>
<span><span class="co">##   number={11},</span></span>
<span><span class="co">##   pages={e91--e91},</span></span>
<span><span class="co">##   year={2014},</span></span>
<span><span class="co">##   publisher={Oxford University Press}</span></span>
<span><span class="co">##  }</span></span>
<span><span class="co">## @article{love2014moderated,</span></span>
<span><span class="co">##   title={Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2},</span></span>
<span><span class="co">##   author={Love, Michael I and Huber, Wolfgang and Anders, Simon},</span></span>
<span><span class="co">##   journal={Genome biology},</span></span>
<span><span class="co">##   volume={15},</span></span>
<span><span class="co">##   number={12},</span></span>
<span><span class="co">##   pages={550},</span></span>
<span><span class="co">##   year={2014},</span></span>
<span><span class="co">##   publisher={Springer}</span></span>
<span><span class="co">##  }</span></span>
<span><span class="co">## @article{law2014voom,</span></span>
<span><span class="co">##   title={voom: Precision weights unlock linear model analysis tools for RNA-seq read counts},</span></span>
<span><span class="co">##   author={Law, Charity W and Chen, Yunshun and Shi, Wei and Smyth, Gordon K},</span></span>
<span><span class="co">##   journal={Genome biology},</span></span>
<span><span class="co">##   volume={15},</span></span>
<span><span class="co">##   number={2},</span></span>
<span><span class="co">##   pages={R29},</span></span>
<span><span class="co">##   year={2014},</span></span>
<span><span class="co">##   publisher={Springer}</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">## @article{liu2015weight,</span></span>
<span><span class="co">##   title={Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses},</span></span>
<span><span class="co">##   author={Liu, Ruijie and Holik, Aliaksei Z and Su, Shian and Jansz, Natasha and Chen, Kelan and Leong, Huei San and Blewitt, Marnie E and Asselin-Labat, Marie-Liesse and Smyth, Gordon K and Ritchie, Matthew E},</span></span>
<span><span class="co">##   journal={Nucleic acids research},</span></span>
<span><span class="co">##   volume={43},</span></span>
<span><span class="co">##   number={15},</span></span>
<span><span class="co">##   pages={e97--e97},</span></span>
<span><span class="co">##   year={2015},</span></span>
<span><span class="co">##   publisher={Oxford University Press}</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">## @article{leek2012sva,</span></span>
<span><span class="co">##   title={The sva package for removing batch effects and other unwanted variation in high-throughput experiments},</span></span>
<span><span class="co">##   author={Leek, Jeffrey T and Johnson, W Evan and Parker, Hilary S and Jaffe, Andrew E and Storey, John D},</span></span>
<span><span class="co">##   journal={Bioinformatics},</span></span>
<span><span class="co">##   volume={28},</span></span>
<span><span class="co">##   number={6},</span></span>
<span><span class="co">##   pages={882--883},</span></span>
<span><span class="co">##   year={2012},</span></span>
<span><span class="co">##   publisher={Oxford University Press}</span></span>
<span><span class="co">##  }</span></span>
<span><span class="co">## @article{newman2015robust,</span></span>
<span><span class="co">##   title={Robust enumeration of cell subsets from tissue expression profiles},</span></span>
<span><span class="co">##   author={Newman, Aaron M and Liu, Chih Long and Green, Michael R and Gentles, Andrew J and Feng, Weiguo and Xu, Yue and Hoang, Chuong D and Diehn, Maximilian and Alizadeh, Ash A},</span></span>
<span><span class="co">##   journal={Nature methods},</span></span>
<span><span class="co">##   volume={12},</span></span>
<span><span class="co">##   number={5},</span></span>
<span><span class="co">##   pages={453--457},</span></span>
<span><span class="co">##   year={2015},</span></span>
<span><span class="co">##   publisher={Nature Publishing Group}</span></span>
<span><span class="co">##  }</span></span></code></pre>
<div class="sourceCode" id="cb197"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] patchwork_1.3.2                 GGally_2.4.0                   </span></span>
<span><span class="co">##  [3] DESeq2_1.49.4                   edgeR_4.7.5                    </span></span>
<span><span class="co">##  [5] limma_3.65.5                    tidySummarizedExperiment_1.19.0</span></span>
<span><span class="co">##  [7] airway_1.29.0                   tidybulk_1.99.4                </span></span>
<span><span class="co">##  [9] ttservice_0.5.3                 SummarizedExperiment_1.39.2    </span></span>
<span><span class="co">## [11] Biobase_2.69.1                  GenomicRanges_1.61.5           </span></span>
<span><span class="co">## [13] Seqinfo_0.99.2                  IRanges_2.43.5                 </span></span>
<span><span class="co">## [15] S4Vectors_0.47.4                BiocGenerics_0.55.1            </span></span>
<span><span class="co">## [17] generics_0.1.4                  MatrixGenerics_1.21.0          </span></span>
<span><span class="co">## [19] matrixStats_1.5.0               ggrepel_0.9.6                  </span></span>
<span><span class="co">## [21] ggplot2_4.0.0                   forcats_1.0.1                  </span></span>
<span><span class="co">## [23] magrittr_2.0.4                  purrr_1.1.0                    </span></span>
<span><span class="co">## [25] tibble_3.3.0                    tidyr_1.3.1                    </span></span>
<span><span class="co">## [27] dplyr_1.1.4                     knitr_1.50                     </span></span>
<span><span class="co">## [29] BiocStyle_2.37.1               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] splines_4.5.1          ggplotify_0.1.3        R.oo_1.27.1           </span></span>
<span><span class="co">##   [4] preprocessCore_1.71.2  XML_3.99-0.19          lifecycle_1.0.4       </span></span>
<span><span class="co">##   [7] Rdpack_2.6.4           rstatix_0.7.2          pbmcapply_1.5.1       </span></span>
<span><span class="co">##  [10] lattice_0.22-7         MASS_7.3-65            backports_1.5.0       </span></span>
<span><span class="co">##  [13] SnowballC_0.7.1        plotly_4.11.0          sass_0.4.10           </span></span>
<span><span class="co">##  [16] rmarkdown_2.30         jquerylib_0.1.4        yaml_2.3.10           </span></span>
<span><span class="co">##  [19] ggtangle_0.0.7         cowplot_1.2.0          pbapply_1.7-4         </span></span>
<span><span class="co">##  [22] DBI_1.2.3              minqa_1.2.8            RColorBrewer_1.1-3    </span></span>
<span><span class="co">##  [25] abind_1.4-8            glmmTMB_1.1.12         R.utils_2.13.0        </span></span>
<span><span class="co">##  [28] msigdbr_25.1.1         yulab.utils_0.2.1      rappdirs_0.3.3        </span></span>
<span><span class="co">##  [31] sandwich_3.1-1         sva_3.57.0             enrichplot_1.29.2     </span></span>
<span><span class="co">##  [34] tokenizers_0.3.0       tidytree_0.4.6         genefilter_1.91.0     </span></span>
<span><span class="co">##  [37] annotate_1.87.0        pkgdown_2.1.3          codetools_0.2-20      </span></span>
<span><span class="co">##  [40] DelayedArray_0.35.3    DOSE_4.3.0             tidyselect_1.2.1      </span></span>
<span><span class="co">##  [43] aplot_0.2.9            farver_2.1.2           lme4_1.1-37           </span></span>
<span><span class="co">##  [46] jsonlite_2.0.0         e1071_1.7-16           ellipsis_0.3.2        </span></span>
<span><span class="co">##  [49] Formula_1.2-5          survival_3.8-3         systemfonts_1.3.1     </span></span>
<span><span class="co">##  [52] tools_4.5.1            treeio_1.33.0          ragg_1.5.0            </span></span>
<span><span class="co">##  [55] Rcpp_1.1.0             glue_1.8.0             SparseArray_1.9.1     </span></span>
<span><span class="co">##  [58] xfun_0.53              mgcv_1.9-3             qvalue_2.41.0         </span></span>
<span><span class="co">##  [61] glmmSeq_0.5.5          withr_3.0.2            numDeriv_2016.8-1.1   </span></span>
<span><span class="co">##  [64] BiocManager_1.30.26    fastmap_1.2.0          boot_1.3-32           </span></span>
<span><span class="co">##  [67] fansi_1.0.6            digest_0.6.37          gridGraphics_0.5-1    </span></span>
<span><span class="co">##  [70] R6_2.6.1               textshaping_1.0.3      GO.db_3.21.0          </span></span>
<span><span class="co">##  [73] RSQLite_2.4.3          R.methodsS3_1.8.2      utf8_1.2.6            </span></span>
<span><span class="co">##  [76] data.table_1.17.8      class_7.3-23           httr_1.4.7            </span></span>
<span><span class="co">##  [79] htmlwidgets_1.6.4      S4Arrays_1.9.1         org.Mm.eg.db_3.21.0   </span></span>
<span><span class="co">##  [82] ggstats_0.11.0         pkgconfig_2.0.3        gtable_0.3.6          </span></span>
<span><span class="co">##  [85] blob_1.2.4             S7_0.2.0               XVector_0.49.1        </span></span>
<span><span class="co">##  [88] clusterProfiler_4.17.0 janeaustenr_1.0.0      htmltools_0.5.8.1     </span></span>
<span><span class="co">##  [91] carData_3.0-5          fgsea_1.35.6           bookdown_0.44         </span></span>
<span><span class="co">##  [94] TMB_1.9.17             scales_1.4.0           png_0.1-8             </span></span>
<span><span class="co">##  [97] reformulas_0.4.1       ggfun_0.2.0            uuid_1.2-1            </span></span>
<span><span class="co">## [100] reshape2_1.4.4         widyr_0.1.5            nlme_3.1-168          </span></span>
<span><span class="co">## [103] curl_7.0.0             nloptr_2.2.1           org.Hs.eg.db_3.21.0   </span></span>
<span><span class="co">## [106] proxy_0.4-27           cachem_1.1.0           zoo_1.8-14            </span></span>
<span><span class="co">## [109] stringr_1.5.2          parallel_4.5.1         AnnotationDbi_1.71.1  </span></span>
<span><span class="co">## [112] desc_1.4.3             pillar_1.11.1          grid_4.5.1            </span></span>
<span><span class="co">## [115] vctrs_0.6.5            ggpubr_0.6.1           car_3.1-3             </span></span>
<span><span class="co">## [118] xtable_1.8-4           evaluate_1.0.5         cli_3.6.5             </span></span>
<span><span class="co">## [121] locfit_1.5-9.12        compiler_4.5.1         rlang_1.1.6           </span></span>
<span><span class="co">## [124] crayon_1.5.3           tidytext_0.4.3         ggsignif_0.6.4        </span></span>
<span><span class="co">## [127] labeling_0.4.3         plyr_1.8.9             fs_1.6.6              </span></span>
<span><span class="co">## [130] ggiraph_0.9.1          stringi_1.8.7          viridisLite_0.4.2     </span></span>
<span><span class="co">## [133] BiocParallel_1.43.4    assertthat_0.2.1       babelgene_22.9        </span></span>
<span><span class="co">## [136] lmerTest_3.1-3         Biostrings_2.77.2      lazyeval_0.2.2        </span></span>
<span><span class="co">## [139] GOSemSim_2.35.1        Matrix_1.7-4           bit64_4.6.0-1         </span></span>
<span><span class="co">## [142] KEGGREST_1.49.1        statmod_1.5.0          rbibutils_2.3         </span></span>
<span><span class="co">## [145] igraph_2.1.4           broom_1.0.10           memoise_2.0.1         </span></span>
<span><span class="co">## [148] bslib_0.9.0            ggtree_3.99.0          fastmatch_1.1-6       </span></span>
<span><span class="co">## [151] bit_4.6.0              gson_0.1.0             ape_5.8-1</span></span></code></pre>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-alhamdoosh2017egsea" class="csl-entry">
Alhamdoosh, M, M Ng, NJ Wilson, et al. 2017. <span>“Combining Multiple
Tools Outperforms Individual Methods in Gene Set Enrichment
Analyses.”</span> <em>Bioinformatics</em> 33 (3): 414–24.
</div>
<div id="ref-anders2010rle" class="csl-entry">
Anders, Simon, and Wolfgang Huber. 2010. <span>“Differential Expression
Analysis for Sequence Count Data.”</span> <em>Genome Biology</em> 11
(10): R106.
</div>
<div id="ref-ashburner2000go" class="csl-entry">
<span class="nocase">Ashburner, Michael, Catherine A Ball, Judith A
Blake, et al.</span> 2000. <span>“Gene Ontology: Tool for the
Unification of Biology.”</span> <em>Nature Genetics</em> 25 (1): 25–29.
</div>
<div id="ref-bullard2010uq" class="csl-entry">
Bullard, James H, Elizabeth Purdom, Kasper D Hansen, and Sandrine
Dudoit. 2010. <span>“Evaluation of Statistical Methods for Normalization
and Differential Expression in mRNA-Seq Experiments.”</span> <em>BMC
Bioinformatics</em> 11 (1): 94.
</div>
<div id="ref-chen2016edgeR" class="csl-entry">
Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. <span>“From Reads
to Genes to Pathways: Differential Expression Analysis of RNA-Seq
Experiments Using Rsubread and the edgeR Quasi-Likelihood
Pipeline.”</span> <em>F1000Research</em> 5.
</div>
<div id="ref-hotelling1933pca" class="csl-entry">
Hotelling, Harold. 1933. <span>“Analysis of a Complex of Statistical
Variables into Principal Components.”</span> <em>Journal of Educational
Psychology</em> 24 (6): 417.
</div>
<div id="ref-kruskal1964mds" class="csl-entry">
Kruskal, Joseph B. 1964. <span>“Multidimensional Scaling by Optimizing
Goodness of Fit to a Nonmetric Hypothesis.”</span>
<em>Psychometrika</em> 29 (1): 1–27.
</div>
<div id="ref-law2014voom" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014.
<span>“Voom: Precision Weights Unlock Linear Model Analysis Tools for
RNA-Seq Read Counts.”</span> <em>Genome Biology</em> 15 (2): R29.
</div>
<div id="ref-liu2015voomweights" class="csl-entry">
Liu, Rui, Anna Z Holik, Shian Su, et al. 2015. <span>“Why Weight?
Modelling Sample and Observational Level Variability Improves Power in
RNA-Seq Analyses.”</span> <em>Nucleic Acids Research</em> 43 (15):
e97–97.
</div>
<div id="ref-love2014deseq2" class="csl-entry">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014.
<span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq
Data with DESeq2.”</span> <em>Genome Biology</em> 15 (12): 550.
</div>
<div id="ref-ma2020glmmseq" class="csl-entry">
<span class="nocase">Ma, Shuang, Bing Zhang, Laura M LaFave, et
al.</span> 2020. <span>“A General Linear Mixed Model for Differential
Expression Analysis of Single-Cell RNA-Seq Data.”</span> <em>Nature
Communications</em> 11 (1): 1–11.
</div>
<div id="ref-macqueen1967kmeans" class="csl-entry">
MacQueen, J. 1967. <span>“Some Methods for Classification and Analysis
of Multivariate Observations.”</span> <em>Proceedings of the Fifth
Berkeley Symposium on Mathematical Statistics and Probability</em> 1:
281–97.
</div>
<div id="ref-mccarthy2009treat" class="csl-entry">
McCarthy, Davis J, and Gordon K Smyth. 2009. <span>“Testing Significance
Relative to a Fold-Change Threshold Is a TREAT.”</span>
<em>Bioinformatics</em> 25 (6): 765–71.
</div>
<div id="ref-newman2015cibersort" class="csl-entry">
Newman, Aaron M, Chih Long Liu, Michael R Green, et al. 2015.
<span>“Robust Enumeration of Cell Subsets from Tissue Expression
Profiles.”</span> <em>Nature Methods</em> 12 (5): 453–57.
</div>
<div id="ref-robinson2010edger" class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010.
<span>“edgeR: A Bioconductor Package for Differential Expression
Analysis of Digital Gene Expression Data.”</span>
<em>Bioinformatics</em> 26 (1): 139–40.
</div>
<div id="ref-subramanian2005gsea" class="csl-entry">
<span class="nocase">Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha,
et al.</span> 2005. <span>“Gene Set Enrichment Analysis: A
Knowledge-Based Approach for Interpreting Genome-Wide Expression
Profiles.”</span> <em>Proceedings of the National Academy of
Sciences</em> 102 (43): 15545–50.
</div>
<div id="ref-zhang2017combatseq" class="csl-entry">
Zhang, Yu, Giovanni Parmigiani, and W Evan Johnson. 2020.
<span>“ComBat-Seq: Batch Effect Adjustment for RNA-Seq Count
Data.”</span> <em>NAR Genomics and Bioinformatics</em> 2 (3): lqaa078.
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
